---
title: "STARR-seq_A.t"
output:
  prettydoc::html_pretty:
    # theme: cayman # 1
    # theme: tactile #2
    # theme: architect #3
    # theme: leonids #4
    theme: hpstr #5
    highlight: github
    # highlight: vignette
editor_options: 
  markdown: 
    wrap: sentence
  chunk_output_type: console
---

```{r setup, include=FALSE, background="transparent", fig.align='center' }
knitr::opts_chunk$set(echo = F, include = T, highlight = T, warning = F, fig.align = "c")
options(knitr.kable.NA = '')
library(pheatmap)
library(grid)
library(data.table)
library(GenomicRanges)
library(tidyverse)
source("D:/SUST/code/TanYongjun_code.R")
```

# 0. Read Me

Only enhancers identified by traditional STARR-seq were used, because a set of varied length mRNAs generated from the same fragment which contain PAS were count as different fragments in data analysis pipeline of iSTARR-seq (Jing Wan).

All result files were saved at `/data5/wanjing/Arabidopsis_enhancer/work2`, raw data were saved at `/data5/wanjing/Arabidopsis_enhancer/data`, copy all of them to the `/scratch` of TaiYi.

```{bash, eval=F}
scp -r work2 bio-tanyj@172.18.6.175:/scratch/2021-02-22/bioTanyj/
```

# 1. Summary of data generated in STARR-seq.

## 1.1 Summary of reads/fragments

Number of raw reads, clean reads, all fragments, and unique fragments in library of cDNA and plasmid.

### .1 Number

*Number of fragments in enhancer calling*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
df <- NULL

for (i in list.files(path = "./R2.enhancer_calling/", pattern = "Number_fragments.+calling.tsv", recursive = T)) {
  df <- fread(paste("./R2.enhancer_calling/", i, sep = ""), header = F) %>%
    mutate(method = str_replace_all(i, "(.+)/.+", "\\1")) %>%
    rbind(., df)
}
names(df) <- c("Number", "id", "method")
df <- df %>%
  mutate(library = NA, 
         strategy = str_extract(id, "(startOnly|strategy1)"))
df$library[str_detect(df$id, "(P1|P2|P_)")] <- "plasmid"
df$library[str_detect(df$id, "(C1|C2|C_)")] <- "cDNA"
df$library[str_detect(df$id, "_dT")] <- "cDNA_dT"
df$library[str_detect(df$id, "_nodT")] <- "cDNA_nodT"
df$replicate <- str_replace_all(df$id, "(RAtEC|RAtEP|RAtEnWC|RAtEnWP)(\\d*)_.+", "\\2") 
df$replicate[df$replicate != 1 & df$replicate != 2] <- "merge"
df %>%
  dplyr::select(-id) %>%
  unique()
```

```{r}
pivot_wider(id_cols = c(method, library, replicate), names_from = "strategy", values_from = "Number") %T>%
  # fwrite(file = "./R2.enhancer_calling/Number_fragments_in_cSTARRseq_iSTARRseq_calling.tsv", sep = "\t")

NA
```

### .2 Mapped reads




## 1.2 Reproducibility between replicates

### .1 calculation

`/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/1.2_CorrelationBetweenBamFiles.lsf`

### .2 plot

binSize = 600bp

```{r}
source("D:/SUST/code/TanYongjun_code.R")
df <- fread("./R1.library_summary/bam_correlation/Bam_files_PearsonCorr_readCounts_600bpbin.tab")
df$V1 <- NULL

row.names(df) <- c("plasmid-1", "plasmid-2", "cDNA-1", "cDNA-2")
colnames(df) <- c("plasmid-1", "plasmid-2", "cDNA-1", "cDNA-2")

p <- pheatmap::pheatmap(df, 
         cluster_rows = F, cluster_cols = F, 
         display_numbers = T, number_color = "grey20", 
         angle_col = 45, number_format = "%.4f", 
         fontsize = 6)
toolkit_tyj$SavePlot(filename_prefix = "./R1.library_summary/bam_correlation/R1_Heatmap_PearsonCorr_600bpbin",
                     plot = p, device = c("png", "pdf"),
                     width = 7, height = 6)
```

Correlation matrix calculated by WanJing

```{r}
df <- fread("./work2_wj/3.library_quality/old/PearsonCorr.tab")
row.names(df) <- df$V1
df$V1 <- NULL
p <- pheatmap(df, 
         cluster_rows = F, cluster_cols = F, 
         display_numbers = T, number_color = "grey20", 
         angle_col = 45, number_format = "%.4f", 
         fontsize_number = 10)
toolkit_tyj$SavePlot(filename_prefix = "./R1.library_summary/bam_correlation/R1_Heatmap_PearsonCorr_WJ",
                     plot = p,
                     width = 12, height = 10)
```

## 1.3 Fragment length and Genome coverage

### .2 Fragment length

Distribution of all `unique` fragments length.
Two scripts: `/scratch/2021-03-02/bioTanyj/At_enhancer/scripts/1.3.2_FragmentsLengthDistribution.lsf` `/scratch/2021-03-02/bioTanyj/At_enhancer/scripts/1.3.2_FragmentsLengthDistribution.R` Using bed files which contain all unique fragments.

### .3 Cumulative coverage of the whole genome

\*Step 1, calculate depth of each base pair.\
`/scratch/2021-03-02/bioTanyj/At_enhancer/scripts/1.3.3_CoverageNoRepeatRegion.lsf`

\*Step 2, plot in R.

```{r}
  source("D:/SUST/code/TanYongjun_code.R")
  depth.all <- NULL
  for (i in list.files(pattern = ".merge.coverage_RefNoRepeat.tmp.txt", 
                       path = "./R1.library_summary/CumulativeCoverage/")) {
    temp <- fread(file = paste("./R1.library_summary/CumulativeCoverage/", i, sep = ""), 
                  sep = "\t", header = F, data.table = F)
    names(temp) <- c("Depth", "Counts")
    temp$Method <- str_replace_all(i, "(.+?)_(.+)\\.merge.co.+", "\\1")
    temp$sample <- str_replace_all(i, "(.+?)_(.+)\\.merge.co.+", "\\2")
    temp$temp <- temp$Depth * temp$Counts
    temp <- plyr::arrange(temp, -Depth)
    temp$Percent <- temp$Counts / sum(temp$Counts)
    temp$Percent <- cumsum(temp$Percent)
    depth.all <- rbind(depth.all, temp)
  }
  

  library(ggformula)
## cSTARR + iSTARR
  depth.all %>%
    # mutate(Sample = factor(Sample, levels = c("plasmid", "cDNA"))) %>%
    ggplot(aes(x = Depth, y = Percent)) + 
      geom_col(color = "grey50", fill = "grey50") +
      # geom_spline() +
      facet_grid(Method ~ sample) +
      xlim(1, 400) + 
      toolkit_tyj$theme_my + 
      scale_y_continuous(breaks = seq(0, 1, by = 0.2), labels = scales::percent) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R1.library_summary/CumulativeCoverage/Cumulative_coverage_in_no-repeated_region",
                       width = 16, height = 12)
  
## cSTARR
  depth.all %>%
    dplyr::filter(Method == "cSTARRseq", 
                  sample != "C", 
                  sample != "P") %>%
    mutate(Replicate = ifelse(str_detect(sample, ".+1$"), "replicate1", "replicate2"),
           Sample = str_replace_all(sample, c("C1" = "cDNA", 
                                              "C2" = "cDNA",
                                              "P1" = "plasmid",
                                              "P2" = "plasmid")),
           Sample = factor(Sample, levels = c("plasmid", "cDNA"))) %>%
    ggplot(aes(x = Depth, y = Percent)) + 
      geom_col(color = "grey50", fill = "grey50") +
      # geom_spline() +
      facet_grid(Sample ~ Replicate) +
      xlim(1, 400) + 
      toolkit_tyj$theme_my + 
      scale_y_continuous(breaks = seq(0, 1, by = 0.2), labels = scales::percent) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R1.library_summary/CumulativeCoverage/Cumulative_coverage_in_no-repeated_region_cSTARRseq",
                       width = 12, height = 8, device = c("png", "svg"))
  rm(temp, i, j, p, df)
```

### .4 Cov of genomic features.

Calcualte the cuverage at the different genomic features (promoter, exon, intron, and UTR region et.al.) overlapped with PAS or not.

*1. Bed files for each genomic regions*

```{r prepare bed files for each genomic regions, eval=F}
# priority of regions: (repeat sequences, CDS, core promoter(+-50bp around TSS), 5'UTR, 3'UTR, 1st intron, 200bp upstrem TSS, intergenic)
  source("D:/SUST/code/TanYongjun_code.R")
# repeat regions (chromosome length match)
  repeat_region <- fread(input = "./work2_wj/3.library_quality/TAIR10_repeat_region_chr.bed") %>%
    dplyr::rename(seqnames = V1, start = V2, end = V3) %>%
    mutate(seqnames = str_replace_all(seqnames, "chr", "")) %>%
    makeGRangesFromDataFrame()

# PAS regions (download from PlantAPAdb, high conficence PAC regions.)
  pas_HQ <- fread("./refgenome/arabidopsis_thaliana.high_confidence.PAC.bed", header = F) %>%
    dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
    makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3")
  quantile(width(pas_HQ), seq(0, 1, 0.05))
  # hist(width(pas_HQ))
  
# genome regions 
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  
  # intron
    # first intron
    intron_all <- intronsByTranscript(txdb, use.names = T) %>% as.data.frame()
    first_intron1 <- intron_all %>%
      dplyr::filter(strand == "+") %>%
      dplyr::arrange(group_name, start)
    first_intron1 <- first_intron1[match(unique(first_intron1$group_name), first_intron1$group_name),]
    first_intron2 <- intron_all %>%
      dplyr::filter(strand == "-") %>%
      dplyr::arrange(group_name, desc(start))
    first_intron2 <- first_intron2[match(unique(first_intron2$group_name), first_intron2$group_name),]
    first_intron <- rbind(first_intron1, first_intron2) %>%
      dplyr::arrange(seqnames, group_name, start) %>%
      na.omit() %>%
      makeGRangesFromDataFrame() %>%
      GenomicRanges::reduce()
    rm(first_intron1, first_intron2)  

    # first intron
    other_intron <- intron_all %>%
      makeGRangesFromDataFrame() %>%
      GenomicRanges::setdiff(., first_intron) %>%
      GenomicRanges::reduce()
    # findOverlaps(first_intron, other_intron)

# deduct overlapped regions based on priority.
  # (repeat sequences, CDS, core promoter(+-50bp around TSS), 5'UTR, 3'UTR, 1st intron, 200bp upstrem TSS, intergenic)
  cds_hg <- cds(txdb) %>%
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE)
  
  pro_core <- promoters(txdb, upstream = 50, downstream = 50) %>% 
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE)
  
  pro_200bp <- promoters(txdb, upstream = 200, downstream = 0) %>% 
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_core, ignore.strand=TRUE)
  
  five_utr <- fiveUTRsByTranscript(txdb) %>% 
    unlist() %>% 
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_core, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_200bp, ignore.strand=TRUE)
  
  three_utr <- threeUTRsByTranscript(txdb) %>% 
    unlist() %>% 
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_core, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_200bp, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., five_utr, ignore.strand=TRUE)
  
  first_intron <- first_intron %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_core, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_200bp, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., five_utr, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., three_utr, ignore.strand=TRUE)
  
  other_intron <- other_intron %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_core, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_200bp, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., five_utr, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., first_intron, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., three_utr, ignore.strand=TRUE) 
  
  intergenic <- IRanges::gaps(cds_hg, start = 1L, end = seqlengths(pro_200bp)) %>%
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., repeat_region, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., cds_hg, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_core, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_200bp, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., five_utr, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., three_utr, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., first_intron, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., other_intron, ignore.strand=TRUE)

  genomic_regions <- list(repeat_region = repeat_region,
                          CDS = cds_hg,
                          promoter_core = pro_core,
                          promoter_200bp = pro_200bp,
                          first_intron = first_intron,
                          other_intron = other_intron,
                          five_UTR = five_utr,
                          three_UTR = three_utr,
                          intergenic = intergenic)
  
### verification of extracted genomic regions.
# Is any overlap among genomic features?
  tb <- combn(names(genomic_regions), 2) %>% t() %>% as.data.frame()
  for (i in 1:nrow(tb)) {
    cat(tb$V1[i], tb$V2[i], "  Number of overlapped regions:  ", NROW(findOverlaps(genomic_regions[[tb$V1[i]]], 
                                 genomic_regions[[tb$V2[i]]])), "\n")
  }
  # findOverlaps(first_intron, other_intron)
  
# the max chromosome position in each genomic features
  purrr::map(genomic_regions, function(x){
    x %>%
      as.data.frame() %>%
      mutate(seqnames = as.character(seqnames)) %>%
      group_by(seqnames) %>%
      summarise(MaxPos = max(end) / 1e6) %>%
      as.data.frame() %>%
      return()
  })
  
# the region border were all repeatly counted?
  x <- genomic_regions %>%
    purrr::reduce(c) %>%
    as.data.frame() %>%
    arrange(seqnames, start)
  
# summary of all genomic regions
  temp <- lapply(genomic_regions, function(x){
    return(sum(width(x)))
  }) %>% unlist()
  
  cat("The total length of these genomic features was ", sum(temp) / 1e6, 
      "MB \n Attention: CDS")
  
  # seqlengths(txdb) %>% sum() / 1e6
  
  # plot
  data.frame(name = names(temp), Length_MB = temp / 1e6) %>%
    ggplot(aes(x = name, y = Length_MB, fill = name)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_fill_npg() + 
      scale_y_log10() + 
      # ggtitle("Length of each genomic features") + 
      xlab("Genomic features") + 
      guides(fill = guide_legend(title = "Genomic Features"))
  
### two set of files containing genomic features overlapped by PAS or not.
  genomic_regions_PAS <- list()
  genomic_regions_noPAS <- list()
  for (i in 1:length(genomic_regions)) {
    id <- names(genomic_regions)[i]
    region_tmp <- genomic_regions[[id]]
    ol <- findOverlaps(region_tmp, pas_HQ)
    PAS_tmp <- region_tmp[unique(ol@from)]
    noPAS_tmp <- region_tmp[-unique(ol@from)]
    
    genomic_regions_PAS <- c(genomic_regions_PAS, list(PAS_tmp))
    names(genomic_regions_PAS)[length(genomic_regions_PAS)] <- paste(id, "_PAS", sep = "")
    genomic_regions_noPAS <- c(genomic_regions_noPAS, list(noPAS_tmp))
    names(genomic_regions_noPAS)[length(genomic_regions_noPAS)] <- paste(id, "_noPAS", sep = "")
  }

#### write to bed file
  for (i in 1:length(genomic_regions)) {
    genomic_regions[[i]] %>%
      as.data.frame() %>%
      dplyr::mutate(seqnames = str_c("chr", as.character(seqnames), sep = "")) %>%
      fwrite(file = paste("./R1.library_summary/Cov_genomic_features/bed_genomic_features/",
             names(genomic_regions)[i], ".bed.gz", sep = ""),
             quote = F, row.names = F, col.names = F, sep = "\t")
    
    genomic_regions_PAS[[i]] %>%
      as.data.frame() %>%
      dplyr::mutate(seqnames = str_c("chr", as.character(seqnames), sep = "")) %>%
      fwrite(file = paste("./R1.library_summary/Cov_genomic_features/bed_genomic_features/",
             names(genomic_regions_PAS)[i], ".bed.gz", sep = ""),
             quote = F, row.names = F, col.names = F, sep = "\t")
    
    genomic_regions_noPAS[[i]] %>%
      as.data.frame() %>%
      dplyr::mutate(seqnames = str_c("chr", as.character(seqnames), sep = "")) %>%
      fwrite(file = paste("./R1.library_summary/Cov_genomic_features/bed_genomic_features/",
             names(genomic_regions_noPAS)[i], ".bed.gz", sep = ""),
             quote = F, row.names = F, col.names = F, sep = "\t")
  }
```

*2. Calculate coverage with bed tools*

```{bash, eval=F}
#!/bin/bash
#BSUB -J RegionCov                  ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 2                  ### ask for number of cores (default: 1)
#BSUB -W 20:00               ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

cd /scratch/2021-03-02/bioTanyj/At_enhancer/R1.library_summary/Cov_genomic_features/bed_genomic_features
for i in `ls | grep -E "bed.gz$"`
do
    echo ">>> ${i}............"
    # # devide into windows; this step seems not needed, because `mean coverage` was calculated in the following step.
    # zcat ${i} | \
    #     /work/bio-tanyj/soft/bedtools makewindows -b stdin -w 1000 \
    #     > ${i%.bed.gz}.windowed.bed

    # calculate coverage
    for j in `ls ../../CumulativeCoverage | grep -E ".STARR.+.bed$"`
    do
        if [ ! -a "../${i%.bed.gz}_${j%.bed}.lock" ]; then
            echo "    `basename ${j}`"
            echo "lock" > ../${i%.bed.gz}_${j%.bed}.lock
            zcat ${i} | /work/bio-tanyj/soft/bedtools coverage -b ../../CumulativeCoverage/${j} -a stdin -mean > ../${i%.bed.gz}_${j%.bed}.bed 
        fi
    done
done

```

*3. Collect results in R*             
The last column of files generated in previous step was mean coverage in each interval.

```{r, eval=F}
# Collect coverage of each intervals for all kinds of genomic regions (only bed files for merged replicates).
#                          TYJ 20201118

setwd("/scratch/2021-02-22/bioTanyj/At_enhancer/cSTARR_seq/R1_library_summary/R1.3_fragment_length_Genome_coverage/genomic_features")
source("/work/bio-tanyj/scripts/TanYongjun_code.R", echo = F)
regionCov <- NULL
for(i in list.files(pattern=".bed$")){
    cat("\nloading:  ", i)
    temp <- fread(file = i, data.table = F)
    temp <- data.frame(id = i, width = temp[,4], coverage = temp[,6])
    regionCov <- rbind(regionCov, temp)
}
table(regionCov$id)
object.size(regionCov)
# save(regionCov, file = "genomic_regions_coverage_all.bin")
fwrite(regionCov, file = "genomic_regions_coverage_all.csv.gz") # faster than save() but generate a little bigger file


# Collect coverage of each intervals for all kinds of genomic regions (only bed files for each replicates).
#                          TYJ 20210415

setwd("/scratch/2021-04-10/bioTanyj/At_enhancer/R1.library_summary/Cov_genomic_features/backup")
source("/work/bio-tanyj/scripts/TanYongjun_code.R", echo = F)
regionCov <- NULL
for(i in list.files(pattern=".bed$")){
    cat("\nloading:  ", i)
    temp <- fread(file = i, data.table = F)
    temp <- data.frame(id = i, width = temp[,4], coverage = temp[,6])
    regionCov <- rbind(regionCov, temp)
}
table(regionCov$id)
object.size(regionCov)
# save(regionCov, file = "genomic_regions_coverage_all.bin")
fwrite(regionCov, file = "genomic_regions_coverage_each_replicates.csv.gz") # faster than save() but generate a little bigger file
```

*4. plot in R*

merged replicates

```{r, eval=F}
source("D:/SUST/code/TanYongjun_code.R")
regionCov <- fread("./R1.library_summary/Cov_genomic_features/genomic_regions_coverage_all.csv.gz")
head(regionCov)
regionCov <- regionCov %>%
  as_tibble() %>%
  mutate(Type = "All", 
         library = ifelse(str_detect(id, "_P\\."), "plasmid", "cDNA"),
         Method = str_replace_all(id, ".+(cSTARRseq|iSTARRseq).+", "\\1"),
         genomic_regions = str_replace_all(id, "(.+)_(cSTARRseq|iSTARRseq).+", "\\1")) 

regionCov$Type[str_detect(regionCov$id, "_PAS")] <- "PAS"
regionCov$Type[str_detect(regionCov$id, "_noPAS")] <- "noPAS"
regionCov$genomic_regions <- str_replace_all(regionCov$genomic_regions, "_(PAS|noPAS)", "")
regionCov <- regionCov %>%
  # dplyr::filter(genomic_regions != "repeat_region") %>%
  mutate(genomic_regions = str_replace_all(genomic_regions, 
                                           c("first_intron" = "1st intron", 
                                             "five_UTR" = "5' UTR", 
                                             "intergenic" = "Intergenic", 
                                             "other_intron" = "Other introns", 
                                             "promoter_200bp" = "200bp upstream TSS", 
                                             "promoter_core" = "+/- 50bp TSS", 
                                             "repeat_region" = "Repeat region",
                                             "three_UTR" = "3' UTR"))) %>%
  mutate(genomic_regions = factor(genomic_regions, 
                                  levels = c("200bp upstream TSS", 
                                             "+/- 50bp TSS",
                                             "5' UTR",
                                             "1st intron",
                                             "CDS",
                                             "Other introns",
                                             "3' UTR",
                                             "Intergenic",
                                             "Repeat region"
                                             )))
# plot
# quantile(regionCov$coverage, probs = seq(0, 1, by = 0.1))

# Genomic regions (cSTARRseq + iSTARRseq)
p <- regionCov %>%
  dplyr::filter(Type == "All") %>%
  ggplot(aes(x = genomic_regions, y = coverage)) + 
    # stat_boxplot(geom = "errorbar", width = 0.4) +
    geom_boxplot(outlier.alpha = 0, aes(color = genomic_regions), 
                 position = position_dodge(width = 0.8), width = 0.6, fill = "transparent") +
    # geom_violin(fill = "transparent", scale = "width") +
    facet_grid(library ~ Method, scales = "free_y") +
    coord_cartesian(ylim = c(0, 450)) +
    ylab("Read depth") + 
    xlab("Genomic regions") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          axis.ticks.x = element_blank(), 
          legend.position = "none") +
    scale_color_npg() +
    toolkit_tyj$theme_my

  # toolkit_tyj$SavePlot(filename_prefix = paste("./R1.library_summary/Cov_genomic_features/",
  #                                              "Coverage_of_genomic_features_20210308", sep = ""),
  #                      width = 12, height = 10, plot = p)
  
# Genomic regions (cSTARRseq)
p <- regionCov %>%
  dplyr::filter(Type == "All", Method == "cSTARRseq") %>%
  ggplot(aes(x = genomic_regions, y = coverage)) + 
    # stat_boxplot(geom = "errorbar", width = 0.4) +
    geom_boxplot(outlier.alpha = 0, aes(color = genomic_regions), 
                 position = position_dodge(width = 0.8), width = 0.6, fill = "transparent") +
    # geom_violin(fill = "transparent", scale = "width") +
    facet_grid(library ~ ., scales = "free") +
    coord_cartesian(ylim = c(0, 450)) +
    ylab("Read depth") + 
    xlab("Genomic regions") +
    theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1), 
          axis.ticks.x = element_blank(), 
          legend.position = "none") +
    scale_color_npg() +
    toolkit_tyj$theme_my

  # toolkit_tyj$SavePlot(filename_prefix = paste("./R1.library_summary/Cov_genomic_features/",
  #                                              "Coverage_of_genomic_features_cSTARRseq_20210415", sep = ""),
  #                      width = 12, height = 10, plot = p, device = c("png", "svg"))
  
# Genomic regions x PAS
p <- regionCov %>%
  ggplot(aes(x = genomic_regions, y = coverage)) + 
    # stat_boxplot(geom = "errorbar", width = 0.4) +
    geom_boxplot(outlier.alpha = 0, aes(color = Type), 
                 position = position_dodge(width = 0.8), 
                 width = 0.6, fill = "transparent", 
                 size = 0.2) +
    # geom_violin(fill = "transparent", scale = "width") +
    facet_grid(library ~ Method, scales = "free_y") +
    coord_cartesian(ylim = c(0, 450)) +
    ylab("Read depth") + 
    xlab("Genomic regions") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          axis.ticks.x = element_blank(), 
          legend.position = "right") +
    scale_color_npg() +
    toolkit_tyj$theme_my

  toolkit_tyj$SavePlot(filename_prefix = paste("./R1.library_summary/Cov_genomic_features/",
                                               "Coverage_of_genomic_features_plus_PAS_20210308", sep = ""),
                       width = 12, height = 8, plot = p)
  
# PAS or not
p <- regionCov %>%
  # dplyr::filter(Type != "All") %>%
  ggplot(aes(x = Type, y = coverage)) + 
    # stat_boxplot(geom = "errorbar", width = 0.4) +
    geom_boxplot(outlier.alpha = 0, aes(color = Type),
                 position = position_dodge(width = 0.8), width = 0.6, fill = "transparent") +
    # geom_violin(fill = "transparent", scale = "width") +
    facet_grid(library ~ Method, scales = "free_y") +
    coord_cartesian(ylim = c(0, 250)) +
    ylab("Read depth") + 
    xlab("Genomic regions") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          axis.ticks.x = element_blank(), 
          legend.position = "none") +
    scale_fill_npg() +
    toolkit_tyj$theme_my

  toolkit_tyj$SavePlot(filename_prefix = paste("./R1.library_summary/Cov_genomic_features/",
                                               "Coverage_of_region_PAS_20210308-3", sep = ""),
                       width = 6, height = 8, plot = p)
```

Each replicates

```{r, eval=F}
source("D:/SUST/code/TanYongjun_code.R")
regionCov <- fread("./R1.library_summary/Cov_genomic_features/genomic_regions_coverage_each_replicates.csv.gz") %>%
  dplyr::filter(str_detect(id, "iSTARRseq", negate = T))
head(regionCov)
regionCov <- regionCov %>%
  mutate(Type = "All", 
         Replicates = ifelse(str_detect(id, ".+1.bed"), "replicate1", "replicate2"),
         library = ifelse(str_detect(id, "_P\\d.+"), "plasmid", "cDNA"),
         Method = str_replace_all(id, ".+(cSTARRseq|iSTARRseq).+", "\\1"),
         genomic_regions = str_replace_all(id, "(.+)_(cSTARRseq|iSTARRseq).+", "\\1")) 

regionCov$Type[str_detect(regionCov$id, "_PAS")] <- "PAS"
regionCov$Type[str_detect(regionCov$id, "_noPAS")] <- "noPAS"
regionCov$genomic_regions <- str_replace_all(regionCov$genomic_regions, "_(PAS|noPAS)", "")
regionCov <- regionCov %>%
  # dplyr::filter(genomic_regions != "repeat_region") %>%
  mutate(genomic_regions = str_replace_all(genomic_regions, 
                                           c("first_intron" = "1st intron", 
                                             "five_UTR" = "5' UTR", 
                                             "intergenic" = "Intergenic", 
                                             "other_intron" = "Other introns", 
                                             "promoter_200bp" = "200bp upstream TSS", 
                                             "promoter_core" = "+/- 50bp TSS", 
                                             "repeat_region" = "Repeat region",
                                             "three_UTR" = "3' UTR"))) %>%
  mutate(genomic_regions = factor(genomic_regions, 
                                  levels = c("200bp upstream TSS", 
                                             "+/- 50bp TSS",
                                             "5' UTR",
                                             "1st intron",
                                             "CDS",
                                             "Other introns",
                                             "3' UTR",
                                             "Intergenic",
                                             "Repeat region"
                                             )))
# plot
# quantile(regionCov$coverage, probs = seq(0, 1, by = 0.1))

# Genomic regions (cSTARRseq each replicates)
p <- regionCov %>%
  dplyr::filter(Type == "All") %>%
  ggplot(aes(x = genomic_regions, y = coverage)) + 
    # stat_boxplot(geom = "errorbar", width = 0.4) +
    geom_boxplot(outlier.alpha = 0, aes(color = genomic_regions), 
                 position = position_dodge(width = 0.8), width = 0.6, fill = "transparent") +
    # geom_violin(fill = "transparent", scale = "width") +
    facet_grid(library ~ Replicates) +
    coord_cartesian(ylim = c(0, 300)) +
    ylab("Read depth") + 
    xlab("Genomic regions") +
    theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1), 
          axis.ticks.x = element_blank(), 
          legend.position = "none") +
    scale_color_npg() +
    toolkit_tyj$theme_my

  toolkit_tyj$SavePlot(filename_prefix = paste("./R1.library_summary/Cov_genomic_features/",
                                               "Coverage_of_genomic_features_cSTARRseq_replicates_20210415", sep = ""),
                       width = 12, height = 10, plot = p, device = c("png", "pdf"))
  

```

### .5 Cov of bins with varied GC content

Two steps: 1.
calculate coverage of each genomic bins (), generate files contain two columns: 1.GC% of bin, 2.
Coverage of bins 2.
plot in R.
All run on server.

# 2. Enhancer identification.

## 2.1 Overlap (Methods x replicates x Strategy)

### 2.1.1 load and filtration

```{r}
source("D:/SUST/code/TanYongjun_code.R")
# collect raw enhancers
en_raw <- list()
for (i in list.files(path = "./R2.enhancer_calling/", pattern = ".Enhancer.+raw.tsv", recursive = T)) {
  id <- str_replace_all(i, ".+/(.+)_raw.tsv", "\\1") %>% str_replace_all(., "Extended", "")
  en_raw[[id]] <- fread(file = paste("./R2.enhancer_calling/", i, sep = ""), header = T) %>%
    mutate(method = ifelse(str_detect(id, "iEnhancer"),"iSTARR-seq", "cSTARR-seq"), 
           ID = id, id = id) %>%
    separate(id, c("Type", "rep", "Strategy"))
}

# write to file
# en_raw %>%
#   purrr::reduce(rbind) %>% 
#   fwrite(file = "./R2.enhancer_calling/All_raw_enhancers_20210307.csv")

# profile of coverage in plasmid and cDNA
en_raw %>% 
  purrr::reduce(rbind) %>%
  pivot_longer(cols = sampleCov:controlCov, names_to = "library", values_to = "Coverage") %>%
  ggplot(aes(x = rep, y = Coverage, color = library)) + 
    geom_violin(fill = "transparent") + 
    geom_boxplot(fill = "transparent", outlier.alpha = 0, width = 0.1, 
                 position = position_dodge(width = 0.89)) +
    toolkit_tyj$theme_my + 
    theme(legend.position = "none") + 
    facet_grid(method ~ .) + 
    scale_color_npg() + 
    scale_y_log10() + 
    ggtitle("Coverage of all raw enhancers")

en_raw %>% 
  purrr::reduce(rbind) %>%
  group_by(ID) %>%
  summarise(sampleCovP0.95 = quantile(sampleCov, 0.95), 
         sampleCovP0.99 = quantile(sampleCov, 0.99), 
         controlCovP0.95 = quantile(controlCov, 0.95), 
         controlCovP0.99 = quantile(controlCov, 0.99))
  

# filtration based on enrichment and coverage.
en_1.2 <- en_raw %>% 
  purrr::reduce(rbind) %>%
  dplyr::filter(enrichment > 1.3, sampleCov < 2700, controlCov < 2700)

# enhancers with enrichment > 1.2
df_lab <- en_1.2 %>%
  group_by(method, rep, Strategy) %>%
  summarise(Num = n()) %>%
  mutate(label = paste("(N=", Num, ")", sep = ""))
en_1.2 %>%
  ggplot(aes(x = rep, y = enrichment, color = rep)) +
    geom_violin(fill = "transparent") +
    geom_boxplot(fill = "transparent", outlier.alpha = 0, width = 0.05) + 
    facet_grid(method ~ Strategy) + 
    scale_color_npg() + 
    scale_y_log10() + 
    theme(legend.position = "none", axis.title.x = element_blank()) + 
    toolkit_tyj$theme_my + 
    geom_label(data = df_lab, aes(x = rep, y = 30, label = label), size = 2, fill = "transparent") +
    labs(title = "Number and activity of enhancers identified by two methods", subtitle = "(enrichment > 1.3)")

# toolkit_tyj$SavePlot("./R2.enhancer_calling/Number_and_activity_of_enhancer_1.3",
#                      width = 12, height = 8)

df_lab %>%
  ggplot(aes(x = rep, y = Num)) +
    geom_col(fill = "grey60") + 
    facet_grid(Strategy ~ method) + 
    toolkit_tyj$theme_my + 
    geom_text(aes(label = Num, y = Num / 2), color = "grey10") + 
    labs(caption = "enrichment > 1.3")
toolkit_tyj$SavePlot("./R2.enhancer_calling/Number_enhancer_1.3",
                     width = 12, height = 8)
```

### 2.1.2 Overlap.

```{r 20210307}
source("D:/SUST/code/TanYongjun_code.R")
en_1.2 <- fread("./R2.enhancer_calling/All_raw_enhancers_20210307.csv") %>%
  dplyr::filter(enrichment > 1.3, sampleCov < 2700, controlCov < 2700) %>%
  split(.$ID) %>%
  map(., GenomicRanges::makeGRangesFromDataFrame, keep.extra.columns = T)

names(en_1.2)

# overlap between dataset
df <- combn(names(en_1.2), m = 2) %>% t() %>% as.data.frame()
df$Num_1 <- df$Num_2 <- df$Num_ov <- NULL
for (i in 1:nrow(df)) {
  cat(i, " ")
  tmp1 <- en_1.2[[df$V1[i]]]
  tmp2 <- en_1.2[[df$V2[i]]]
  ol <- findOverlaps(tmp1, tmp2, minoverlap = 300)
  df$Num_1[i] <- length(tmp1)
  df$Num_2[i] <- length(tmp2)
  df$Num_ov[i] <- length(unique(ol@from)) # number of overlapped regions
}
df$Percent_1 <- df$Num_ov / df$Num_1
df$Percent_2 <- df$Num_ov / df$Num_2

df <- df %>%
  separate(V1, into = c("Type_1", "rep_1", "Strategy_1", sep = "_")) %>%
  separate(V2, into = c("Type_2", "rep_2", "Strategy_2", sep = "_"))

# fwrite(df, file = "./R2.enhancer_calling/Overlaps_enhancers_1.2.csv")
```

```{r 20210415}
source("D:/SUST/code/TanYongjun_code.R")
en_1.2 <- fread("./R2.enhancer_calling/All_raw_enhancers_20210307.csv") %>%
  dplyr::filter(enrichment > 1.3, sampleCov < 2700, controlCov < 2700) %>%
  split(.$ID) %>%
  map(., GenomicRanges::makeGRangesFromDataFrame, keep.extra.columns = T)

names(en_1.2)

# overlap between dataset
df <- combn(names(en_1.2), m = 2) %>% t() %>% as.data.frame()
df$Num_1 <- df$Num_2 <- df$Num_ov <- NULL
for (i in 1:nrow(df)) {
  cat(i, " ")
  tmp1 <- en_1.2[[df$V1[i]]]
  tmp2 <- en_1.2[[df$V2[i]]]
  ol <- findOverlaps(tmp1, tmp2, minoverlap = 300)
  df$Num_1[i] <- length(tmp1)
  df$Num_2[i] <- length(tmp2)
  df$Num_ov[i] <- length(unique(ol@from)) # number of overlapped regions
}
df$Percent_1 <- df$Num_ov / df$Num_1
df$Percent_2 <- df$Num_ov / df$Num_2

df <- df %>%
  separate(V1, into = c("Type_1", "rep_1", "Strategy_1", sep = "_")) %>%
  separate(V2, into = c("Type_2", "rep_2", "Strategy_2", sep = "_"))

# fwrite(df, file = "./R2.enhancer_calling/Overlaps_enhancers_1.3.csv")
```

> *Enhancers identified with merged replicates through strategy1 were used in following analysis*

```{r iEnhancer and cEnhancers}
source("D:/SUST/code/TanYongjun_code.R")
en_1.2 <- fread("./R2.enhancer_calling/All_raw_enhancers_20210307.csv") %>%
  dplyr::filter(enrichment > 1.2, sampleCov < 2700, controlCov < 2700) %>%
  split(.$ID) %>%
  map(., GenomicRanges::makeGRangesFromDataFrame, keep.extra.columns = T)

## strategy 1 ----
ol <- findOverlaps(en_1.2$cEnhancer_merge_strategy1, en_1.2$iEnhancer_merge_strategy1, minoverlap = 300)
enC <- as.data.frame(en_1.2$cEnhancer_merge_strategy1)
enI <- as.data.frame(en_1.2$iEnhancer_merge_strategy1)
enC$Type <- "cSTARRseq-specific"
enI$Type <- "iSTARRseq-specific"
enC$Type[unique(ol@from)] <- "cSTARRseq-common"
enI$Type[unique(ol@to)] <- "iSTARRseq-common"

en_merge <- rbind(enC, enI)
# fwrite(en_merge, file = "./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv")

## strategy 2, (startOnly fragments) ----
ol <- findOverlaps(en_1.2$cEnhancer_merge_strategy2, en_1.2$iEnhancer_merge_strategy2, minoverlap = 300)
enC <- as.data.frame(en_1.2$cEnhancer_merge_strategy2)
enI <- as.data.frame(en_1.2$iEnhancer_merge_strategy2)
enC$Type <- "cSTARRseq-specific"
enI$Type <- "iSTARRseq-specific"
enC$Type[unique(ol@from)] <- "cSTARRseq-common"
enI$Type[unique(ol@to)] <- "iSTARRseq-common"

en_merge <- rbind(enC, enI)
# fwrite(en_merge, file = "./R2.enhancer_calling/Enhancers_merged_strategy2_1.2_20210307.csv")

```

### 2.1.3 Correlation between En identified in Rep1 and rep2.
Only 1716 enhancers repeatedly identified in two replicates were plot.

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(ggrastr)
en_rep <- fread("./R2.enhancer_calling/All_raw_enhancers_20210307.csv") %>%
  dplyr::filter(enrichment > 1.3, sampleCov < 2700, controlCov < 2700,
                method == "cSTARR-seq", Strategy == "strategy1", 
                rep != "merge") %>%
  split(.$ID) %>%
  map(., GenomicRanges::makeGRangesFromDataFrame, keep.extra.columns = T)

ol <- findOverlaps(en_rep$cEnhancer_rep1_strategy1, 
                   en_rep$cEnhancer_rep2_strategy1, 
                   minoverlap = 300)

df <- data.frame(replicate1 = as.data.frame(en_rep$cEnhancer_rep1_strategy1)[ol@from,9], 
                 replicate2 = as.data.frame(en_rep$cEnhancer_rep2_strategy1)[ol@to,9])
p <- df %>%
  ggplot(aes(x = replicate1, y = replicate2)) + 
    # geom_point(alpha = 1/3) + 
      geom_text(x = 2, y = 5, label = expression(italic(r)==0.94),
                size = 6/.pt, check_overlap = T) +
    geom_point_rast(alpha = 2/3, raster.dpi = 300, pch = 1) +
    geom_smooth(method = 'lm', formula = y ~ x) +
    # scale_y_log10() +
    # scale_x_log10() +
    xlab("Enhancer activity (rep 1)") + 
    ylab("Enhancer activity (rep 2)") +
    # ggtitle("Correlation between enhancer enrichment \nidentified in two replictes") + 
    toolkit_tyj$theme_my


toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/Correlation_between_replicates_1716enhancers",
                     width = 6, height = 6, device = c("pdf", "png"), plot = p)

```

## 2.2 Number and Activity of identified enhancers.

### .1 iSTARRseq

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>% 
    dplyr::filter(method == "iSTARR-seq", seqnames != "chrC", seqnames != "chrM")
# map(enhancer, function(x){fivenum(as.numeric(x))})

# Activity of enhancer on each chromosome
enhancer %>%
  ggplot(aes(x = seqnames, y = enrichment, color = seqnames)) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0, fill = "transparent") + 
    geom_violin(fill = "transparent") +
    coord_cartesian(ylim = c(1.2, 10)) +
    scale_y_log10() + 
    toolkit_tyj$theme_my + 
    # ggtitle("Activity of enhancers identified on each chromosome") + 
    xlab("Chromosome") + 
    ylab("Enrichment (iSTARR-seq)") + 
    scale_color_npg() + 
    theme(legend.position = "none")

toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/iSTARRseq_1.2_Activity_of_enhancers_located_on_each_chromosome",
                     width = 12, height = 8)

# number and activity median
df <- enhancer %>%
  group_by(seqnames) %>%
  summarise(MedianEnrich = median(enrichment), NumberOfEnhancer = n()) %>%
  mutate(seqnames = as.character(str_replace_all(seqnames, "chr", "")))

# Density of enhancer on each chromosome.
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  chrLen <- seqinfo(txdb) %>% 
    as.data.frame() %>%
    mutate(seqnames = row.names(.))
  full_join(chrLen, df) %>%
    mutate(Density = NumberOfEnhancer * 1e6 / seqlengths) %>%
    ggplot(aes(x = seqnames, y = Density)) + 
      geom_col(fill = "grey60") + 
      toolkit_tyj$theme_my + 
      ylab("Number of enhancers in each MB") + 
      xlab(label = "Chromosome")
 toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/iSTARRseq_1.2_Density_of_enhancers_located_on_each_chromosome",
                     width = 12, height = 8) 
 
# histone of enhancer activity
fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>%
  mutate(bins = cut(enrichment, breaks = c(seq(1, 3, 0.5), Inf))) %>%
  group_by(bins, method) %>%
  summarise(Num = n()) %>%
  ggplot(aes(x = bins, y = Num)) +
    geom_col(fill = "grey60") + 
    toolkit_tyj$theme_my +
    facet_grid(method ~ ., scales = "free_y")
toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/Distribution_of_enhancer_activity_1.2",
                   width = 12, height = 8)
```

### .2 cSTARRseq

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>% 
    dplyr::filter(method == "cSTARR-seq", 
                  # seqnames != "chrC", seqnames != "chrM", 
                  enrichment > 1.3)
# map(enhancer, function(x){fivenum(as.numeric(x))})

# Activity of enhancer on each chromosome
enhancer %>%
  ggplot(aes(x = seqnames, y = enrichment, color = seqnames)) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0, fill = "transparent") + 
    geom_violin(fill = "transparent") +
    coord_cartesian(ylim = c(1.2, 7)) +
    scale_y_log10() + 
    toolkit_tyj$theme_my + 
    # ggtitle("Activity of enhancers identified on each chromosome") + 
    xlab("Chromosome") + 
    ylab("Enrichment (STARR-seq)") + 
    scale_color_npg() + 
    theme(legend.position = "none")
toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/cSTARRseq_1.3_Activity_of_enhancers_located_on_each_chromosome",
                     width = 12, height = 8)

# number and activity median
df <- enhancer %>%
  group_by(seqnames) %>%
  summarise(MedianEnrich = median(enrichment), NumberOfEnhancer = n()) %>%
  mutate(seqnames = as.character(str_replace_all(seqnames, "chr", "")))

# Density of enhancer on each chromosome.
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  chrLen <- seqinfo(txdb) %>% 
    as.data.frame() %>%
    mutate(seqnames = row.names(.))
  full_join(chrLen, df) %>%
    mutate(Density = NumberOfEnhancer * 1e6 / seqlengths) %>%
    dplyr::filter(str_detect(seqnames, "(Mt|Pt|M)", negate = T)) %>%
    ggplot(aes(x = seqnames, y = Density)) + 
      geom_col(fill = "grey60") + 
      toolkit_tyj$theme_my + 
      ylab("Number of enhancers in each MB") + 
      xlab(label = "Chromosome") + 
      geom_text(aes(x = seqnames, y = Density / 2, label = round(Density, digits = 2)))
 toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/cSTARRseq_1.3_Density_of_enhancers_located_on_each_chromosome",
                     width = 12, height = 8) 
 
# histogram of enhancer activity
 enhancer %>%
   dplyr::mutate(Bins = cut(enrichment, breaks = c(1.3, 1.4, 1.5, 1.6, 1.7, 2, 3, 4, Inf))) %>%
   dplyr::group_by(Bins) %>%
   summarise(Num = n()) %>%
   ggplot(aes(x = enrichment)) + 
    geom_col(aes(x = Bins, y = Num), fill = "grey60") + 
    toolkit_tyj$theme_my + 
    xlab("Enhancer activity") + 
    ylab("Number")
  # toolkit_tyj$SavePlot("./R2.enhancer_calling/Activity_cSTARRseq_1.3", width = 12, height = 8) 
```

### .3 Merged 20210713
```{r}
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("./R2.enhancer_calling/At_enhancer_raw_20210713.tsv")
  enhancer %>%
    ggplot(aes(x = enrichment)) + 
      geom_histogram(bins = 50, fill = "grey55") +
      toolkit_tyj$theme_my
  quantile(enhancer$enrichment, probs = seq(0, 1, 0.05))
  table(cut(enhancer$enrichment, breaks = c(0,1, 1.1,1.2,1.3, 2,100)))
  
  NA
```

### .3 Merged 20210909
```{r}
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("./R2.enhancer_calling/calling_merged_20210909/At_cSTARRseq_enhancer_raw_20210909.tsv")
  # enhancer %>%
  #   ggplot(aes(x = enrichment)) + 
  #     geom_histogram(bins = 50, fill = "grey55") +
  #     toolkit_tyj$theme_my
  # quantile(enhancer$enrichment, probs = seq(0, 1, 0.05))
  table(cut(enhancer$enrichment, breaks = c(0,1, 1.1,1.2,1.3,100)))

```


## 2.3 Enhancer and PAS.

```{r}

### each type of enhancers (cSTARR-seq-specific, iSTARR-seq, common.) -----
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>%
    dplyr::filter(Type != "cSTARRseq-common") %>%
  dplyr::mutate(Type = str_replace_all(Type, ".+common", "Common"))
pas_HQ <- fread("./refgenome/arabidopsis_thaliana.high_confidence.PAC.bed", header = F) %>%
  dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
  dplyr::mutate(V1 = str_c("chr", V1, sep = "")) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3", strand.field = "V5")
enhancer_ls <- enhancer %>%
  split(.$Type) %>%
  map(., makeGRangesFromDataFrame, keep.extra.column = T)
ol_ls <- map_int(enhancer_ls, function(x){
  ol <- findOverlaps(x, pas_HQ)
  return(length(unique(ol@from)))
})

num_type <- table(enhancer$Type) %>% as.data.frame()
df <- cbind(num_type, ol_ls) %>%
  mutate(Rate = ol_ls / Freq, 
         Type = Var1, 
         label = str_c(percent(Rate), "\n(", ol_ls, "/", Freq, ")", sep = ""))
# plot
df %>%
  ggplot(aes(x = Type, y = Rate, fill = Type)) + 
    geom_col() + 
    toolkit_tyj$theme_my + 
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          axis.title.x = element_blank()) + 
    scale_fill_npg() + 
    geom_text(aes(y = Rate / 2, x = Type, label = label), size = 2) + 
    labs(caption = "Percent of enhancers overlapped with PAS") + 
    ylab("Percent")
# toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/Percent_of_enhancers_FourType_strategy1_overlapped_PAS_20210309", 
#                      width = 6, height = 8)

### iSTARRseq and cSTARR-seq
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv")
pas_HQ <- fread("./refgenome/arabidopsis_thaliana.high_confidence.PAC.bed", header = F) %>%
  dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
  dplyr::mutate(V1 = str_c("chr", V1, sep = "")) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3", strand.field = "V5")
enhancer_ls <- enhancer %>%
  split(.$method) %>%
  map(., function(x){
    x %>% 
      arrange(pVal, desc(enrichment)) %>%
      head(4000) %>%
      return()
  }) %>%
  map(., makeGRangesFromDataFrame, keep.extra.column = T)
ol_ls <- map_int(enhancer_ls, function(x){
  ol <- findOverlaps(x, pas_HQ)
  return(length(unique(ol@from)))
})

num_type <- enhancer_ls %>% 
  map(., as.data.frame) %>%
  purrr::reduce(rbind) %>%
  group_by(method) %>%
  summarise(Freq = n())

df <- cbind(num_type, ol_ls) %>%
  mutate(Rate = ol_ls / Freq, 
         Type = method, 
         label = str_c(percent(Rate), "\n(", ol_ls, "/", Freq, ")", sep = ""))
# plot
df %>%
  ggplot(aes(x = Type, y = Rate, fill = Type)) + 
    geom_col() + 
    toolkit_tyj$theme_my + 
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          axis.title.x = element_blank()) + 
    scale_fill_npg() + 
    geom_text(aes(y = Rate / 2, x = Type, label = label), size = 2) + 
    labs(caption = "Percent of enhancers overlapped with PAS") + 
    ylab("Percent")
# toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/Percent_of_enhancers_top4000_TwoType_strategy1_overlapped_PAS_20210309",
#                      width = 6, height = 8)

NA
```

## 2.4 Plot (pyGenometrack)

1.iSTARR-seq can identify enhancers located near PAS.
2.iSTARR-seq can identify enhancers more acurately.
3.enhancers identified by iSTARR-seq and cSTARR-seq.

### .0 enrichment file 
calculate enrichment using functions modified from BasicSTARRseq.

```{r}

```



### .1 bed to bw

format data in bed file.(scientific numeric was not accept for `bedItemOverlapCount`)

```{r, eval=F}
source("/work/bio-tanyj/soft/TanYongjun_code.R")
setwd("/scratch/2021-07-02/bioTanyj/At_enhancer/R2.enhancer_calling/cSTARR-seq")
options(scipen = 200)
for(i in list.files(pattern = ".+dd40_filtered_uniq_strategy1.bed$")){
    cat(i, "\n")
    timestamp()
    fread(i) %>%
        arrange(V1, V2) %>%
        # mutate(V2 = format(V2, scientific = F), 
        #        V3 = format(V3, scientific = F)) %>%
        fwrite(str_replace(i, ".bed", "_tidy.bed"), col.names = F, sep = "\t")
}

# df %>%
#   head(1000) %>%
#   # mutate(V2 = format(V2, scientific = F), 
#                # V3 = format(V3, scientific = F)) %>% str()
#         fwrite(str_replace(i, ".bed", "_tidy.bed"), col.names = F, sep = "\t")
```

```{bash, eval = F}
#!/bin/bash
#BSUB -J bed2bw                  ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 15                  ### ask for number of cores (default: 1)
#BSUB -W 15:00               ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

#--------------------------------------------------------------------------------------#

# convert bed file into bigwig for visualization in IGV or other software.
cd /scratch/2021-03-02/bioTanyj/At_enhancer/R2.enhancer_calling/bw_file
referenceChrom="/scratch/2021-03-02/bioTanyj/At_enhancer/refgenome/TAIR10_chr_size_enlarge.tsv" # some region may contain out of bouder regions.

for i in `ls | grep -E "_notExtend_tidy.bed$"`
do
    {
    echo -e "Generating ${i%_tidy.bed}.bedGraph........"
    sort -k1,1 ${i} | /work/bio-tanyj/soft/bedItemOverlapCount.bib TAIR10 -chromSize=$referenceChrom stdin > ${i%_tidy.bed}.bedGraph
    echo -e "       completed!"
    /work/bio-tanyj/soft/bedGraphToBigWig.bib ${i%_tidy.bed}.bedGraph $referenceChrom ${i%_tidy.bed}.bw
    rm ${i%_tidy.bed}.bedGraph
    echo -e "${i%_tidy.bed}.bw Generated \n\n"
    }&
done
wait

```

### .2 Bed files of enhancer tracks

Generate bed files for each enhancer set.
(for `pyGenomeTracks`)

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_IntronicEn_TE_20220424.csv.gz")

for (i in unique(enhancer$method)) {
  enhancer %>%
    dplyr::filter(method == i) %>%
    dplyr::select(1:3) %>%
    fwrite(file = paste("./R2.enhancer_calling/plot_pyGenomeTracks/", i, ".bed", sep = ""), 
           sep = "\t", col.names = F)
}

```

### .3 plot with pyGenomeTracks
#### .1 iSTARR, cSTARR, replicates (202103)

*Generate origin configure files used by pyGenomeTracks* parameters related to figures were contained in this configuration file.

```{r, eval = F}
# add 2bp to each PAS
fread("./refgenome/arabidopsis_thaliana.high_confidence.PAC.bed", header = F) %>%
  mutate(V2 = V2 - 1, V3 = V3 + 1) %>%
  fwrite("./refgenome/arabidopsis_thaliana.high_confidence.PAC.add2bp.bed", sep = "\t", col.names = F)

```

```{bash, eval = F}
cd /scratch/2021-03-02/bioTanyj/At_enhancer/R2.enhancer_calling/bw_file

conda activate macs2 # pyGenomeTracks install in this env

# gff file to gtf
gffread TAIR10_GFF3_genes_transposons_rename_chr.gff -T -o TAIR10_GFF3_genes_transposons_rename_chr.gtf

# make track file
make_tracks_file --trackFiles \
  cSTARR-seq.bed \
  iSTARR-seq.bed  \
  arabidopsis_thaliana.high_confidence.PAC.add2bp.bed \
  cSTARRseq_P.bw \
  cSTARRseq_P1.bw \
  cSTARRseq_P2.bw \
  cSTARRseq_C.bw \
  cSTARRseq_C1.bw \
  cSTARRseq_C2.bw \
  iSTARRseq_P.bw \
  iSTARRseq_P1.bw \
  iSTARRseq_P2.bw \
  iSTARRseq_C.bw \
  iSTARRseq_C1.bw \
  iSTARRseq_C2.bw \
  RAtEC_add40_filtered_uniq_strategy1_notExtend.bw \
  RAtEC1_add40_filtered_uniq_strategy1_notExtend.bw \
  RAtEC2_add40_filtered_uniq_strategy1_notExtend.bw \
  TAIR10_GFF3_genes_transposons_rename_chr.gtf \
  NC2018_high.mcool \
  -o enhancer_cSTARR_iSTARR_backup.ini

# make track file
make_tracks_file --trackFiles \
  cSTARR-seq.bed \
  iSTARR-seq.bed  \
  arabidopsis_thaliana.high_confidence.PAC.add2bp.bed \
  cSTARRseq_P.bw \
  cSTARRseq_C.bw \
  iSTARRseq_P.bw \
  RAtEC_add40_filtered_uniq_strategy1_notExtend.bw \
  TAIR10_GFF3_genes_transposons_rename_chr.gtf \
  NC2018_high.mcool \
  -o enhancer_cSTARR_iSTARR_backup_1.ini
```

*Select regions and plot*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv")
pas_HQ <- fread("./refgenome/arabidopsis_thaliana.high_confidence.PAC.bed", header = F) %>%
  dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
  dplyr::mutate(V1 = str_c("chr", V1, sep = "")) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3", strand.field = "V5")
ol <- findOverlaps(makeGRangesFromDataFrame(enhancer), 
                   pas_HQ)
enhancer$PAS <- "noPAS"
enhancer$PAS[unique(ol@from)] <- "PAS"
# fwrite(enhancer, file = "./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_PAS_20210309.csv")

# select
table(enhancer$Type, enhancer$PAS)
enlarge_size <- 8000
Num <- 200

for (i in unique(enhancer$Type)) {
  for (j in unique(enhancer$PAS)) {
    set.seed(123456)
    enhancer %>%
      dplyr::filter(Type == i, PAS == j) %>%
      arrange(desc(enrichment)) %>%
      dplyr::mutate(ID = paste(seqnames, ":", start - 6000, "-", end + 6000, sep = "")) %>%
      select(1:3) %>%
      mutate(start = start - enlarge_size, end = end + enlarge_size) %>%
      slice_sample(n = Num) %>%
      fwrite(paste("./R2.enhancer_calling/plot_pyGenomeTracks/",
             i, "_", j, ".bed", sep = ""), 
             col.names = F, sep = "\t")
  }
}

```

```{bash, Demo of plot, eval = F}
cd /scratch/2021-03-02/bioTanyj/At_enhancer/R2.enhancer_calling/plot_pyGenomeTracks

## .hic to .cool
hicConvertFormat --matrices GSE104333_HCT116_Rao-2017-untreated_combined_30.hic --outFileName GSE104333_HCT116_Rao-2017-untreated_combined_30.cool --inputFormat hic --outputFormat cool # resolution not determined, .mcool file will generated.
hicConvertFormat --matrices GSE63525_K562_combined_30.hic --outFileName GSE63525_K562_combined_30.cool --inputFormat hic --outputFormat cool # resolution not determined, .mcool file will generated.

# single region
pyGenomeTracks \
  --tracks ../bw_file/enhancer_cSTARR_iSTARR_backup_1.ini  \
  --region chr4:10272758-10281357 \
  --trackLabelFraction 0.12 \
  --width 16 --height 12 --fontSize 8  --dpi 600 \
  --title chr4:10272758-10281357  \
  -o test7.png

# a bed file
pyGenomeTracks \
  --tracks ../bw_file/enhancer_cSTARR_iSTARR.ini  \
  --BED iSTARRseq-specific_PAS.bed \
  --trackLabelFraction 0.12 \
  --width 16 --height 12 --fontSize 8  --dpi 600 \
  -o ./random_selected/iSTARRseq-specific_PAS.png
  
# a set of bed files
for i in `ls | grep -E ".+PAS.bed"`
do
# bed
  {
  pyGenomeTracks \
    --tracks ../bw_file/enhancer_cSTARR_iSTARR.ini  \
    --BED ${i} \
    --trackLabelFraction 0.12 \
    --width 16 --height 12 --fontSize 8  --dpi 600 \
    -o ./random_selected/${i%.bed}.png
  }&
done
wait

```

#### .2 cSTARR (1.3)
Two kind of plots:
   1. for illustrating STARR-seq; tracks: plasmid/cDNA of rep1, rep2, and merged.
   2. for illustrating enhancers belong to each Kmeans clusters. tracks: merged plasmid/cDNA, DHS, ChIP-seq, Methy.

*Generate origin configure files used by pyGenomeTracks* parameters related to figures were contained in this configuration file.



```{bash, eval = F}
cd /scratch/2021-03-02/bioTanyj/At_enhancer/R2.enhancer_calling/bw_file

conda activate macs2 # pyGenomeTracks install in this env

# gff file to gtf
gffread TAIR10_GFF3_genes_transposons_rename_chr.gff -T -o TAIR10_GFF3_genes_transposons_rename_chr.gtf

# make track file
make_tracks_file --trackFiles \
  cSTARR-seq.bed \
  iSTARR-seq.bed  \
  arabidopsis_thaliana.high_confidence.PAC.add2bp.bed \
  cSTARRseq_P.bw \
  cSTARRseq_P1.bw \
  cSTARRseq_P2.bw \
  cSTARRseq_C.bw \
  cSTARRseq_C1.bw \
  cSTARRseq_C2.bw \
  iSTARRseq_P.bw \
  iSTARRseq_P1.bw \
  iSTARRseq_P2.bw \
  iSTARRseq_C.bw \
  iSTARRseq_C1.bw \
  iSTARRseq_C2.bw \
  RAtEC_add40_filtered_uniq_strategy1_notExtend.bw \
  RAtEC1_add40_filtered_uniq_strategy1_notExtend.bw \
  RAtEC2_add40_filtered_uniq_strategy1_notExtend.bw \
  TAIR10_GFF3_genes_transposons_rename_chr.gtf \
  NC2018_high.mcool \
  -o enhancer_cSTARR_iSTARR_backup.ini

# make track file
make_tracks_file --trackFiles \
  cSTARR-seq.bed \
  iSTARR-seq.bed  \
  arabidopsis_thaliana.high_confidence.PAC.add2bp.bed \
  cSTARRseq_P.bw \
  cSTARRseq_C.bw \
  iSTARRseq_P.bw \
  RAtEC_add40_filtered_uniq_strategy1_notExtend.bw \
  TAIR10_GFF3_genes_transposons_rename_chr.gtf \
  NC2018_high.mcool \
  -o enhancer_cSTARR_iSTARR_backup_1.ini
```

*Select regions and plot*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv")
pas_HQ <- fread("./refgenome/arabidopsis_thaliana.high_confidence.PAC.bed", header = F) %>%
  dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
  dplyr::mutate(V1 = str_c("chr", V1, sep = "")) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3", strand.field = "V5")
ol <- findOverlaps(makeGRangesFromDataFrame(enhancer), 
                   pas_HQ)
enhancer$PAS <- "noPAS"
enhancer$PAS[unique(ol@from)] <- "PAS"
# fwrite(enhancer, file = "./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_PAS_20210309.csv")

# select
table(enhancer$Type, enhancer$PAS)
enlarge_size <- 8000
Num <- 200

for (i in unique(enhancer$Type)) {
  for (j in unique(enhancer$PAS)) {
    set.seed(123456)
    enhancer %>%
      dplyr::filter(Type == i, PAS == j) %>%
      arrange(desc(enrichment)) %>%
      dplyr::mutate(ID = paste(seqnames, ":", start - 6000, "-", end + 6000, sep = "")) %>%
      select(1:3) %>%
      mutate(start = start - enlarge_size, end = end + enlarge_size) %>%
      slice_sample(n = Num) %>%
      fwrite(paste("./R2.enhancer_calling/plot_pyGenomeTracks/",
             i, "_", j, ".bed", sep = ""), 
             col.names = F, sep = "\t")
  }
}

```

```{bash, Demo of plot, eval = F}
cd /scratch/2021-03-02/bioTanyj/At_enhancer/R2.enhancer_calling/plot_pyGenomeTracks

## .hic to .cool
hicConvertFormat --matrices GSE104333_HCT116_Rao-2017-untreated_combined_30.hic --outFileName GSE104333_HCT116_Rao-2017-untreated_combined_30.cool --inputFormat hic --outputFormat cool # resolution not determined, .mcool file will generated.
hicConvertFormat --matrices GSE63525_K562_combined_30.hic --outFileName GSE63525_K562_combined_30.cool --inputFormat hic --outputFormat cool # resolution not determined, .mcool file will generated.

# single region
pyGenomeTracks \
  --tracks ../bw_file/enhancer_cSTARR_iSTARR_backup_1.ini  \
  --region chr4:10272758-10281357 \
  --trackLabelFraction 0.12 \
  --width 16 --height 12 --fontSize 8  --dpi 600 \
  --title chr4:10272758-10281357  \
  -o test7.png

# a bed file
pyGenomeTracks \
  --tracks ../bw_file/enhancer_cSTARR_iSTARR.ini  \
  --BED iSTARRseq-specific_PAS.bed \
  --trackLabelFraction 0.12 \
  --width 16 --height 12 --fontSize 8  --dpi 600 \
  -o ./random_selected/iSTARRseq-specific_PAS.png
  
# a set of bed files
for i in `ls | grep -E ".+PAS.bed"`
do
# bed
  {
  pyGenomeTracks \
    --tracks ../bw_file/enhancer_cSTARR_iSTARR.ini  \
    --BED ${i} \
    --trackLabelFraction 0.12 \
    --width 16 --height 12 --fontSize 8  --dpi 600 \
    -o ./random_selected/${i%.bed}.png
  }&
done
wait

```


## 2.6 ActivityCluster

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_RPKM_20210321.csv.gz")

# classify into several part
enhancer$ClusterActivity <- NA
df <- NULL
for (i in unique(enhancer$method)) {
  tmp <- enhancer %>% dplyr::filter(method == i)
  if (str_detect(i, "STARR")) {
    tmp1 <- tmp %>% 
      dplyr::filter(enrichment > 1.3)
    breaks_ls <- c(0, quantile(tmp1$enrichment, probs = seq(0, 1, 0.2))[2:5], 1000)
    tmp1$ClusterActivity <- cut(tmp1$enrichment, breaks = breaks_ls) %>%
      as.numeric() %>% str_c("Part", ., sep = "")
    tmp <- rbind(tmp1, dplyr::filter(tmp, enrichment <= 1.3))
  }
  df <- rbind(df, tmp)
}

# fwrite(df, "Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz")
```

## 2.7 validation (Sun JL)

### 2.7.1 collect regions

Two methods and two batches validation experimens were performed.
cSTARR-seq (old-STARRseq): QPCR iSTARR-seq : QPCR + GUS

*Blast*

```{bash}
###########################################
# Local blast and subset determined region
###########################################
# genomic DNA

#0,# install blast & add to PATH, add path of blastdb to $BLASTDB
#1,#make datebase
makeblastdb -in TAIR10.fa  -input_type fasta -dbtype nucl -title TAIR10.fa -parse_seqids

#2,#blastn
blastn -db TAIR10.fa -query all_primers_seq.fa -out primer.html -html -evalue 5 -word_size 7

blastn -db TAIR10.fa -query all_primers_seq.fa -out primer.tsv -outfmt '7 qseqid sseqid sallacc pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen' -evalue 5 -word_size 7

```

*Collect and filtration*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
blast_all <- read.table("./validation/blast/primer.tsv", comment = "#")
colnames(blast_all) <- c("query", "subject", "subjectacc", "identity", "alignment length", "mismatches", 
                         "gap opens", "q. start", "q. end", "s. start", "s. end", "evalue", "bit score", "query length") %>%
  str_to_title(.) %>%
  str_replace_all(., " ", "")

# filtration, just keep hit with the smallest Evalue
length(unique(blast_all$Query))
blast_all_flt <- blast_all %>%
  group_by(Query) %>%
  arrange(Query, Evalue) %>%
  mutate(MaxBitScore = max(BitScore), 
         Keep = ifelse(BitScore == MaxBitScore, "Yes", "No") ) %>%
  dplyr::filter(Keep == "Yes") %>%
  dplyr::mutate(CRE_id = str_replace_all(Query, "_[FR]", ""), 
                Primer = ifelse(str_detect(Query, "_F_"), "Forward", "Reverse"), 
                tmp = str_c(Subject, S.Start, S.End, sep = "_"), 
                Num = n())

# fwrite(blast_all_flt, file = "./validation/Blast_results_all_20210427.csv")

# multi position will found for some primers, keep position of paired-primers based on product size mannually.
blast_all_mk <- fread("./validation/Blast_results_marked_20210427.csv") %>%
  dplyr::filter(Keep == "Yes")

table(blast_all_mk$CRE_id) %>% sort()
length(unique(blast_all_mk$CRE_id))
length(unique(blast_all_flt$CRE_id))
 # two primers (two regions/CRE) were discarded because two regions with similar length will be amplified.

table(blast_all_mk$Query) %>% sort() %>% rev()
blast_all_mk <- blast_all_mk %>%
  pivot_wider(id_cols = CRE_id, names_from = Primer, values_from = tmp) %>%
  cbind(., str_split_fixed(.$Forward, "_", n = 3)) %>%
  cbind(., str_split_fixed(.$Reverse, "_", n = 3))

names(blast_all_mk)[c(4:9)] <- c("chr_F", "start_F", "end_F", "chr_R", "start_R", "end_R")
blast_all_mk$Forward <- blast_all_mk$Reverse <- NULL
blast_all_mk$chr <- apply(blast_all_mk[,str_detect(colnames(blast_all_mk), "chr")], 
                               1, 
                               function(x){str_c(unique(x), collapse = ",")})
blast_all_mk$start <- apply(blast_all_mk[,str_detect(colnames(blast_all_mk), "start")], 
                               1, 
                               function(x){min(x)}) %>%
  as.numeric()
blast_all_mk$end <- apply(blast_all_mk[,str_detect(colnames(blast_all_mk), "end")], 
                               1, 
                               function(x){max(x)}) %>%
  as.numeric()
blast_all_mk$width <- blast_all_mk$end - blast_all_mk$start

head(blast_all_mk)
unique(blast_all_mk$CRE_id)

blast_all_mk <- blast_all_mk %>%
  dplyr::mutate(Method = str_replace_all(CRE_id, "(.+)_(.+)_(.+)", "\\1"), 
                Type = str_replace_all(CRE_id, "(.+)_(.+)_(.+)", "\\2"))


# fwrite(blast_all_mk, "./validation/Formated_all_validated_CREs_20210427.csv")
```

### 2.7.2 QPCR

*Overlap between enhancer and validated regions*

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")

# load
  regions <- fread("./validation/Formated_all_validated_CREs_20210427.csv") %>%
    arrange(chr, start, end) %>%
    dplyr::select(!matches("(_F|_R)"))
  regionsGR <- makeGRangesFromDataFrame(regions, seqnames.field = "chr", keep.extra.columns = T)
  enhancers <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq", enrichment > 1.3) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, "")) %>%
    arrange(seqnames, start, end)
  enhancerGR <- makeGRangesFromDataFrame(enhancers)

# overlap between validated regions with STARRseq enhancer (new).
  MinOverlapSize <- 500
  ol <- findOverlaps(regionsGR, enhancerGR, minoverlap = MinOverlapSize)
  df <- cbind(regions[ol@from,], enhancers[ol@to,]) # two enhancers were both validated with QPCR and GUS.
  cat("Number of overlap regions:", length(unique(ol@from)))
  
  regions$HitNewEn <- "No"
  regions$HitNewEn[unique(ol@from)] <- "Yes"
  regions$STARR_enrichment <- NA
  regions$STARR_enrichment[ol@from] <- enhancers$enrichment[ol@to]
  
  table(regions$Type, regions$HitNewEn)
  regions <- regions %>%
    arrange(chr, start)
  
# summary
  regions %>%
    group_by(Method, HitNewEn, Type) %>%
    count() %>%
    pivot_wider(names_from = HitNewEn, values_from = n, values_fill = 0) %>%
    fwrite(file = "./validation/Summary_QPCR_GUS_20210427.csv")
    
# add QPCR results
  exp_qpcr <- fread("./validation/all_QPCR_results.tsv")
  regions <- full_join(regions, exp_qpcr)

# iSTARR enhancers overlapped by cSTARR-seq enhancer were seem as cSTARR-seq enhancer
  regions$Method[regions$HitNewEn == "Yes"] <- "Enhancer"

# correlation between QPCR and STARRseq
  regions %>%
    dplyr::filter(!is.na(QPCR), HitNewEn == "Yes") %>%
    ggplot(aes(x = QPCR, y = STARR_enrichment)) + 
      geom_smooth(method = "lm", se = F, alpha = 1/3) +
      geom_point() + 
      toolkit_tyj$theme_my
  toolkit_tyj$SavePlot(filename_prefix = "./validation/Plot_correlation_QPCR_cSTARRseq_20210427", 
                       width = 8, height = 6, units = "cm", device = "png")

# compare activity of enhancers identified by different methods
  tmp <- regions %>%
    dplyr::filter(str_detect(Method, "(Enhancer|DHS|H3K4me1)"), 
                  !is.na(QPCR)) %>%
    dplyr::mutate(Method = factor(Method, levels = c("DHS", "H3K4me1", "Enhancer")))
  
  tmp_label <- tmp %>%
    group_by(Method) %>%
    summarise(Median = round(median(QPCR, na.rm = T), digits = 2), 
              Num = n()) %>%
    dplyr::mutate(label = str_c(Method, "\nMedian=", Median, "\nN=", Num, sep = ""))
  
  tmp %>%
    ggplot(aes(x = Method, y = QPCR)) + 
      geom_violin(fill  = "transparent") + 
      geom_boxplot(fill = "transparent", width = 0.04, outlier.alpha = 0) +
      geom_beeswarm(alpha = 3/4, shape = 16, aes(color = ifelse(QPCR > 1, "A", "B"))) + 
      toolkit_tyj$theme_my + 
      theme(legend.position = "none", axis.title.x = element_blank()) + 
      scale_color_npg() + 
      geom_signif(test = "wilcox.test", comparisons = list(c("DHS", "Enhancer"), 
                                                           c("H3K4me1", "Enhancer")), 
                  y_position = c(4.4, 4.2), textsize = 2, size = 0.2) + 
      # geom_label(data = tmp_label, aes(x = Method, y = 4.5, label = label),
      #            label.size = 0.2, color = "black", fill = "transparent", 
      #            size = 1.5) + 
      scale_x_discrete(limits = tmp_label$Method, labels = tmp_label$label)
  toolkit_tyj$SavePlot(filename_prefix = "./validation/Compare_QPCR_DHS_H3K4me1_20210427", 
                       width = 8, height = 6, device = "png")
```

### 2.7.3 GUS
not analyzed.
```{r}

```


## 2.7 validation (202106)
Ideally, investigate enhancer activity of 20 enhancers and 5 negative controls.
Types: 
 1. Genomic features (intergenic, intronic, exonic).
 2. Activity level.
 3. DHS or not
 4. clustered or not.

Methods:
 1. Direction or position (upstream/dowstream)
 
### 2.7.1 select for validation with luciferase.
```{r}
## load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::mutate(annotationSlim = str_replace_all(annotation, "(.+) \\(.+\\)", "\\1")) %>%
    dplyr::filter(seqnames != "Mt")
    unique(enhancer$annotationSlim)
## DHS or not
  dhs <- fread("./refgenome/GSE34318_dhsites_region.bed/GSE34318_dhsites_region.bed", header = F, 
                       sep = "\t", sep2 = "\\S", fill = T) %>%
    dplyr::mutate(label = str_c(V4, V5, V6, V7, sep = "_"),
                  V1 = str_replace_all(V1, "Chr", "")) %>%
    dplyr::filter(str_detect(label, "wtleaf")) %>%
    dplyr::select(1:3) %>%
    makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3")
  ol <- findOverlaps(dhs, makeGRangesFromDataFrame(enhancer), minoverlap = 300)
  enhancer$DHSPeak <- "noDHS"
  enhancer$DHSPeak[ol@to] <- "DHS"
  
## select (DHS + activity)
  table(enhancer$DHSPeak, enhancer$ClusterType)
  table(enhancer$ClusterActivity)
  
  NumEachClu <- 10
  SeedTmp <- 1234567
  RandomDF <- NULL
  for (i in unique(enhancer$ClusterActivity)) {
    set.seed(SeedTmp)
    RandomDF <- enhancer %>%
      dplyr::filter(ClusterActivity == i) %>%
      slice_sample(n = NumEachClu) %>%
      rbind(., RandomDF)
  }
  table(RandomDF$DHSPeak)
  table(RandomDF$ClusterType)
  table(RandomDF$annotationSlim)
  table(RandomDF$ClusterActivity)
  
  RandomDF <- RandomDF %>%
    dplyr::mutate(ClusterType = str_replace_all(RandomDF$ClusterType, "_enhancer", ""),
                  ClusterActivity = str_replace_all(ClusterActivity, "Part", "ActBin")) %>%
    dplyr::mutate(seqnames = str_c("Chr", seqnames, sep = "")) %>%
    dplyr::mutate(ID = str_c(ClusterActivity, ClusterType, DHSPeak, 
                             annotationSlim, enrichment, seqnames, start, end, sep = "_")) %>%
    dplyr::select(seqnames:end, ClusterActivity, ClusterType, DHSPeak, enrichment, annotationSlim, ID)

## control
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  set.seed(SeedTmp)
  ControlDF <- makeGRangesFromDataFrame(enhancer) %>%
    toolkit_tyj$shuffle(., txdb) %>%
    as.data.frame() %>%
    slice_sample(n = 20) %>%
    as.data.frame() %>%
    dplyr::mutate(seqnames = str_c("Chr", seqnames, sep = "")) %>%
    dplyr::mutate(ID = str_c("Control", seqnames, start, end, sep = "_")) %>%
    dplyr::select(seqnames:end, ID)
  
## Activity of selected enhancer
  enhancer %>%
    dplyr::mutate(ClusterActivity = str_replace_all(ClusterActivity, "Part", "ActBin")) %>%
    ggplot(aes(x = ClusterActivity, y = enrichment)) + 
      geom_boxplot(fill = "transparent", width = 0.1) + 
      geom_violin(fill = "transparent", scale = "width") + 
      toolkit_tyj$theme_my + 
      scale_y_log10() + 
      geom_beeswarm(data = RandomDF, aes(x = ClusterActivity, y = enrichment), color = "red", size = 0.5, shape = 1) + 
      ggtitle("Enrichment of selected enhancers")
  toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/validation_NiuLJ/Enrichment_of_selected_enhancers", 
                       width = 8, height = 6)
  
## write to file
  df <- full_join(RandomDF, ControlDF) %>%
    dplyr::mutate(annotationSlim = factor(annotationSlim, levels = c("Exon", "Promoter", "Intron", "Distal Intergenic"))) %>%
    arrange(desc(ClusterActivity), annotationSlim, seqnames, start)
  fwrite(df, file = "./R2.enhancer_calling/validation_NiuLJ/selected_for_validation_20210605.csv",
         row.names = F, col.names = T, sep = ",")

## to fasta
  # The difference between TAIR9 and TAIR10 was only the annotation, not the genome sequence.
  library(BSgenome.Athaliana.TAIR.TAIR9)
  dfGR <- makeGRangesFromDataFrame(df) %>%
    GenomicRanges::resize(width = 2600, fix = "center", use.names = T)
  selectedFA <- BSgenome::getSeq(BSgenome.Athaliana.TAIR.TAIR9, dfGR)
  names(selectedFA) <- df$ID
  
  writeXStringSet(selectedFA, filepath = "./R2.enhancer_calling/validation_NiuLJ/selected_for_validation_20210605.fa")

```

### 2.7.2 primer design

#### .1 input for Primer3
```{r}

library(tcltk)
library(Biostrings)
seqs <- readDNAStringSet("./R2.enhancer_calling/validation_NiuLJ/selected_for_validation_20210605.fa")
names(seqs) <- str_replace_all(names(seqs), "\\s", "_")
for (i in 1:length(seqs)) {
  p3.paras <- list(
      SEQUENCE_ID=names(seqs)[[i]],
      SEQUENCE_TEMPLATE=as.character(seqs[[i]]),
      SEQUENCE_TARGET="1000,600",
      PRIMER_TASK="generic",
      PRIMER_PICK_LEFT_PRIMER=1,
      PRIMER_PICK_INTERNAL_OLIGO=0,
      PRIMER_PICK_RIGHT_PRIMER=1,
      PRIMER_NUM_RETURN=5,
      PRIMER_OPT_SIZE=20,
      PRIMER_MIN_SIZE=18,
      PRIMER_MAX_SIZE=25,
      PRIMER_MAX_NS_ACCEPTED=0,
      PRIMER_PRODUCT_SIZE_RANGE="700-1500",
      P3_FILE_FLAG=1,
      SEQUENCE_INTERNAL_EXCLUDED_REGION="1000,600",
      PRIMER_EXPLAIN_FLAG=1,
      PRIMER_PAIR_MAX_DIFF_TM=2,
      PRIMER_PICK_ANYWAY=1,
      PRIMER_MIN_GC=40,
      PRIMER_MAX_GC=60,
      PRIMER_MIN_TM=55,
      PRIMER_MAX_TM=65
    )
  p3.paras <- paste(names(p3.paras), '=', p3.paras, sep='')
  p3.paras[length(p3.paras) + 1] <- "="
  writeLines(p3.paras, paste("./R2.enhancer_calling/validation_NiuLJ/primer3_input/", names(seqs)[[i]], ".P3input", sep = ""))
}

```

#### .2 parse Primer3 output
generate primer table.
```{r}
primerInfo <- function(x){
  x <- str_replace_all(x, c("\\d+\\s.+ PRIMER" = "PRIMER",
                            ".+ PRIMER" = "PRIMER")) %>%
    str_replace_all(., c("^\\s+" = "",
                         "\\s+" = "\t")) %>%
    str_split(pattern = "\\t") %>%
    return()
}
df <- NULL
for (m in list.files(path = "./R2.enhancer_calling/validation_NiuLJ/primer3_output/", 
                     pattern = ".+P3output$")) {
    cat(">  ", m, "\n")
    p3.results <- readLines(str_c("./R2.enhancer_calling/validation_NiuLJ/primer3_output/", m, sep = ""))
    
    LeftLs <- grep("LEFT PRIMER", p3.results)
    rightLs <- grep("RIGHT PRIMER", p3.results)
    infoLs <- grep("PRODUCT SIZE:", p3.results)
    
    
    for(i in 1:length(LeftLs)){
      # cat(">>> \n", primerInfo(p3.results[LeftLs[i]]), "\n", primerInfo(p3.results[rightLs[i]]), "\n", p3.results[infoLs[i]], "\n")
      productSize <- str_replace_all(p3.results[infoLs[i]], ".*PRODUCT SIZE: (\\d+).+", "\\1")
      df <- rbind(df, 
                  unlist(c(str_replace_all(m, ".P3output", ""), "Forward", 
                           primerInfo(p3.results[LeftLs[i]]), i, productSize))) %>%
        rbind(., unlist(c(str_replace_all(m, ".P3output", ""), "Reverse", 
                          primerInfo(p3.results[rightLs[i]]), i, productSize)))
    }

}
df <- as.data.frame(df)
colnames(df) <- c("ID", "type", "name", "start", "primer_length", "TM", "GC", "any_th", "3'TH", "hairpin", "seq", "primer_ID", "length")

df <- df %>%
  dplyr::mutate(across(matches("start|primer_length|length"), as.numeric),
                end = ifelse(type == "Forward", start + primer_length, start),
                start = ifelse(type == "Forward", start, start - primer_length))

df %>%
  dplyr::select(ID:start, end, primer_length:length) %>%
  fwrite( "./R2.enhancer_calling/validation_NiuLJ/Primer_list_20210711.csv", quote = F)
```


### 2.7.3 blast and sequence extract
truncate sequence from the genomic sequence based on the primer blast results.

*blast primers*
use parameter `-evalue 5 -word_size 7` in blastn for short sequence.
```{shell}
#!/bin/bash
#BSUB -J blast                  ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 1                  ### ask for number of cores (default: 1)
#BSUB -W 00:20             ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

##-------------------------------------------------------------------------------------##
# blast primer to genome sequence.
#   Steps:
#      1. make blastdb
#      2. blast
#      3. collect results.
#                                                       Tan,Yongjun 20210924
#----------------------------------------------------------------------------------------

source activate trinity

# makeblastdb
# #1,#make datebase
# cd /scratch/2021-09-24/bioTanyj/At_enhancer/refgenome

# makeblastdb -in TAIR10_no_repeat.fa  -input_type fasta -dbtype nucl -title TAIR10_no_repeat.fa -parse_seqids

# sequences
cd /scratch/2021-09-24/bioTanyj/At_enhancer/vlidation
refPath="/scratch/2021-09-24/bioTanyj/At_enhancer/refgenome/TAIR10_no_repeat.fa"
blastn -db ${refPath} -query primers_seq_20210924.fa -evalue 5 -word_size 7 -out primers_seq_20210924.fa.html -html
blastn -db ${refPath} -query primers_seq_20210924.fa -evalue 5 -word_size 7 -out primers_seq_20210924.fa.tsv -outfmt '7 qseqid sseqid sallacc pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen'

```


*Collect and filtration*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
blast_all <- read.table("./validation/primers_20210924/primers_seq_20210924.fa.tsv", comment = "#")
colnames(blast_all) <- c("query", "subject", "subjectacc", "identity", "alignment length", "mismatches", 
                         "gap opens", "q. start", "q. end", "s. start", "s. end", "evalue", "bit score", "query length") %>%
  str_to_title(.) %>%
  str_replace_all(., " ", "")

# filtration, just keep hit with the smallest Evalue
length(unique(blast_all$Query))
blast_all_flt <- blast_all %>%
  group_by(Query) %>%
  arrange(Query, Evalue) %>%
  mutate(MaxBitScore = max(BitScore), 
         Keep = ifelse(BitScore == MaxBitScore, "Yes", "No") ) %>%
  dplyr::filter(Keep == "Yes") %>%
  dplyr::mutate(CRE_id = str_replace_all(Query, "_[FR]", ""), 
                Primer = ifelse(str_detect(Query, "_F_"), "Forward", "Reverse"), 
                tmp = str_c(Subject, S.Start, S.End, sep = "_"), 
                Num = n())

# generate bed
seq_region <- blast_all_flt %>%
  dplyr::mutate(CRE_id = str_replace_all(Query, "(.+)-.", "\\1")) %>%
  group_by(CRE_id, Subject) %>%
  summarise(start = min(c(S.Start, S.End)), 
            end = max(c(S.Start, S.End))) %>%
  dplyr::rename(seqnames = Subject) %>%
  ungroup() %>%
  dplyr::mutate(seqnames = str_replace_all(seqnames, "chr", "Chr"))

# generate fasta
  # The difference between TAIR9 and TAIR10 was only the annotation, not the genome sequence.
  library(BSgenome.Athaliana.TAIR.TAIR9)
  dfGR <- makeGRangesFromDataFrame(seq_region)
  selectedFA <- BSgenome::getSeq(BSgenome.Athaliana.TAIR.TAIR9, dfGR)
  names(selectedFA) <- str_c(seq_region$CRE_id, seq_region$seqnames, seq_region$start, seq_region$end, sep = "_")
  
  writeXStringSet(selectedFA, filepath = "./validation/primers_20210924/genomic_seq_20210924.fa")
```

## 2.7 validation (202205)

### 2.7.1 select regions
  TE-derived enhancers (LTR/Gypsy & DNA super family)
  Exonic enhancer 
  random control
  ePromoter (overlap TSS or not)
  important genes.
  
  random seed was set. regions may overlap with TSS were also marked.
  
Attention: 
  Two annotation methods were different, annotatePeaks() only report one gene based on priority, findoverlap()-based method will report all genes with regions overlapped with enhancer (larger than parameter: minOverlap).
  
```{r}
    rm(list = ls())
    source("D:/SUST/code/TanYongjun_code.R")
    enhancers <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
      dplyr::mutate(Features = str_replace_all(annotation, "^([^\\()]+)\\s\\(.+$", "\\1")) %>%
      dplyr::select(-c(ATACseq:Dim.5)) %>%
      dplyr::rename(gene_id = geneId) %>%
      dplyr::filter(seqnames != "Mt")
    gene_regions <- fread("./R4.Distribution_of_enhancers/Exonic_enhancer/List_gene_CDS_hit_Enhancer_EnResize200bp_MinOverlap50_20220508.csv") %>%
      as_tibble()
    library(TxDb.Athaliana.BioMart.plantsmart28)
    txdb <- TxDb.Athaliana.BioMart.plantsmart28
    core_pro <- promoters(TxDb.Athaliana.BioMart.plantsmart28, upstream = 1, downstream = 1, use.names = T) #downstream 400: because the largest region amplified by PCR was 1000bp.
    TAIR_gene <- fread("./refgenome/TAIR_genes_info_merge_20220510.csv.gz")
    
    
  # validated in batch1.
    en_batch1 <- fread("./validation/拟南芥第一批验证_20220508.tsv") %>%
      dplyr::mutate(seqnames = str_replace_all(seqnames, "Chr", "")) %>%
      left_join(., dplyr::select(enhancers, seqnames:end, annotation, matches("^TE"), Features))
    en_batch1GR <- makeGRangesFromDataFrame(en_batch1)
    en_batch1 %>%
      dplyr::filter(`20220330GUS` == "Y") %>%
      makeGRangesFromDataFrame() %>%
      findOverlaps(., core_pro) %>%
      as.data.frame()
  
  # discard enhancers which overlap TSS.
    ol <- makeGRangesFromDataFrame(enhancers) %>%
      findOverlaps(., core_pro)
    enhancers_noTSS <- enhancers[-unique(ol@from),]
  
  ## Key genes (with enhancers located in promoter and distal intergenic, distance to TSS less than 3k)
    # not used, because no key genes with enhancer located in promoter/distal intergenic ( dist2TSS < 3k, not TSS/UTR/CDS/intron)
    select_KeyGene <- enhancers_noTSS %>%
      dplyr::filter(str_detect(Features, "Promoter|Distal Intergenic|Downstream"),
                    abs(distanceToTSS) < 3000) %>%
      left_join(., dplyr::select(TAIR_gene, gene_id, Num_citation, SYMBOL_ls, mut_phe) %>% distinct()) %>%
      dplyr::filter(Num_citation > 4) %>%
      arrange(desc(Num_citation)) %>%
      head(n = 40) %>%
      dplyr::select(seqnames:end, Features, TE, TE_SuperFamily, TE_Family, Num_citation, SYMBOL_ls, mut_phe,
                    gene_id, Num_citation, SYMBOL_ls, mut_phe) %>%
      dplyr::mutate(Group = "KeyGeneFlanking")
    
  # TE-derived
    set.seed(12345)
    select_TE <- enhancers %>%
      dplyr::filter(str_length(TE) > 2,
                    str_detect(TE_SuperFamily, "^DNA$|LTR/Gypsy")) %>%
      group_by(TE_SuperFamily) %>%
      slice_sample(n = 50) %>%
      dplyr::select(seqnames:end, Features, TE, TE_SuperFamily, TE_Family) %>%
      dplyr::mutate(Group = "TE")
    table(select_TE$TE_SuperFamily, select_TE$Features)
  
    TE_DNA <- enhancers %>%
      dplyr::filter(TE_SuperFamily == "DNA")
    
  # CDS enhancers (no Promoter)
  select_CDS <- gene_regions %>%
    dplyr::filter(Features == "CDS", Num_citation > 100) %>%
    dplyr::mutate(CRE = str_replace_all(CRE_ls, "(.+);.+", "\\1")) %>%
    dplyr::select(gene_id, Feature_id, SYMBOL_ls, Num_citation, CRE, mut_phe) %>%
    tidyr::separate(., col = CRE, into = c("seqnames", "start", "end"), sep = "_") %>%
    dplyr::rename(Features = Feature_id) %>%
    dplyr::mutate(Group = "eCDS",
                  start = as.numeric(start), 
                  end = as.numeric(end))
  
  # Random (no enhancer and no TSS)
    set.seed(12345)
    select_random <- select_CDS %>%
      makeGRangesFromDataFrame() %>%
      toolkit_tyj$shuffle(peak.gr = ., TxDb = txdb, RandomSeed = 123456) 
    ol1 <- findOverlaps(select_random, makeGRangesFromDataFrame(enhancers))
    ol2 <- findOverlaps(select_random, core_pro)
    
    set.seed(12345)
    select_random <- select_random[-unique(c(ol1@from, ol2@from))] %>%
      as_tibble() %>%
      dplyr::filter(seqnames != "Mt") %>%
      slice_sample(n = 20) %>%
      dplyr::select(-c(width, strand)) %>%
      dplyr::mutate(Group = "Random", seqnames = as.character(seqnames))
  
  # ePromoter
    set.seed(12345)
    select_ePro <- enhancers %>%
      as_tibble() %>%
      dplyr::filter(Features == "Promoter", str_length(TE) < 2) %>%
      dplyr::select(seqnames:end, matches("^TE"), Features) %>%
      dplyr::mutate(Group = "ePromoter") 
    
    select_ePro <- select_ePro %>%
      slice_sample(n = 40)
      
    table(select_ePro$Features)
  
  selected_all <- full_join(select_CDS, select_ePro)%>%
    full_join(., select_random) %>%
    # full_join(., select_KeyGene) %>%
    full_join(., select_TE)
  
# overlap TSS or not
  ol <- findOverlaps(makeGRangesFromDataFrame(selected_all), core_pro)
  selected_all$TSS <- "N"
  selected_all$TSS[unique(ol@from)] <- "Y"
  table(selected_all$Group, selected_all$TSS)
  
  # discard enhancers overlaped with TSS (except ePromoter)
  selected_all <- selected_all %>%
    dplyr::filter(!(TSS == "Y" & Group != "ePromoter")) %>%
    dplyr::mutate(Group = ifelse(TSS == "Y", str_c(Group, "_TSS"), Group),
                  seqnames = str_c("Chr", seqnames), 
                  En_id = str_c("En", Group, seqnames, start, end, sep = "_"))
  table(selected_all$Group, selected_all$TSS)

# discard enhancers validated in batch1
  ol <- findOverlaps(en_batch1GR, selected_all %>% 
                           dplyr::mutate(seqnames = str_replace_all(seqnames, "Chr", "")) 
                     %>% makeGRangesFromDataFrame(.))
  selected_all <- selected_all[-unique(ol@to),]
  
# summary
  table(selected_all$Group, selected_all$TSS)
  
# save to file
  # table
    fwrite(selected_all, file = "./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220510.csv")
  
  # fasta
    # The difference between TAIR9 and TAIR10 was only the annotation, not the genome sequence.
    library(BSgenome.Athaliana.TAIR.TAIR9)
    dfGR <- makeGRangesFromDataFrame(selected_all) %>%
      GenomicRanges::resize(width = 2600, fix = "center", use.names = T)
    selectedFA <- BSgenome::getSeq(BSgenome.Athaliana.TAIR.TAIR9, dfGR)
    names(selectedFA) <- selected_all$En_id
    
    writeXStringSet(selectedFA, 
                    filepath = "./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220510.fa")
  
```

### 2.7.2 primer design
 Step 1-5 can acomplished in `/scratch/2022-05-05/bioTanyj/At_enhancer/scripts/old_scripts_20210908/2.7.2_Primer_design_all_in_one.R`
#### .1 input for Primer3
```{r}

library(tcltk)
library(Biostrings)
seqs <- readDNAStringSet("./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220509.fa")

for (i in 1:length(seqs)) {
  p3.paras <- list(
      SEQUENCE_ID=names(seqs)[[i]],
      SEQUENCE_TEMPLATE=as.character(seqs[[i]]),
      SEQUENCE_TARGET="1000,600",
      PRIMER_TASK="generic",
      PRIMER_PICK_LEFT_PRIMER=1,
      PRIMER_PICK_INTERNAL_OLIGO=0,
      PRIMER_PICK_RIGHT_PRIMER=1,
      PRIMER_NUM_RETURN=5,
      PRIMER_OPT_SIZE=20,
      PRIMER_MIN_SIZE=18,
      PRIMER_MAX_SIZE=25,
      PRIMER_MAX_NS_ACCEPTED=0,
      PRIMER_PRODUCT_SIZE_RANGE="700-1500",
      P3_FILE_FLAG=1,
      SEQUENCE_INTERNAL_EXCLUDED_REGION="1000,600",
      PRIMER_EXPLAIN_FLAG=1,
      PRIMER_PAIR_MAX_DIFF_TM=2,
      PRIMER_PICK_ANYWAY=1,
      PRIMER_MIN_GC=40,
      PRIMER_MAX_GC=60,
      PRIMER_MIN_TM=55,
      PRIMER_MAX_TM=65
    )
  p3.paras <- paste(names(p3.paras), '=', p3.paras, sep='')
  p3.paras[length(p3.paras) + 1] <- "="
  writeLines(p3.paras, paste("./R2.enhancer_calling/validation_batch2/primer3_input/", names(seqs)[[i]], ".P3input", sep = ""))
}

```

#### .2-5 run Primer3 on Taiyi.
```{r}
#-#####################################################################
#  Primer design for Enhancer validtion.
#     Steps:
#       1. design primters using Primer3.
#       2. predict putative products based on blastN.
#       3. Restriction endonuclease recognition sites (RERS).
#       4. Output results.
#                                       Yongjun Tan, 20211008, 20220509
#-#####################################################################

# set parapeters ---------------------------
    setwd("/scratch/2022-05-21/bioTanyj/At_enhancer/R2.enhancer_calling/primer_validation_20220510")
    source("/work/bio-tanyj/soft/TanYongjun_code.R")
    library(Biostrings)
    seqPath <- "slected_Enhancers_eCDS_ePro_TE_Random_20220510.fa"
    Path2P3Input <- "Primer_Input"
    Path2P3Output <- "Primer_Output"
    PrimerFileNamePrefix <- "Primer_all_raw_20220510"
    
    # for blast
    # refPath <- "/scratch/2022-05-21/bioTanyj/At_enhancer/refgenome/TAIR10_no_repeat.fa" # db for blast.
    refPath <- "/scratch/2022-05-21/bioTanyj/At_enhancer/refgenome/TAIR10.fa" # db for blast.

    # endocuclease recongnision sites
    enzy_site <- c("CCCGGG", "GGATCC") %>% DNAStringSet()
    names(enzy_site) <- c("Enzy1", "Enzy2")
    enzy_name <- c("sma1", "bamh1")

# Functions -----------------------------

# prepare input files of Primer3 from DNAStringSets object
fa2P3Input <- function(seqPath, Path2P3Input){
    system(command = str_c("mkdir ", Path2P3Input), ignore.stderr = T)
    seqs <- readDNAStringSet(seqPath)
    for (i in 1:length(seqs)) {
    p3.paras <- list(
        SEQUENCE_ID=names(seqs)[[i]],
        SEQUENCE_TEMPLATE=as.character(seqs[[i]]),
        SEQUENCE_TARGET="1020,580",
        PRIMER_TASK="generic",
        PRIMER_PICK_LEFT_PRIMER=1,
        PRIMER_PICK_INTERNAL_OLIGO=0,
        PRIMER_PICK_RIGHT_PRIMER=1,
        PRIMER_NUM_RETURN=5,
        PRIMER_OPT_SIZE=22,
        PRIMER_MIN_SIZE=18,
        PRIMER_MAX_SIZE=25,
        PRIMER_MAX_NS_ACCEPTED=0,
        PRIMER_PRODUCT_SIZE_RANGE="400-1000",
        P3_FILE_FLAG=1,
        SEQUENCE_INTERNAL_EXCLUDED_REGION="1020,580",
        PRIMER_EXPLAIN_FLAG=1,
        PRIMER_PAIR_MAX_DIFF_TM=2,
        PRIMER_PICK_ANYWAY=1,
        PRIMER_MIN_GC=40,
        PRIMER_MAX_GC=60,
        PRIMER_MIN_TM=55,
        PRIMER_MAX_TM=65
        )
    p3.paras <- paste(names(p3.paras), '=', p3.paras, sep='')
    p3.paras[length(p3.paras) + 1] <- "="
    writeLines(p3.paras, paste("./", Path2P3Input, "/", names(seqs)[[i]], ".P3input", sep = ""))
    }
}

# run primer3 (wait=F: parallel!, need to run with furrr)
RunPrimer3 <- function(Path2P3Input, Path2P3Output){
    library(furrr)
    system(command = str_c("mkdir ", Path2P3Output))
    ## for edition
    # for(i in list.files(path = Path2P3Input, pattern = ".+P3input")){
    #     system(command = str_c(
    #         "/work/bio-tanyj/miniconda3/envs/trinity/bin/primer3_core ",
    #         " ./", Path2P3Input, "/", i, 
    #         " --format_output --out=./", Path2P3Output, "/", str_replace_all(i, "P3input", "P3output"),
    #         sep = ""
    #     ), wait = T)
    #     system(str_c("rm *.rev"))
    #     system(str_c("rm *.for"))
    # }

    ## parallel with furrr
    plan(multisession, workers = 10)
    future_map(list.files(path = Path2P3Input, pattern = ".+P3input"),
        function(x){
        system(command = str_c(
            "/work/bio-tanyj/miniconda3/envs/trinity/bin/primer3_core ",
            " ./", Path2P3Input, "/", x, 
            " --format_output --out=./", Path2P3Output, "/", str_replace_all(x, "P3input", "P3output"),
            sep = ""
        ), wait = T)
        }
    )
    plan(sequential)
    system(str_c("rm *.rev"))
    system(str_c("rm *.for"))
}

# collect primer3 results, (tsv; fa file for blast)
CollectPrimers <- function(PrimerFileNamePrefix){
    .primerInfo <- function(x){
        x <- str_replace_all(x, c("\\d+\\s.+ PRIMER" = "PRIMER",
                                ".+ PRIMER" = "PRIMER")) %>%
        str_replace_all(., c("^\\s+" = "",
                            "\\s+" = "\t")) %>%
        str_split(pattern = "\\t") %>%
        return()
    }
    
    df <- NULL
    for (m in list.files(path = Path2P3Output, 
                        pattern = ".+P3output$")) {
        cat(">  ", m, "\n")
        p3.results <- readLines(str_c(Path2P3Output, "/", m, sep = ""))
        
        LeftLs <- grep("LEFT PRIMER", p3.results)
        rightLs <- grep("RIGHT PRIMER", p3.results)
        infoLs <- grep("PRODUCT SIZE:", p3.results)
        
        
        if (length(infoLs) > 0) {
            for(i in 1:length(LeftLs)){
            # cat(">>> \n", .primerInfo(p3.results[LeftLs[i]]), "\n", .primerInfo(p3.results[rightLs[i]]), "\n", p3.results[infoLs[i]], "\n")
            productSize <- str_replace_all(p3.results[infoLs[i]], ".*PRODUCT SIZE: (\\d+).+", "\\1")
            df <- rbind(df, 
                        unlist(c(str_replace_all(m, ".P3output", ""), "Forward", 
                                .primerInfo(p3.results[LeftLs[i]]), i, productSize))) %>%
                rbind(., unlist(c(str_replace_all(m, ".P3output", ""), "Reverse", 
                                .primerInfo(p3.results[rightLs[i]]), i, productSize)))
            }
        }
    
    }
    df <- as.data.frame(df)
    colnames(df) <- c("ID", "type", "name", "start", "primer_length", "TM", "GC", "any_th", "3'TH", "hairpin", "seq", "primer_ID", "length")
    
    df <- df %>%
        dplyr::mutate(across(matches("start|primer_length|length"), as.numeric),
                    end = ifelse(type == "Forward", start + primer_length, start),
                    start = ifelse(type == "Forward", start, start - primer_length))

    ## save to file
    df %>%
        dplyr::select(ID:start, end, primer_length:length) %>%
        fwrite(str_c(PrimerFileNamePrefix, ".csv"), quote = F)
    
    library(Biostrings)
    primers_fa <- BStringSet(df$seq)
    names(primers_fa) <- str_c(df$ID, df$primer_ID, str_replace_all(df$type, "orward|everse", ""), sep = "_")
    
    writeXStringSet(primers_fa, filepath = str_c(PrimerFileNamePrefix, ".fa"))

}

# blast primers to genome sequences
BlastPrimer2Genome <- function(PrimerFileNamePrefix, refPath){
    system("source activate trinity") # env contain the blast
    # makeblastdb (if not)
    # #1,#make datebase
    # makeblastdb -in TAIR10.fa -input_type fasta -dbtype nucl -title TAIR10.fa -parse_seqids

    # blast
    system(command = str_c(
        "/work/bio-tanyj/miniconda3/envs/trinity/bin/blastn ",
        " -db ", refPath,
        " -query ", PrimerFileNamePrefix, ".fa",
        " -evalue 5 -word_size 7 -out ", PrimerFileNamePrefix, ".blast.tsv ",
        "-outfmt '7 qseqid sseqid sallacc pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen'"
    ))
}

# count putative PCR products based on blastN results.
CountProductPCR <- function(seqPath){
    # load table generated by blastN and format.
        blast_all <- read.table(file = str_c(PrimerFileNamePrefix, ".blast.tsv"), comment.char = "#") %>%
            as.data.frame()

        colnames(blast_all) <- c("query", "subject", "subjectacc", "identity", "alignment length", "mismatches", 
                                "gap opens", "q. start", "q. end", "s. start", "s. end", "evalue", "bit score", "query length") %>%
            str_to_title(.) %>%
            str_replace_all(., " ", "")

        blast_all <- blast_all %>%
            dplyr::mutate(CRE_primer_set = str_replace_all(Query, "(.+)_([FR])", "\\1"),
                        primer = str_replace_all(Query, "(.+)_([FR])", "\\2"),
                        strand = ifelse(S.Start > S.End, "-", "+")) %>%
            dplyr::filter((Q.End == QueryLength & strand == "+") |
                            (Q.Start == 1 & strand == "-")) %>%
            dplyr::select(-Query, -Subjectacc, -Identity)

        nrow(blast_all)
        table(blast_all$CRE_primer_set, blast_all$primer)
        head(blast_all)

    # mark nonspecific amplification
        # only mark when putative products less than 5000.
        # primer-pair ID, putative products and length.
        products_PCR <- split(blast_all, blast_all$CRE_primer_set) %>%
            # x <- blast_all_l[[6]] # test
            map_df(., .id = "id", function(x){
                # list of chr which both hit by F and R primers.
                chr_ls <- x %>%
                    group_by(Subject, primer, strand) %>%
                    count() %>%
                    group_by(Subject, strand) %>%
                    count() %>%
                    group_by(Subject) %>%
                    count() %>%
                    dplyr::filter(n > 1) %>%
                    ungroup() %>%
                    dplyr::select(Subject) %>%
                    unlist() %>%
                    as.vector() %>%
                    unique()

                # search putative PCR products on each chr.
                product_len <- NULL
                for(i in chr_ls){
                    chr_tmp <- x %>%
                        dplyr::filter(Subject == i)
                    # product length (distance) between primer pos on strand + and -.
                    negtive_pos <- chr_tmp$S.Start[chr_tmp$strand == "-"] %>% unlist %>% as.vector %>% sort
                    positive_pos <- chr_tmp$S.Start[chr_tmp$strand == "+"] %>% unlist %>% as.vector %>% sort
                    len_tmp <- map(positive_pos, function(x){negtive_pos - x + 1}) %>%
                        purrr::reduce(.f = c)
                    lem_tmp <- len_tmp[len_tmp > 0 & len_tmp < 5000]
                    product_len <- c(product_len, lem_tmp)
                }

                # return summary of nonspecific amplication. 
                #  (only length of top ten products were reported)
                return(tibble(Num_fragments = length(product_len),
                    Len_Top10fragments = str_c(sort(product_len) %>% head(., n = 10), collapse = ";")))

            })
        # save
        fwrite(products_PCR, file = str_c("Number_of_PCR_products_for_each_primer_pair_", 
                                          PrimerFileNamePrefix, ".csv"), sep = "\t")
}

# count restriction endonuclease recongnision sites, classity RERS based on its position according to primer.
RERS <- function(seqPath, PrimerFileNamePrefix, enzy_name){
    library(Biostrings)

    # search recognition sites in CREs. (only fragments contain recognition site were contained in the table)
    df <- NULL
    tmp <- readDNAStringSet(filepath = seqPath)
    for (j in names(enzy_site)) {
        df <- as.list(vmatchPattern(enzy_site[[j]], tmp)) %>%
            map_dfr(., as.data.frame, .id = "ID") %>%
            dplyr::mutate(enzy_site = j) %>%
            as_tibble() %>%
            rbind(., df)
    }

    rm(j)
    
    df_t <- df %>%
        # dplyr::mutate(enzy_site = str_replace_all(enzy_site, "_c", "")) %>%
        pivot_wider(id_cols = ID, names_from = enzy_site, values_from = start, values_fill = list())
    
    # position between primer and RERS.
    # Situations: 1. no RERS; 2. RERS located between paired-primers but not in core region (one or both side); 3. RERS located in core regions.
    RERS_site_classify <- function(PCR_start, PCR_end, RERS_site){
        # mark three situations
        RERS_site_class <- rep("None", length(PCR_start))
        for (i in 1:length(PCR_start)) {
        RERS_site_class[i] <- "None"
        RERS_PCR_region <- sum(unlist(RERS_site[i]) > PCR_start[i] & unlist(RERS_site[i]) < PCR_end[i])
        RERS_core_region <- sum(unlist(RERS_site[i]) > 1000 & unlist(RERS_site[i]) < 1600)
        if(RERS_PCR_region > 0) {RERS_site_class[i] <- "PCR_region"}
        if(RERS_core_region > 0) {RERS_site_class[i] <- "core_region"}
        }
        return(RERS_site_class)
    }

    primers_all <- fread(str_c(PrimerFileNamePrefix, ".csv")) %>%
        as_tibble() %>%
        dplyr::mutate(pos = ifelse(type == "Reverse", end, start)) %>%
        pivot_wider(id_cols = c(ID, primer_ID), names_from = type, values_from = pos) %>%
        dplyr::rename(PCR_start = Forward, PCR_end = Reverse) %>%
        left_join(., df_t) %>%
        dplyr::mutate(Enzy1_type = RERS_site_classify(.$PCR_start, .$PCR_end, .$Enzy1),
                    Enzy2_type = RERS_site_classify(.$PCR_start, .$PCR_end, .$Enzy2)) %>%
        dplyr::mutate(Enzy1 = sapply(.$Enzy1, function(x){str_c(unlist(x), collapse = ";")}),
                    Enzy2 = sapply(.$Enzy2, function(x){str_c(unlist(x), collapse = ";")}))
    colnames(primers_all) <- str_replace_all(colnames(primers_all),c("Enzy1" = enzy_name[1],
                                                                     "Enzy2" = enzy_name[2]))

    primers_all %>%
        fwrite(file = str_replace_all(seqPath, ".fa", "_Restriction_enzyme_recongnision_sites.csv"))
}


# Step1 primer design -------------------------
    # generate primer3 input files from seqs
    timestamp()
    cat(">>> Prepare Primer3 input files\n")
    fa2P3Input(seqPath, Path2P3Input)

    # Design primers with Primer3
    timestamp()
    cat(">>> Design primers\n")
    RunPrimer3(Path2P3Input, Path2P3Output)

    # collect Primers
    timestamp()
    cat(">>> Collect primers\n")
    CollectPrimers(PrimerFileNamePrefix)

# Step2 putative products -----------------------
    # blast to genome
    timestamp()
    cat(">>> blast primers to genome \n")
    BlastPrimer2Genome(PrimerFileNamePrefix, refPath)

    # collect blast reuslts (count number of putative PCR products which less than 5K)
    timestamp()
    cat(">>> Count putative PCR products \n")
    CountProductPCR(PrimerFileNamePrefix)

# Step3 Restriction endonuclease recognition sites identification -------------
    timestamp()
    cat(">>> Count Endonuclease recognition sites \n")
    RERS(seqPath, PrimerFileNamePrefix, enzy_name)

# Step4 merge results ---------------------
    timestamp()
    cat(">>> Merge all results to a table \n")
    # summary of PCR amplified fragments
    primer_nonspec <- fread(str_c("Number_of_PCR_products_for_each_primer_pair_", 
                                          PrimerFileNamePrefix, ".csv")) %>%
        as_tibble() %>%
        dplyr::mutate(ID = str_replace_all(id, "(.+)_\\d", "\\1"),
                    primer_ID = str_replace_all(id, ".+_(\\d)", "\\1") %>% as.integer()) %>%
        dplyr::select(-id)
    table(primer_nonspec$ID)
    
    # sequence of primers
    specific_primers <- fread(str_c(PrimerFileNamePrefix, ".csv")) %>%
        as_tibble() %>%
        left_join(., primer_nonspec) %>%
        dplyr::mutate(Num_fragments = ifelse(is.na(Num_fragments), 0, Num_fragments),
                    specific = ifelse(Num_fragments == 1, "Y", "N")) # NA means no hit identified in blast
    
    # sum(is.na(specific_primers$specific))
    
    # restriction enzy recognision site
    RERS_info <- fread(str_replace_all(seqPath, ".fa", "_Restriction_enzyme_recongnision_sites.csv")) %>%
        as_tibble()
    specific_primers <- left_join(specific_primers, RERS_info) %>%
        dplyr::mutate(Features = str_replace_all(ID, "En_(.+)_Chr.+", "\\1"))

    fwrite(specific_primers,
            file = str_c(str_replace_all(seqPath, ".fa", "_primers_all.csv")))

```


#### .6 select primers.

```{r}
# summary of PCR amplified fragments
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  region_info <- read_csv("./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220510.csv") %>%
    dplyr::rename(ID = En_id) %>%
    dplyr::select(ID, TSS, contains("TE"), Group, Num_citation, SYMBOL_ls, gene_id, mut_phe)
  
  primers_all <- read_csv("./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220510_primers_all.csv") %>%
    dplyr::mutate(RERS_type = ifelse(sma1_type == "PCR_region" | bamh1_type == "PCR_region", "PCR_region", "None"),
                  RERS_type = ifelse(sma1_type == "core_region" | bamh1_type == "core_region", "core_region", RERS_type)) %>%
    left_join(., region_info) %>%
    dplyr::select(-name)
  
# mark fragments with TSS located in amplified fragments. (mannually validated using enhancers located in the first CDS of PHYB/AT2G18790)
  primers_tmp <- primers_all %>%
    dplyr::mutate(ExtendCREStart = str_replace_all(ID, ".+_Chr\\d+_(\\d+)_\\d+", "\\1") %>% as.numeric(),
                  start = ExtendCREStart -1000 + start,
                  end = ExtendCREStart -1000 + end) %>%
    dplyr::select(ID, start, end, primer_ID) %>%
    group_by(ID, primer_ID) %>%
    dplyr::summarise(start = min(start), end = max(end)) %>%
    dplyr::mutate(seqnames = str_extract(ID, "Chr\\d"),
                  seqnames = str_replace_all(seqnames, "Chr", ""))
  primers_tmpGR <- makeGRangesFromDataFrame(primers_tmp)
  core_pro <- promoters(TxDb.Athaliana.BioMart.plantsmart28, upstream = 1, downstream = 1, use.names = T) #downstream 400: because the largest region amplified by PCR was 1000bp.
  ol <- findOverlaps(primers_tmpGR, core_pro)
  primers_tmp$PcrHitTSS <- "N"
  primers_tmp$PcrHitTSS[unique(ol@from)] <- "Y"
  primers_tmp <- primers_tmp %>%
    dplyr::select(ID, primer_ID, PcrHitTSS)
  primers_all <- left_join(primers_all, primers_tmp)
  table(primers_all$Group, primers_all$PcrHitTSS)
  
  # discard fragments with TSS (except Group: ePromoter_TSS)
  primers_all <- primers_all %>%
    dplyr::filter(!(Group != "ePromoter_TSS" & PcrHitTSS == "Y"))
  
  
# summary
  # only specific region can be amplified by the primer pair.
  primers_all %>%
    # dplyr::select(specific, bamh1_type, sma1_type, Features) %>%
    dplyr::filter(RERS_type != "core_region") %>%
    group_by(ID, Features, specific) %>%
    count() %>%
    group_by(Features, specific) %>%
    count() %>%
    pivot_wider(names_from = specific, values_from = n) %>%
    kable(caption = "Number of regions with sucessfull design primers")
  
# select (primers target unique position, if no putative products, choice the primer pair with the shortest PCR products.)
  # Random, ePromoter
  primers_select <- primers_all %>%
    dplyr::filter(str_detect(Features, "Random|ePromoter"), specific == "Y", RERS_type != "core_region") %>%
    group_by(ID) %>%
    arrange(Features, ID, primer_ID, length) %>%
    slice_head(n = 2) %>%
    group_by(Features) #%>%
    # slice_head(n = 10)
    
  # eCDS and KeyGeneFlanking
  primers_select <- primers_all %>%
    dplyr::filter(str_detect(Features, "eCDS|KeyGeneFlanking"), specific == "Y", RERS_type != "core_region") %>%
    group_by(ID) %>%
    arrange(desc(Num_citation), ID, primer_ID, length) %>%
    slice_head(n = 2) %>% 
    ungroup() %>%
    rbind(., primers_select)
  
  # TE
  primers_select <- primers_all %>%
    dplyr::filter(str_detect(Features, "TE"), RERS_type != "core_region") %>%
    group_by(ID) %>%
    arrange(ID, desc(Num_citation), primer_ID, length) %>%
    slice_head(n = 2) %>%
    group_by(TE_SuperFamily) %>%
    arrange(length) %>%
    # slice_head(n = 20) %>% 
    rbind(., primers_select) %>%
    arrange(Group, TE_SuperFamily, desc(Num_citation), length)
  
  primers_select %>%
    dplyr::filter(type == "Forward") %>%
    group_by(Group, specific, PcrHitTSS) %>%
    count() %>%
    pivot_wider(names_from = PcrHitTSS, values_from = n, values_fill = 0) %>%
    kable(caption = "Number of regions (primers selected) hit TSS or not")
  
  primers_select %>%
    dplyr::mutate(type = str_replace_all(type, "orward|everse", ""),
                  PrimterID = str_c(ID, primer_ID, type, sep = "_")) %>%
    dplyr::relocate(Group, Features, ID, gene_id, type, TE_SuperFamily, TE_Family, TE) %>%
    fwrite(., file = "./R2.enhancer_calling/validation_batch2/Selected_primers_for_GUS_20220510.csv")
    
```

## 2.7 validation (20220523)

### 2.7.1 select regions
  TE-derived enhancers (LTR/Gypsy & DNA super family)
  
```{r}
    rm(list = ls())
    source("D:/SUST/code/TanYongjun_code.R")
    enhancers <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
      dplyr::mutate(Features = str_replace_all(annotation, "^([^\\()]+)\\s\\(.+$", "\\1")) %>%
      dplyr::select(-c(ATACseq:Dim.5)) %>%
      dplyr::rename(gene_id = geneId) %>%
      dplyr::filter(seqnames != "Mt")
    gene_regions <- fread("./R4.Distribution_of_enhancers/Exonic_enhancer/List_gene_CDS_hit_Enhancer_EnResize200bp_MinOverlap50_20220508.csv") %>%
      as_tibble()
    library(TxDb.Athaliana.BioMart.plantsmart28)
    txdb <- TxDb.Athaliana.BioMart.plantsmart28
    core_pro <- promoters(TxDb.Athaliana.BioMart.plantsmart28, upstream = 1, downstream = 1, use.names = T) #downstream 400: because the largest region amplified by PCR was 1000bp.
    TAIR_gene <- fread("./refgenome/TAIR_genes_info_merge_20220510.csv.gz")
    
    
  # validated in batch1.
    en_batch1 <- fread("./validation/拟南芥第一批验证_20220508.tsv") %>%
      dplyr::mutate(seqnames = str_replace_all(seqnames, "Chr", "")) %>%
      left_join(., dplyr::select(enhancers, seqnames:end, annotation, matches("^TE"), Features))
    en_batch1GR <- makeGRangesFromDataFrame(en_batch1)
    en_batch1 %>%
      dplyr::filter(`20220330GUS` == "Y") %>%
      makeGRangesFromDataFrame() %>%
      findOverlaps(., core_pro) %>%
      as.data.frame()
  
  # validated in batch2 (20220510)
    en_batch2 <- fread("./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220510.csv")
    
  # discard enhancers which overlap TSS.
    ol <- makeGRangesFromDataFrame(enhancers) %>%
      findOverlaps(., core_pro)
    enhancers_noTSS <- enhancers[-unique(ol@from),]
  
```

### 2.7.2 primer design

#### .1-5 design
 Step 1-5 can acomplished in `/scratch/2022-05-05/bioTanyj/At_enhancer/scripts/old_scripts_20210908/2.7.2_Primer_design_all_in_one.R`

#### .6 select primers.

```{r}
# summary of PCR amplified fragments
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  region_info <- read_csv("./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220510.csv") %>%
    dplyr::rename(ID = En_id) %>%
    dplyr::select(ID, TSS, contains("TE"), Group, Num_citation, SYMBOL_ls, gene_id, mut_phe)
  
  primers_all <- read_csv("./R2.enhancer_calling/validation_batch2/slected_Enhancers_eCDS_ePro_TE_Random_20220523_primers_all.csv") %>%
    dplyr::mutate(RERS_type = ifelse(sma1_type == "PCR_region" | bamh1_type == "PCR_region", "PCR_region", "None"),
                  RERS_type = ifelse(sma1_type == "core_region" | bamh1_type == "core_region", "core_region", RERS_type)) %>%
    left_join(., region_info) %>%
    dplyr::select(-name)
  
# mark fragments with TSS located in amplified fragments. (mannually validated using enhancers located in the first CDS of PHYB/AT2G18790)
  primers_tmp <- primers_all %>%
    dplyr::mutate(ExtendCREStart = str_replace_all(ID, ".+_Chr\\d+_(\\d+)_\\d+", "\\1") %>% as.numeric(),
                  start = ExtendCREStart -1000 + start,
                  end = ExtendCREStart -1000 + end) %>%
    dplyr::select(ID, start, end, primer_ID) %>%
    group_by(ID, primer_ID) %>%
    dplyr::summarise(start = min(start), end = max(end)) %>%
    dplyr::mutate(seqnames = str_extract(ID, "Chr\\d"),
                  seqnames = str_replace_all(seqnames, "Chr", ""))
  primers_tmpGR <- makeGRangesFromDataFrame(primers_tmp)
  core_pro <- promoters(TxDb.Athaliana.BioMart.plantsmart28, upstream = 1, downstream = 1, use.names = T) #downstream 400: because the largest region amplified by PCR was 1000bp.
  ol <- findOverlaps(primers_tmpGR, core_pro)
  primers_tmp$PcrHitTSS <- "N"
  primers_tmp$PcrHitTSS[unique(ol@from)] <- "Y"
  primers_tmp <- primers_tmp %>%
    dplyr::select(ID, primer_ID, PcrHitTSS)
  primers_all <- left_join(primers_all, primers_tmp)
  table(primers_all$Group, primers_all$PcrHitTSS)
  rm(primers_tmp, ol, core_pro, primers_tmpGR)
  
  # discard fragments with TSS (except Group: ePromoter_TSS)
  primers_all <- primers_all %>%
    dplyr::filter(!(Group != "ePromoter_TSS" & PcrHitTSS == "Y"))
  
  # label primers selected in 20220510
  primers_all <- readxl::read_xlsx("./R2.enhancer_calling/validation_batch2/Selected_primers_for_GUS_20220510.xlsx", sheet = 2) %>%
    dplyr::select(primer_ID, Group, Features, ID) %>%
    dplyr::mutate(select0510 = "Y") %>%
    distinct() %>%
    left_join(primers_all, .)

# summary
  # only specific region can be amplified by the primer pair.
  primers_all %>%
    # dplyr::select(specific, bamh1_type, sma1_type, Features) %>%
    dplyr::filter(RERS_type != "core_region") %>%
    group_by(ID, Features, specific, select0510) %>%
    count() %>%
    group_by(Features, specific, select0510) %>%
    count() %>%
    pivot_wider(names_from = specific, values_from = n, values_fill = 0) %>%
    kable(caption = "Number of regions with sucessfull design primers")
  
# select TE derived enhancers for validation
  primers_TE <- primers_all %>%
    dplyr::filter(Features == "TE",
                  RERS_type == "None",
                  # specific == "Y",
                  str_count(Len_Top10fragments, ";") < 5
                  ) %>%
    dplyr::group_by(ID, primer_ID, TE_SuperFamily) %>%
    dplyr::arrange(TE_SuperFamily, ID, Num_fragments, primer_ID) %>%
    slice_head(n = 2)
  
  primers_TE %>%
    group_by(ID, Features, specific, select0510, TE_SuperFamily) %>%
    count() %>%
    group_by(Features, specific, select0510,TE_SuperFamily) %>%
    count() %>%
    pivot_wider(names_from = specific, values_from = n, values_fill = 0) %>%
    kable(caption = "Number of regions with sucessfull design primers")
  
# write to file
  primers_TE %>%
    dplyr::mutate(type = str_replace_all(type, "orward|everse", ""),
                  PrimerID = str_c(ID, primer_ID, type, sep = "_")) %>%
    arrange(Features, TE_SuperFamily, ID, Num_fragments) %>%
    dplyr::relocate(Group, Features, ID, PrimerID, gene_id, type, TE_SuperFamily, TE_Family, TE) %>%
    fwrite(., file = "./R2.enhancer_calling/validation_batch2/Selected_primers_for_GUS_TE_Enhancer_20220523.csv")
    
```

## 2.8 reported enhancers (NON were found!)
*NO overlap were observed between 45 reported enhancers and cSTARRseq1.3 enhancers (N=4423)!*
*NO overlap were observed between 45 reported enhancers and all cSTARRseq enhancers (N=7623)!*

```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enReported <- fread("./refgenome/At_enhancers_reported_TYJ_collect_20210519.tsv")
  enhancers <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq")
  
  enRepGR <- enReported %>%
    makeGRangesFromDataFrame(seqnames.field = "chr", keep.extra.columns = T)
  enhancersGR <- makeGRangesFromDataFrame(enhancers)
  
  ol <- findOverlaps(enRepGR, enhancersGR)
  ol

```



# 3. STARRseq vs DHS vs H3K4me1

Definition of DHS overlapped with STARR-seq enhancer: the mid point of one region located in other region.
(Identification overlap reciprocally to exclude false negtive caused by varied region width)

## 3.1 Intronic enhancers
Intronic enhancer: 80% of DHS overlaped with a intron.
Criterial used in "Genomic editing of intronic enhancers unveils their role in fine-tuning tissue-specific gene expression in Arabidopsis thaliana.", but the list of all intronic enhancers were not provided.
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")

## DHS
# GSE34318, Zhang et.al, 2012, PC.
dhs <- fread("./refgenome/GSE34318_dhsites_region.bed/GSE34318_dhsites_region.bed", header = F, 
                     sep = "\t", sep2 = "\\S", fill = T) %>%
  dplyr::mutate(label = str_c(V4, V5, V6, V7, sep = "_"),
                V1 = str_replace_all(V1, "Chr", "")) %>%
  dplyr::filter(str_detect(label, "wtleaf")) %>%
  dplyr::select(1:3) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3") %>%
  sort()

# extract introns 
library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28
intron_all <- intronsByTranscript(txdb) %>% 
  unlist() %>%
  GenomicRanges::reduce()

# overlaped regions
dhsIntron <- GenomicRanges::intersect(dhs, intron_all, ignore.strand = T) %>%
  sort()
dhsIntronDF <- as.data.frame(dhsIntron) %>%
  arrange(seqnames, start)
dhsDF <- as.data.frame(dhs) %>%
  arrange(seqnames, start)

# infor of overlaped reigons in each DHS.
ol <- findOverlaps(dhs, dhsIntron)
dhsDF$ol_width <- 0
dhsDF$ol_width[ol@from] <- dhsIntronDF$width[ol@to]
dhsDF$RatioIntron <- dhsDF$ol_width / dhsDF$width

IntronicEnhancer <- dhsDF %>%
  dplyr::filter(RatioIntron > 0.8) # follow criterial used in the article.
hist(IntronicEnhancer$ol_width)

# fwrite(IntronicEnhancer, file = "./refgenome/Intronic_DHS_enhancer_seedling_20210604.csv.gz", 
#        sep = ",", quote = F, row.names = F, col.names = T)

```



## 3.2 Enhancer ol DHS/MHS peaks.

### 3.2.1 DHS/MHS of leaf
DHSs were called by Wanjing using MACS with data download from Zhang et.al, 2012 (14 days leaves).
MHSs were download from GSM4230476 (14 days seedling tissues).

```{r}
source("D:/SUST/code/TanYongjun_code.R")
## DHS and MHS

# call by WanJing (Not used)
# dhs <- fread("./work2_wj/DHS/macs_res_merge/DNase_merged_peaks.xls") %>%
#   makeGRangesFromDataFrame() %>%
#   sort() 

# GSE34318, Zhang et.al, 2012, PC.
dhs <- fread("./refgenome/GSE34318_dhsites_region.bed/GSE34318_dhsites_region.bed", header = F, 
                     sep = "\t", sep2 = "\\S", fill = T) %>%
  dplyr::mutate(label = str_c(V4, V5, V6, V7, sep = "_"),
                V1 = str_replace_all(V1, "Chr", "chr")) %>%
  dplyr::filter(str_detect(label, "wtleaf")) %>%
  dplyr::select(1:3) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3")

mhs <- fread("./work2_wj/MHSs_GSM4230476/GSM4230476_mhs.bed.gz") %>%
  mutate(V1 = str_replace_all(V1, "Chr", "chr")) %>%
  makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3") %>%
  sort()

# width of DHS and MHS
  data.frame(width = c(width(dhs), width(mhs)), 
             Type = rep(c("DHS", "MHS"), c(length(dhs), length(mhs)))) %>%
    
    ggplot(aes(x = width, color = Type, fill = Type)) + 
    geom_histogram(position = "dodge") + 
    toolkit_tyj$theme_my +
    # coord_cartesian(xlim = c(0, 1000)) +
    ggtitle("Distribution of MHSs/DHSs regions width") + 
    scale_x_continuous(breaks = seq(0, 5500, by = 500)) + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.position = c(0.9, 0.8), 
          legend.key.size = unit(0.4, units = "cm")) + 
    scale_fill_npg()
  # toolkit_tyj$SavePlot(filename_prefix = "./R3.enhancer_DistalDHS_H3K4me1/Distribution_of_DHS_MHSs_length",
  #                    width = 8, height = 6, device = "png")

## STARR-seq enhancer
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>%
  dplyr::filter(method == "cSTARR-seq", enrichment > 1.3) %>%
  makeGRangesFromDataFrame(keep.extra.columns = T)
# enhancer <- fread("./work2_wj/4.call_enhancer/oldSTARR1.2.bed") %>%
#   makeGRangesFromDataFrame(keep.extra.columns = T, seqnames.field = "V1", start.field = "V2", end.field = "V3")


## overlap (center)
  # DHS and MHS
  MinOverlapSize <- 100
  ol <- findOverlaps(dhs, mhs)
  ol1 <- findOverlaps(enhancer, dhs, minoverlap = MinOverlapSize)
  ol2 <- findOverlaps(enhancer, mhs, minoverlap = MinOverlapSize)
  
  cat("Number of DHSs overlapped with MHSs was: ", percent(length(unique(ol@from)) / length(dhs)), 
        length(unique(ol@from)), "/ length(dhs); ", length(dhs) - length(unique(ol@from)), "not overlapped with MHSs.\n", 
      "Number of MHSs overlapped with DHSs was: ", percent(length(unique(ol@to)) / 70046), 
        length(unique(ol@to)), "/ 70046; ", 70046 - length(unique(ol@to)), " not overlapped with DHSs.")
  
  cat("Number of enhancers overlapped with MHSs was: ", percent(length(unique(ol2@from)) / 4443), 
        length(unique(ol@from)), "/ 4443\n", 
      "Number of enhancers overlapped with DHSs was: ", percent(length(unique(ol1@from)) / 4443), 
        length(unique(ol@to)), "/ 4443\n")
  enhancer <- as.data.frame(enhancer) %>%
    mutate(DHS = "N", MHS = "N")
  enhancer$DHS[unique(ol1@from)] <- "Y"
  enhancer$MHS[unique(ol2@from)] <- "Y"
  
  # summary of enhancers overlppad with DHSs.
  table(ifelse(enhancer$DHS == "Y", "DHS", "noDHS")) %>%
    as.data.frame() %>%
    ggplot(aes(x = Var1, y = Freq, fill = Var1))+ 
      geom_col() +
      toolkit_tyj$theme_my + 
      scale_fill_npg() + 
      theme(legend.position = "none") + 
      geom_text(aes(x = Var1, y = Freq/2, label = Freq), size = 2) +
      ggtitle("Number of enhancers overlapped with DHSs", subtitle = paste("Overlap size was ", MinOverlapSize, "bp")) + 
      xlab("Type")
  # toolkit_tyj$SavePlot(filename_prefix = paste("./R3.enhancer_DistalDHS_H3K4me1/Number_of_cSTARRseq_enhancers1.3_ol_with_DHS_", 
  #                                            MinOverlapSize, "bp", sep = ""),
  #                    width = 8, height = 6, device = "png")

    
  # summary of enhancers overlapped with DHS/MHS
  df <- table(paste(ifelse(enhancer$DHS == "Y", "DHS", "Null"), 
              ifelse(enhancer$MHS == "Y", "MHS", "Null"), sep = "-")) %>%
    as.data.frame() %>%
    dplyr::rename(Combn = Var1, Number = Freq)
  df %>%
    ggplot(aes(x = Combn, y = Number, fill = Combn)) + 
      geom_col() +
      toolkit_tyj$theme_my + 
      scale_fill_npg() + 
      theme(legend.position = "none") + 
      geom_text(aes(x = Combn, y = Number/2, label = Number), size = 2) +
      ggtitle("Number of enhancers overlapped with \nDHSs or MHSs", subtitle = paste("Overlap size was ", MinOverlapSize, "bp")) + 
      xlab("Type")
  # toolkit_tyj$SavePlot(filename_prefix = paste("./R3.enhancer_DistalDHS_H3K4me1/Number_of_cSTARRseq_enhancers1.3_ol_with_DHS_or_MHS_", 
  #                                              MinOverlapSize, "bp", sep = ""),
  #                      width = 8, height = 6, device = "png")


```

### 3.2.2 Tissues (DHS)

#### .1 Findoverlap
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")

## load DHS
  # GSE34318, Zhang et.al, 2012, PC.
  dhs <- fread("./refgenome/GSE34318_dhsites_Zhang2012/GSE34318_dhsites_region.bed", header = F, 
                       sep = "\t", sep2 = "\\S", fill = T) %>%
    dplyr::mutate(label = str_c(V4, V5, V6, V7, sep = "_"),
                  V1 = str_replace_all(V1, "Chr", "chr")) %>%
    dplyr::rename(seqnames = V1, start = V2, end = V3)
  
  dhs_leafZhu <- dhs %>%
    dplyr::filter(str_detect(label, "wtleaf")) %>%
    dplyr::select(1:3) %>%
    dplyr::mutate(ID = "leaf_Zhang2012")
  
  dhs_flowerZhu <- dhs %>%
    dplyr::filter(str_detect(label, "wtflower")) %>%
    dplyr::select(1:3) %>%
    dplyr::mutate(ID = "flower_Zhang2012")
  
  # GSM1139913, Pajoro et.al, 2014, Genome Biology.
  dhs <- NULL
  for (i in list.files(path = "./refgenome/DHS_flower_states_GenomeBiology2014/", 
                       pattern = "")) {
    cat(i, "   ")
    dhs <- fread(str_c("./refgenome/DHS_flower_states_GenomeBiology2014/", i, sep = "")) %>%
      dplyr::mutate(ID = str_replace_all(i, "GSM11.+_DNase_I_(.+).txt.gz", "\\1")) %>%
      dplyr::mutate(ID = str_c("flower_", ID, "_Pajoro2014"), 
                    chr = str_c("chr", chr, sep = "")) %>%
      dplyr::select(chr, `peak start`, `peak end`, "ID") %>%
      dplyr::filter(!is.na(chr)) %>%
      dplyr::rename(seqnames = chr, start = `peak start`, end = `peak end`) %>%
      rbind(., dhs)
  }
  
  dhsAll <- rbind(dhs, dhs_leafZhu, dhs_flowerZhu)
  rm(dhs, dhs_leafZhu, dhs_flowerZhu, i)
  table(dhsAll$ID)
  dhsAllGR <- dhsAll %>%
    na.omit() %>%
    as.data.frame() %>%
    base::split(., .$ID) %>%
    purrr::map(makeGRangesFromDataFrame)
  
  # dhsAllGR %>%
  #   map(., width) %>%
  #   map_dfr(., quantile, probs = seq(0, 1, 0.1), .id = "DHS datasets") %>%
  #   fwrite(., file = "./R3.enhancer_DistalDHS_H3K4me1/Width_quantile_of_DHSs_At_flower_leaf.csv")
    
  ## bed file for GAT
  
  # dhsAll %>%
  #   fwrite("./R3.enhancer_DistalDHS_H3K4me1/GAT_EnhancerSTARRseq_DHS/dhs_all_for_GAT_20210609.bed.gz",
  #          col.names = F, row.names = F, sep = "\t")
  
  # for (i in unique(dhsAll$ID)) {
  #   dhsAll %>%
  #     dplyr::filter(ID == i) %>%
  #     dplyr::select(1:3) %>%
  #     fwrite(file = str_c("./R3.enhancer_DistalDHS_H3K4me1/GAT_EnhancerSTARRseq_DHS/DHS_",i, ".bed", sep = ""), 
  #            col.names = F, row.names = F, sep = "\t")
  # }
  
## enhancer
  enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>%
    dplyr::filter(method == "cSTARR-seq", enrichment > 1.3, seqnames != "chrM")
  enhancerGR <- makeGRangesFromDataFrame(enhancer)
  
  # bed file for GAT
  # enhancer %>%
  #   dplyr::select(seqnames:end, method) %>%
  #   fwrite("./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT_DHS_STARRseq/Enhancer_STARRseq_1.3_for_GAT_20210609.bed.gz", 
  #          col.names = F, row.names = F, sep = "\t")
  
## Count overlap
  # length of DHSs
  map_df(dhsAllGR, function(x){quantile(width(x), probs = seq(0, 1, 0.1))}, .id = "name") %>%
    kable(caption = "Width of ACRs")
  
  for (overLen in c(1, 20, 100, 300)) {
    cat(str_c(">>> Overlap length ", overLen, "\n", sep = ""))
    map_dfr(dhsAllGR, 
            function(x){
                        # ol <- findOverlaps(GenomicRanges::resize(enhancerGR, width = 200, fix = "center"),
                        #                    x, minoverlap = overLen)
                        ol <- findOverlaps(enhancerGR,
                                           x, minoverlap = overLen)
                        # ol <- findOverlaps(enhancerGR, 
                        #                    GenomicRanges::resize(x, width = 1, fix = "center"), minoverlap = overLen)
                        NumEnInDHS <- length(unique(ol@from))
                        NumEn <- length(enhancerGR)
                        return(data.frame(NumEn = NumEn, 
                                          NumEnInDHS = NumEnInDHS, 
                                          Percent = NumEnInDHS / NumEn))
            }, 
      .id = "datasets") %>%
      dplyr::mutate(label = str_c(percent(Percent, accuracy = 0.01), " (", NumEnInDHS, "/", NumEn, ")", sep = "")) %>%
      ggplot(aes(y = datasets, x = Percent, fill = datasets)) + 
        geom_col() + 
        scale_fill_npg() + 
        theme(legend.position = "none") + 
        toolkit_tyj$theme_my + 
        geom_text(aes(x = Percent / 2, label = label), size = 1.5) + 
        # labs(caption = str_c("overlap ", overLen, "bp", sep = "")) + 
        ylab("DHS datasets") + 
        scale_x_continuous(labels = scales::percent) +
        xlab("Percentage of enhancers overlapped with DHSs")
    
    toolkit_tyj$SavePlot(filename_prefix = paste("./R3.enhancer_DistalDHS_H3K4me1/Percent_of_Enhancer_STARRseq_500bp_in_DHS_OverLap_length_",
                                                 overLen, "bp", sep = ""),
                         width = 8, height = 6, device = "png")
}
  
## mark enhancers overlapped with DHSs (Leaf, zhang, 2012)
  ol <- findOverlaps(enhancerGR, dhsAllGR$leaf_Zhang2012)
  unique(ol@from) %>% length()
  enhancer$DHS_leaf_Zhang <- "no_DHS"
  enhancer$DHS_leaf_Zhang[unique(ol@from)] <- "DHS"
  table(enhancer$DHS_leaf_Zhang)
  
  enhancer %>%
    dplyr::mutate(seqnames = str_replace_all(seqnames, "chr", ""),
                  CRE_id = str_c(seqnames, start, end, sep = "_")) %>%
    dplyr::select(seqnames:end, CRE_id, DHS_leaf_Zhang) %>%
    fwrite(., file = "./R3.enhancer_DistalDHS_H3K4me1/enhancer_STARRseq_4327_hit_DHS_leaf_zhang2012.csv")
  NA  
```

#### .2 GAT
`/scratch/2021-06-03/bioTanyj/At_enhancer/scripts/3.2.2_GAT_EnhancerSTARRseq_DHS_overlap.sh`

*Plot*
```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./R3.enhancer_DistalDHS_H3K4me1/GAT_EnhancerSTARRseq_DHS/GAT_results_DHS_overlap_Enhancer-1_20210609.tsv")

# plot
  # enrichment, fold change
  df %>%
    dplyr::mutate(SigLab = toolkit_tyj$returnAsterisk(qvalue), 
                  label = str_c(percent(percent_overlap_size_track/100, accuracy = 0.01), "\n", SigLab)) %>%
    ggplot(aes(y = annotation, x = l2fold)) + 
      geom_col(aes(fill = annotation)) + 
      toolkit_tyj$theme_my + 
      scale_fill_npg() + 
      geom_text(aes(x = l2fold/2, y = annotation, label = label), size = 2) + 
      theme(legend.position = "none")
  
  # toolkit_tyj$SavePlot("./R3.enhancer_DistalDHS_H3K4me1/GAT_EnhancerSTARRseq_DHS/Enrichment_Enhancer_STARRseq_in_DHS_20210609", width = 8, height = 6)
    
  # percent of CRE overlapped with DHS (observed and expected)
  tmp <- df %>%
    dplyr::select(annotation, observed, expected, fold, qvalue, track_size) %>%
    pivot_longer(cols = observed:expected) %>%
    dplyr::mutate(Percent = value / track_size,
                  annotation = str_replace_all(annotation, c("no_treat" = "notreat",
                                                              "_Pajoro2014" = "\nPajoro2014",
                                                             "_Zhang" = "\nZhang")), 
                  name = factor(name, levels = c("observed", "expected")))
  tmpSigLabel <- tmp %>%
    dplyr::mutate(SigLabel = toolkit_tyj$returnAsterisk(qvalue), 
                  fold = round(fold, digits = 2), 
                  label = str_c("FC=", fold, "\n", SigLabel)) %>%
    dplyr::select(annotation, label) %>%
    unique() %>%
    dplyr::mutate(start = "expected", end = "observed", 
                  y = 0.19)
  

  ggplot(tmp, aes(x = name, y = Percent, fill = name)) + 
    geom_col() + 
    geom_signif(data = tmpSigLabel, 
                aes(xmin = start, xmax = end,
                    annotations = label, y_position = y), 
                manual = T, inherit.aes = F, 
                color = "grey35", textsize = 6/.pt, 
                size = 0.2) +
    facet_grid(. ~ annotation) + 
    toolkit_tyj$theme_my + 
    scale_fill_npg() + 
    scale_y_continuous(labels = scales::percent, 
                       expand = expansion(mult = c(0,0), add = c(0, 0.025))) + 
    theme(legend.key.size = unit(0.3, units = "cm"), 
          legend.position = "none", 
          legend.margin = margin(-5, 0, 0, 0), 
          axis.title.x = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), panel.spacing.x = unit(0.05, "cm")) + 
    ylab("Percent of enhancer located in chromatin accessible regions") + 
    geom_text(aes(label = percent(Percent, accuracy = 0.01)), angle = 90, 
              hjust = 1.2, color = "grey25", size = 6/.pt)
  
  toolkit_tyj$SavePlot("./R3.enhancer_DistalDHS_H3K4me1/GAT_EnhancerSTARRseq_DHS/Percent_Enhancer_STARRseq_in_DHS_20210609", width = 10, height = 8)

```


## 3.3 Overlap (STARRseq, DHS, H3K4me1)

### .DHSs-enhancer

Methods: 1.
DHSs located 1.5kb away from TSS (intergenic region).
A total of 7515 identified putative leaf enhancers were identified.
-------Zhu et.al, 2015, Plant Cell.
2.
DHSs located 1kb away from TSS.
these DHSs can be classified into two categries: intergenic and intragenic enhancers.
Yan et.al use this strategy to identify enhancers in function in flower morphogenesis of A.t.
Yan et.al, 2019, Nat Comm.

```{r}
source("D:/SUST/code/TanYongjun_code.R")
dhs_enhancer <- NULL
for (i in list.files(path = "./R3.enhancer_DistalDHS_H3K4me1/", pattern = "DHS_enhancer.+2015_PC.csv")) {
  temp <- fread(paste("./R3.enhancer_DistalDHS_H3K4me1/", i, sep = ""), skip = 3) %>%
    mutate(Type = str_replace_all(i, "_Zhu_2015.+", ""))
  dhs_enhancer <- rbind(dhs_enhancer, temp)
}
dhs_enhancer$method <- "Enhancer_Zhu2015"
dhs_enhancer$width = dhs_enhancer$End - dhs_enhancer$Start
dhs_enhancer <- dhs_enhancer[, c(1:3, 7:9)]
names(dhs_enhancer) <- c("seqnames", "start", "end",  "Type","method", "width")
dhs_enhancer$seqnames <- str_replace_all(dhs_enhancer$seqnames, "Chr", "chr")
```

### .H3K4me1-enhancer

Methods: 1.
H3K4me1 peaks 2.
DHSs + Ratio of H3K4me1 and H3K4me3 \>2.
A total of 1919 putative enhancers were identified by Wang et.al, 2019, RNA.

```{r}
H3K4_enhancer <- fread("./R3.enhancer_DistalDHS_H3K4me1/DHS_H3K4_enhancer_Wang_2019RNA_GSE99406_RAW/GSM2643596_PE.forward_reverse.pA.RRP41_1m.bed") %>%
  dplyr::select(1:3) %>%
  dplyr::mutate(Type = "PE", method = "Enhancer_Wang2019")
names(H3K4_enhancer)[1:3] <- c("seqnames", "start", "end")
H3K4_enhancer$width <- H3K4_enhancer$end  - H3K4_enhancer$start
table(H3K4_enhancer$seqnames)
```

### .merge

```{r}
enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv") %>%
  dplyr::filter(method == "cSTARR-seq" & enrichment > 1.3) %>%
  full_join(., dhs_enhancer) %>%
  full_join(., H3K4_enhancer) %>%
  dplyr::filter(Type != "DHS_enhancer_flower_specific") %>%
  mutate(strand = "*")

## width of enhancers
enhancer %>%
  dplyr::mutate(width = end - start) %>%
  ggplot(aes(x = width)) + 
    geom_histogram(aes(fill = method), position = "dodge")

## number
table(enhancer$method) %>%
  as.data.frame() %>%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
    geom_col() + 
    scale_fill_npg() + 
    geom_text(aes(x = Var1, y = Freq/2, label = Freq)) + 
    toolkit_tyj$theme_my + 
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          axis.title.y  = element_blank()) + 
    coord_flip() + 
    xlab("Number") +
    ggtitle("Number of identified enhancers") + 
    labs(caption = "PE : putative enhancers, DHS + (H3K4me1/H3K4me3 > 2)")
# toolkit_tyj$SavePlot(filename_prefix = "./R3.enhancer_DistalDHS_H3K4me1/Number_of_enhancer_STARRseq_DHS_PE_20210311", 
                     # width = 12, height = 8)
# fwrite(enhancer, file = "./Enahncers_DHS_cSTARRseq1.3_H3K4_20210410.csv")
```

### .plot (iSTARR, cSTARR)

*The total number of enhancer used in Upset plot was not equal to the sum of all kind of enhancers.* Because merged regions of enhancers were used as unique enhancer list to generate overlap table.

```{r, iSTARR-seq & cSTARRseq}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread(file = "./Enahncers_DHS_STARRseq_H3K4_20210311.csv") %>%
  dplyr::mutate(ID = str_c(method, seqnames, start, sep = "_"))
enhancerGR <- enhancer %>%
  split(.$method) %>%
  map(GenomicRanges::makeGRangesFromDataFrame, keep.extra.columns = T)


## Count overlap between enhancers identified by varied methods.
df <- combn(names(enhancerGR), m = 2) %>% t() %>% as.data.frame()
df$Num_1 <- df$Num_2 <- df$Num_ov <- NULL
for (i in 1:nrow(df)) {
  n1 <- df$V1[i]
  n2 <- df$V2[i]
  cat(">>>>  ", n1, "<=>", n2, ";\n")
  tmp1 <- enhancerGR[[n1]]
  tmp2 <- enhancerGR[[n2]]
  ol <- findOverlaps(tmp1, tmp2, minoverlap = 300)
  
  # count
  df$Num_1[i] <- length(tmp1)
  df$Num_2[i] <- length(tmp2)
  df$Num_ov[i] <- length(unique(ol@from)) # number of overlapped regions
}
df$Percent_1 <- df$Num_ov / df$Num_1
df$Percent_2 <- df$Num_ov / df$Num_2

# fwrite(df, file = "./R3.enhancer_DistalDHS_H3K4me1/Overlaps_enhancers_STARR_DHS_H3K4.csv")


## the generate overlap information table: row for ehancer; column for method; value 1 and 0 represent overlapped or not.
# prepare table for overlap information
over_table <- enhancer %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce() %>%
  as.data.frame() %>%
  mutate(`cSTARR-seq` = 0, `iSTARR-seq` = 0, Enhancer_Zhu2015 = 0, Enhancer_Wang2019 = 0)
over_tableGR <- makeGRangesFromDataFrame(over_table)
for (i in 1:length(enhancerGR)) {
  id_tmp <- names(enhancerGR)[i]
  ol <- findOverlaps(over_tableGR, 
                     GenomicRanges::resize(enhancerGR[[i]], width = 1, fix = "center"))
  over_table[unique(ol@from),
             str_detect(colnames(over_table), id_tmp)] <- 1
}

# upsetR
# discard oerlapped enhancers which were repeatedly count in the table.
library(UpSetR)
require(ggplot2); require(plyr); require(gridExtra); require(grid);

colSums(over_table[,6:9])
png("./R3.enhancer_DistalDHS_H3K4me1/UpsetPlot_enhancers_overlap.png", width = 12, height = 8, units = "cm", res = 600)
upset(over_table, 
      sets = unique(enhancer$method), 
      order.by = c("freq", "degree"), 
      decreasing = c(TRUE,FALSE), 
      queries = list(list(query = intersects, params = list("iSTARR-seq", "cSTARR-seq"), 
                          color = pal_npg()(1), active = T)))
dev.off() # run in R console!

# regions hit by varied kind of enhancers.
# over_table$Sum = apply(over_table[,6:9], 1, sum)

```

### . upset plot (cSTARR 1.3)
```{r, cSTARRseq 1.3}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread(file = "./Enahncers_DHS_cSTARRseq1.3_H3K4_20210410.csv") %>%
  dplyr::mutate(ID = str_c(method, seqnames, start, sep = "_"))
enhancerGR <- enhancer %>%
  split(.$method) %>%
  map(GenomicRanges::makeGRangesFromDataFrame, keep.extra.columns = T)

## Count overlap between enhancers identified by varied methods.
df <- combn(names(enhancerGR), m = 2) %>% t() %>% as.data.frame()
df$Num_1 <- df$Num_2 <- df$Num_ov <- NULL
for (i in 1:nrow(df)) {
  n1 <- df$V1[i]
  n2 <- df$V2[i]
  cat(">>>>  ", n1, "<=>", n2, ";\n")
  tmp1 <- enhancerGR[[n1]]
  tmp2 <- enhancerGR[[n2]]
  ol <- findOverlaps(tmp1, tmp2, minoverlap = 300)
  
  # count
  df$Num_1[i] <- length(tmp1)
  df$Num_2[i] <- length(tmp2)
  df$Num_ov[i] <- length(unique(ol@from)) # number of overlapped regions
}
df$Percent_1 <- df$Num_ov / df$Num_1
df$Percent_2 <- df$Num_ov / df$Num_2

fwrite(df, file = "./R3.enhancer_DistalDHS_H3K4me1/Overlaps_300bp_cenhancers1.3_STARR_DHS_H3K4.csv")


## the generate overlap information table: row for enhancer; column for method; value 1 and 0 represent overlapped or not.
# prepare table for overlap information
over_table <- enhancer %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce() %>%
  as.data.frame() %>%
  mutate(`cSTARR-seq` = 0, Enhancer_Zhu2015 = 0, Enhancer_Wang2019 = 0)
over_tableGR <- makeGRangesFromDataFrame(over_table)
for (i in 1:length(enhancerGR)) {
  id_tmp <- names(enhancerGR)[i]
  ol <- findOverlaps(over_tableGR, 
                     GenomicRanges::resize(enhancerGR[[i]], width = 1, fix = "center"))
  over_table[unique(ol@from),
             str_detect(colnames(over_table), id_tmp)] <- 1
}

# upsetR
# discard oerlapped enhancers which were repeatedly count in the table.
library(UpSetR)
require(ggplot2); require(plyr); require(gridExtra); require(grid);

colSums(over_table[,6:9])
png("./R3.enhancer_DistalDHS_H3K4me1/UpsetPlot_enhancers_cSTARRseq1.3_overlap.png", width = 12, height = 8, units = "cm", res = 600)
upset(over_table, 
      sets = unique(enhancer$method), 
      order.by = c("freq", "degree"), 
      decreasing = c(TRUE,FALSE))
dev.off() # run in R console!

# regions hit by varied kind of enhancers.
# over_table$Sum = apply(over_table[,6:9], 1, sum)

```

### .GAT

Did overlap between enhancers identified by different methods higher than expectation?
Run with `GAT`.

##### .1 202104
```{r}
##################################################################################
#  Did overlap between enhancers identified by different methods higher than expectation?
#     perform permutation with GAT.
#                         Tanyongjun 20210418
##################################################################################


setwd("/scratch/2021-04-10/bioTanyj/At_enhancer/R3.STARR_DHS_H3K4me1")
source("/work/bio-tanyj/soft/TanYongjun_code.R")
cre_all <- fread("./R5.chromatin_states/CRE_6column_bed_for_GAT_peak_overlap_20210416.bed.gz") %>%
  dplyr::filter(str_detect(V4, "(cSTARR-seq|Enhancer|Random)"))
table(cre_all$V4)

# to bed file
for (i in unique(cre_all$V4)) {
  cre_all %>%
    dplyr::filter(V4 == i) %>%
    fwrite(paste(i, ".bed", sep = ""), sep = "\t", col.names = F)
}

# perform GAT
Num_threads <- 40
Num_sample <- "100000"

df <- combn(unique(cre_all$V4), 2) %>%
  t() %>%
  as.data.frame()

for (i in seq_len(nrow(df))) {
  pair_id <- str_c(df[i,], collapse = "_vs_")
  if(!file.exists(paste(pair_id, ".tsv", sep = "")) &
    !file.exists(paste(pair_id, ".analyzing", sep = ""))){
      write(NA, paste(pair_id, ".analyzing", sep = ""))
      cat("   ", pair_id, "; ", format(Sys.time()), "\n")

      # run GAT
      system(paste("/work/bio-tanyj/miniconda3/envs/peaks_overlap/bin/gat-run.py ",
                  " --segment-file=", "./", df$V1[i],
                  " --workspace-file=/scratch/2021-04-10/bioTanyj/At_enhancer/refgenome/TAIR10_chr_size_noCM_forGAT.tsv ",
                  " --annotation-file=", "./", df$V2[i], 
                  " -t ", Num_threads, " --random-seed=123456 ",
                  " --num-samples=", Num_sample,
                  " --log=all_logs.log ",
                  " > ", pair_id, ".tsv", sep = ""))

      # remove lock file
      if(file.exists(paste(pair_id, ".tsv", sep = "")) &
        file.exists(paste(pair_id, ".analyzing", sep = ""))){
        system(paste("rm ", pair_id, ".analyzing", sep = ""))
      } # if
    } # if
}

```

Collect results

```{r}
source("D:/SUST/code/TanYongjun_code.R")
df <- NULL
for (i in list.files(path = "./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT/", 
                     pattern = ".+\\.tsv")) {
  df <- fread(str_c("./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT/", i, sep = "")) %>%
    dplyr::mutate(track = i) %>%
    rbind(df, .)
}
df <- df %>%
  dplyr::mutate(CRE1 = str_replace_all(track, "(.+)_vs_(.+).tsv", "\\1"), 
                CRE2 = str_replace_all(track, "(.+)_vs_(.+).tsv", "\\2"))
df %>%
  fwrite(file = "./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT/Overlap_summary.csv", 
         sep = ",", quote = F)
```

##### .2 202106
`/scratch/2021-06-03/bioTanyj/At_enhancer/scripts/3.3_overlap_enhancers_GAT.R`
`/scratch/2021-06-03/bioTanyj/At_enhancer/scripts/3.3_overlap_enhancers_GAT.sh`

Collect results

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
df <- NULL
for (i in list.files(path = "./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT_Enhancer/", 
                     pattern = ".+\\.tsv")) {
  df <- fread(str_c("./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT_Enhancer/", i, sep = "")) %>%
    dplyr::mutate(track = i) %>%
    rbind(df, .)
}
df <- df %>%
  dplyr::mutate(CRE1 = str_replace_all(track, "(.+)_vs_(.+).tsv", "\\1"), 
                CRE2 = str_replace_all(track, "(.+)_vs_(.+).tsv", "\\2"),
                qvalue = p.adjust(pvalue, "BH"))
df %>%
  fwrite(file = "./R3.enhancer_DistalDHS_H3K4me1/overlap_GAT_Enhancer/Overlap_summary.csv", 
         sep = ",", quote = F)
```
# 4. Distribution

## 4.1 XXX Genomic features (txdb-based)
*Attention:*
Percent of CRE located on the gene body or not was maily affected by the total number of genes in TxDb object. Obvious difference was observed when used GFF3 download from ENSEMBL and MSU.

### .0 region size 
```{r}
dfTair10 <- fread("./refgenome/TAIR10_GFF3_genes_transposons.gff.gz", sep = "\t")
dfTair10$V6 <- dfTair10$V5 - dfTair10$V4
unique(dfTair10$V3)
dfTair10 %>%
  dplyr::filter(V3 != "chromosome") %>%
  ggplot(aes(y = V3, x = V6)) + 
    # geom_violin() + 
    geom_boxplot() + 
    scale_x_log10()

dfTair10 %>%
  group_by(V3) %>%
  summarise(Num = n(), TotalLen = sum(V6) / 1e6, MedianLen = median(V6), MeanLen = mean(V6))
```


### .1 each type as a whole

#### .202105
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(ChIPseeker)

# parameters
use_peak <- T
# enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv")
enhancer <- fread("Enahncers_DHS_STARRseq_H3K4_20210311.csv") %>%
  dplyr::filter(!(method == "cSTARR-seq" & enrichment <= 1.3))
table(enhancer$method)
library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28


### annotation and plot
# unique(cis_elements$ID)
anno_list <- list()
anno_stats <- NULL
enhancer$ID <- enhancer$method
enhancer$seqnames <- str_replace_all(enhancer$seqnames, "chr", "")
enhancer$seqnames <- str_replace_all(enhancer$seqnames, "M", "Mt")
enhancer$strand <- "*"

# add random selected region in list
set.seed(12345)
random <- enhancer %>%
  makeGRangesFromDataFrame(keep.extra.columns = T) %>%
  toolkit_tyj$shuffle(TxDb = txdb) %>%
  as.data.frame() %>%
  mutate(ID = "Random", Type = "Random", method = "Random")
enhancer <- full_join(enhancer, random)

# add PAS in list
pas_HQ <- fread("./refgenome/PAC/arabidopsis_thaliana.high_confidence.PAC.add2bp.bed", header = F) %>%
  dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
  dplyr::rename(seqnames = V1, start = V2, end = V3, width = V4, strand = V5) %>%
  mutate(method = "PAS", Type = "PAS", ID = "PAS", Strategy = "PAS",
         start = as.integer(start), end = as.integer(end), width = end - start)
enhancer <- full_join(enhancer, pas_HQ)

if (use_peak) {
  enhancer$startBK <- enhancer$start
  enhancer$endBK <- enhancer$end
  enhancer$start <- floor((enhancer$startBK + enhancer$endBK) / 2) 
  enhancer$end <- ceiling((enhancer$startBK + enhancer$endBK) / 2)
}


## annotation
for (i in unique(enhancer$ID)) {
  cat(">>> ", i, "\n")
  cis1 <- enhancer %>% filter(ID == i)
  cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
  options(ChIPseeker.downstreamDistance = 1000) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
  options(verbose = F)
  peakanno <- annotatePeak(cis1_gr, 
                           tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                           TxDb = txdb, verbose = F, overlap = "all")
  
  # collect results
    temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
    anno_list <- c(anno_list, peakanno)
    names(anno_list)[length(anno_list)] <- i
    
    temp_stats$ID <- i
    anno_stats <- rbind(anno_stats, temp_stats)
    
  # pie plot
    # p <- temp_stats %>%
    #   mutate(label = paste(Feature, " (", round(Rate * 100, digits = 2), 
    #                        "%)", sep = "")) %>%
    #   mutate(label = factor(label, levels = label)) %>%
    #   ggplot(aes(x = "", y = Rate, fill = label)) + 
    #   geom_bar(stat = "identity", color = "black") + 
    #   coord_polar("y", start = 0, direction = -1) + toolkit_tyj$theme_my + 
    #   theme(legend.title = element_blank(), axis.text = element_blank(), 
    #         axis.title = element_blank(), axis.ticks = element_blank(), 
    #         panel.border = element_rect(color = "transparent"), 
    #         legend.key.size = unit(0.5, units = "cm")) + 
    #   scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
    #                                brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
    #   ggtitle(i)
    # # print(p)
    # toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/Enhancer_distribution_",
    #                                              i, sep = ""),
    #                      plot = p, device = "png", width = 16, height = 10)
    
  # pie plot with function from ChipSeeker
    # png(filename = paste("./R4.Distribution_of_enhancers/Enhancer_distribution_Pie_up2000_", i, ".png",sep = ""),
    #     width = 16, height = 10, res = 600, bg = "transparent", units = "cm")
    #   plotAnnoPie(peakanno)
    # dev.off()
    
}
rm(p, i, cis1, cis1_gr, temp_stats, peakanno)

# save.image(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210525.bin")
```

```{r plot and summary}
  rm(list = ls())
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210525.bin")
  # load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210415.bin")
  # load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up1500_down0_iSTARR_cSTARR_20210508.bin")
  source("D:/SUST/code/TanYongjun_code.R")
  ## merge Downstream(>1bp) into Distal Intergenic.
  anno_stats <- anno_stats %>%
    dplyr::mutate(Feature = as.character(Feature)) %>%
    dplyr::mutate(Feature = str_replace_all(Feature, 
                                            c("Downstream \\(1\\-2kb\\)" = "Distal Intergenic" ,
                                              "Downstream \\(2\\-3kb\\)" = "Distal Intergenic"))) %>%
    dplyr::mutate(Feature = factor(Feature, levels = c("Promoter (2-3kb)", 
                                                        "Promoter (1-2kb)",
                                                        "Promoter (<=1kb)",
                                                        "Promoter",
                                                        "5' UTR",
                                                        "1'st exon",
                                                        "1'st intron",
                                                        "Other exon",
                                                        "Other intron",
                                                        "Last intron",
                                                        "Last exon",
                                                        "3' UTR",
                                                        "Downstream (<1kb)",
                                                        "Distal Intergenic"))) %>%
    group_by(Feature, ID) %>%
    dplyr::summarise(Frequency = sum(Frequency), Rate = sum(Rate))
  
  ## write to tsv file
  anno_stats %>%
    dplyr::mutate(Species = "Arabidopsis thaliana") %>%
    fwrite(str_c("./R4.Distribution_of_enhancers/A.thaliana_Enhancer",
           "_annostates_up500_down0_STARR_20210526.tsv", sep = ""), 
           quote = F, col.names = T, row.names = F, sep = "\t")
  
  ## bar plot (all)
    anno_stats %>%
      dplyr::filter(ID != "iSTARRseq-common") %>%
      mutate(ID = str_replace_all(ID, ".+common", "Common")) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.5, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210525",
                         width = 8, height = 8)

  ## bar plot 20210329
    anno_stats %>%
      dplyr::filter(ID != "iSTARR-seq", ID != "PAS") %>%
      mutate(ID = str_replace_all(ID, c("cSTARR-seq" = "Enhancer_STARRseq"))) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.5, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR_20210525",
                         width = 8, height = 8)
    
  ## enrichment profile (Enhancer / Random) 20210508
    anno_stats %>%
      dplyr::filter(ID != "iSTARR-seq", ID != "PAS") %>%
      mutate(ID = str_replace_all(ID, c("cSTARR-seq" = "Enhancer_STARRseq"))) %>%
      pivot_wider(id_cols = Feature, values_from = Rate, names_from = ID) %>%
      dplyr::mutate(Enhancer_STARRseq = Enhancer_STARRseq / Random, 
                    Enhancer_Zhu2015 = Enhancer_Zhu2015 / Random,
                    Enhancer_Wang2019 = Enhancer_Wang2019 / Random) %>%
      dplyr::select(-Random) %>%
      pivot_longer(cols = Enhancer_STARRseq:Enhancer_Zhu2015, names_to = "CRE", values_to = "Enrichment") %>%
      dplyr::mutate(Enrichment = log2(Enrichment)) %>%
      ggplot(aes(x = Feature, y = Enrichment, fill = CRE)) + 
        geom_col(position = position_dodge(), width = 0.7) + 
        toolkit_tyj$theme_my + 
        scale_fill_npg() + 
        # facet_grid(. ~ Feature)
        theme(legend.position = "bottom", 
              axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
              legend.key.size = unit(0.3, units = "cm")) + 
        guides(fill = guide_legend(title = NULL, nrow = 1))
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR_enrichment_20210525",
                         width = 8, height = 6)
    
  ## summary table
    anno_stats %>%
      mutate(value = paste(Frequency, " (", round(Rate * 100, 4), "%)", sep = "")) %>%
      dplyr::select(-Frequency, -Rate) %>%
      pivot_wider(id_cols = Feature, names_from = ID, values_from = value) %T>%
      write.csv(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210525.csv",
                quote = F) %>%
      kable()
      
```

#### .202106 Add IntronicEnhancer
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(ChIPseeker)
library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28

# parameters
use_peak <- T

# enhancer <- fread("./R2.enhancer_calling/Enhancers_merged_strategy1_1.2_20210307.csv")
enhancer <- fread("Enahncers_DHS_STARRseq_H3K4_20210311.csv") %>%
  dplyr::filter(!(method == "cSTARR-seq" & enrichment <= 1.3), seqnames != "chrM")
table(enhancer$ID)

### annotation and plot
# unique(cis_elements$ID)
anno_list <- list()
anno_stats <- NULL
enhancer$ID <- enhancer$method
enhancer$seqnames <- str_replace_all(enhancer$seqnames, "chr", "")
enhancer$seqnames <- str_replace_all(enhancer$seqnames, "M", "Mt")
enhancer$strand <- "*"

# add random selected region in list
set.seed(12345)
random <- enhancer %>%
  head(n = 4000) %>%
  makeGRangesFromDataFrame(keep.extra.columns = T) %>%
  toolkit_tyj$shuffle(TxDb = txdb) %>%
  GenomicRanges::resize(width = 600, fix = "center") %>%
  GenomicRanges::reduce() %>%
  as.data.frame() %>%
  dplyr::filter(width == 600) %>%
  mutate(ID = "Random", Type = "Random", method = "Random")
enhancer <- full_join(enhancer, random)

# add Intronic enhancer
IntronicEnhancer <- fread("./refgenome/Intronic_DHS_enhancer_seedling_20210604.csv.gz") %>%
  dplyr::select(seqnames:end) %>%
  dplyr::mutate(method = "IntronicEnhancer", 
                ID = "IntronicEnhancer", 
                seqnames = as.character(seqnames))
enhancer <- full_join(enhancer, IntronicEnhancer)
table(enhancer$method)

# add PAS in list
pas_HQ <- fread("./refgenome/PAC/arabidopsis_thaliana.high_confidence.PAC.add2bp.bed", header = F) %>%
  dplyr::filter(V1 != "Mt", V1 != "Pt") %>%
  dplyr::rename(seqnames = V1, start = V2, end = V3, width = V4, strand = V5) %>%
  mutate(method = "PAS", Type = "PAS", ID = "PAS", Strategy = "PAS",
         start = as.integer(start), end = as.integer(end), width = end - start)
enhancer <- full_join(enhancer, pas_HQ)

if (use_peak) {
  enhancer$startBK <- enhancer$start
  enhancer$endBK <- enhancer$end
  enhancer$start <- floor((enhancer$startBK + enhancer$endBK) / 2) 
  enhancer$end <- ceiling((enhancer$startBK + enhancer$endBK) / 2)
}


## annotation
for (i in unique(enhancer$ID)) {
  cat(">>> ", i, "\n")
  cis1 <- enhancer %>% filter(ID == i)
  cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
  options(ChIPseeker.downstreamDistance = 500) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
  options(verbose = F)
  peakanno <- annotatePeak(cis1_gr, 
                           level = "gene", 
                           tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                           TxDb = txdb, verbose = F, overlap = "all")
  
  # collect results
    temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
    anno_list <- c(anno_list, peakanno)
    names(anno_list)[length(anno_list)] <- i
    
    temp_stats$ID <- i
    anno_stats <- rbind(anno_stats, temp_stats)
    
  # pie plot
    # p <- temp_stats %>%
    #   mutate(label = paste(Feature, " (", round(Rate * 100, digits = 2), 
    #                        "%)", sep = "")) %>%
    #   mutate(label = factor(label, levels = label)) %>%
    #   ggplot(aes(x = "", y = Rate, fill = label)) + 
    #   geom_bar(stat = "identity", color = "black") + 
    #   coord_polar("y", start = 0, direction = -1) + toolkit_tyj$theme_my + 
    #   theme(legend.title = element_blank(), axis.text = element_blank(), 
    #         axis.title = element_blank(), axis.ticks = element_blank(), 
    #         panel.border = element_rect(color = "transparent"), 
    #         legend.key.size = unit(0.5, units = "cm")) + 
    #   scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
    #                                brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
    #   ggtitle(i)
    # # print(p)
    # toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/Enhancer_distribution_",
    #                                              i, sep = ""),
    #                      plot = p, device = "png", width = 16, height = 10)
    
  # pie plot with function from ChipSeeker
    # png(filename = paste("./R4.Distribution_of_enhancers/Enhancer_distribution_Pie_up2000_", i, ".png",sep = ""),
    #     width = 16, height = 10, res = 600, bg = "transparent", units = "cm")
    #   plotAnnoPie(peakanno)
    # dev.off()
    
}


# rm(p, i, cis1, cis1_gr, temp_stats, peakanno)
# save.image(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20220516.bin")

# # list of Enhancer
# enhancer %>%
#   dplyr::mutate(ID = str_replace_all(ID, c("cSTARR-seq" = "Enhancer_STARRseq", 
#                                            "IntronicEnhancer" = "Enhancer_Meng2021"))) %>%
#   fwrite(., "./Enhancer_all_20220420.csv.gz")

```

```{r plot and summary}
  rm(list = ls())
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")
  source("D:/SUST/code/TanYongjun_code.R")
  ## merge Downstream(>1bp) into Distal Intergenic.
  anno_stats <- anno_stats %>%
    dplyr::mutate(Feature = as.character(Feature)) %>%
    dplyr::mutate(Feature = str_replace_all(Feature, 
                                            c("Downstream \\(1\\-2kb\\)" = "Distal Intergenic" ,
                                              "Downstream \\(2\\-3kb\\)" = "Distal Intergenic"))) %>%
    dplyr::mutate(Feature = factor(Feature, levels = c("Promoter (2-3kb)", 
                                                        "Promoter (1-2kb)",
                                                        "Promoter (<=1kb)",
                                                        "Promoter",
                                                        "5' UTR",
                                                        "1'st exon",
                                                        "1'st intron",
                                                        "Other exon",
                                                        "Other intron",
                                                        "Last intron",
                                                        "Last exon",
                                                        "3' UTR",
                                                        "Downstream (<1kb)",
                                                        "Distal Intergenic"))) %>%
    group_by(Feature, ID) %>%
    dplyr::summarise(Frequency = sum(Frequency), Rate = sum(Rate)) %>%
    dplyr::mutate(ID = str_replace_all(ID, c("cSTARR-seq" = "Enhancer_STARRseq",
                                             "IntronicEnhancer" = "Enhancer_Meng2021")))
  
  # ## write to tsv file
  # anno_stats %>%
  #   dplyr::mutate(Species = "Arabidopsis thaliana") %>%
  #   fwrite(str_c("./R4.Distribution_of_enhancers/A.thaliana_Enhancer",
  #          "_annostates_up500_down0_STARR_20210605.tsv", sep = ""), 
  #          quote = F, col.names = T, row.names = F, sep = "\t")
  
  ## bar plot (all)
    anno_stats %>%
      dplyr::filter(ID != "iSTARRseq-common") %>%
      mutate(ID = str_replace_all(ID, ".+common", "Common")) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.5, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605",
    #                      width = 8, height = 8)

  ## bar plot 20210329
    anno_stats %>%
      dplyr::filter(ID != "iSTARR-seq", ID != "PAS") %>%
      mutate(ID = str_replace_all(ID, c("cSTARR-seq" = "Enhancer_STARRseq"))) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.5, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR_IntronicEnhancer_20210605",
                         width = 8, height = 8)
    
```


### .2 enhancer Kmeans cluster

```{r}
## prepare data sets
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28

  # load
  enhancer <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_RPKM_20210321.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq", enrichment > 1.3) %>%
    dplyr::mutate(clusterK4 = paste("Cluster-", clusterK4, sep = ""),
                  clusterK7 = paste("Cluster-", clusterK7, sep = "")) %>%
    full_join(., mutate(., clusterK4 = "All", clusterK7 = "All"))
  Random <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_RPKM_20210321.csv.gz") %>%
    dplyr::filter(method == "Random") %>%
    dplyr::mutate(., clusterK4 = "Random", clusterK7 = "Random")
  
  enhancer <- full_join(enhancer, Random) 
  
  table(enhancer$clusterK4)
  table(enhancer$clusterK7)
  
## annotation and plot
  anno_list <- list()
  anno_stats <- NULL
  for (i in match(c("clusterK4", "clusterK7"), colnames(enhancer))) {
    df <- enhancer %>%
          dplyr::select(seqnames:end, i)
    names(df)[4] <- "ID"
    prefix <- names(enhancer)[i]
    cat("\n>>>>>", prefix, "\n")
        # annotation peak
    for (j in unique(df$ID)) {
      cat("    ", j, "\n")
      cis1 <- df %>% filter(ID == j)
      cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
      options(ChIPseeker.downstreamDistance = 500) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
      options(verbose = F)
      peakanno <- ChIPseeker::annotatePeak(cis1_gr, 
                               tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                               TxDb = txdb, verbose = F, overlap = "all")
      
      # collect results
        temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
        anno_list <- c(anno_list, peakanno)
        names(anno_list)[length(anno_list)] <- paste(prefix, j, sep = "_")
        
        temp_stats$ID <- paste(prefix, j, sep = "_")
        anno_stats <- rbind(anno_stats, temp_stats)
        
      # pie plot
        # p <- temp_stats %>%
        #   mutate(label = paste(Feature, " (", round(Rate * 100, digits = 2), 
        #                        "%)", sep = "")) %>%
        #   mutate(label = factor(label, levels = label)) %>%
        #   ggplot(aes(x = "", y = Rate, fill = label)) + 
        #   geom_bar(stat = "identity", color = "black") + 
        #   coord_polar("y", start = 0, direction = -1) + toolkit_tyj$theme_my + 
        #   theme(legend.title = element_blank(), axis.text = element_blank(), 
        #         axis.title = element_blank(), axis.ticks = element_blank(), 
        #         panel.border = element_rect(color = "transparent"), 
        #         legend.key.size = unit(0.5, units = "cm")) + 
        #   scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
        #                                brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        #   ggtitle(j) + 
        #   labs(caption = prefix)
        # # print(p)
        # toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/", prefix,
        #                                              "_", j, "_20210323", sep = ""),
        #                      plot = p, device = "png", width = 16, height = 10)
    }
  }

rm(p, i, cis1, cis1_gr, temp_stats, peakanno, j, x, prefix, enhancer, Random, df)
# save.image(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR1.3_cluster_20210415.bin")
```

```{r plot and summary}
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR1.3_cluster_20210415.bin")
  ## bar plot
    anno_stats %>%
      dplyr::filter(str_detect(ID, "clusterK7", negate = T)) %>%
      mutate(ID = str_replace_all(ID, c("clusterK4_" = "", 
                                        "All" = "All enhancers"))) %>%
      dplyr::filter(ID != "Cluster-0") %>%
      mutate(ID = factor(ID, levels = c(paste("Cluster-", 1:4, sep = ""), "All enhancers", "Random"))) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR1.3_cluster_20210415",
    #                      width = 16, height = 12)


  ## summary table
    anno_stats %>%
      mutate(value = paste(Frequency, " (", round(Rate * 100, 4), "%)", sep = "")) %>%
      dplyr::select(-Frequency, -Rate) %>%
      pivot_wider(id_cols = Feature, names_from = ID, values_from = value) %T>%
      # write.csv(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210415.csv",
      #           quote = F) %>%
      kable()

## DHS enhancer + enhancer lcuster, 20210329
    load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR1.3_cluster_20210415.bin")
    anno_stats_cluster <- anno_stats
    load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210415.bin")
    anno_stats <- anno_stats %>%
      dplyr::filter(str_detect(ID, "(Enhancer_Wang2019|Enhancer_Zhu2015)"))
    
    full_join(anno_stats, anno_stats_cluster) %>%
      dplyr::filter(str_detect(ID, "clusterK7", negate = T)) %>%
      mutate(ID = str_replace_all(ID, c("clusterK4_" = "", 
                                        "All" = "Enhancer"))) %>%
      dplyr::filter(ID != "Cluster-0") %>%
      mutate(ID = factor(ID, levels = c(paste("Cluster-", 1:4, sep = ""), 
                                        "Enhancer", "Enhancer_Wang2019", "Enhancer_Zhu2015",
                                        "Random"))) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_DHS_enhancer_cluster_20210415",
    #                      width = 16, height = 12)
```

### .3 Activity cluster

```{r}
## prepare data sets
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28

  # load
  enhancer <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq", enrichment > 1.3) 
  Random <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz") %>%
    dplyr::filter(method == "Random") %>%
    dplyr::mutate(., ClusterActivity = "Random")
  
  enhancer <- full_join(enhancer, Random)  %>%
    full_join(dplyr::mutate(enhancer, ClusterActivity = "All"), .)
  
  table(enhancer$ClusterActivity)
  
## annotation and plot
  anno_list <- list()
  anno_stats <- NULL
  
  df <- enhancer %>%
        dplyr::select(seqnames:end, ClusterActivity)
  names(df)[4] <- "ID"
  
  # annotation peak
  for (j in unique(df$ID)) {
    cat("    ", j, "\n")
    cis1 <- df %>% filter(ID == j)
    cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
    options(ChIPseeker.downstreamDistance = 500) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
    options(verbose = F)
    peakanno <- ChIPseeker::annotatePeak(cis1_gr, 
                             tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                             TxDb = txdb, verbose = F, overlap = "all")
    
    # collect results
      temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
      anno_list <- c(anno_list, peakanno)
      names(anno_list)[length(anno_list)] <- j
      
      temp_stats$ID <- j
      anno_stats <- rbind(anno_stats, temp_stats)
  }

rm(p, i, cis1, cis1_gr, temp_stats, peakanno, j, x, prefix, enhancer, Random, df)
# save.image(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up5000_down0_cSTARR1.3_ActCluster_20210415.bin")
```

```{r plot and summary}
  rm(list = ls())
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up5000_down0_cSTARR1.3_ActCluster_20210415.bin")
  ## bar plot
    anno_stats %>%
      mutate(ID = factor(ID, levels = c(paste("Part", 1:5, sep = ""), "All", "Random"))) %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent)
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR1.3_ActivityCluster_20210415",
                         width = 16, height = 12)


  ## summary table
    anno_stats %>%
      mutate(value = paste(Frequency, " (", round(Rate * 100, 4), "%)", sep = "")) %>%
      dplyr::select(-Frequency, -Rate) %>%
      pivot_wider(id_cols = Feature, names_from = ID, values_from = value) %T>%
      write.csv(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_cSTARR1.3_ActivityCluster_20210415.csv.gz",
                quote = F) %>%
      kable()

```

### .4 D.m, H.m, and O.s

*Distribution of enhancers identifiedn in Drosophila*
```{r prepare datasets}
    rm(list = ls())
    library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
    txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene
    
  ## load
    source("D:/SUST/code/TanYongjun_code.R")
    enhancer <- fread("../homo_cis/database/fruitfly_S2/S2_enhancer_arnold_supp_table07.bed") %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(ID = "enhancer")
    random <- NULL
    for (i in 1:5) {
      random <- toolkit_tyj$shuffle(makeGRangesFromDataFrame(enhancer), TxDb = txdb) %>%
        as.data.frame() %>%
        rbind(., random)
    }
    random <- random[,1:3]
    random$ID <- "Random"
    colnames(random) <- colnames(enhancer)
    cre_all <- rbind(enhancer, random)
    colnames(cre_all) <- c("seqnames", "start", "end", "ID")
    
  ## annotation and plot peak
    anno_list <- list()
    anno_stats <- NULL
    for (i in unique(cre_all$ID)) {
      cat(">>> ", i, "\n")
      cis1 <- cre_all %>% filter(ID == i)
      cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
      options(ChIPseeker.downstreamDistance = 1000) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
      options(verbose = F)
      peakanno <- ChIPseeker::annotatePeak(cis1_gr, 
                               tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                               TxDb = txdb, verbose = F, overlap = "all")
      
      # collect results
        temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
        anno_list <- c(anno_list, peakanno)
        names(anno_list)[length(anno_list)] <- i
        
        temp_stats$ID <- i
        anno_stats <- rbind(anno_stats, temp_stats)
        
    }
  
## plot
    anno_stats <- anno_stats %>%
    dplyr::mutate(Feature = as.character(Feature)) %>%
    dplyr::mutate(Feature = str_replace_all(Feature, 
                                            c("Downstream \\(1\\-2kb\\)" = "Distal Intergenic" ,
                                              "Downstream \\(2\\-3kb\\)" = "Distal Intergenic"))) %>%
    dplyr::mutate(Feature = factor(Feature, levels = c("Promoter (2-3kb)", 
                                                        "Promoter (1-2kb)",
                                                        "Promoter (<=1kb)",
                                                        "Promoter",
                                                        "5' UTR",
                                                        "1'st exon",
                                                        "1'st intron",
                                                        "Other exon",
                                                        "Other intron",
                                                        "Last intron",
                                                        "Last exon",
                                                        "3' UTR",
                                                        "Downstream (<1kb)",
                                                        "Distal Intergenic"))) %>%
    group_by(Feature, ID) %>%
    dplyr::summarise(Frequency = sum(Frequency), Rate = sum(Rate))
  
  ## bar plot (all)
    anno_stats %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.3, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent) + 
        labs(caption = "Drosophila")
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Drosophila_Enhancer_ChIPseeker_Peak_up500_down0_STARR_20210526",
                         width = 8, height = 8)
    
  ## save to file (for plot)
    anno_stats %>%
      dplyr::mutate(Species = "Drosophila melanogaster") %>%
      fwrite("./R4.Distribution_of_enhancers/Drosophila_Enhancer_annostates_up500_down0_STARR_20210526.tsv", 
             quote = F, col.names = T, row.names = F, sep = "\t")
```

*Distribution of enhancers identified in Human K562*
```{r prepare datasets}
    rm(list = ls())
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
    library(liftOver)
    library(ChIPseeker)
    chFile = import.chain(system.file(package="liftOver", "extdata", "hg38Tohg19.over.chain"))
    
  ## load
    source("D:/SUST/code/TanYongjun_code.R")
    repID <- c("ENCFF394DBM", "ENCFF717VJK")[2]
    enhancer <- fread(paste("../homo_cis/database/K562_STARRseq_enhancer_", repID, ".bed.gz", sep = "")) %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(ID = "enhancer")
    # liftover
    names(enhancer)[1:3] <- c("seqnames", "start", "end")
    enhancer <- enhancer %>% 
      makeGRangesFromDataFrame() %>%
      liftOver(., chFile) %>%
      unlist() %>%
      as.data.frame() %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(ID = "enhancer")

    random <- NULL
    for (i in 1:2) {
      random <- toolkit_tyj$shuffle(makeGRangesFromDataFrame(enhancer), TxDb = txdb) %>%
        as.data.frame() %>%
        rbind(., random)
    }
    random <- random[,1:3]
    random$ID <- "Random"
    colnames(random) <- colnames(enhancer)
    cre_all <- rbind(enhancer, random)
    colnames(cre_all) <- c("seqnames", "start", "end", "ID")
    
  ## annotation and plot peak
    anno_list <- list()
    anno_stats <- NULL
    for (i in unique(cre_all$ID)) {
      cat(">>> ", i, "\n")
      cis1 <- cre_all %>% filter(ID == i)
      cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
      options(ChIPseeker.downstreamDistance = 1000) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
      options(verbose = F)
      peakanno <- annotatePeak(cis1_gr, 
                               tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                               TxDb = txdb, verbose = F, overlap = "all")
      
      # collect results
        temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
        anno_list <- c(anno_list, peakanno)
        names(anno_list)[length(anno_list)] <- i
        
        temp_stats$ID <- i
        anno_stats <- rbind(anno_stats, temp_stats)
        
    }
  
## plot
    anno_stats <- anno_stats %>%
    dplyr::mutate(Feature = as.character(Feature)) %>%
    dplyr::mutate(Feature = str_replace_all(Feature, 
                                            c("Downstream \\(1\\-2kb\\)" = "Distal Intergenic" ,
                                              "Downstream \\(2\\-3kb\\)" = "Distal Intergenic"))) %>%
    dplyr::mutate(Feature = factor(Feature, levels = c("Promoter (2-3kb)", 
                                                        "Promoter (1-2kb)",
                                                        "Promoter (<=1kb)",
                                                        "Promoter",
                                                        "5' UTR",
                                                        "1'st exon",
                                                        "1'st intron",
                                                        "Other exon",
                                                        "Other intron",
                                                        "Last intron",
                                                        "Last exon",
                                                        "3' UTR",
                                                        "Downstream (<1kb)",
                                                        "Distal Intergenic"))) %>%
    group_by(Feature, ID) %>%
    dplyr::summarise(Frequency = sum(Frequency), Rate = sum(Rate))
  
  ## bar plot (all)
    anno_stats %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.3, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent) + 
        labs(caption = str_c("Human", "_K562-", repID, sep = ""))
    
    toolkit_tyj$SavePlot(filename_prefix = str_c("./R4.Distribution_of_enhancers/HumanK562_Enhancer_",
                         repID, "_ChIPseeker_Peak_up500_down0_STARR_20210526", sep = ""),
                         width = 8, height = 8)
    
  ## save to file (for plot)
    anno_stats %>%
      dplyr::mutate(Species = "Homo sapiens (K562)") %>%
      fwrite(str_c("./R4.Distribution_of_enhancers/HumanK562_Enhancer_", repID, 
             "_annostates_up500_down0_STARR_20210526.tsv", sep = ""), 
             quote = F, col.names = T, row.names = F, sep = "\t")
```

*Distribution of enhancers identified in Rice*
```{r prepare datasets}
    rm(list = ls())
    library(GenomicFeatures)

 # aquire TxDb from Biomart/ensembl (run with R 4.0.3, or error will occur for some function change in "dplyr")
    # mart <- biomaRt::useMart(biomart="plants_mart",
    #                          host="plants.ensembl.org")
    # # biomaRt::listDatasets(mart)
    # txdb <- makeTxDbFromBiomart(dataset="osativa_eg_gene", host = "plants.ensembl.org", biomart = "plants_mart")
    
    # make TxDb from GFF is a better choice, because network error may occure in "makeTxDbFromBiomart".
    # add seqinfo for toolkit_tyj$shuffle()
    library(BSgenome.Osativa.MSU.MSU7)
    # seqinfo(BSgenome.Osativa.MSU.MSU7) %>%
    # as.data.frame()
    # seqInfoRice <- Seqinfo(seqnames = c(str_c("Chr", c(1:12, "M", "P", "Sy", "Un"), sep = "")), 
    #                          seqlengths = c(43270923, 35937250, 36413819, 35502694,
    #                                         29958434, 31248787, 29697621, 28443022,
    #                                         23012720, 23207287, 29021106, 27531856,
    #                                         490520, 134525), 
    #                          isCircular = c(rep(F, 12), T, T), genome = "IRGSP 1.0")
    txdb <- makeTxDbFromGFF("./refgenome/MSU_rice_all.gff3.gz", 
                            chrominfo = seqinfo(BSgenome.Osativa.MSU.MSU7))
    # txdb1 <- makeTxDbFromGFF("./refgenome/Oryza_sativa.IRGSP-1.0.51.chr.gff3.gz") # less genes annotated in this file, lead more than half of enhancers annotated as distal intergenic.
    
  ## load
    source("D:/SUST/code/TanYongjun_code.R")
    enhancer <- fread("./refgenome/GSE121231_Nipponbare_merged_peaks.bed.gz") %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(ID = "enhancer", 
                    V1 = str_replace_all(V1, c("chr0" = "Chr", 
                                           "chr" = "Chr")))
    colnames(enhancer) <- c("seqnames", "start", "end", "ID")
    
    random <- NULL
    for (i in 1:5) {
      random <- toolkit_tyj$shuffle(makeGRangesFromDataFrame(enhancer), TxDb = txdb) %>%
        as.data.frame() %>%
        rbind(., random)
    }
    random <- random[,1:3]
    random$ID <- "Random"
    colnames(random) <- colnames(enhancer)
    cre_all <- rbind(enhancer, random)
    colnames(cre_all) <- c("seqnames", "start", "end", "ID")
    
  ## annotation and plot peak
    anno_list <- list()
    anno_stats <- NULL
    for (i in unique(cre_all$ID)) {
      cat(">>> ", i, "\n")
      cis1 <- cre_all %>% filter(ID == i)
      cis1_gr <- makeGRangesFromDataFrame(df = cis1, keep.extra.columns = T)
      options(ChIPseeker.downstreamDistance = 1000) # this option just affect name of label used in "peakanno@annoStat", but did not affect the feature definition.
      options(verbose = F)
      peakanno <- ChIPseeker::annotatePeak(cis1_gr, 
                               tssRegion = c(-500, 0), # downstream set to 0, or it will cover the UTR and 1'st exon
                               TxDb = txdb, verbose = F, overlap = "all")
      
      # collect results
        temp_stats <- toolkit_tyj$ChipSeekerAnnoStat(peakanno) # annoStats with custom function
        anno_list <- c(anno_list, peakanno)
        names(anno_list)[length(anno_list)] <- i
        
        temp_stats$ID <- i
        anno_stats <- rbind(anno_stats, temp_stats)
        
    }
  
## plot
    anno_stats <- anno_stats %>%
    dplyr::mutate(Feature = as.character(Feature)) %>%
    dplyr::mutate(Feature = str_replace_all(Feature, 
                                            c("Downstream \\(1\\-2kb\\)" = "Distal Intergenic" ,
                                              "Downstream \\(2\\-3kb\\)" = "Distal Intergenic"))) %>%
    dplyr::mutate(Feature = factor(Feature, levels = c("Promoter (2-3kb)", 
                                                        "Promoter (1-2kb)",
                                                        "Promoter (<=1kb)",
                                                        "Promoter",
                                                        "5' UTR",
                                                        "1'st exon",
                                                        "1'st intron",
                                                        "Other exon",
                                                        "Other intron",
                                                        "Last intron",
                                                        "Last exon",
                                                        "3' UTR",
                                                        "Downstream (<1kb)",
                                                        "Distal Intergenic"))) %>%
    group_by(Feature, ID) %>%
    dplyr::summarise(Frequency = sum(Frequency), Rate = sum(Rate))
  
  ## bar plot (all)
    anno_stats %>%
      ggplot() +
        geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
        scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                       brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
        # facet_grid(.~Cis_element, scales = "free", space = "free") + 
        toolkit_tyj$theme_my + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              legend.key.size = unit(0.3, units = "cm")) +
        ylab("Percent") + 
        scale_y_continuous(labels = scales::percent) + 
        labs(caption = "Rice")
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Rice_MSUV7_Enhancer_ChIPseeker_Peak_up500_down0_STARR_20210526",
                         width = 8, height = 8)
    
  ## save to file (for plot)
    anno_stats %>%
      dplyr::mutate(Species = "Oryza Sativa") %>%
      fwrite("./R4.Distribution_of_enhancers/Rice_MSUV7_Enhancer_annostates_up500_down0_STARR_20210526.tsv", 
             quote = F, col.names = T, row.names = F, sep = "\t")
```

*Plot in one figure*
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
statsDf <- NULL
for (i in list.files("./R4.Distribution_of_enhancers/", pattern = ".+_annostates_up500.+tsv")) {
  statsDf  <- fread(str_c("./R4.Distribution_of_enhancers/", i, sep = "")) %>%
    rbind(., statsDf)
}
table(statsDf$Species)

# plot
statsDf %>%
    dplyr::mutate(Feature = factor(Feature, levels = c("Promoter (2-3kb)", 
                                                      "Promoter (1-2kb)",
                                                      "Promoter (<=1kb)",
                                                      "Promoter",
                                                      "5' UTR",
                                                      "1'st exon",
                                                      "1'st intron",
                                                      "Other exon",
                                                      "Other intron",
                                                      "Last intron",
                                                      "Last exon",
                                                      "3' UTR",
                                                      "Downstream (<1kb)",
                                                      "Distal Intergenic"))) %>%
    dplyr::filter(str_detect(ID, "(iSTARR|Enhancer_|PAS)", negate = T)) %>%
    dplyr::mutate(str_replace_all(ID, c("cSTARR-seq" = "Enhancer_STARRseq", 
                                        "enhancer" = "Enhancer_STARRseq"))) %>%
    dplyr::mutate(Species = factor(Species, levels = c("Arabidopsis thaliana", "Oryza Sativa",
                                                       "Drosophila melanogaster", "Homo sapiens (K562)"))) %>%
    ggplot() +
    geom_col(aes(x = ID, y = Rate, fill = Feature)) + 
    scale_fill_manual(values = c(brewer.pal(12, "Paired"), 
                                   brewer.pal(8, "Dark2"))[c(1:5, 7:15, 17:20)]) + 
    facet_grid(.~ Species, scales = "free", space = "free") +
    toolkit_tyj$theme_my + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.key.size = unit(0.3, units = "cm"), 
          legend.margin = margin(0,0,0,-0.1, unit = "cm")) +
    ylab("Percent") + 
    scale_y_continuous(labels = scales::percent)
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Four_Species_enhancer_distribution_ChIPseeker_Peak_up500_down0_STARR_20210526",
                         width = 14, height = 8)

```

#### .1 Summary of genome of these species.
```{r}
    rm(list = ls())
    library(GenomicFeatures)
    library(BSgenome.Osativa.MSU.MSU7)
    library(TxDb.Athaliana.BioMart.plantsmart28)
    library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    txdbRice <- makeTxDbFromGFF("./refgenome/MSU_rice_all.gff3.gz", 
                            chrominfo = seqinfo(BSgenome.Osativa.MSU.MSU7))
    
    allTxDb <- list(Oryza_sativa = txdbRice, 
                    Arabidopsis_thaliana = TxDb.Athaliana.BioMart.plantsmart28, 
                    Drosophila_melanogaster = TxDb.Dmelanogaster.UCSC.dm3.ensGene, 
                    Homo_sapiens = TxDb.Hsapiens.UCSC.hg19.knownGene)
    
    # Percent of genome sequence covered by gene (TSS-TTS)
    df <- map_dfr(allTxDb, function(x){
      c(genes(x) %>% length(), 
        genes(x) %>% width() %>% sum(),
        seqlengths(x) %>% sum()
       )
    }, .id = "Species")
```


## .4.1 GAT (genomic features)
perform permutation to identify overlap between enhancer and genomic features (and lncRNA)

#### .1 prepare .bed file

##### A.t
*genomic features*
> Attention: 5' UTR may cover the part of or the whole 1'st exon, and the 5'UTR of some gene may more than one exon. Do not use '5'UTR' and 1'st exon at the same time.

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  
## .bed for genomic regions: 
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28

  tmp <- toolkit_tyj$GetGenoFeaturesRegion(txdb, ProDefine = c(-500, 0), DownStreamLen = 500)
  
  # bed
  tmp$Features %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames)) %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/At_GenomicFeatures_20210605.bed.gz",
           sep = "\t", row.names = F, col.names = F)
  # for (i in unique(tmp$Features$Features)) {
  #   tmp$Features %>%
  #     dplyr::filter(Features == i) %>%
  #     dplyr::mutate(seqnames = str_c("chr", seqnames)) %>%
  #     fwrite(file = paste("./R4.Distribution_of_enhancers/GAT_overlap/At_", i, 
  #                         "_regions_20210605.bed.gz", sep = ""), 
  #            sep = "\t", row.names = F, col.names = F)
  # }

  # summary 
  tmp$RegionSummary %>%
    fwrite("./R4.Distribution_of_enhancers/GAT_overlap/At_Summary_features_20210605.csv")

## bed for CREs
  cre_all <- fread("./R7.TFs/bed_files/at_CREs_all_20210416.bed.gz")
  IntronicEnhancer <- fread("./refgenome/Intronic_DHS_enhancer_seedling_20210604.csv.gz") %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""),
                  V4 = "Enhancer_Meng2021") %>%
    dplyr::select(seqnames, start, end, V4)
  colnames(IntronicEnhancer) <- str_c("V", 1:4)
  cre_all <- rbind(cre_all, IntronicEnhancer)
  
  table(cre_all$V4)
  cre_all <- cre_all %>%
    dplyr::filter(str_detect(V4, "(CorePro|Intron|Exon|GeneBody|pEnhancer|sEnhancer)", negate = T))
  unique(cre_all$V4)
  cre_all %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/At_CRE_selected_20210605.bed.gz", 
           sep = "\t", row.names = F, col.names = F)
    
```

##### O.s
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  
## .bed for genomic regions: 
  library(BSgenome.Osativa.MSU.MSU7)
  txdb <- makeTxDbFromGFF("./refgenome/MSU_rice_all.gff3.gz", 
                          chrominfo = seqinfo(BSgenome.Osativa.MSU.MSU7))
  
# # chromosome region for GAT.
#   seqlengths(txdb) %>%
#     as.data.frame() %>%
#     fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Os_chr_size_noCM_forGAT.tsv",
#            sep = "\t", row.names = T, col.names = F)
  
  tmp <- toolkit_tyj$GetGenoFeaturesRegion(txdb, ProDefine = c(-500, 0), DownStreamLen = 500)
  
  # bed
  tmp$Features %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Os_GenomicFeatures_20210605.bed.gz",
           sep = "\t", row.names = F, col.names = F)
  # for (i in unique(tmp$Features$Features)) {
  #   tmp$Features %>%
  #     dplyr::filter(Features == i) %>%
  #     fwrite(file = paste("./R4.Distribution_of_enhancers/GAT_overlap/Os_", i, 
  #                         "_regions_20210605.bed.gz", sep = ""), 
  #            sep = "\t", row.names = F, col.names = F)
  # }
  
  # summary 
  tmp$RegionSummary %>%
    fwrite("./R4.Distribution_of_enhancers/GAT_overlap/Os_Summary_features_20210605.csv")
  
## bed for CREs(enhancer)
  enhancer <- fread("./refgenome/GSE121231_Nipponbare_merged_peaks.bed.gz") %>%
    dplyr::select(1:3) %>%
    dplyr::mutate(ID = "Os_enhancer", 
                  V1 = str_replace_all(V1, c("chr0" = "Chr", 
                                         "chr" = "Chr")))
  colnames(enhancer) <- c("seqnames", "start", "end", "ID")
  enhancer %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Os_CRE_selected_20210605.bed.gz", 
           sep = "\t", row.names = F, col.names = F)
```

##### D.m

```{r prepare datasets}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  
## .bed for genomic regions: 
  library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
  txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene

# # chromosome region for GAT.
  # seqlengths(txdb) %>%
  #   as.data.frame() %>%
  #   fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Dm_chr_size_noCM_forGAT.tsv",
  #          sep = "\t", row.names = T, col.names = F)
  
  tmp <- toolkit_tyj$GetGenoFeaturesRegion(txdb, ProDefine = c(-500, 0), DownStreamLen = 500)
  
  # bed
  tmp$Features %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Dm_GenomicFeatures_20210605.bed.gz",
           sep = "\t", row.names = F, col.names = F)
  # for (i in unique(tmp$Features$Features)) {
  #   tmp$Features %>%
  #     dplyr::filter(Features == i) %>%
  #     fwrite(file = paste("./R4.Distribution_of_enhancers/GAT_overlap/Dm_", i, 
  #                         "_regions_20210605.bed.gz", sep = ""), 
  #            sep = "\t", row.names = F, col.names = F)
  # }
  
  # summary 
  tmp$RegionSummary %>%
    fwrite("./R4.Distribution_of_enhancers/GAT_overlap/Dm_Summary_features_20210605.csv")
  
## bed for CREs(enhancer)
  enhancer <- fread("../homo_cis/database/fruitfly_S2/S2_enhancer_arnold_supp_table07.bed") %>%
    dplyr::select(1:3) %>%
    dplyr::mutate(ID = "Dm_enhancer")
  colnames(enhancer) <- c("seqnames", "start", "end", "ID")
  enhancer %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Dm_CRE_selected_20210605.bed.gz", 
           sep = "\t", row.names = F, col.names = F)

```


##### H.s

```{r prepare datasets}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  
## .bed for genomic regions: 
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
  # chromosome region for GAT.
  # seqlengths(txdb) %>% 
  #   as.data.frame() %>% 
  #   fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Hs_chr_size_noCM_forGAT.tsv", 
  #          sep = "\t", row.names = T, col.names = F)
  
  tmp <- toolkit_tyj$GetGenoFeaturesRegion(txdb, ProDefine = c(-500, 0), DownStreamLen = 500)
  
  # bed
  tmp$Features %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Hs_GenomicFeatures_20210605.bed.gz",
           sep = "\t", row.names = F, col.names = F)
  # for (i in unique(tmp$Features$Features)) {
  #   tmp$Features %>%
  #     dplyr::filter(Features == i) %>%
  #     dplyr::mutate(seqnames = str_c("chr", seqnames)) %>%
  #     fwrite(file = paste("./R4.Distribution_of_enhancers/GAT_overlap/Hs_", i, 
  #                         "_regions_20210605.bed.gz", sep = ""), 
  #            sep = "\t", row.names = F, col.names = F)
  # }
  # 
  # summary 
  tmp$RegionSummary %>%
    fwrite("./R4.Distribution_of_enhancers/GAT_overlap/Hs_Summary_features_20210605.csv")
  
## bed for CREs(enhancer)
  library(liftOver)
  library(ChIPseeker)
  chFile = import.chain(system.file(package="liftOver", "extdata", "hg38Tohg19.over.chain"))
  repID <- c("ENCFF394DBM", "ENCFF717VJK")[2]
  enhancer <- fread("../homo_cis/database/STARRseq_enhancer_ENCODE/STARRseq_enhancer_K562_ENCFF045TVA.bed.gz") %>%
    dplyr::select(1:3) %>%
    dplyr::mutate(ID = "enhancer")
  # liftover
  names(enhancer)[1:3] <- c("seqnames", "start", "end")
  enhancer <- enhancer %>% 
    makeGRangesFromDataFrame() %>%
    liftOver(., chFile) %>%
    unlist() %>%
    as.data.frame() %>%
    dplyr::select(1:3) %>%
    dplyr::mutate(ID = "enhancer")
  colnames(enhancer) <- c("seqnames", "start", "end", "ID")
  enhancer %>%
    fwrite(file = "./R4.Distribution_of_enhancers/GAT_overlap/Hs_CRE_selected_20210605.bed.gz", 
           sep = "\t", row.names = F, col.names = F)

```


### .2 run GAT
`/scratch/2021-06-03/bioTanyj/At_enhancer/scripts/4.1.1.GenomicFeatures_overlap_GAT.sh`
`/scratch/2021-06-03/bioTanyj/At_enhancer/scripts/4.1.1.GenomicFeatures_overlap_GAT.R`

### .3 Collecr results

##### .1 202105 (only A.t)
Only A.t
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- NULL
  for (i in list.files(path = "./R4.Distribution_of_enhancers/GAT_overlap/", 
                       pattern = ".+regions_20210509_add.tsv")) {
    df <- fread(paste("./R4.Distribution_of_enhancers/GAT_overlap/", 
                      i, sep = "")) %>%
      dplyr::mutate(track = str_replace_all(i, "(.+)_regions.+", "\\1")) %>%
      rbind(., df)
  }
  
  df$track <- factor(str_replace_all(df$track, c("500" = "", 
                                                 "Distal" = "", 
                                                 "Pro" = "Promoter")), 
                     levels = c("Promoter", "FiveUTR", "FirstExon", "FirstIntron",
                                          "OtherExon", "OtherIntron", "LastExon", "LastIntron",
                                          "ThreeUTR", "DownStream", "Intergenic", "TSS2TTS"))
  df$annotation <- str_replace_all(df$annotation, "^Enhancer$", "Enhancer_STARRseq")
# adjust pvalue
  df$qvalue <- p.adjust(df$pvalue, method = "BH")
  
# each enhancers as a whole.
  df %>%
    dplyr::filter(str_detect(annotation, "^Random.+", negate = T), 
                  str_detect(annotation, "(Zhu|Wang|^Enhancer_STARRseq$)")) %>%
    dplyr::mutate(sigLable = ifelse(qvalue < 0.001, "*", "ns"), 
                  lablePos = ifelse(l2fold >0, l2fold + 0.2, l2fold - 0.2)) %>%
    ggplot(aes(x = annotation, y = l2fold, fill = annotation)) + 
      geom_col(position = "dodge", width = 0.8) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_fill_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.key.size = unit(0.35, units = "cm"), 
            axis.text.x = element_blank(), 
            axis.ticks.x = element_blank()) + 
      geom_text(aes(x = annotation, y = lablePos, label = sigLable), size = 2) + 
      facet_wrap("track", nrow = 2)
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/GAT_overlap/Enhancer_whole_enrichment", 
  #                      width = 12, height = 8)
  
# activity cluster
  df %>%
    dplyr::filter(str_detect(annotation, "^Random.+", negate = T), 
                  str_detect(annotation, "(ActivityC|^Enhancer_STARR)")) %>%
    dplyr::mutate(sigLable = ifelse(qvalue < 0.001, "*", "ns"), 
                  lablePos = ifelse(l2fold >0, l2fold + 0.2, l2fold - 0.2)) %>%
    ggplot(aes(x = annotation, y = l2fold, fill = annotation)) + 
      geom_col(position = "dodge", width = 0.8) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_fill_npg() + 
      scale_color_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.key.size = unit(0.35, units = "cm"), 
            axis.text.x = element_blank(), 
            axis.ticks.x = element_blank()) + 
      geom_text(aes(label = sigLable, y = lablePos), size = 1.5) + 
      facet_wrap(facets = "track", nrow = 2)
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/GAT_overlap/ActivityCluster_enrichment", 
  #                      width = 12, height = 8)
  
# Kmeans cluster
  df %>%
    dplyr::filter(str_detect(annotation, "^Random.+", negate = T), 
                  str_detect(annotation, "(Kmean|^Enhancer_STARR)")) %>%
    dplyr::mutate(sigLable = ifelse(qvalue < 0.001, "*", "ns"), 
                  lablePos = ifelse(l2fold >0, l2fold + 0.2, l2fold - 0.2)) %>%
    ggplot(aes(x = annotation, y = l2fold, fill = annotation)) + 
      geom_col(position = "dodge", width = 0.8) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_fill_npg() + 
      scale_color_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.key.size = unit(0.35, units = "cm"), 
            axis.text.x = element_blank(), 
            axis.ticks.x = element_blank()) + 
      geom_text(aes(label = sigLable, y = lablePos), size = 1.5) + 
      facet_wrap(facets = "track", nrow = 2)
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/GAT_overlap/KmeansCluster_enrichment", 
  #                      width = 12, height = 8)
  NA
```

##### .2 202106 (At, Os, Dm, Hs)
At, Os, Dm, Hs
calculate percent of CREs located in genomic features based on length (CREs were not resized).

*Collect*
```{r}
## load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- NULL
  for (i in list.files(path = "./R4.Distribution_of_enhancers/GAT_overlap/", 
                       pattern = ".+_ol_GenomicFeatures_GAT_20210606.tsv")) {
    df <- fread(paste("./R4.Distribution_of_enhancers/GAT_overlap/", 
                      i, sep = "")) %>%
      dplyr::mutate(species = str_replace_all(i, "(.+)_CRE.+", "\\1")) %>%
      rbind(., df)
  }
  # table(df$species, df$annotation)
  # unique(df$annotation)
  df$annotation <- factor(str_replace_all(df$annotation, c("500" = "", 
                                                 "Distal" = "", 
                                                 "Pro" = "Promoter")), 
                     levels = c("Promoter", "FiveUTR", "FirstExon", "FirstIntron",
                                          "OtherExon", "OtherIntron", "LastExon", "LastIntron",
                                          "ThreeUTR", "DownStream", "Intergenic", "TSS2TTS"))
  df$annotation <- str_replace_all(df$annotation, "^Enhancer$", "Enhancer_STARRseq")
  
# p adjust 
  df <- df %>%
    group_by(species) %>%
    dplyr::mutate(padjBonferroni = p.adjust(pvalue, method = "bonferroni"),
                  padjBH = p.adjust(pvalue, method = "BH"))
```

###### .Stack bar plot
```{r}
### Plot Percent
  # calculate percentage of each genomic features (according to all features).
  tmp <- df %>%
    dplyr::filter(str_detect(track, "(Random)", negate = T), 
                  str_detect(annotation, "TSS2TTS", negate = T)) %>%
    group_by(species, track) %>%
    dplyr::mutate(Percent_expect = expected / sum(expected), # 计算各feature相对基因组的比例。
                  Percent_observed = observed / sum(observed)) %>% # 计算真实与CRE重叠各feature占基因组的比例。
    dplyr::select(species, track, annotation, Percent_expect, Percent_observed) %>%
    pivot_longer(cols = matches("Percent.+"))

  tmpBK <- tmp %>%
    dplyr::filter(name == "Percent_expect") %>%
    group_by(species, annotation) %>%
    dplyr::summarise(value = mean(value)) %>%
    dplyr::mutate(name = "Percent_expect")
  
  tmp <- tmp %>%
    dplyr::filter(name == "Percent_observed") %>%
    full_join(., tmpBK) %>%
    dplyr::mutate(track = ifelse(is.na(track), "Expected", track), 
                  annotation = factor(annotation, levels = c("Promoter", "FiveUTR", "FirstExon", "FirstIntron",
                                          "OtherExon", "OtherIntron", "LastExon", "LastIntron",
                                          "ThreeUTR", "DownStream", "Intergenic", "TSS2TTS")), 
                  species1 = str_replace_all(species, c("At" = "A.thaliana", "Os" = "O.sativa", "Dm" = "D.melanogaster","Hs" = "H.sapiens")),
                  species1 = factor(species1, levels = c("A.thaliana", "O.sativa", "D.melanogaster", "H.sapiens")),
                  species = factor(species, levels = c("At", "Os", "Dm", "Hs")))
  
  # (All species + All CRE groups)
  tmp %>%
    ggplot(aes(x = track, fill = annotation, y = value)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ species, scales = "free", space = "free") + 
      # scale_fill_npg() + 
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      labs(caption = "Length-based") + 
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            axis.title.x = element_blank())
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Distribution_All_CREs_four_species_lengthBased_20210607",
  #                      width = 12, height = 8)
  
  # # write percent of CRE located in Genomic features
  # tmp %>%
  #   pivot_wider(names_from = annotation, values_from = value) %>%
  #   arrange(species, track, name) %>%
  #   fwrite("./R4.Distribution_of_enhancers/GAT_overlap/Percent_CRE_ol_Genomic_features_LengthBased_20210607.csv")
  
   # (All species, Enhancer)
  tmp %>%
    dplyr::filter(str_detect(track, "(Kmeans|Activity)", negate = T), 
                  str_detect(track, "(hancer$|Expected)")) %>%
    dplyr::mutate(track = str_replace_all(track, ".+hancer", "Enhancer")) %>%
    ggplot(aes(x = track, fill = annotation, y = value)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ species1, scales = "free", space = "free") + 
      # scale_fill_npg() + 
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      ylab("Percentage") +
      scale_y_continuous(labels = scales::percent) +
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, -3, 0, -9), 
            strip.text = element_text(size = 5, face = "italic"), 
            panel.spacing = unit(0.05, "cm"),
            axis.title.x = element_blank(), legend.title = element_blank())

  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Distribution_EnhancerSTARRseq_four_species_lengthBased_20210607",
  #                      width = 8, height = 6)
  
  # only A.t (different methods)
  tmp %>%
    dplyr::filter(species == "At", 
                  str_detect(track, "(Activit|Kmeans)", negate = T)) %>%
    dplyr::mutate(track = str_replace_all(track, c("^Enhancer$" = "Enhancer_STARRseq",
                                                   "Expected" = "Random")), 
                  track = factor(track, levels = c("Enhancer_STARRseq", "Enhancer_Zhu2015", 
                                                   "Enhancer_Wang2019", "Enhancer_Meng2021", 
                                                   "Random"))) %>%
    ggplot(aes(x = track, fill = annotation, y = value)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      # facet_grid(. ~ species, scales = "free", space = "free") + 
      # scale_fill_npg() + 
      scale_y_continuous(labels = scales::percent) +
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      ylab("Percentage") +
      guides(fill = guide_legend(title = "Genomic features")) +
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")})+
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            axis.title.x = element_blank())
  
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Distribution_Enhancer_Methods_At_lengthBased_20220426",
                       width = 8, height = 6)
    
  # only A.t (Kmeans cluster)
  tmp %>%
    dplyr::filter(species == "At", 
                  str_detect(track, "(Kmeans|^Expected|^Enhancer$)", negate = F)) %>%
    dplyr::mutate(track = str_replace_all(track, c("KmeansCluster" = "C",
                                                   "Expected" = "Random"))) %>%
    dplyr::mutate(track = str_replace_all(track, "^Enhancer$", "En_STARRseq"), 
                  track = factor(track, levels = c(str_c("C", 1:4), 
                                                   "En_STARRseq", 
                                                   "Random"))) %>%
    ggplot(aes(x = track, fill = annotation, y = value)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      # facet_grid(. ~ species, scales = "free", space = "free") + 
      # scale_fill_npg() + 
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      scale_y_continuous(labels = scales::percent) + ylab("Percentage") +
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            axis.title.x = element_blank())
  
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Distribution_At_Enhancer_KmeansCluster_lengthBased_20210607",
                       width = 8, height = 6)
    
  # only A.t (Activity cluster)
  tmp %>%
    dplyr::filter(species == "At", 
                  str_detect(track, "(Activity|^Expected|^Enhancer$)", negate = F)) %>%
    dplyr::mutate(track = str_replace_all(track, "^Enhancer$", "Enhancer_STARRseq"), 
                  track = factor(track, levels = c(str_c("ActivityCluster", 1:5), 
                                                   "Enhancer_STARRseq", 
                                                   "Expected"))) %>%
    ggplot(aes(x = track, fill = annotation, y = value)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      # facet_grid(. ~ species, scales = "free", space = "free") + 
      # scale_fill_npg() + 
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      labs(caption = "Length-based") + 
      scale_y_continuous(labels = scales::percent) + ylab("Percentage") +
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            axis.title.x = element_blank())
  
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Distribution_At_Enhancer_ActivityCluster_lengthBased_20210607",
                       width = 8, height = 6)
```

###### .Enrichment profile
```{r}
tmp <- df %>%
  dplyr::mutate(sigLable = toolkit_tyj$returnAsterisk(qvalue), 
                track = str_replace_all(track, c("Os_enhancer" = "Enhancer_STARRseq", 
                                                 "Dm_enhancer" = "Enhancer_STARRseq", 
                                                 "^enhancer$" = "Enhancer_STARRseq",
                                                 "^Enhancer$" = "Enhancer_STARRseq"))) %>%
  dplyr::filter(str_detect(track, "(^Random.+)", negate = T)) %>%
  dplyr::mutate(annotation = factor(annotation, levels = rev(c("Promoter", "FiveUTR", "FirstExon", "FirstIntron",
                                          "OtherExon", "OtherIntron", "LastExon", "LastIntron",
                                          "ThreeUTR", "DownStream", "Intergenic", "TSS2TTS"))), 
                species1 = str_replace_all(species, c("At" = "A.thaliana", "Os" = "O.sativa", "Dm" = "D.melanogaster","Hs" = "H.sapiens")),
                species1 = factor(species1, levels = c("A.thaliana", "O.sativa", "D.melanogaster", "H.sapiens")),
                species = factor(species, levels = c("At", "Os", "Dm", "Hs")), 
                l2fold = round(l2fold, digits = 2), 
                label = str_c(l2fold, "\n", sigLable, sep = ""),
                label1 = str_c(fold, "\n", sigLable, sep = ""))

## write to table
# tmp %>%
#   dplyr::select(-label) %>%
#   arrange(species, track, annotation) %>%
#   fwrite("./R4.Distribution_of_enhancers/GAT_overlap/All_GAT_results_20210607.csv", 
#          row.names = F)

## Enhancers in all species
  tmp %>%
    dplyr::filter(str_detect(track, "(Activity|Kmeans|Zhu|Meng|Wang)", negate = T)) %>%
    ggplot(aes(y = annotation, fill = annotation, x = l2fold)) + 
      geom_vline(xintercept = 0, linetype = 1, color = "grey45") +
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~species, space = "free") + 
      geom_hline(yintercept = 0, linetype  = 2, color = "grey45") +
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45", "grey55")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            legend.position = "none") +
      geom_text(aes(x = l2fold / 2, y = annotation, label = label1), size = 5/.pt) + 
      xlab(expression(log[2](Fold~change)))
      
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enrichment_Enhancers_four_species_lengthBased_fold_20210607",
  #                      width = 10, height = 8)
  
## At, different methods.
  tmp %>%
    dplyr::filter(str_detect(track, "(STARR|Zhu|Meng|Wang)"), 
                  species == "At") %>%
    dplyr::mutate(track = str_replace_all(track, "Enhancer_", "En_"),
                  track = factor(track, levels = c("En_STARRseq", "En_Zhu2015", 
                                                   "En_Wang2019", "En_Meng2021")),
                  track = toolkit_tyj$Format_En_lev(track)) %>%
    ggplot(aes(y = annotation, fill = annotation, x = l2fold)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~track, space = "free") + 
      geom_hline(yintercept = 0, linetype  = 2, color = "grey45") +
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45", "grey55")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            legend.position = "none", 
            panel.spacing = unit(0.08, "cm")) +
      geom_text(aes(x = l2fold / 2, y = annotation, label = label1), size = 5/.pt) + 
      xlab(expression(log[2](Fold~change)))
      
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enrichment_Enhancers_At_Methods_lengthBased_fold_20210607",
                       width = 8, height = 6)
  
  tmp %>%
    dplyr::filter(str_detect(track, "(STARR|Zhu|Meng|Wang)"), 
                  species == "At") %>%
    dplyr::mutate(track = str_replace_all(track, "Enhancer_", "En_"),
                  track = factor(track, levels = c("En_STARRseq", "En_Zhu2015", 
                                                   "En_Wang2019", "En_Meng2021")),
                  annotation = toolkit_tyj$Aten_FormatFeatures(annotation),
                  track = toolkit_tyj$Format_En_lev(track)) %>%
    ggplot(aes(x = annotation, fill = l2fold, y = l2fold)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(track ~ ., scales = "free_y") + 
      geom_hline(yintercept = 0, linetype  = 2, color = "grey45") +
      # scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45", "grey55")) +
      # scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1],
      #                      mid = brewer.pal(11, "RdYlGn")[6],
      #                      low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) +
      scale_fill_gradientn(colours = brewer.pal(11, "RdYlGn")[c(1,6, 7:11)] %>% rev) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      guides(fill = guide_colorbar(title = expression(log[2](Fold~change)))) +
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            legend.position = "bottom",
            axis.title.x = element_blank(),
            panel.spacing = unit(0.04, "cm"), 
            strip.text = element_text(size = 6),
            # legend.title = element_blank(),
            legend.box.margin = margin(-10,0,0,0)) +
      geom_text(aes(y = l2fold / 2, x = annotation, label = sigLable), size = 6/.pt) + 
      ylab("Relative enhancer enrichment")
      
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enrichment_Enhancers_At_Methods_lengthBased_fold_style1_20210607",
                       width = 8, height = 8)
  
## At, Kmeans cluster
  tmp %>%
    dplyr::filter(str_detect(track, "(STARR|Kmeans)"), 
                  species == "At") %>%
    dplyr::mutate(track = str_replace_all(track, c("KmeansCluster" = "C", "Enhancer_" = "En_")),
                  track = factor(track, levels = c(str_c("C", 1:4, sep = ""), "En_STARRseq"))) %>%
    ggplot(aes(y = annotation, fill = l2fold, x = l2fold)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~track, space = "free") + 
      geom_hline(yintercept = 0, linetype  = 2, color = "grey45") +
      # scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45", "grey55")) +
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0, n.breaks = 6 ) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      # labs(caption = "Length-based") + 
      ylab("Genomic features") +
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -8), 
            legend.position = "right", 
            legend.title = element_blank(),
            panel.spacing = unit(0.08, "cm")) +
      geom_text(aes(x = l2fold / 2, y = annotation, label = sigLable), size = 5/.pt) + 
      xlab(expression(log[2](Fold~change))) + 
      scale_x_continuous(breaks = seq(-5, 2.5, by = 2.5))
      
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enrichment_Enhancers_At_KmeansCluster_lengthBased_fold_20210607",
                       width = 9, height = 6)
  
## At, Activity cluster
  tmp %>%
    dplyr::filter(str_detect(track, "(STARR|Activity)"), 
                  species == "At") %>%
    ggplot(aes(y = annotation, fill = annotation, x = l2fold)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~track, space = "free") + 
      geom_hline(yintercept = 0, linetype  = 2, color = "grey45") +
      scale_fill_manual(values = c(pal_npg("nrc")(10), "grey45", "grey55")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      # ggtitle("Distribution of CREs according to genomic features") + 
      labs(caption = "Length-based") + 
      theme(legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(0, 0, 0, -5), 
            legend.position = "none") +
      geom_text(aes(x = l2fold / 2, y = annotation, label = label1), size = 1.8) + 
      xlab(expression(log[2](Fold~change)))
      
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enrichment_Enhancers_At_ActivityCluster_lengthBased_fold_20210607",
                       width = 15.5, height = 8)
```


## 4.2 TTS-TSS.

### .1 each type as a whole

```{r}
# calculate the relative bins according to gene body.
rm(list = ls())
load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20220516.bin")
source("D:/SUST/code/TanYongjun_code.R")
anno_table <- anno_list %>% 
  map(., as.data.frame) %>%
  purrr::reduce(rbind) %>%
  toolkit_tyj$relativeBins2TSS(flank.region = 5000, bins_flank = 50, bins_gene = 50, flank_scale = 1)

### cSTARRseq-specific, common, iSTARR-specific, PAS, Random.----------
# normalized to random
df <- anno_table %>%
  dplyr::select(method:Type, relativeBins) %>%
  dplyr::mutate(Type = str_replace_all(Type, "-", "_")) %>%
  dplyr::filter(!is.na(relativeBins)) %>%
  group_by(relativeBins, Type) %>%
  dplyr::summarise(Num = n()) %>%
  pivot_wider(id_cols = relativeBins, names_from = Type, 
              values_from = Num, values_fill = 0)

df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
  as.data.frame() %>%
  dplyr::rename(relativeBins = V1) %>%
  mutate(PAS = PAS - Random, 
         cSTARRseq_specific = cSTARRseq_specific - Random, 
         iSTARRseq_specific = iSTARRseq_specific - Random, 
         iSTARRseq_common  = iSTARRseq_common - Random, 
         cSTARRseq_common  = cSTARRseq_common - Random, 
         DHS_enhancer_common  = DHS_enhancer_common - Random, 
         DHS_enhancer_leaf_specific  = DHS_enhancer_leaf_specific - Random, 
         PE  = PE - Random, 
         Random = 0) %>%
  pivot_longer(cols = -relativeBins, 
               names_to = "Type", 
               values_to = "Count")
  

# plot all
library(ggformula)
df %>%
  dplyr::filter(Type != "Random", Type != "cSTARRseq_common") %>%
  # dplyr::mutate(Type = str_replace_all(Type, ".+common", "Common")) %>%
  ggplot() + 
    toolkit_tyj$theme_my +
    geom_rect(xmin = 0, xmax = 50, ymin = -3.4, ymax = -3, fill = "grey80") +
    geom_text(x = 25, y = -3.2, label = "Gene body", size = 2.5, color = "grey30") +
    geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
    # geom_line() + 
    geom_spline(aes(x = relativeBins, color = Type, y = Count)) +
    # scale_color_manual(values = pal_npg(palette = "nrc")(9)[c(1, 2, 6, 4)]) +
    # scale_linetype_manual(values = c(1,1,1,2)) +
    scale_color_npg() +
    xlab(label = "Distance to TSS/TTS (kb)") + 
    ylab(label = "Relative density") + 
    scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                       labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                  "2.5", "5")) +
    theme(legend.key.size = unit(0.4, units = "cm"), 
        legend.position = c(0.83, 0.77),
        # legend.position = "none",
        legend.title = element_blank(), 
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) 

# toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_iSTARRseq_cSTARRseq_DHS_overlap",
#                      width = 12, height = 8)


### cSTARRseq, iSTARRseq, PAS, Random------------
# normalized to random
df <- anno_table %>%
  dplyr::select(method:Type, relativeBins) %>%
  dplyr::mutate(Type = str_replace_all(method, "-", "_")) %>%
  dplyr::filter(!is.na(relativeBins)) %>%
  group_by(relativeBins, Type) %>%
  dplyr::summarise(Num = n()) %>%
  pivot_wider(id_cols = relativeBins, names_from = Type, 
              values_from = Num, values_fill = 0)
df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
  as.data.frame() %>%
  dplyr::rename(relativeBins = V1) %>%
  mutate(PAS = PAS - Random, 
         `cSTARR_seq` = `cSTARR_seq` - Random, 
         `iSTARR_seq` = `iSTARR_seq` - Random, 
         Enhancer_Zhu2015 = Enhancer_Zhu2015 - Random,
         Enhancer_Wang2019 = Enhancer_Wang2019 - Random, 
         Random = 0) %>%
  pivot_longer(cols = -relativeBins, 
               names_to = "Type", 
               values_to = "Count")
  

# plot all
library(ggformula)
df %>%
  dplyr::filter(Type != "Random", Type != "cSTARRseq_common") %>%
  dplyr::mutate(Type = str_replace_all(Type, ".+common", "Common")) %>%
  ggplot() + 
    toolkit_tyj$theme_my +
    geom_rect(xmin = 0, xmax = 50, ymin = -3.2, ymax = -2.8, fill = "grey80") +
    geom_text(x = 25, y = -3, label = "Gene body", size = 2.5, color = "grey30") +
    geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
    # geom_line() + 
    geom_spline(aes(x = relativeBins, color = Type, y = Count, linetype = Type)) +
    # scale_color_manual(values = pal_npg(palette = "nrc")(9)[c(1, 2, 3, 4, 6)]) +
    # scale_linetype_manual(values = c(1,1,1,1,2)) +
    xlab(label = "Distance to TSS/TTS (kb)") + 
    ylab(label = "Relative density") + 
    scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                       labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                  "2.5", "5")) +
    theme(legend.key.size = unit(0.4, units = "cm"), 
        legend.position = c(0.83, 0.77),
        # legend.position = "none",
        legend.title = element_blank(), 
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) 

# toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_iSTARRseq_cSTARRseq_DHS",
#                      width = 12, height = 8)

# plot cSTARR + two DHS enhancer sets.
library(ggformula)
df %>%
  dplyr::filter(Type != "Random", Type != "iSTARR_seq", Type != "PAS") %>%
  dplyr::mutate(Type = str_replace_all(Type, c("cSTARR_seq" = "Enhancer_STARRseq"))) %>%
  ggplot() + 
    toolkit_tyj$theme_my +
    geom_rect(xmin = 0, xmax = 50, ymin = -3.2, ymax = -2.8, fill = "grey80") +
    geom_text(x = 25, y = -3, label = "Gene body", size = 2.5, color = "grey30") +
    geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
    # geom_line() + 
    geom_spline(aes(x = relativeBins, color = Type, y = Count, linetype = Type)) +
    scale_color_manual(values = pal_npg(palette = "nrc")(9)[c(1, 2, 3, 4, 6)]) +
    scale_linetype_manual(values = c(1,1,1,1,2)) +
    xlab(label = "Distance to TSS/TTS (kb)") + 
    ylab(label = "Relative density") + 
    scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                       labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                  "2.5", "5")) +
    theme(legend.key.size = unit(0.4, units = "cm"), 
        legend.position = c(0.83, 0.77),
        # legend.position = "none",
        legend.title = element_blank(), 
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) 

# toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_cSTARRseq_DHS",
#                      width = 8, height = 6)

```

*Add intronic enhancer 202106*
```{r}
# calculate the relative bins according to gene body.
rm(list = ls())
load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")
# load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_iSTARR_cSTARR_20210415.bin")
source("D:/SUST/code/TanYongjun_code.R")
anno_table <- anno_list %>% 
  map(., as.data.frame) %>%
  purrr::reduce(rbind) %>%
  toolkit_tyj$relativeBins2TSS(flank.region = 5000, bins_flank = 50, bins_gene = 50, flank_scale = 1) %>%
  dplyr::mutate(method = str_replace_all(method, c("IntronicEnhancer" = "Enhancer_Meng2021", 
                                           "cSTARR-seq" = "Enhancer_STARRseq")))

### Enhancer, PAS, Random------------
# normalized to random
  df <- anno_table %>%
    dplyr::select(method:Type, relativeBins) %>%
    dplyr::mutate(Type = str_replace_all(method, "-", "_")) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, Type) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = Type, 
                values_from = Num, values_fill = 0) %>%
    pivot_longer(cols = matches("(Enhancer|PAS|STARR|Random)")) %>%
    pivot_wider(id_cols = c(relativeBins), names_from = name, values_from = value) 
  
  df <- cbind(df$relativeBins, scale(df[,2:ncol(df)]))%>%
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    pivot_longer(cols = matches("(Enhancer|PAS|STARR)")) %>%
    dplyr::mutate(NormedValue = value - Random)
  
  # plot all
  library(ggformula)
  df %>%
    dplyr::filter(str_detect(name, "(iSTARR|PAS)", negate = T)) %>%
    dplyr::mutate(name = factor(name, levels = c("Enhancer_STARRseq", "Enhancer_Zhu2015", 
                                                 "Enhancer_Wang2019", "Enhancer_Meng2021", 
                                                 "PAS", "iSTARR-seq", "Random"))) %>%
    dplyr::filter(str_detect(name, "Random|STARR|Enhancer")) %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.2, ymax = -2.8, fill = "grey80") +
      geom_text(x = 25, y = -3, label = "Gene body", size = 1.8, color = "grey30") +
      geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      # geom_line() +
      geom_spline(aes(x = relativeBins, color = name, y = NormedValue, linetype = name)) +
      # scale_color_manual(values = pal_npg(palette = "nrc")(9)[c(1, 2, 3, 4, 6)]) +
      scale_color_npg() +
      scale_linetype_manual(values = c(1,2,2,2,2)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.83, 0.22),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
          legend.key.height = unit(0.25, units = "cm")) 
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_STARR_DHS_20210606",
  #                      width = 8, height = 6)

  
# Not normalized
  df <- anno_table %>%
    dplyr::select(method:Type, relativeBins) %>%
    dplyr::mutate(Type = str_replace_all(method, "-", "_")) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, Type) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = Type, 
                values_from = Num, values_fill = 0) %>%
    pivot_longer(cols = matches("(Enhancer|PAS|STARR|Random)")) %>%
    pivot_wider(id_cols = c(relativeBins), names_from = name, values_from = value) 
  
  df <- cbind(df$relativeBins, scale(df[,2:ncol(df)]))%>%
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    pivot_longer(cols = matches("(Enhancer|PAS|STARR|Random)")) 
  
  # plot all
  library(ggformula)
  df %>%
    dplyr::filter(str_detect(name, "(iSTARR|PAS)", negate = T)) %>%
    dplyr::mutate(name = factor(name, levels = c("Enhancer_STARRseq", "Enhancer_Zhu2015", 
                                                 "Enhancer_Wang2019", "Enhancer_Meng2021", 
                                                 "PAS", "iSTARR-seq", "Random"))) %>%
    dplyr::filter(str_detect(name, "Random|STARR")) %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.2, ymax = -2.8, fill = "grey80") +
      geom_text(x = 25, y = -3, label = "Gene body", size = 1.8, color = "grey30") +
      # geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      # geom_line() +
      geom_spline(aes(x = relativeBins, color = name, y = value, linetype = name)) +
      # scale_color_manual(values = pal_npg(palette = "nrc")(9)[c(1, 2, 3, 4, 6)]) +
      scale_color_manual(values = toolkit_tyj$AtEn_YNRandom_color[c(1,2)]) +
      scale_linetype_manual(values = c(1,2,2,2,2)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.83, 0.82),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
          legend.key.height = unit(0.25, units = "cm")) 
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_notNormed_TTS_TSS_STARR_DHS_20210606",
                       width = 8, height = 6)

```
### .2 enhancer cluster

```{r}
## calculate the relative bins according to gene body, and prepare.
  rm(list = ls())
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")
  anno_list <- anno_list[c("cSTARR-seq", "Random")]
  source("D:/SUST/code/TanYongjun_code.R")
  
  # load cluster information
  enhancer <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_RPKM_20210321.csv.gz") %>%
    dplyr::select(seqnames, startBK, endBK, method, clusterK4, clusterK7) %>%
    split(.$method)
  for (i in names(anno_list)) {
    anno_list[[i]] <- full_join(as.data.frame(anno_list[[i]]), enhancer[[i]])
  }
  
  # list to table
  anno_table <- anno_list %>% 
    map(., as.data.frame) %>%
    purrr::reduce(rbind) %>% 
    dplyr::filter(str_detect(method, "(cSTARR-seq|Random)")) %>%
    toolkit_tyj$relativeBins2TSS(flank.region = 5000, bins_flank = 50, bins_gene = 50, flank_scale = 1) %>%
    mutate(clusterK4 = as.character(clusterK4), clusterK7 = as.character(clusterK7)) %>%
    full_join(., dplyr::mutate(., clusterK4 = "All", 
                       clusterK7 = "All")) %>%
    dplyr::filter(!(method == "Random"& clusterK4 == 0)) %>%
    dplyr::mutate(clusterK4 = ifelse(method == "Random", "Random", clusterK4),
                  clusterK7 = ifelse(method == "Random", "Random", clusterK7))


### each cluster of enhancer K=4 (deduct random)
  # normalized to random
  df <- anno_table %>%
    dplyr::select(clusterK4, relativeBins) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, clusterK4) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = clusterK4, 
                values_from = Num, values_fill = 0)
  df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    mutate(`C0` = `0` - Random,
           `C1` = `1` - Random,
           `C2` = `2` - Random,
           `C3` = `3` - Random,
           `C4` = `4` - Random,
           `All` = `All` - Random,
           Random = 0) %>%
    pivot_longer(cols = -relativeBins, 
                 names_to = "Type", 
                 values_to = "Count") %>%
    dplyr::filter(str_detect(Type, "(All|Random|C)"))
    
  
  # plot
  library(ggformula)
  df %>%
    dplyr::filter(Type != "Random", Type != "C0") %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.4, ymax = -3, fill = "grey80") +
      geom_text(x = 25, y = -3.2, label = "Gene body", size = 2.5, color = "grey30") +
      geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      geom_spline(aes(x = relativeBins, color = Type, y = Count, linetype = Type)) +
      # geom_line(aes(x = relativeBins, color = Type, y = Count)) +
      scale_color_manual(values = c("grey45", pal_npg(palette = "nrc")(4))) +
      # scale_color_npg() +
      scale_linetype_manual(values = c(2, 1,1,1,1)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.83, 0.27),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_cSTARRseq1.3_clusterK4_20210322",
                       width = 8, height = 6)

### each cluster of enhancer K=4 (not normed)
  df <- anno_table %>%
    dplyr::select(clusterK4, relativeBins) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, clusterK4) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = clusterK4, 
                values_from = Num, values_fill = 0)
  df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    # mutate(`Cluster-0` = `0` - Random,
    #        `Cluster-1` = `1` - Random,
    #        `Cluster-2` = `2` - Random,
    #        `Cluster-3` = `3` - Random,
    #        `Cluster-4` = `4` - Random,
    #        `All` = `All` - Random,
    #        Random = 0) %>%
    dplyr::rename(C0 = `0`,
                  C1 = `1`,
                  C2 = `2`,
                  C3 = `3`,
                  C4 = `4`,
                  All = All) %>%
    pivot_longer(cols = -relativeBins, 
                 names_to = "Type", 
                 values_to = "Count") %>%
    dplyr::filter(str_detect(Type, "(All|Random|C)"))
    
  
  # plot
  library(ggformula)
  df %>%
    dplyr::filter(str_detect(Type, "All|C0", negate = T)) %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.4, ymax = -3, fill = "grey80") +
      geom_text(x = 25, y = -3.2, label = "Gene body", size = 2.5, color = "grey30") +
      # geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      geom_spline(aes(x = relativeBins, color = Type, y = Count, linetype = Type)) +
      # geom_line(aes(x = relativeBins, color = Type, y = Count)) +
      scale_color_manual(values = c(toolkit_tyj$AtEn_KmeanC_color, "grey35")) +
      # scale_color_npg() +
      scale_linetype_manual(values = c(1,1,1,1,2)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.83, 0.77),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_NotNormed_TTS_TSS_cSTARRseq1.3_clusterK4_20210322",
                       width = 8, height = 6)

```

### .3 Activity cluster

```{r}
## calculate the relative bins according to gene body, and prepare.
  rm(list = ls())
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")
    anno_list <- anno_list[c("cSTARR-seq", "Random")]
  source("D:/SUST/code/TanYongjun_code.R")
  
  # load cluster information
  enhancer <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz") %>%
    dplyr::select(seqnames, startBK, endBK, method, ClusterActivity) %>%
    split(.$method)
  for (i in names(anno_list)) {
    anno_list[[i]] <- full_join(as.data.frame(anno_list[[i]]), enhancer[[i]])
  }
  
  # list to table
  anno_table <- anno_list %>% 
    purrr::reduce(rbind) %>%
    dplyr::filter(str_detect(method, "(cSTARR-seq|Random)"), 
                  !(method == "cSTARR-seq" & enrichment <= 1.3)) %>%
    dplyr::mutate(ClusterActivity = ifelse(is.na(ClusterActivity), "Random", ClusterActivity)) %>%
    toolkit_tyj$relativeBins2TSS(flank.region = 5000, bins_flank = 50, bins_gene = 50, flank_scale = 1) %>%
    full_join(., dplyr::filter(., ClusterActivity != "Random") %>% dplyr::mutate(ClusterActivity = "All")) 

### each Activity cluster
  # normalized to random
  df <- anno_table %>%
    dplyr::select(ClusterActivity, relativeBins) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, ClusterActivity) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = ClusterActivity, 
                values_from = Num, values_fill = 0)
  df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    mutate(`Part1` = `Part1` - Random,
           `Part2` = `Part2` - Random,
           `Part3` = `Part3` - Random,
           `Part4` = `Part4` - Random,
           `Part5` = `Part5` - Random,
           `All` = `All` - Random,
           Random = 0) %>%
    pivot_longer(cols = -relativeBins, 
                 names_to = "Type", 
                 values_to = "Count") %>%
    dplyr::filter(str_detect(Type, "(All|Random|Part)"))
    
  
  # plot
  library(ggformula)
  df %>%
    dplyr::mutate(Type = factor(Type, levels = c("Part1", "Part2", "Part3", "Part4", "Part5", "All", "Random"))) %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.1, ymax = -2.8, fill = "grey80") +
      geom_text(x = 25, y = -2.95, label = "Gene body", size = 1, color = "grey30") +
      # geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      geom_spline(aes(x = relativeBins, color = Type, y = Count, linetype = Type)) +
      # geom_line(aes(x = relativeBins, color = Type, y = Count)) +
      scale_color_manual(values = c(carto_pal(5, "Temps"), "grey35", "grey70")) +
      # scale_color_npg() +
      scale_linetype_manual(values = c(1,1,1,1,1,2,3)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.89, 0.32),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_cSTARRseq1.3_ClusterActivity_deduct_Random_20210509",
                       width = 8, height = 6)

```

### .4 TS/HK enhancer

```{r}
## calculate the relative bins according to gene body, and prepare.
  rm(list = ls())
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")
  source("D:/SUST/code/TanYongjun_code.R")
    
  # load enhancer info (located near tissue-specific or house-keeping genes)
  tshk_enhancer <- fread("./R7.TFs/bed_files/Enhancer_TSHK_gene_up500_down500_20220517.bed.gz") %>%
    dplyr::rename(seqnames = V1, start = V2, end = V3, Gene_type = V4)
  
  anno_table <- anno_list[c("cSTARR-seq", "Random")] %>%
    map_dfr(., as.data.frame) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames), 
                  start = startBK, end = endBK) %>%
    left_join(., tshk_enhancer)
  table(anno_table$method, anno_table$Gene_type)
  
  # list to table
  anno_table <- anno_table %>% 
    toolkit_tyj$relativeBins2TSS(flank.region = 5000, bins_flank = 50, bins_gene = 50, flank_scale = 1) %>%
    mutate(Gene_type = ifelse(method == "Random", "Random", Gene_type),
           Gene_type = ifelse(is.na(Gene_type), "Distal_enhancer", Gene_type))
  table(anno_table$method, anno_table$Gene_type)

### each kind of enhancer (deduct random)
  # normalized to random
  df <- anno_table %>%
    dplyr::select(Gene_type, relativeBins) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, Gene_type) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = Gene_type, 
                values_from = Num, values_fill = 0)
  df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    mutate(`Distal_enhancer` = `Distal_enhancer` - Random,
           `TSenhancer` = `TSenhancer` - Random,
           `HKenhancer` = `HKenhancer` - Random,
           Random = 0) %>%
    pivot_longer(cols = -relativeBins, 
                 names_to = "Type", 
                 values_to = "Count") %>%
    dplyr::mutate(Type = str_replace_all(Type, c("HKenhancer" = "House-keeping",
                                                    "TSenhancer" = "Tissue-specific",
                                                    "Distal_enhancer" = "Distal-enhancer")),
           Type = factor(Type, levels = c("House-keeping", "Tissue-specific", 
                                                    "Distal-enhancer", "Random")))
    
  
  # plot
  library(ggformula)
  df %>%
    dplyr::filter(Type != "Random") %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.4, ymax = -3, fill = "grey80") +
      geom_text(x = 25, y = -3.2, label = "Gene body", size = 2.5, color = "grey30") +
      geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      geom_spline(aes(x = relativeBins, color = Type, y = Count)) +
      # geom_line(aes(x = relativeBins, color = Type, y = Count)) +
      scale_color_manual(values = toolkit_tyj$AtEn_TSHKDisRan_color) +
      # scale_color_npg() +
      # scale_linetype_manual(values = c(2, 1,1,1,1)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.83, 0.27),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_TS_HK_enhancer_deduct_Random_20210322",
                       width = 8, height = 6)

  # not normed
  df <- anno_table %>%
    dplyr::select(Gene_type, relativeBins) %>%
    dplyr::filter(!is.na(relativeBins)) %>%
    group_by(relativeBins, Gene_type) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = relativeBins, names_from = Gene_type, 
                values_from = Num, values_fill = 0)
  df <- cbind(df$relativeBins, scale(as.matrix(df[,2:ncol(df)]))) %>% 
    as.data.frame() %>%
    dplyr::rename(relativeBins = V1) %>%
    # mutate(`Distal_enhancer` = `Distal_enhancer` - Random,
    #        `TSenhancer` = `TSenhancer` - Random,
    #        `HKenhancer` = `HKenhancer` - Random,
    #        Random = 0) %>%
    pivot_longer(cols = -relativeBins, 
                 names_to = "Type", 
                 values_to = "Count") %>%
    dplyr::mutate(Type = str_replace_all(Type, c("HKenhancer" = "House-keeping",
                                                    "TSenhancer" = "Tissue-specific",
                                                    "Distal_enhancer" = "Distal-enhancer")),
           Type = factor(Type, levels = c("House-keeping", "Tissue-specific", 
                                                    "Distal-enhancer", "Random")))
    
  
  # plot
  library(ggformula)
  df %>%
    # dplyr::filter(Type != "Random") %>%
    ggplot() + 
      toolkit_tyj$theme_my +
      geom_rect(xmin = 0, xmax = 50, ymin = -3.4, ymax = -3, fill = "grey80") +
      geom_text(x = 25, y = -3.2, label = "Gene body", size = 2.5, color = "grey30") +
      # geom_hline(yintercept = 0, linetype = 3, color = "grey40") +
      geom_spline(aes(x = relativeBins, color = Type, y = Count, linetype = Type)) +
      # geom_line(aes(x = relativeBins, color = Type, y = Count)) +
      scale_color_manual(values = toolkit_tyj$AtEn_TSHKDisRan_color) +
      # scale_color_npg() +
      scale_linetype_manual(values = c(1,1,1,2)) +
      xlab(label = "Distance to TSS/TTS (kb)") + 
      ylab(label = "Relative density") + 
      scale_x_continuous(breaks = c(-50, -25, 0, 10, 20, 30, 40, 50, 75, 100),
                         labels = c("-5", "-2.5", "TSS", "20%", "40%", "60%", "80%", "TTS",
                                    "2.5", "5")) +
      theme(legend.key.size = unit(0.4, units = "cm"), 
          legend.position = c(0.83, 0.77),
          # legend.position = "none",
          legend.title = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
      
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Enhancer_distribution_TTS_TSS_TS_HK_enhancer_Not_deduct_Random_20210322",
                       width = 8, height = 4.5)

```




## xxx 4.3 TAD

*TAD and loop were identified use Arrowhead and HICCUPS respectively at resoluton 500bp and 1000bp.* *Not used, because the identified TAD were not obvious in heatmap of Hic* Distribution of enhancers/promoters according to TAD border.
The list of housekeeping genes were aquired from article (sup data S4) pulished by Cheng et.al (<https://doi.org/10.1111/tpj.13415>; Araport11: a complete reannotation of the Arabidopsis thaliana reference genome, 2016, the plant journal).

```{r}
## prepare data set
  source("D:/SUST/code/TanYongjun_code.R")
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up3000_down0_iSTARR_cSTARR_20210308.bin")
  
# promoter (House-keeping / Tissue-specific genes)
  pro_core <- promoters(TxDb.Athaliana.BioMart.plantsmart28, upstream = 50, downstream = 50, use.names = T) %>%
    as.data.frame() %>%
    dplyr::select(-tx_id) %>%
    dplyr::rename(rep = tx_name) %>%
    dplyr::mutate(method = "Promoter", ID = "Promoter", Type = "Promoter")
  houseKeeping <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv")
  pro_core$Type <- "TissueSpecific"
  pro_core$Type[pro_core$rep %in% houseKeeping$`Transcript ID`] <- "HouseKeeping"
  # table(pro_core$Type)
  
# recover width of cre
  enhancer$start <- enhancer$startBK
  enhancer$end <- enhancer$endBK
  enhancer$width <- enhancer$end - enhancer$start
  enhancer$start[enhancer$width < 1 ] <- enhancer$start[enhancer$width < 1] - 1
  enhancer$end[enhancer$width < 1 ] <- enhancer$end[enhancer$width < 1] + 1
  enhancer$width <- enhancer$end - enhancer$start
  boxplot(enhancer$width ~ enhancer$method)
  
# merge
  all_cre <- full_join(enhancer, pro_core)
  table(all_cre$method)
  table(all_cre$seqnames)
  # fwrite(all_cre, file = "./Enahncers_DHS_STARRseq_H3K4_Pro_Random_20210313.csv.gz")

# # TAD
#   tad500 <- fread("./refgenome/NC2018_high_arrowhead_domain_500_TYJ/500_blocks.bedpe") %>%
#     dplyr::select(-8) %>%
#     dplyr::mutate(chr1 = str_c("chr", chr1, sep = ""))
#   tad1000 <- fread("./refgenome/NC2018_high_arrowhead_domain_1000_TYJ/1000_blocks.bedpe") %>%
#     dplyr::select(-8) %>%
#     dplyr::mutate(chr1 = str_c("chr", chr1, sep = ""))
#   tad <- rbind(tad500, tad1000) %>%
#     makeGRangesFromDataFrame(seqnames.field = "chr1", start.field = "x1", 
#                              end.field = "x2", keep.extra.columns = T)
#   hist(width(tad), main = "Width of TAD")
#   tad %>%
#     as.data.frame() %>% 
#     dplyr::select(1:3) %>%
#     fwrite(file = "./refgenome/TAD_500_1000_merged.bed", sep = "\t", col.names = F)
#   NA
  
```

## xxx 4.4 loop locus

*TAD and loop were identified use Arrowhead and HICCUPS respectively at resoluton 500bp and 1000bp.* Distribution of enhancers/promoters according to loop locus.

## 4.5 Interactions

Generate the decay curve of interactions between CREs to flanking regions.
*The following code were modified from code used in anlaysing of human CREs* Did CRE act as hub in long-range interactions?
`straw` can used to extract data from `.hic`, it was write in C++ and python and can be use in shell, python, and R (return data.frame).

##### - 4.5.1 count interactions in .hic file using `Straw.`

*Attention: the binstart need covered by region difined in chr:xxx:xxx, a optimal choice was set this region equal to the binsize.*

```{r straw-R demo code}
# run straw through R API.
remotes::install_github("aidenlab/straw/R")
hic.data.frame <- strawr::straw("observed", "KR", "/path/to/file.hic", "11", "11", "BP", 10000)
```

*Did silencer/enhancer/promoter which located near to (\<10kb) other kind of CREs need to be exclude in calculation?*

> Use all CREs Interaction between CRE and flanking region (\<1MB) `/scratch/2021-03-11/bioTanyj/At_enhancer/scripts/4.5.1_Interactions_between_CRE_flanking_regions.R`

##### - Collect results, calculate Median/Mean interactions in each bins.

less long-range interactions were calculated between CREs and distal bins.
add cluster/Segway information based on position.
(nearest CRE)

*all CREs*

```{r}
source("/work/bio-tanyj/soft/TanYongjun_code.R")
setwd("/scratch/2021-03-21/bioTanyj/At_enhancer/R4.Distribution")

# collect
summary_all <- list()
for (i in list.files("./", pattern = ".+extracted.tab$")) {
  cat(format(Sys.time()), "  ", i, "\n")
  tmp <- fread(paste("./", i, sep = "")) %>%
    mutate(Distance = abs(x - y)) %>%
    dplyr::select(-x, -y) %>%
    group_by(Distance) %>%
    summarise(Median = median(counts, na.rm = T), 
              Mean = mean(counts, na.rm = T), 
              Num = n()) %>%
    mutate(ID = str_replace_all(i, "_straw_extracted.tab", ""))
  summary_all <- c(summary_all, list(tmp))
}
summary_all <- purrr::reduce(summary_all, rbind)

save(summary_all, file = "Summary_of_interactions_CREs_vs_flanking_regions_20210329.bin")

```

##### - plot

*all CREs*

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
load("./R4.Distribution_of_enhancers/Hic_interaction_CRE_vs_Flanking/Summary_of_interactions_CREs_vs_flanking_regions_20210329.bin")

# each kind CRE as a whole
summary_all %>%
  dplyr::filter(str_detect(ID, "(^cSTARR-seq$|^Promoter$|Enhancer_Zhu2015|Enhancer_Wang2019)")) %>%
  # mutate(cis_element = str_replace_all(cis_element, "silencer", "Silencer"),
  #        cis_element = factor(cis_element, levels = c("ActiveEnhancer", "PoisedEnhancer",
  #                                                     "Promoter", "Silencer", "Random"))) %>%
  ggplot(aes(x = Distance, y = Median, color = ID, linetype = ID)) + 
    # geom_point() + 
    geom_line(size = 0.2, alpha = 0.8) + 
    toolkit_tyj$theme_my + 
    # facet_grid(. ~ cell_line) + 
    scale_color_manual(values = c(pal_npg()(9)[c(8, 5, 3, 4)], "grey45")) +
    scale_linetype_manual(values = c("solid", "solid", "solid", "solid",  "dotted")) + 
    scale_x_log10() +
    scale_y_log10() +
    labs(color = "Cis-elements", linetype = "Cis-elements") + 
    theme(legend.position = "bottom") + 
    ylab("Interactions") + 
    xlab("Distance (bp)") + 
    # coord_cartesian(xlim = c(1, 50000)) +
    ggtitle("Interactions between CREs and flanking regions (all CREs)")
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Hic_interaction_CRE_vs_Flanking/Interactions_CREs_vs_flanking_regions_20210329", 
                       width = 16, height = 12, bg = "white")
```

## 4.6 SuperEnhancer
### .1 identification
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
## load CREs 
  enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
    dplyr::filter(!(str_detect(method, "STARR", negate = F) & enrichment <= 1.3)) %>%
    dplyr::mutate(V4 = str_replace_all(ClusterActivity, "Part", "Act"),
                  KmeansCluster = str_c("C", clusterK4))
  df <- enhancer %>% 
    dplyr::filter(method == "cSTARR-seq")
    
## define clustered enhancer with distance less than 600
  MaxDist <- 600 # between border of two adjacent enhancers.
  NumIndividualEnhancer <- 4
  
  # clustered enhancers (Number of individual enhancers >1)
  dfGR_reduce <- df %>% 
    makeGRangesFromDataFrame() %>%
    GenomicRanges::resize(width = (600 + MaxDist), fix = "center") %>%
    GenomicRanges::reduce()
  ol <- findOverlaps(makeGRangesFromDataFrame(df), dfGR_reduce)
  superEn <- dfGR_reduce[width(dfGR_reduce) > (600 + MaxDist)] %>%
    as.data.frame() %>%
    dplyr::mutate(seqnames = as.character(seqnames))

 # filtration based on number of individual enhancers in SuperEnhancer.
  ol1 <- findOverlaps(makeGRangesFromDataFrame(superEn), 
                      makeGRangesFromDataFrame(df))
  NumIndEnCount <- ol1@from %>%
    table() %>%
    as.data.frame()
  superEn$NumIndEn <- NA
  superEn$NumIndEn[NumIndEnCount[,1]] <- NumIndEnCount[,2]
  superEn <- superEn %>%
    dplyr::filter(NumIndEn >= NumIndividualEnhancer)
  
  # Number of enhancers in SuperEnhancer
  p <- superEn %>%
    ggplot(aes(x = NumIndEn)) + 
      geom_histogram() + 
      theme() + 
      toolkit_tyj$theme_my + 
      ggtitle("Number of enhancers in each SuperEnhancer") + 
      xlab("Number of enhancers") + 
      labs(caption = str_c("Max distance between center of two adjacent enhancers:", MaxDist, 
                           "bp\nMin number of enhancers in one superenhancer:", NumIndividualEnhancer))
  p
  # toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/SuperEnhancer/Number_of_enhancers_in_SuperEnhancer_MaxDist_",
  #                                              MaxDist, "bp_Number_", NumIndividualEnhancer, sep = ""),
  #                      width = 8, height = 6)
  

  # resize super enhancer
  superEn <- superEn %>%
    dplyr::mutate(start = start + MaxDist / 2, 
                  end = end - MaxDist /2)
  
  # superEn %>%
  #   dplyr::mutate(seqnames =str_c("chr", seqnames, sep = "")) %>%
  #   fwrite(file = paste("./R4.Distribution_of_enhancers/SuperEnhancer/List_of_SuperEnhancer",
  #                       MaxDist, "bp_Number_", NumIndividualEnhancer, 
  #                       ".bed", sep = ""), 
  #          col.names = F, row.names = F, sep = "\t")
  
  # df %>%
  #   dplyr::select(1:3) %>%
  #   fwrite(file = "./R4.Distribution_of_enhancers/SuperEnhancer/Enhancer_STARRseq_1.3.bed", 
  #          row.names = F, col.names = F, sep = "\t")
```

### .2 properties

#### . ChrStates, Kmeans
```{r}
## Properties of SuperEnhancer
  # genome distribution of super enhancer
  centromereAt <- fread("./refgenome/TAIR9_GFF3_assemblies.gff", skip = 1, header = F) %>%
    dplyr::filter(str_detect(V9, "CEN")) %>%
    dplyr::group_by(V1) %>%
    summarise(start = mean(V4) / 1e6, 
              end = mean(V5) / 1e6) %>%
    dplyr::mutate(V1 = as.numeric(str_replace_all(V1, "Chr", "")), 
                  width = end - start)
  
  library(TxDb.Athaliana.BioMart.plantsmart28)
  chr_info <- data.frame(x = names(seqlengths(TxDb.Athaliana.BioMart.plantsmart28)), 
                         length = seqlengths(TxDb.Athaliana.BioMart.plantsmart28) / 1e6)
  p <- ggplot(data = superEn) + 
    geom_col(data = chr_info, aes(x = x, y = length), fill = "grey75", 
             width = 0.1) +
    # geom_vline(aes(xintercept = seqnames),  linetype  = 2, color = "grey45") +
    geom_tile(aes(x = V1, y = (start + end) / 2, width = 0.13, height = 0.5), 
              data = centromereAt, fill = "grey40") +
    # geom_tile(aes(x = seqnames, y = (start + end) / 2e6, height = width / 3000, 
    #               width = 0.1), 
    #           alpha = 1/2, color = "black") +
    geom_point(aes(x = seqnames, y = (start + end) / 2e6), 
          alpha = 2/3, color = "steelblue", shape = "-", size = 6) +
    toolkit_tyj$theme_my + 
    scale_y_reverse() + 
    xlab("Chromosome") +
    ggtitle("Position of Supencer Enhancers") + 
    labs(caption = str_c("Max distance between center of two adjacent enhancers:", MaxDist, 
                         "bp\nMin number of enhancers in one superenhancer:", NumIndividualEnhancer))
  p
  # toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/SuperEnhancer/Distribution_of_Clustered_enhancer_",
  #                                              MaxDist, "bp_Number_", NumIndividualEnhancer, sep = ""),
  #                      width = 12, height = 8, plot = p)

  # Width of super enhancer
  p <- superEn %>%
    dplyr::mutate(bins = cut(width, breaks = c(seq(3000, 6000, by = 500), "Inf"))) %>%
    group_by(bins) %>%
    summarise(Num = n()) %>%
    ggplot(aes(x = bins, y = Num)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      xlab("Width of clustered Enhancers") + 
      ylab("Number") + 
      ggtitle("Width of clustered enhancer")
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/SuperEnhancer/Width_of_Clustered_enhancer_",
  #                                              MaxDist, "bp_Number_", NumIndividualEnhancer, sep = ""),
  #                      width = 8, height = 6, plot = p)
  
  # ChrState of super enhancer
  df$ClusterType <- "Dispersed"
  superEnGR <- makeGRangesFromDataFrame(superEn)
  dfGR <- makeGRangesFromDataFrame(df)
  ol2 <- findOverlaps(dfGR, superEnGR, minoverlap = 1)
  df$ClusterType[ol2@from] <- "Clustered"
  table(df$ClusterType)
  sum(superEn$NumIndEn)
  # fwrite(df, "Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz")
  
  ## add additional chromatin states.
  cre_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210416.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq") %>%
    unique() %>%
    makeGRangesFromDataFrame(keep.extra.columns = T) %>%
    resize(width = 600, fix = "center") %>%
    as.data.frame() %>%
    dplyr::select(-width, -strand) %>%
    pivot_wider(id_cols = c(seqnames, start, end, method), names_from = "chrState", values_from = "score") %>%
    dplyr::mutate(seqnames = str_replace_all(seqnames, "chr", ""))
    
  
  dfNew <- df %>%
    dplyr::select(seqnames:endBK, Dim.1:ClusterType) %>%
    left_join(., cre_score)
  dfNew <- dfNew %>%
    dplyr::select(seqnames:endBK, ATACseq:RNAPOII, H2AW:H4K16ac, Dim.1:ClusterType)
  
  # plot
  library(ggsignif)

  p <- dfNew %>% 
    dplyr::select(ClusterType, enrichment, ATACseq:H4K16ac) %>%
    pivot_longer(cols = enrichment:H4K16ac) %>%
    dplyr::filter(str_detect(name, "ATAC|enrichment|FAIRE", negate = T)) %>%
    dplyr::mutate(name = factor(str_replace_all(name, c("H2A" = "H2A.",
                                                        "^H2A.$" = "H2A")),
                                levels = str_replace_all(toolkit_tyj$AtEpi_order, 
                                                         c("H2A" = "H2A.",
                                                          "^H2A.$" = "H2A")))) %>%
    ggplot(aes(x = ClusterType, y = value, color = ClusterType)) + 
      geom_boxplot(fill = "transparent", outlier.alpha = 0, width = 0.2) +
      geom_violin(fill = "transparent") + 
      facet_wrap("name", scales = "free_y", ncol = 5) + 
      theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1), 
            legend.position = "none") + 
      toolkit_tyj$theme_my +
      scale_color_manual(values = toolkit_tyj$AtEn_CluDis_color) +
      scale_y_log10(expand = expansion(mult = c(0.1, 0.28))) + 
      geom_signif(comparisons = list(c("Clustered", "Dispersed")), 
                  color = "grey30", vjust = -0.3, test = "wilcox.test", size = 0.2,
                  textsize = 6/.pt)  + 
      # labs(caption = str_c("Max distance between boundaries of two adjacent enhancers:", MaxDist, 
      #                      "bp\nMin number of enhancers in one superenhancer:", NumIndividualEnhancer)) +
      xlab("Type")
  p
  
  toolkit_tyj$SavePlot(filename_prefix = paste("./R4.Distribution_of_enhancers/SuperEnhancer/Chrstates_cluster_vs_dispersed_enhancers_",
                                               MaxDist, "bp_Number_", NumIndividualEnhancer, sep = ""),
                       width = 12, height = 10, plot = p)
```

#### . Cluster
```{r}
  # Number of Clustered/Dispersed enhancers in different ActivityCluster
  p <- table(df$ClusterType, df$V4) %>%
    as.data.frame() %>%
    ggplot(aes(x = Var1, y = Freq, fill = Var1)) + 
      geom_col(alpha = 2/3) + 
      facet_grid(. ~ Var2) + 
      scale_fill_manual(values = toolkit_tyj$AtEn_CluDis_color) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "none", 
            axis.title.x = element_blank()) + 
      ylab("Number") +
      # ggtitle("Number of two kind of enhancers in five ActivityClusters") + 
      # labs(caption = str_c("Max distance between boundaries of two adjacent enhancers:", MaxDist, 
      #                      "bp\nMin number of enhancers in one superenhancer:", NumIndividualEnhancer)) + 
      toolkit_tyj$theme_my
  p
  
  toolkit_tyj$SavePlot(paste("./R4.Distribution_of_enhancers/SuperEnhancer/Number_of_Clustered_Dispersed_enhancer_vs_ActivityCluster_",
                             MaxDist, "bp_Number_", NumIndividualEnhancer, sep = ""),
                       width = 6, height = 5.5, plot = p)
  
  # Number of Clustered/Dispersed enhancers in different KmeansCluster
  p <- table(df$ClusterType, df$KmeansCluster) %>%
    as.data.frame() %>%
    dplyr::filter(Var2 != "C0") %>%
    ggplot(aes(x = Var1, y = Freq, fill = Var1)) + 
      geom_col(alpha = 2/3) + 
      scale_fill_manual(values = toolkit_tyj$AtEn_CluDis_color) +
      facet_grid(. ~ Var2) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "none", 
            axis.title.x = element_blank()) + 
      # ggtitle("Number of two kind of enhancers in KmeansClusters") + 
      # labs(caption = str_c("Max distance between boundaries of two adjacent enhancers:", MaxDist, 
      #                      "bp\nMin number of enhancers in one superenhancer:", NumIndividualEnhancer))+ 
      ylab("Number") +
      toolkit_tyj$theme_my
  p
  
  toolkit_tyj$SavePlot(paste("./R4.Distribution_of_enhancers/SuperEnhancer/Number_of_Clustered_Dispersed_enhancer_vs_KmeansCluster_K4_",
                             MaxDist, "bp_Number_", NumIndividualEnhancer, sep = ""),
                       width = 6, height = 6, plot = p)
```


#### . TEs
Results illustrated in this part was calculated based on overlap (center of TE in enhancer or center of enhancer in TE).
20210513
```{r}
## overlap with TEs?
  tmp <- df %>%
    dplyr::mutate(TE_SuperFamily = ifelse(str_length(TE_SuperFamily) > 1, 
                                             "withTE", "nonTE")) %>%
    dplyr::group_by(TE_SuperFamily, ClusterType) %>%
    summarise(Num = n()) %>%
    pivot_wider(names_from = TE_SuperFamily, values_from = Num)
  
  # fwrite(tmp, file = "./R4.Distribution_of_enhancers/SuperEnhancer/Num_SuperEnhancer_ol_TEs_contingency_table_20210513.csv")
  # write.table(paste("The p value of fisher's exact test was ", fisher.test(tmp[,2:3])$p.val, sep = ""), 
  #        file = "./R4.Distribution_of_enhancers/SuperEnhancer/Num_SuperEnhancer_ol_TEs_contingency_table_20210513.csv", 
  #        append = T, row.names = F, col.names = F)
  
  # plot
  p <- tmp %>%
    dplyr::mutate(Total = nonTE + withTE,
                  Rate = withTE / Total, 
                  label = str_c(percent(Rate, accuracy = 0.01), "\n(", withTE, "/", Total, ")")) %>%
    ggplot(aes(x = ClusterType, y = Rate, fill = ClusterType)) + 
      geom_col(alpha = 2/3) + 
      scale_fill_manual(values = toolkit_tyj$AtEn_CluDis_color) +
      geom_text(aes(x = ClusterType, y = Rate / 2, label = label), size = 6/.pt) + 
      geom_signif(annotations = paste("Fisher's exact test\np=",
                                      formatC(fisher.test(tmp[,2:3])$p.val, digits = 1), sep = ""), 
                  xmin = 1, xmax = 2,
                  y_position = 0.89, textsize = 6/.pt, size = 0.2) + 
      toolkit_tyj$theme_my + 
      theme(legend.position = "none", axis.title.x = element_blank()) + 
      coord_cartesian(ylim = c(0, 1)) + 
      ylab("Percent of two enhancer types overlaped with TE") +
      scale_y_continuous(labels = scales::percent)
  p
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/SuperEnhancer/Percent_of_clustered_dispersed_enhancer_ol_TEs_FisherExacttest",
                       width = 5.25, height = 5, plot = p)
    
# Percent of clustered/dispersed enhancers in enhancer groups based on TE family.------------
  tmp <- df %>%
    dplyr::mutate(TE_SuperFamily = ifelse(str_length(TE_SuperFamily) > 1, 
                                             TE_SuperFamily, "nonTE")) %>%
    dplyr::group_by(TE_SuperFamily, ClusterType) %>%
    summarise(Num = n()) %>%
    pivot_wider(names_from = ClusterType, values_from = Num, values_fill = 0) %>%
    dplyr::mutate(Total = Clustered + Dispersed) %>%
    pivot_longer(cols = Clustered:Dispersed, names_to = "Type", values_to = "Value") %>%
    dplyr::mutate(Rate = Value / Total) 
  
  order_tmp <- tmp %>%
    dplyr::filter(Type == "Clustered") %>%
    arrange(desc(Rate)) %>%
    dplyr::select(TE_SuperFamily) %>%
    unlist()
  order_tmp <- c("nonTE", order_tmp[str_detect(order_tmp, "nonTE", negate = T)])
  
  # bar plot (Percent)
  tmp <- tmp %>%
    dplyr::mutate(TE_SuperFamily = factor(TE_SuperFamily, levels = order_tmp), 
                  label = str_c(percent(Rate, accuracy = 0.01), " (", Value, "/", Total, ")", sep = ""),
                  xlab = str_c(TE_SuperFamily, "\n(", Total, ")")) %>%
    arrange(TE_SuperFamily)
  tmp %>%
    
    ggplot(aes(x = TE_SuperFamily, y = Rate, fill = Type)) + 
      geom_col(alpha = 2/3) + 
      scale_fill_manual(values = toolkit_tyj$AtEn_CluDis_color) +
      # geom_text(aes(label = label, y = TE_SuperFamily, x = 0.5), 
      #           data = dplyr::filter(tmp, Type == "Clustered"), 
      #           color = "grey10", size = 6/.pt) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.3, units = "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      # ggtitle("Percent of clustered and dispersed enhancers in each enhancer group") + 
      toolkit_tyj$theme_my + 
      scale_y_continuous(labels = scales::percent) +
      scale_x_discrete(labels = unique(tmp$xlab))
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/SuperEnhancer/SuperEnhancer_ol_TEs_Rate_20210513",
                       width = 8, height = 6)
  
  # bar plot (Number)
  tmp <- tmp %>%
    dplyr::mutate(TextPos = Total + 300) %>%
    dplyr::mutate(TextPos = ifelse(TextPos > 2500, 2500, TextPos),
                  xlab = str_c(TE_SuperFamily, "\n(", Total, ")"))
  tmp %>%
    ggplot(aes(y = TE_SuperFamily, x = Value, fill = Type)) + 
      geom_col(alpha = 4/5) + 
      scale_fill_manual(values = toolkit_tyj$AtEn_CluDis_color) +
      geom_text(aes(label = label, y = TE_SuperFamily, x = TextPos), 
                data = dplyr::filter(tmp, Type == "Clustered_enhancer"), 
                color = "grey10", size = 2) +
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.3, units = "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      xlab("Number of enhancers") +
      ggtitle("Number of clustered and dispersed enhancers\nin each enhancer group")
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/SuperEnhancer/SuperEnhancer_ol_TEs_Number_20210513",
                       width = 7, height = 8)
  
# Percent of Clustered and dispersed enhancer overlaped with TE superfamily and family (In Part 4.7 TEs).

  
```


### .3 RPKM
```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancers <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz")
  gene_tab <- fread("./R6.expression/Gene_list_with_CREs_500TSS100_300TTS300_20210322.csv.gz") # Enhancer_STARRseq (>1.3)
  
# define genes flanking to CREs
  gene_tab <- gene_tab %>%
    dplyr::select(1:10, Gene_type)
  enhancers$ID <- str_c("cSTARR-seq", enhancers$seqnames, enhancers$start, 
                        "K4=", enhancers$clusterK4, "K7=", enhancers$clusterK7, 
                        sep = "_")
  match(unique(gene_tab$cre_GeneBody), enhancers$ID)
  
  # remove enhancers identified by Zhu and Wang in gene_tab
  RmDHSenhancer <- function(x){
    str_replace_all(x, "Enhancer[^#]+", "#") %>%
      str_replace_all("^#+$", "") %>%
      str_replace_all("^$", "NO") %>%
      return()
  }
  # test
  # gene_tab$cre_GeneBody %>% head(n = 100)
  # gene_tab$cre_GeneBody %>% RmDHSenhancer() %>% head(n = 100)
  
  gene_tab$cre_GeneBody <- RmDHSenhancer(gene_tab$cre_GeneBody)
  gene_tab$cre_TSS <- RmDHSenhancer(gene_tab$cre_TSS)
  gene_tab$cre_TTS <- RmDHSenhancer(gene_tab$cre_TTS)
  
  # quantile(str_count(gene_tab$cre_TTS, "#"))
  
  # reshape
  gene_tab <- gene_tab %>%
    separate(., cre_GeneBody, into = paste("CRE_Genebody", 1:10, sep = "-"), sep = "#", fill = "left", extra = "merge") %>%
    separate(., cre_TSS, into = paste("CRE_TSS", 1:10, sep = "-"), sep = "#", fill = "left", extra = "merge") %>%
    separate(., cre_TTS, into = paste("CRE_TTS", 1:10, sep = "-"), sep = "#", fill = "left", extra = "merge")
  
  gene_tab <- gene_tab %>%
    pivot_longer(cols = matches("^CRE.+"), names_to = "region") %>%
    dplyr::mutate(value = str_replace_na(value, "NO"), 
                  value = ifelse(str_length(value) > 1, value, "NO"), 
                  region  = str_replace_all(region, "(.+)\\-\\d+", "\\1")) %>%
    unique() %>%
    dplyr::rename(ID = value)
  
  gene_tab <- enhancers %>%
    dplyr::select(ID, ClusterType) %>%
    left_join(gene_tab, .)
  
  gene_tab$ClusterType <- ifelse(is.na(gene_tab$ClusterType), "non-enhancer", gene_tab$ClusterType)
  table(gene_tab$ClusterType)
  # fwrite(gene_tab, "./R4.Distribution_of_enhancers/SuperEnhancer/gene_table_clustered_dispersed_enhancer_info_20210518.csv.gz")
  
  # count CRE in TSS, TTS, and gene body.
  df <- gene_tab %>%
    group_by(gene_id, region, ClusterType, RPKM) %>%
    summarise(Num = n()) %>% 
    # dplyr::filter(ClusterType != "non-enhancer") %>%
    group_by(gene_id, ClusterType, RPKM) %>%
    summarise(Num = sum(Num)) %>%
    pivot_wider(names_from = ClusterType, values_from = Num, values_fill = 0) %>%
    dplyr::mutate(Dispersed_Type = ifelse(Dispersed_enhancer > 0, "Dispersed", ""), 
                  Clustered_Type = ifelse(Clustered_enhancer > 0, "Clustered", ""),
                  NoEnhancer = ifelse(`non-enhancer` > 0, "NoEnhancer", "") ) %>%
    dplyr::mutate(Type = str_c(Clustered_Type, Dispersed_Type, NoEnhancer, sep = "-"), 
                  Type = str_replace_all(Type, c("^--NoEnhancer$" = "NoEnhancer", 
                                                 "^-Dispersed-NoEnhancer$" = "Dispersed", 
                                                 "^Clustered-Dispersed-NoEnhancer$" = "Clustered-Dispersed", 
                                                 "^Clustered--NoEnhancer$" = "Clustered")))
  table(df$Type)
  
# plot
  df$RPKM[is.na(df$RPKM)] <- 0
  df_label <- df %>%
    group_by(Type) %>%
    summarise(Num = n(), Median = median(RPKM, na.rm = T), 
              Mean = mean(RPKM, na.rm = T)) %>%
    dplyr::mutate(label = str_c("Mean=", round(Mean, digits = 2), "\nN=", Num, sep = ""))
  
  p <- df %>%
    dplyr::filter(Type != "Clustered-Dispersed") %>%
    ggplot(aes(x = Type, y = RPKM + 1, color = Type)) + 
      geom_label(data = dplyr::filter(df_label, Type != "Clustered-Dispersed"), 
             aes(x = Type, y = 2000, label = label), 
             fill = "transparent", size = 1.5, 
             nudge_y = 0.2, label.padding = unit(0.05, "cm"), 
             fill = "white", 
             alpha = 0.8) + 
      geom_violin(fill = "transparent") + 
      geom_boxplot(fill ="transparent", outlier.alpha = 0, width = 0.04) +
      toolkit_tyj$theme_my + 
      theme(legend.position = "none") + 
      scale_color_npg() + 
      geom_signif(comparisons = list(c("Dispersed", "Clustered"), 
                                     c("Dispersed", "NoEnhancer"),
                                     c("NoEnhancer", "Clustered")), 
                  test = "wilcox.test",
                  step_increase = 0.07, 
                  color = "grey40", size = 0.3,
                  textsize = 2) +
      labs(caption = "Wilcox test") + 
      ylab("RPKM") + 
      xlab("Gene type") +
      scale_y_log10() + 
      scale_x_discrete(labels = c("Clustered\nenhancer", "Dispersed\nenhancer", "No Enhancer"))

  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/SuperEnhancer/Expression_genes_clustered_dispersed_enhancers", 
                       width = 8, height = 6, plot = p)
```

### .4 flanking genes.
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  gene_tab <- fread("./R4.Distribution_of_enhancers/SuperEnhancer/gene_table_clustered_dispersed_enhancer_info_20210518.csv.gz")
  table(gene_tab$ClusterType)
  gene_SuEn <- gene_tab %>%
    dplyr::filter(ClusterType == "Clustered_enhancer", 
                  seqnames != "Mt")
  length(unique(gene_SuEn$gene_id))
  unique(gene_SuEn$gene_id) %>% str_c(collapse = ",")
  
  # load discription of these genes.
  geneDescription <- fread("./refgenome/gene_description_20131231.txt/gene_description_20140101.tsv", sep = "\t")
  geneDescription$gene_id <- str_replace_all(geneDescription$V1, "(AT.+)\\..+", "\\1")
  
  gene_SuEn %>%
    left_join(., geneDescription) %>%
    fwrite("./R4.Distribution_of_enhancers/SuperEnhancer/Genes_with_clustered_enhancers_discription.csv")
```

### .5 DNA-derived enhancers
nearly all enhancers derived from the DNA super family located around the cemtromere of chromosome1!
And these enhancers was nearly all derived from the ATREP18 family which belong to the DNA super family.

```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  te <- fread("./refgenome/TAIR10_Transposable_Elements_20210510_aquired.txt") %>%
    dplyr::mutate(seqnames = str_replace_all(Transposon_Name, "AT(\\d)TE.+", "\\1"), 
                  strand = ifelse(orientation_is_5prime == "TRUE", "+", "-"), 
                  start = Transposon_min_Start, 
                  end = Transposon_max_End)
  table(te$Transposon_Super_Family, te$seqnames)
  te_DNA <- te %>%
    dplyr::filter(Transposon_Super_Family == "DNA")
  table(te_DNA$seqnames, te_DNA$Transposon_Family)
  
  te_DNA %>%
    ggplot() + 
    geom_bar(aes(x = seqnames), fill = "grey45") + 
      toolkit_tyj$theme_my +
      facet_wrap("Transposon_Family") +
      xlab("Chromosome") +
      ggtitle("Number of TEs belong to each families of the DNA super family")
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/SuperEnhancer/Family_size_of_the_DNA_super_family",
  #                      width = 10, height = 10, device = "png")
  
  # centromere of TAIR10
  centromereAt <- fread("./refgenome/TAIR9_GFF3_assemblies.gff", skip = 1, header = F) %>%
    dplyr::filter(str_detect(V9, "CEN")) %>%
    dplyr::group_by(V1) %>%
    summarise(start = mean(V4) / 1e6, 
              end = mean(V5) / 1e6) %>%
    dplyr::mutate(V1 = as.numeric(str_replace_all(V1, "Chr", "")), 
                  width = end - start)
  
  # Enhancers
  enhancers <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::mutate(Features = str_replace_all(annotation, "^([^\\()]+)\\s\\(.+$", "\\1")) %>%
    dplyr::select(-c(ATACseq:Dim.5)) %>%
    dplyr::rename(gene_id = geneId)
  
  en_cl <- enhancers %>%
    dplyr::filter(ClusterType == "Clustered_enhancer")
  table(en_cl$seqnames, en_cl$TE_Family, en_cl$TE_SuperFamily)
  
```



## 4.7 TEs
list of Transposable elements was download from TAIR.

### 4.7.1 findOverlap(Percentage)
*Percentage of enhancer overlaped with TE were calculated based on criteria:*
  # criterion for overlap: for TE < 600bp, half of TE overlapped with enhancer; for TE > 600 bp, center of enhancer located in TE.
Percentage of CRE hit by TEs (1bp overlap, or length) were also calculated by GAT in S4.7.2.

#### .1 TEs and findoverlap
```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  te <- fread("./refgenome/TAIR10_Transposable_Elements_20210510_aquired.txt") %>%
    dplyr::mutate(seqnames = str_replace_all(Transposon_Name, "AT(\\d)TE.+", "\\1"), 
                  strand = ifelse(orientation_is_5prime == "TRUE", "+", "-"), 
                  start = Transposon_min_Start, 
                  end = Transposon_max_End)
  
# summary (super family)---------------
  makeGRangesFromDataFrame(te) %>% GenomicRanges::reduce() %>% width() %>% sum(.)/1e6 # total length
  hist(te$end - te$start)
  quantile(te$end - te$start, probs = seq(0, 1, 0.05))
  
  # total length of each TE super family
  tmp <- te %>%
    dplyr::mutate(width = end - start) %>%
    group_by(Transposon_Super_Family) %>%
    summarise(Number = n(), `TotalLength (MB)` = sum(width) / 1e6, MedianLength = median(width)) %>%
    arrange(desc(Number))
  
  tmp %>%
    dplyr::mutate(Transposon_Super_Family = factor(Transposon_Super_Family, levels = rev(.$Transposon_Super_Family))) %>%
    pivot_longer(cols = Number:MedianLength) %>%
    ggplot(aes(y = Transposon_Super_Family, x = value)) + 
      geom_col(fill = "grey55") + 
      facet_grid(. ~ name, scales = "free_x") + 
      toolkit_tyj$theme_my +
      geom_text(aes(label = value), size = 1.5, nudge_x = 0)
  
  toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Number_TotalLength_of_TEs_superFamily",
                       width = 12, height = 6, device = "png")

  # length distribution of each  TE Super Family.
  te %>%
    group_by(Transposon_Super_Family) %>%
    dplyr::mutate(Num = n(),
                  width = end - start) %>%
    ungroup() %>%
    dplyr::mutate(Transposon_Super_Family = fct_reorder(.f = Transposon_Super_Family, .fun = max, .x = Num, .desc = F)) %>%
    ggplot(aes(x = width, y = Transposon_Super_Family)) + 
      geom_violin(scale = "width", size = 0.2) + 
      geom_boxplot(width = 0.2, outlier.shape = NA, size = 0.2) +
      # facet_grid(. ~ Transposon_Super_Family, scales = "free_x") + 
      toolkit_tyj$theme_my + 
      scale_x_log10() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) #+ 
      # ggtitle("Width of TEs of each super family")
  
  # toolkit_tyj$SavePlot("./R4.Distribution_of_enhancers/TEs/Violin_width_TEs_Super_Family",
  #                      width = 5, height = 6, device = "png")
  
  # size of each family
  summaryFamily <- te %>%
    group_by(Transposon_Super_Family, Transposon_Family) %>%
    count() %>% 
    arrange(desc(n))
  
  
  # Summary table of each TE superfamily
    te %>%
      dplyr::mutate(width = end - start) %>%
      group_by(Transposon_Super_Family, Transposon_Family) %>%
      dplyr::mutate(FamilyNum = n(), FamilyLength = sum(width), FamilyMeanLength = mean(width)) %>%
      group_by(Transposon_Super_Family) %>%
      dplyr::mutate(SuperFamilyNum = n(), SuperFamilyLength = sum(width), SuperFamilyMeanLength = mean(width)) %>%
      dplyr::select(matches("(Super|Family)")) %>%
      distinct() %>%
      arrange(desc(FamilyLength)) %T>%
      # fwrite("./R4.Distribution_of_enhancers/TEs/Summary_TEs_superfamily_family.csv", quote = F) %>%
      kable()
    
# summary (family)---------------
  
  # total length of each TE super family
  tmp <- te %>%
    dplyr::mutate(width = end - start) %>%
    dplyr::filter(str_detect(Transposon_Super_Family, "^DNA$|LTR/Gypsy")) %>%
    group_by(Transposon_Super_Family, Transposon_Family) %>%
    summarise(Number = n(), `TotalLength (MB)` = sum(width) / 1e6, MedianLength = median(width)) %>%
    arrange(desc(Number))
  
  tmp %>%
    dplyr::mutate(Transposon_Family = factor(Transposon_Family, levels = rev(.$Transposon_Family))) %>%
    pivot_longer(cols = Number:MedianLength) %>%
    ggplot(aes(y = Transposon_Family, x = value)) + 
      geom_col(fill = "grey55") + 
      facet_wrap(c("Transposon_Super_Family", "name"), scales = "free") + 
      toolkit_tyj$theme_my +
      geom_text(aes(label = value), size = 1.5, nudge_x = 0)
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Number_TotalLength_of_TEs_Family_DNA_LTRGypsy",
  #                      width = 14, height = 12, device = "png")

  # length distribution of each  TE Super Family.
  te %>%
    dplyr::filter(str_detect(Transposon_Super_Family, "^DNA$|LTR/Gypsy")) %>%
    group_by(Transposon_Family) %>%
    dplyr::mutate(Num = n(),
                  width = end - start) %>%
    ungroup() %>%
    dplyr::mutate(Transposon_Family = fct_reorder(.f = Transposon_Family, .fun = max, .x = Num, .desc = F)) %>%
    ggplot(aes(x = width, y = Transposon_Family)) + 
      geom_violin(scale = "width", size = 0.2) + 
      geom_boxplot(width = 0.2, outlier.shape = NA, size = 0.2) +
      facet_grid(Transposon_Super_Family ~ ., scales = "free_y", space = "free") +
      toolkit_tyj$theme_my + 
      scale_x_log10() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) #+ 
      # ggtitle("Width of TEs of each super family")
  
  # toolkit_tyj$SavePlot("./R4.Distribution_of_enhancers/TEs/Violin_width_TEs_Family_DNA_LTRGypsy",
  #                      width = 6, height = 12, device = "png")
  
```

```{r}
# criterion for overlap: for TE < 600bp, half of TE overlapped with enhancer; for TE > 600 bp, center of enhancer located in TE.

# load
  ## resize some TE
  te_resize <- te %>%
    dplyr::mutate(width = end - start, 
                  center = end /2 + start /2) %>%
    dplyr::mutate(start = ifelse(width < 600, round(center) - 300, start), 
                  end = ifelse(width < 600, round(center) + 300, end)) %>%
    dplyr::mutate(width = end - start)
  quantile(te_resize$width)


  ## CRE
  teGR <- makeGRangesFromDataFrame(te_resize, keep.extra.columns = T)
  cre_all <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz")
  # add intronic enhancers
  cre_all <- fread("./R7.TFs/bed_files/Enhancers_Meng2021_20210630.bed.gz") %>%
    dplyr::rename(seqnames = V1, start = V2, end = V3, method = V4) %>%
    dplyr::mutate(ID = method, seqnames = str_replace_all(seqnames, "chr", "")) %>%
    full_join(., cre_all)

  cre_all_GR <- makeGRangesFromDataFrame(cre_all, keep.extra.columns = T)
  
# findoverlap
  ol <- findOverlaps(teGR, resize(cre_all_GR, width = 2, fix = "center", use.names = T))
  
  # CRE
  cre_all$TE <- cre_all$TE_Family <- cre_all$TE_SuperFamily <- NA
  cre_all$TE_Family[ol@to] <- te$Transposon_Family[ol@from]
  cre_all$TE_SuperFamily[ol@to] <- te$Transposon_Super_Family[ol@from]
  cre_all$TE[ol@to] <- te$Transposon_Name[ol@from]
  # fwrite(cre_all, file = "./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz")
  # fwrite(cre_all, file = "./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_IntronicEn_TE_20220424.csv.gz")
```

#### .3 CRE hit by TEs (Overlap)


##### .summary
```{r, percentage}
### summary of enhancers hit TEs ----------
  table(cre_all$method)
  cre_all <- cre_all %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3)) %>%
    dplyr::mutate(method = str_replace_all(method, c("cSTARR-seq" = "Enhancer_STARRseq")))
  
  # clustered / dispersed enhancer
  enhancers <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::select(seqnames, start, end, ClusterType)
  cre_all <- cre_all %>%
    left_join(., enhancers)
  # table(df$method, df$ClusterType)
  
# percent of CRE hit TEs.
  cre_all %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(method, hitTE) %>%
    count() %>%
    pivot_wider(names_from = "hitTE", values_from = "n") %>%
    dplyr::mutate(Rate = Y / (Y + N)) %>%
    ggplot(aes(y = method, x = Rate)) + 
      geom_col(fill = "grey55") + 
      geom_text(aes(x = Rate / 2, label = percent(Rate, accuracy = 0.01)), size = 5/.pt) + 
      toolkit_tyj$theme_my + 
      scale_x_continuous(labels = scales::percent) + 
      labs(caption = "Percentage of CRE overlapped with TEs")
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Percent_of_CREs_hit_TEs_20220425",
  #                      width = 8, height = 6)

# percent of enhancer ActivityCluster hit by TEs.
  cre_all %>%
    dplyr::filter(str_detect(method, "(STARRseq|Random)")) %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(ClusterActivity, hitTE) %>%
    count() %>%
    pivot_wider(names_from = "hitTE", values_from = "n") %>%
    dplyr::mutate(Rate = Y / (Y + N), 
                  ClusterActivity = ifelse(str_length(ClusterActivity) > 2, ClusterActivity, "Random"),
                  ClusterActivity = str_replace_all(ClusterActivity, "Part", "Act")) %>%
    ggplot(aes(y = ClusterActivity, x = Rate)) + 
      geom_col(fill = "grey55") + 
      geom_text(aes(x = Rate / 2, label = round(Rate, digits = 2)), size = 2) + 
      toolkit_tyj$theme_my
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Percent_of_ActivityCluster_hit_TEs_20220425",
  #                      width = 8, height = 6)
  
# percent of enhancer KmeansCluster hit by TEs.
  cre_all %>%
    dplyr::filter(str_detect(method, "(STARRseq)"), clusterK4 != 0) %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(clusterK4 , hitTE) %>%
    count() %>%
    pivot_wider(names_from = "hitTE", values_from = "n") %>%
    dplyr::mutate(Rate = Y / (Y + N), 
                  clusterK4  = str_c("C", clusterK4, sep = "")) %>%
    ggplot(aes(y = clusterK4 , x = Rate)) + 
      geom_col(fill = "grey55") + 
      geom_text(aes(x = Rate / 2, label = percent(Rate, accuracy = 0.01)), size = 2) + 
      toolkit_tyj$theme_my + 
      ylab("Kmeans cluster of all Enhancer_STARRseq") + 
      scale_x_continuous(labels = scales::percent)
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Percent_of_KmeansCluster_hit_TEs_20220425",
  #                      width = 8, height = 6)
  
```

##### .SuperFamily & family
```{r}
### SuperFamily and family (each CRE type) ----------
  
## TOP super families, stack bar plot of enhancers overlap with TEs
  tmp <- cre_all %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(method, TE_SuperFamily, hitTE) %>%
    count() %>%
    group_by(method) %>%
    dplyr::mutate(Sum = sum(n)) %>%
    dplyr::mutate(Rate = n / Sum, 
                  TE_SuperFamily = ifelse(is.na(TE_SuperFamily), "nonTE", TE_SuperFamily)) %>%
    arrange(method, desc(Rate))
  
  # ## write to csv file
  # tmp %>%
  #   # dplyr::mutate(Rate = str_c(round(Rate, digits = 4), "(", n, "/", Sum, ")", sep = "")) %>%
  #   dplyr::mutate(method = str_c(method, "(", Sum, ")",sep = "")) %>%
  #   pivot_wider(id_cols = TE_SuperFamily, names_from = method,
  #               values_from = Rate, values_fill = NA) %>%
  #   fwrite("./R4.Distribution_of_enhancers/TEs/Summary_enhancers_overlap_TE_SuperFamilies_20220425.csv")
  
  # fill top TE superfamilies
  topFamily <- tmp %>%
    group_by(TE_SuperFamily) %>%
    summarise(MeanRate = mean(Rate)) %>%
    arrange(desc(MeanRate)) %>%
    slice_head(n = 9) %>%
    dplyr::select(TE_SuperFamily) %>%
    unlist()
  
  tmp$TE_SuperFamily[!(tmp$TE_SuperFamily %in% topFamily)] <- "Other"
  
  tmp %>%
    group_by(method, TE_SuperFamily) %>%
    summarise(Rate = sum(Rate)) %>%
    dplyr::filter(str_detect(method, "(Random|Enhancer_)"), 
                  TE_SuperFamily != "nonTE") %>%
    dplyr::mutate(TE_SuperFamily = factor(TE_SuperFamily, 
                                          levels = c(topFamily[str_detect(topFamily, "nonTE", negate = T)], 
                                                     "Other", "nonTE")),
                  method = factor(method, levels = toolkit_tyj$AtEn_level)) %>%
    ggplot(aes(x = method, y = Rate, fill = TE_SuperFamily)) + 
      geom_col() + 
      scale_fill_manual(values = c(pal_npg("nrc")(8), "grey65")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            axis.title.x = element_blank(),
            legend.key.size = unit(0.3, units = "cm"),
            legend.margin = margin(0, 0, 0, -5)) + 
      guides(fill = guide_legend(title = "Super Family")) +
      ylab("Percentage") +
      scale_y_continuous(labels = scales::percent) +
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) +
      # labs(caption = "Percent of regions overlapped by TEs") +
      toolkit_tyj$theme_my
  
  # toolkit_tyj$SavePlot("./R4.Distribution_of_enhancers/TEs/Percent_enhancers_ol_TESuperfamily_style1_20220425",
  #                      width = 6, height = 6)
  
## TOP families, stack bar plot of enhancers overlap with TEs
  tmp <- cre_all %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(method, TE_Family, hitTE) %>%
    count() %>%
    group_by(method) %>%
    dplyr::mutate(Sum = sum(n)) %>%
    dplyr::mutate(Rate = n / Sum, 
                  TE_Family = ifelse(is.na(TE_Family), "nonTE", TE_Family)) %>%
    arrange(method, desc(Rate))
  
  # ## write to csv file
  # tmp %>%
  #   # dplyr::mutate(Rate = str_c(round(Rate, digits = 4), "(", n, "/", Sum, ")", sep = "")) %>%
  #   dplyr::mutate(method = str_c(method, "(", Sum, ")",sep = "")) %>%
  #   pivot_wider(id_cols = TE_Family, names_from = method,
  #               values_from = Rate, values_fill = NA) %>%
  #   fwrite("./R4.Distribution_of_enhancers/TEs/Summary_enhancers_overlap_TE_Families_20220425.csv")
  
  # fill top TE superfamilies
  topFamily <- tmp %>%
    dplyr::filter(method == "Enhancer_STARRseq") %>%
    arrange(desc(Rate)) %>%
    slice_head(n = 13) %>%
    as.data.frame() %>%
    dplyr::select(TE_Family) %>%
    unlist()
  
  tmp$TE_Family[!(tmp$TE_Family %in% topFamily)] <- "Other"
  
  tmp %>%
    # group_by(method, TE_Family) %>%
    # summarise(Rate = sum(Rate)) %>%
    dplyr::filter(str_detect(method, "(Random|Enhancer_)"), 
                  TE_Family != "nonTE") %>%
    dplyr::mutate(TE_Family = factor(TE_Family, 
                                          levels = c(topFamily[str_detect(topFamily, "(nonTE|Other)", negate = T)], 
                                                     "Other", "nonTE")),
                  method = factor(method, levels = toolkit_tyj$AtEn_level)) %>%
    ggplot(aes(x = method, y = Rate, fill = TE_Family)) + 
      geom_col() + 
      scale_fill_manual(values = c(brewer.pal(8, "Dark2"), brewer.pal(4, "Paired"), "grey65")) +
      toolkit_tyj$theme_my + 
      guides(fill = guide_legend(title = "Family")) +
      ylab("Percentage") +
      scale_y_continuous(labels = scales::percent) +
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            axis.title.x = element_blank(),
            legend.key.size = unit(0.3, units = "cm"),
            legend.margin = margin(0, 0, 0, -5))
  
  # toolkit_tyj$SavePlot("./R4.Distribution_of_enhancers/TEs/Percent_enhancers_ol_TEfamily_style1_20220425",
  #                      width = 6, height = 6)

### SuperFamily and family (Clustered/dispersed enhancer) ----------
  
## TOP super families, stack bar plot of enhancers overlap with TEs
  # new ID column
  Enhancer_STARRseq <- cre_all %>% dplyr::filter(method == "Enhancer_STARRseq")
  cre_all <- cre_all %>%
    dplyr::mutate(method = ifelse(method == "Enhancer_STARRseq", ClusterType, method)) %>%
    full_join(., Enhancer_STARRseq)
  
  tmp <- cre_all %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(method, TE_SuperFamily, hitTE) %>%
    count() %>%
    group_by(method) %>%
    dplyr::mutate(Sum = sum(n)) %>%
    dplyr::mutate(Rate = n / Sum, 
                  TE_SuperFamily = ifelse(is.na(TE_SuperFamily), "nonTE", TE_SuperFamily)) %>%
    arrange(method, desc(Rate))
  
  # ## write to csv file
  # tmp %>%
  #   # dplyr::mutate(Rate = str_c(round(Rate, digits = 4), "(", n, "/", Sum, ")", sep = "")) %>%
  #   dplyr::mutate(method = str_c(method, "(", Sum, ")",sep = "")) %>%
  #   pivot_wider(id_cols = TE_SuperFamily, names_from = method,
  #               values_from = Rate, values_fill = NA) %>%
  #   fwrite("./R4.Distribution_of_enhancers/TEs/Summary_enhancers_overlap_TE_SuperFamilies_20220425.csv")
  
  # fill top TE superfamilies
  topFamily <- tmp %>%
    group_by(TE_SuperFamily) %>%
    summarise(MeanRate = mean(Rate)) %>%
    arrange(desc(MeanRate)) %>%
    slice_head(n = 9) %>%
    dplyr::select(TE_SuperFamily) %>%
    unlist()
  
  tmp$TE_SuperFamily[!(tmp$TE_SuperFamily %in% topFamily)] <- "Other"
  
  tmp %>%
    group_by(method, TE_SuperFamily) %>%
    summarise(Rate = sum(Rate)) %>%
    dplyr::filter(str_detect(method, "(Random|Enhancer_STARR|enhancer)"), 
                  TE_SuperFamily != "nonTE") %>%
    dplyr::mutate(TE_SuperFamily = factor(TE_SuperFamily, 
                                          levels = c(topFamily[str_detect(topFamily, "nonTE", negate = T)], 
                                                     "Other", "nonTE"))) %>%
    ggplot(aes(x = method, y = Rate, fill = TE_SuperFamily)) + 
      geom_col() + 
      scale_fill_manual(values = c(pal_npg("nrc")(8), "grey65")) +
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            axis.title.x = element_blank(), legend.key.size = unit(0.3, units = "cm")) + 
      guides(fill = guide_legend(title = "Super Family")) +
      ylab("Percentage") +
      scale_y_continuous(labels = scales::percent) +
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) +
      labs(caption = "Percent of regions overlapped by TEs")
  
  # toolkit_tyj$SavePlot("./R4.Distribution_of_enhancers/TEs/Percent_ClusterDispersed_enhancers_ol_TESuperfamily_style1_20220425",
  #                      width = 6, height = 6)
  
## TOP families, stack bar plot of enhancers overlap with TEs
  tmp <- cre_all %>%
    dplyr::mutate(hitTE = ifelse(is.na(TE), "N", "Y")) %>%
    group_by(method, TE_Family, hitTE) %>%
    count() %>%
    group_by(method) %>%
    dplyr::mutate(Sum = sum(n)) %>%
    dplyr::mutate(Rate = n / Sum, 
                  TE_Family = ifelse(is.na(TE_Family), "nonTE", TE_Family)) %>%
    arrange(method, desc(Rate))
  
  # ## write to csv file
  # tmp %>%
  #   # dplyr::mutate(Rate = str_c(round(Rate, digits = 4), "(", n, "/", Sum, ")", sep = "")) %>%
  #   dplyr::mutate(method = str_c(method, "(", Sum, ")",sep = "")) %>%
  #   pivot_wider(id_cols = TE_Family, names_from = method,
  #               values_from = Rate, values_fill = NA) %>%
  #   fwrite("./R4.Distribution_of_enhancers/TEs/Summary_enhancers_overlap_TE_Families_20220425.csv")
  
  # fill top TE superfamilies
  topFamily <- tmp %>%
    dplyr::filter(method == "Enhancer_STARRseq") %>%
    arrange(desc(Rate)) %>%
    slice_head(n = 13) %>%
    as.data.frame() %>%
    dplyr::select(TE_Family) %>%
    unlist()
  
  tmp$TE_Family[!(tmp$TE_Family %in% topFamily)] <- "Other"
  
  tmp %>%
    # group_by(method, TE_Family) %>%
    # summarise(Rate = sum(Rate)) %>%
    dplyr::filter(str_detect(method, "(Random|Enhancer_STARR|enhancer)"), 
                  TE_Family != "nonTE") %>%
    dplyr::mutate(TE_Family = factor(TE_Family, 
                                          levels = c(topFamily[str_detect(topFamily, "(nonTE|Other)", negate = T)], 
                                                     "Other", "nonTE"))) %>%
    ggplot(aes(x = method, y = Rate, fill = TE_Family)) + 
      geom_col() + 
      scale_fill_manual(values = c(brewer.pal(8, "Dark2"), brewer.pal(4, "Paired"), "grey65")) +
      toolkit_tyj$theme_my + 
      guides(fill = guide_legend(title = "Family")) +
      ylab("Percentage") +
      scale_y_continuous(labels = scales::percent) +
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            axis.title.x = element_blank(), legend.key.size = unit(0.3, units = "cm"))
  
  # toolkit_tyj$SavePlot("./R4.Distribution_of_enhancers/TEs/Percent_enhancers_ClusterDispersed_ol_TEfamily_style1_20220425",
  #                      width = 6, height = 6)

```

#### .4 TEs hit by Enhancer_STARR
```{r}
# findoverlap between TEs and Enhancer_STARR
  enhancer_starr <- cre_all  %>% 
    dplyr::filter(method == "Enhancer_STARRseq", enrichment > 1.3)
  ol2 <- findOverlaps(teGR, enhancer_starr %>% makeGRangesFromDataFrame() %>% resize(width = 2, fix = "center"))
  te$Enhancer_STARR <- "N"
  te$Enhancer_STARR[ol2@from] <- "Y"
  
# TE super family
  tmp <- te %>%
    group_by(Transposon_Super_Family, Enhancer_STARR) %>%
    count() %>%
    pivot_wider(names_from = Enhancer_STARR, values_from = n, values_fill = 0) %>%
    dplyr::mutate(Rate = Y / (Y + N), 
                  Sum = Y + N) %>%
    arrange(desc(Rate)) 
  # tmp %>% fwrite("./R4.Distribution_of_enhancers/TEs/Summary_TE_Super_family_with_STARRenhancer_activity.csv")
  
  tmp <- tmp %>% 
   dplyr::mutate(Total = Y + N, label = str_c("(", Y, "/", Total, ")")) %>%
    arrange(desc(Rate))
  
  tmp %>%
    dplyr::mutate(Transposon_Super_Family = factor(Transposon_Super_Family, levels = rev(.$Transposon_Super_Family))) %>%
    ggplot(aes(x = Rate, y = Transposon_Super_Family)) + 
      geom_col(fill = "grey55") + 
      toolkit_tyj$theme_my + 
      geom_text(aes(label = label, x = Rate + 0.007), size = 1.2)
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Summary_TE_Superfamily_with_STARRenhancer_activity", 
  #                      width = 8, height = 6)

# TE family
  tmp <- te %>%
    group_by(Transposon_Family, Enhancer_STARR) %>%
    count() %>%
    pivot_wider(names_from = Enhancer_STARR, values_from = n, values_fill = 0) %>%
    dplyr::mutate(Rate = Y / (Y + N), 
                  Sum = Y + N) %>%
    arrange(desc(Rate))
  # tmp %>% fwrite("./R4.Distribution_of_enhancers/TEs/Summary_TE_family_with_STARRenhancer_activity.csv")
  
  tmp <- tmp %>%
    dplyr::mutate(Total = Y + N, label = str_c("(", Y, "/", Total, ")")) %>%
    dplyr::filter(Total > 50, Rate > 0.05) %>%
    arrange(desc(Rate)) 
  tmp$Transposon_Family <-  factor(tmp$Transposon_Family, levels = rev(tmp$Transposon_Family))
  tmp %>% 
    ggplot(aes(x = Rate, y = Transposon_Family)) + 
      geom_col(fill = "grey55") + 
      toolkit_tyj$theme_my + 
      geom_text(aes(label = label, x = Rate - 0.025), size = 1.2) + 
      labs(caption = "only keep TE family with overlap Rate > 0.05 and Number of region > 50")
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Summary_TE_family_with_STARRenhancer_activity", 
  #                      width = 8, height = 10)
  
    
```

#### .x CREs for GAT (for 4.7.2_2021)
##### 202105 (CRE peak, not used)
resize CRE to peak, Enhancer_Meng2021 was not contained. only Activity cluster and Kmeans cluster.
```{r, 20210510 resized, eval=F}
if (F) {
# bed for GAT
  # as a whole
  resize(cre_all_GR, width = 2, fix = "center", use.names = T) %>%
    as.data.frame() %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
                  str_detect(seqnames, "(Mt|Pt)", negate = T)) %>%
    dplyr::select(seqnames, start, end, method) %>%
    dplyr::filter(str_detect(method, "(PAS|Promoter)", negate = T)) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    fwrite("./R4.Distribution_of_enhancers/TEs/CRE_all_peaks_forGAT_20210510.bed.gz", 
           sep = "\t", col.names = F, row.names = F)
  
  # kmeans cluster
  resize(cre_all_GR, width = 2, fix = "center", use.names = T) %>%
    as.data.frame() %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
                  str_detect(seqnames, "(Mt|Pt)", negate = T)) %>%
    dplyr::filter(str_detect(method, "cSTARR-seq")) %>%
    dplyr::select(seqnames, start, end,clusterK4) %>%
    dplyr::mutate(clusterK4 = str_c("KmeansCluster-", clusterK4, sep = "")) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    fwrite("./R4.Distribution_of_enhancers/TEs/CRE_all_peaks_forGAT_20210510.bed.gz", 
       sep = "\t", col.names = F, row.names = F, append = T)
  
  # activity cluster
  resize(cre_all_GR, width = 2, fix = "center", use.names = T) %>%
    as.data.frame() %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
                  str_detect(seqnames, "(Mt|Pt)", negate = T)) %>%
    dplyr::filter(str_detect(method, "cSTARR-seq")) %>%
    dplyr::select(seqnames, start, end, ClusterActivity ) %>%
    dplyr::mutate(ClusterActivity = str_replace_all(ClusterActivity, "Part", "ActivityCluster-")) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    fwrite("./R4.Distribution_of_enhancers/TEs/CRE_all_peaks_forGAT_20210510.bed.gz", 
       sep = "\t", col.names = F, row.names = F, append = T)
  
  # TEs (resized) for GAT
  library(TxDb.Athaliana.BioMart.plantsmart28)
  teGR %>%
    as.data.frame() %>%
    dplyr::select(seqnames, start, end, Transposon_Super_Family) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""),
                  Transposon_Super_Family = str_replace_all(Transposon_Super_Family, "/", "_")) %>%
    fwrite("./R4.Distribution_of_enhancers/TEs/TEs_resize_forGAT_superFamily_20210510.bed.gz", 
       sep = "\t", col.names = F, row.names = F)
    
}
```

##### 202106
not resize, four types of enahncers, with all subgroups of Enhancer_STARRseq.
```{r, 20210629 not resized}
if (F) {
  ## TEs (origin size) for GAT (20210629)
    # all 
    te %>%
      dplyr::select(seqnames, start, end) %>%
      dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
                    V4 = "TE_TAIR10") %>%
      fwrite("./R4.Distribution_of_enhancers/TEs/TEs_forGAT_allTE_20210629.bed.gz", 
             sep = "\t", col.names = F, row.names = F)
    
    # super family
    te %>%
      dplyr::select(seqnames, start, end, Transposon_Super_Family) %>%
      dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
                    Transposon_Super_Family = str_replace_all(Transposon_Super_Family, "/", "_")) %>%
      fwrite("./R4.Distribution_of_enhancers/TEs/TEs_forGAT_superFamily_20210629.bed.gz", 
             sep = "\t", col.names = F, row.names = F)
    
    # family
    LsFamily20 <- summaryFamily$Transposon_Family[summaryFamily$n > 20]
    te %>%
      dplyr::select(seqnames, start, end, Transposon_Family ) %>%
      dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
                    Transposon_Family  = str_replace_all(Transposon_Family , "/", "_")) %>%
      arrange(Transposon_Family) %>%
      dplyr::filter(Transposon_Family %in% LsFamily20) %>%
      fwrite("./R4.Distribution_of_enhancers/TEs/TEs_forGAT_Family20copies_20210629.bed.gz", 
             sep = "\t", col.names = F, row.names = F)
    
  ## CREs
    cre_tmp <- fread("Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz")
    
    df <- NULL
    # enhancers identified by different methods
    df <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
      dplyr::filter(str_detect(method, "(cSTARR-seq|Enhancer)"),
                    !(method == "cSTARR-seq" & enrichment < 1.3)) %>%
      dplyr::select(seqnames, start, end, method) %>%
      dplyr::mutate(method = str_replace_all(method, "cSTARR-seq", "Enhancer_STARRseq"))
    df <- fread("./refgenome/Intronic_DHS_enhancer_seedling_20210604.csv.gz") %>%
      dplyr::select(seqnames:end) %>%
      dplyr::mutate(method = "Enhancer_Meng2021", 
                    seqnames = as.character(seqnames)) %>%
      rbind(df, .)
    
    # Kmeans cluster
    df <- cre_tmp %>%
      dplyr::select(seqnames, start, end, clusterK4) %>%
      dplyr::mutate(clusterK4 = str_c("KmeansCluster", clusterK4, sep = "")) %>%
      dplyr::rename(method = clusterK4) %>%
      rbind(., df)
    
    # Activity Cluster
    df <- cre_tmp %>%
      dplyr::select(seqnames, start, end, ClusterActivity) %>%
      dplyr::mutate(ClusterActivity = str_replace_all(ClusterActivity, "Part", "ActivityCluster")) %>%
      dplyr::rename(method = ClusterActivity) %>%
      rbind(., df)
    
    # clustered/dispersed enhancer
    df <- cre_tmp %>%
      dplyr::select(seqnames, start, end, ClusterType) %>%
      dplyr::rename(method = ClusterType) %>%
      rbind(., df)
    
    table(df$method, df$seqnames)
    
    
    df %>%
      dplyr::filter(method != "KmeansCluster0") %>%
      dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
      fwrite("./R4.Distribution_of_enhancers/TEs/CREs_forGAT_WholeKmeanActivityClustered_20210629.bed.gz", 
             sep = "\t", col.names = F, row.names = F)
}
```



### 4.7.2 GAT (enrichment)
calculate l2fold and pvalue using GAT ().
TE (super family, resized)
CRE (whole, kmeanscluster, ActivityCluster)

#### . 1 bed for GAT
.2 of 4.7.1 (2021)

#### . 2 run GAT
`/scratch/2021-05-03/bioTanyj/At_enhancer/scripts/4.7.2_CRE_TE_overlap_GAT.sh`
`/scratch/2021-05-03/bioTanyj/At_enhancer/scripts/4.7.2_CRE_TE_overlap_GAT.R`

#### . 3 GAT results (Heatmap).

##### . 202105
```{r, 20210510 }
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./R4.Distribution_of_enhancers/TEs/CRE_ol_TEs_GAT_20210510.tsv") %>%
    dplyr::mutate(SigLabel = toolkit_tyj$returnAsterisk(qvalue),
                  label = str_c(l2fold, " (", SigLabel, ")"),
                  annotation = str_replace_all(annotation, "cSTARR-seq", "Enhancer_STARRseq"),
                  track = str_c(track, " (n=", track_nsegments, ")", sep = ""))
  
# each CRE as a whole
  table(df$annotation)
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "(Enhancer_|Random)")) %>%
    pivot_wider(id_cols = track, names_from = annotation, values_from = label)
  # fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CREWhole_20210511.csv")
  
  # plot
  df %>%
    dplyr::filter(str_detect(annotation, "(Enhancer_|Random)")) %>%
    dplyr::mutate(annotation = str_replace_all(annotation, "_", "\n")) %>%
    ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
      geom_tile() + 
      geom_text(aes(label = label), size = 1.5) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    xlab("CRE") + ylab("Transposon super family") + 
    guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CREWhole_20210511", 
  #                      width = 8.5, height = 6, units = "cm")
  # 
# each CRE Kmeans cluster
  table(df$annotation)
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "(Kmeans|Enhancer_STARRseq|Random)")) %>%
    pivot_wider(id_cols = track, names_from = annotation, values_from = label)
  # fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CRE_KmeansCluster_20210511.csv")
  
  # plot
  df %>%
    dplyr::filter(str_detect(annotation, "(Enhancer_STARR|Random|Kmeans)")) %>%
    dplyr::mutate(annotation = str_replace_all(annotation, c("_" = "\n",
                                                             "Kmeans" = "Kmeans\n"))) %>%
    ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
      geom_tile() + 
      geom_text(aes(label = label), size = 1.5) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    xlab("CRE") + ylab("Transposon super family") + 
    guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CREKmeans_20210511", 
  #                      width = 10, height = 7, units = "cm")
  
# each CRE activity cluster
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "(Activity|Enhancer_STARRseq|Random)")) %>%
    pivot_wider(id_cols = track, names_from = annotation, values_from = label)
  # fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CRE_ActivityCluster_20210511.csv")
  
  # plot
  df %>%
    dplyr::filter(str_detect(annotation, "(Enhancer_STARR|Random|Activity)")) %>%
    dplyr::mutate(annotation = str_replace_all(annotation, c("_" = "\n",
                                                             "Activity" = "Activity\n"))) %>%
    ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
      geom_tile() + 
      geom_text(aes(label = label), size = 1.5) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    xlab("CRE") + ylab("Transposon super family") + 
    guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CREActivity_20210511", 
  #                      width = 10.5, height = 7, units = "cm")
```

##### . 202106

```{r, 20210629 }
# load
  rm(list = ls())
  te_info <- fread("./refgenome/TAIR10_Transposable_Elements_20210510_aquired.txt") %>%
    dplyr::mutate(seqnames = str_replace_all(Transposon_Name, "AT(\\d)TE.+", "\\1"), 
                  strand = ifelse(orientation_is_5prime == "TRUE", "+", "-"), 
                  start = Transposon_min_Start, 
                  end = Transposon_max_End) %>%
    dplyr::select(matches("Family")) %>%
    distinct()

  source("D:/SUST/code/TanYongjun_code.R")
  df <- NULL
  for (i in list.files(path = "./R4.Distribution_of_enhancers/TEs/",
                       pattern = "CRE_ol_TEs.+20210629.tsv")) {
    id <- str_replace_all(i, "CRE_ol_TEs_GAT_(.+)20210629.tsv", "\\1")
    df <- fread(str_c("./R4.Distribution_of_enhancers/TEs/", i, sep = "")) %>%
      dplyr::mutate(TE_lev = id,
                    SigLabel = toolkit_tyj$returnAsterisk(qvalue),
                    label = str_c(round(fold, digits = 2), " (", SigLabel, ")"),
                    annotation = str_replace_all(annotation, "cSTARR-seq", "Enhancer_STARRseq"),
                    track = str_c(track, " (n=", track_nsegments, ")", sep = ""),
                    track = str_replace_all(track, "TE_TAIR10", "All_TEs")) %>%
      rbind(., df)
  }
  table(df$annotation, df$TE_lev)
```

###### . each CRE
```{r}
## each CRE as a whole-----------------
  
  ## all TEs + TE super family.
    # plot
    df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "Family20copies") %>%
      dplyr::mutate(annotation = factor(annotation, levels = toolkit_tyj$AtEn_level)) %>%
      ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
        geom_tile() + 
        geom_text(aes(label = label), size = 6/.pt, color = "grey15") + 
        scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                             mid = brewer.pal(11, "RdYlGn")[6], 
                             low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
      toolkit_tyj$theme_my + 
      xlab("CRE") + ylab("Transposon super family") + 
      guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
      theme(legend.key.width = unit(0.4, units = "cm"), 
            legend.margin = margin(-1, -1, -1, -6), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")})
    
    # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_vs_CREWhole_20220425",
    #                      width = 8, height = 8, units = "cm")
    
    # table
    table(df$annotation)
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "Family20copies") %>%
      pivot_wider(id_cols = track, names_from = annotation, values_from = label)
    # fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_superFamily_vs_CREWhole_20220425.csv")
  

  ## all TEs + TE families (> 20 copy).
    # subset top TE families enriched in enhancers
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "superFamily",
                    annotation == "Enhancer_STARRseq", SigLabel != "ns")
    
    # plot
    df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "superFamily",
                    track %in% tmp$track) %>%
      dplyr::mutate(annotation = factor(annotation, levels = toolkit_tyj$AtEn_level),
                    Transposon_Family = str_replace_all(track, "(.+) \\(.+", "\\1")) %>%
      left_join(., te_info) %>%
      dplyr::mutate(Transposon_Super_Family = ifelse(is.na(Transposon_Super_Family), "All TEs", Transposon_Super_Family)) %>%
      # count(Transposon_Super_Family)
      ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
        geom_tile() + 
        geom_text(aes(label = label), size = 5/.pt, color = "grey15") + 
        scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                             mid = brewer.pal(11, "RdYlGn")[6], 
                             low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
        toolkit_tyj$theme_my + 
        xlab("CRE") + ylab("Transposon super family") + 
        guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
        theme(legend.key.width = unit(0.4, units = "cm"), 
              legend.margin = margin(-1, -1, -1, -6), 
              axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
              strip.text.y = element_text(angle = 0),
              panel.spacing = unit(0.05, "cm")) + 
        scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) + 
        facet_grid(Transposon_Super_Family ~ ., scales = "free", space = "free") +
        labs(caption = "Only TF families significantly enriched or depleted \nin Enhancer_STARRseq were illustrated.")
    
    # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_family_vs_CREWhole_20220425",
    #                      width = 9, height = 16, units = "cm")
    
    # table
    table(df$annotation)
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "superFamily") %>%
      pivot_wider(id_cols = track, names_from = annotation, values_from = label) %>%
      dplyr::mutate(Transposon_Family = str_replace_all(track, "(.+) \\(.+", "\\1")) %>%
      left_join(., te_info)
    # fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_Family_vs_CREWhole_20220425.csv")
    
```

###### . subgroups
```{r}
## each cluster of enahancer_STARRseq-----------------
  table(df$annotation, df$TE_lev)
  
  ## all TEs + TE super family.
    # plot
    df %>%
      dplyr::filter(str_detect(annotation, "(Zhu|Wang|Meng)", negate = T), 
                    TE_lev != "Family20copies") %>%
      dplyr::mutate(annotation = str_replace_all(annotation, 
                                                 c("ActivityCluster" = "Act",
                                                   "_enhancer" = "",
                                                   "KmeansCluster" = "C")),
                    Type_CRE = "All",
                    Type_CRE = ifelse(str_detect(annotation, "Act"), "Activity group", Type_CRE),
                    Type_CRE = ifelse(str_detect(annotation, "^C\\d"), "Kmeans group", Type_CRE),
                    Type_CRE = ifelse(str_detect(annotation, "Clustered|Dispersed"), "Clustered/Dispersed", Type_CRE),
                    Type_CRE = factor(Type_CRE, levels = c("All", "Activity group", "Kmeans group", "Clustered/Dispersed")),
                    # label = ifelse(SigLabel != "ns", label, "-")
                    ) %>%
      ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
        geom_tile() + 
        geom_text(aes(label = label), size = 5/.pt, color = "grey15") + 
        scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                             mid = brewer.pal(11, "RdYlGn")[6], 
                             low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
      toolkit_tyj$theme_my + 
      xlab("CRE") + ylab("Transposon super family") + 
      guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
      theme(legend.key.width = unit(0.4, units = "cm"), 
            legend.margin = margin(-1, -1, -1, -6), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            panel.spacing = unit(0.1, "cm")) + 
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) + 
      facet_grid(. ~ Type_CRE, space = "free", scales = "free")
    
    # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_SuperFamily_vs_CRE_cluster_20220425",
    #                      width = 16, height = 8, units = "cm")
    
    # table
    table(df$annotation)
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Zhu|Wang|Meng)", negate = T), 
                    TE_lev != "Family20copies") %>%
      pivot_wider(id_cols = track, names_from = annotation, values_from = label)
    fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_superFamily_vs_CRE_cluster_20220425.csv")
        
        
  ## all TEs + TE families (> 20 copy).
    # subset TE families enriched in enhancer_STARRseq
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "superFamily",
                    annotation == "Enhancer_STARRseq", SigLabel != "ns")
    
    # plot
    df %>%
      dplyr::filter(str_detect(annotation, "(Zhu|Wang|Meng)", negate = T), 
                    TE_lev != "superFamily",
                    track %in% tmp$track) %>%
      dplyr::mutate(annotation = str_replace_all(annotation, 
                                                 c("ActivityCluster" = "Act",
                                                   "_enhancer" = "",
                                                   "KmeansCluster" = "C")),
                    Type_CRE = "All",
                    Type_CRE = ifelse(str_detect(annotation, "Act"), "Activity group", Type_CRE),
                    Type_CRE = ifelse(str_detect(annotation, "^C\\d"), "Kmeans group", Type_CRE),
                    Type_CRE = ifelse(str_detect(annotation, "Clustered|Dispersed"), "Clustered/Dispersed", Type_CRE),
                    Type_CRE = factor(Type_CRE, levels = c("All", "Activity group", "Kmeans group", "Clustered/Dispersed")),
                    # label = ifelse(SigLabel != "ns", label, "-") %>%
                    Transposon_Family = str_replace_all(track, "(.+) \\(.+", "\\1")
                    ) %>%
      left_join(., te_info) %>%
      dplyr::mutate(Transposon_Super_Family = ifelse(is.na(Transposon_Super_Family), "All TEs", Transposon_Super_Family)) %>% 
      ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
        geom_tile() + 
        geom_text(aes(label = label), size = 5/.pt, color = "grey15") + 
        scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                             mid = brewer.pal(11, "RdYlGn")[6], 
                             low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
      toolkit_tyj$theme_my + 
      xlab("CRE") + ylab("Transposon super family") + 
      guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
      theme(legend.key.width = unit(0.4, units = "cm"), 
            legend.margin = margin(-1, -1, -1, -6), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            panel.spacing = unit(0.1, "cm"),
            strip.text.y = element_text(angle = 0)) + 
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) + 
      facet_grid( Transposon_Super_Family ~ Type_CRE, space = "free", scales = "free")
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_Family_vs_CRE_cluster_20220425",
                         width = 17.5, height = 16, units = "cm")
    
    # table
    table(df$annotation)
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Zhu|Wang|Meng)", negate = T), TE_lev != "superFamily") %>%
      pivot_wider(id_cols = track, names_from = annotation, values_from = label) %>%
      dplyr::mutate(Transposon_Family = str_replace_all(track, "(.+) \\(.+", "\\1")) %>%
      left_join(., te_info)
    fwrite(tmp, file = "./R4.Distribution_of_enhancers/TEs/GAT_results_TE_Family_vs_CRE_cluster_20220425.csv")
    
```



###### . (Whole + Cluster) X (TE + family)

TE superfamily/family with significantly enriched/depleted in one were included in plot.

```{r}
    
    # prepare data(facet_x, facet_y, add total of each TE super family)
    tmp <- df %>%
      dplyr::filter(str_detect(annotation, "(Enhancer_|Random)"), TE_lev != "superFamily",
                    annotation == "Enhancer_STARRseq", SigLabel != "ns") # only keep significantly enriched/depleted TF families or superfamilies.
    
    # plot
    df %>%
      dplyr::filter(#str_detect(annotation, "(Zhu|Wang|Meng)", negate = T), 
                    TE_lev != "superFamily",
                    track %in% tmp$track) %>%
      dplyr::mutate(annotation = str_replace_all(annotation, 
                                                 c("ActivityCluster" = "Act",
                                                   "_enhancer" = "",
                                                   "KmeansCluster" = "C")),
                    Type_CRE = "All",
                    Type_CRE = ifelse(str_detect(annotation, "Act"), "Activity group", Type_CRE),
                    Type_CRE = ifelse(str_detect(annotation, "^C\\d"), "Kmeans group", Type_CRE),
                    Type_CRE = ifelse(str_detect(annotation, "Clustered|Dispersed"), "Clustered/Dispersed", Type_CRE),
                    Type_CRE = factor(Type_CRE, levels = c("All", "Activity group", "Kmeans group", "Clustered/Dispersed")),
                    # label = ifelse(SigLabel != "ns", label, "-") %>%
                    Transposon_Family = str_replace_all(track, "(.+) \\(.+", "\\1")
                    ) %>%
      left_join(., te_info) %>%
      dplyr::mutate(Transposon_Super_Family = ifelse(is.na(Transposon_Super_Family), "All TEs", Transposon_Super_Family)) %>% 
      ggplot(aes(x = annotation, y = track, fill = l2fold)) + 
        geom_tile() + 
        geom_text(aes(label = label), size = 5/.pt, color = "grey15") + 
        scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                             mid = brewer.pal(11, "RdYlGn")[6], 
                             low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
      toolkit_tyj$theme_my + 
      xlab("CRE") + ylab("Transposon super family") + 
      guides(fill = guide_colorbar(title = expression(log[2](Fold)))) + 
      theme(legend.key.width = unit(0.4, units = "cm"), 
            legend.margin = margin(-1, -1, -1, -6), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            panel.spacing = unit(0.1, "cm"),
            strip.text.y = element_text(angle = 0)) + 
      scale_x_discrete(labels = function(x){str_replace_all(x, "_", "\n")}) + 
      facet_grid( Transposon_Super_Family ~ Type_CRE, space = "free", scales = "free")
    
    toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/TEs/Heatmap_All_CRE_all_TEs_GAT_enrich",
                         width = 18, height = 18)
```



## 4.8 exonic-enhancer
How to define exonic-enhancer/cds-enhancer?

  1. intronic-enhancer: 80% of DHS sequence overlapeed with an intron. `Meng et.al., 2021, the Plant Cell.`
  2. eExon: intersect between DHS and exon(infact, use protein-coding region). `Stergachis et.al., 2013, Science`
  3. synonymous constrain region of Exon. `Li et al., 2011, Genome research`
  
  
  
### .1 summary of Intron, UTR, and CDS.
 Overlap regions between different features were not excluded.
```{r}
  rm(list = ls())
  library(furrr)
  plan(multisession, workers = 4)

# load enhancer annotation (Distribution of enhancers according to genomic features. generated in 4.1)
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")  
  rm(list = ls(pattern = "cis|enhancer|^i$|pas|peak|random|temp|Enhancer|ChIP"))
  source("D:/SUST/code/TanYongjun_code.R")
  
# genomic regions
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  ClassifyType <- function(x){
    if(nrow(x) == 1){
      x$Type <- "OnlyOne"
    }else{
      x$Type <- "Other"
      if(x$strand == "+"){
        x <- x %>%
          arrange(start)
      }else{
        x <- x %>%
          arrange(desc(end))
      }
      x$Type[1] <- "First"
      x$Type[nrow(x)] <-  "Last"
    }
    return(x)
  }
  
  CDS_all <- cds(txdb, columns = c("gene_id", "cds_id")) %>% # This function will return CDS in all transcripts!
    as_tibble() %>%
    # slice_tail(., n = 3000) %>%
    dplyr::mutate(gene_id = map_chr(gene_id, function(x){str_c(x, collapse = ";")}),
                  Features = "CDS") %>%
    group_by(gene_id) %>%
    split(., .$gene_id) %>%
    future_map_dfr(., function(x){
      makeGRangesFromDataFrame(x) %>%
        GenomicRanges::reduce(.) %>%
        as_tibble(.) %>%
        ClassifyType(.) %>%
        return()
    }, .id = "gene_id") %>%
    dplyr::mutate(Features = "CDS")
  
  
  Five_UTR <- fiveUTRsByTranscript(txdb, use.names = T) %>%
    unlist() %>%
    as_tibble() %>%
    # slice_tail(., n = 3000) %>%
    dplyr::mutate(gene_id = str_replace_all(exon_name,"(AT.+)\\..+\\..+$", "\\1")) %>%
    dplyr::rename(id = exon_id) %>%
    dplyr::select(-c(exon_name, exon_rank,id)) %>%
    split(., .$gene_id) %>%
    future_map_dfr(., function(x){
      makeGRangesFromDataFrame(x) %>%
        GenomicRanges::reduce(.) %>%
        as_tibble(.) %>%
        dplyr::mutate(Type = "Five_UTR") %>%
        return()
    }, .id = "gene_id") %>%
    dplyr::mutate(Features = "Five_UTR")
    
  Three_UTR <- threeUTRsByTranscript(txdb, use.names = T) %>%
    unlist() %>%
    as_tibble() %>%
    # slice_tail(., n = 3000) %>%
    dplyr::mutate(gene_id = str_replace_all(exon_name,"(AT.+)\\..+\\..+$", "\\1")) %>%
    dplyr::rename(id = exon_id) %>%
    dplyr::select(-c(exon_name, exon_rank,id)) %>%
    split(., .$gene_id) %>%
    future_map_dfr(., function(x){
      makeGRangesFromDataFrame(x) %>%
        GenomicRanges::reduce(.) %>%
        as_tibble(.) %>%
        dplyr::mutate(Type = "Three_UTR") %>%
        return()
    }, .id = "gene_id") %>%
    dplyr::mutate(Features = "Three_UTR")
  
  Intron_all <- intronsByTranscript(txdb, use.names = T) %>%
    as_tibble() %>%
    # slice_tail(., n = 3000) %>%
    dplyr::mutate(gene_id = str_replace_all(group_name,"(AT.+)\\..+$", "\\1"),
                  Features = "Intron") %>%
    group_by(gene_id) %>%
    split(., .$gene_id) %>%
    future_map_dfr(., function(x){
      makeGRangesFromDataFrame(x) %>%
        GenomicRanges::reduce(.) %>%
        as_tibble(.) %>%
        ClassifyType(.) %>%
        return()
    }, .id = "gene_id") %>%
    dplyr::mutate(Features = "Intron")
  
  plan(sequential)
  
  Region_all <- full_join(ungroup(CDS_all), ungroup(Five_UTR)) %>%
    full_join(., ungroup(Three_UTR)) %>%
    full_join(., ungroup(Intron_all)) %>%
    dplyr::mutate(Feature_id = ifelse(Features == Type,
                                      Features,
                                      str_c(Features, Type, sep = "_")),
                  Feature_id = factor(Feature_id, levels = c("CDS_First", "CDS_Other", "CDS_Last", "CDS_OnlyOne",
                                                             "Intron_First", "Intron_Other", "Intron_Last", "Intron_OnlyOne",
                                                             "Five_UTR", "Three_UTR")))
  
  table(Region_all$Features, Region_all$Type)
  table(Region_all$Feature_id)
  # saveRDS(Region_all, file = "./R4.Distribution_of_enhancers/Exonic_enhancer/Regions_all_tidy_20220507.rds")
  rm(Three_UTR, Five_UTR, Intron_all, CDS_all)
  
  
  
  # Length and number of regions
  Region_all <- read_rds("./R4.Distribution_of_enhancers/Exonic_enhancer/Regions_all_tidy_20220507.rds")
  options(scipen = 200)
  tmp_lab <- Region_all %>%
        group_by(Feature_id, Features, Type) %>%
        dplyr::summarise(MedianWidth = median(width), MeanWidth = mean(width), 
                         TotalLength = sum(width)/1e6,
                         Num = n()) %>%
        dplyr::mutate(label = str_c("Median=", round(MedianWidth), "\nMean=", round(MeanWidth), "\nNum=", Num))

  ggplot() + 
    geom_histogram(color = "transparent", aes(x = width, fill = Features), data = Region_all) +
    # geom_density(fill = "transparent") +
    toolkit_tyj$theme_my + 
    scale_x_log10() +
    scale_fill_npg() +
    facet_wrap("Feature_id", scales = "free_y", dir = "v") + 
    xlab("Length") +
    ylab("Count") +
    ggtitle("Length distribution of each kind of genomic regions") + 
    theme(legend.key.size = unit(0.3, "cm"),
          legend.position = c(0.9, 0.2),
          legend.margin = margin(0,0,0,-9)) +
    geom_abs_text(data = tmp_lab, aes(xpos = 0.2, ypos = 0.75, label = label), size = 5/.pt) + 
    labs(caption = "Overlapped regions were not excluded\nIntrons for diff transcripts of the same gene were not reduced.")
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R4.Distribution_of_enhancers/Exonic_enhancer/Length_distributin_genomic_regions_20220506",
  #                      width = 14, height = 12, device = "png")
  
  
```

### .2 Identification (findoverlap)
  Definition:
    Overlap between region and enhancer larger than 50bp.
  
  output:
    1. Enhancer list with overlapped regions.
    2. regions list with overlapped enhancers.
  
```{r}
  rm(list = ls())
  Region_all <- read_rds("./R4.Distribution_of_enhancers/Exonic_enhancer/Regions_all_tidy_20220507.rds")
  source("D:/SUST/code/TanYongjun_code.R")
  MinOverlapSize <- 50 # regions whose width less than MinOverlapSize will be discarded.
  ResizeEnhancer <- 600 # resize enhancer to determined size (i.e. the core region of enhancer).
  
# CRE list
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")  
  rm(list = ls(pattern = "cis|enhancer|^i$|pas|peak|random|temp|^Enhancer|ChIP"))
  
# identify exonic enhancer (annotatePeak() was center-based)
  Region_allLS <- Region_all %>%
    dplyr::mutate(region_id = str_c(gene_id, Feature_id, start, end, strand, width, sep = "_")) %>%
    split(., .$Feature_id) 
  Region_allGR <- Region_allLS %>%
    map(., makeGRangesFromDataFrame)
  
  # total length of each region
  Region_allGR %>%
    map_int(., function(x){
      GenomicRanges::reduce(x) %>%
        width(.) %>%
        sum(.) %>%
        return()
    })
  
  En_STARR <- anno_list$`cSTARR-seq`@anno %>% 
    as_tibble() %>%
    dplyr::mutate(start = startBK, end = endBK, 
                  queryHits = 1:nrow(.),
                  CRE_id = str_c(seqnames, start, end, sep = "_")) %>%
    dplyr::select(-c(startBK, endBK, ID, method, Type, rep, Strategy, strand, geneStrand, geneChr, width))
  En_STARR_GR <- makeGRangesFromDataFrame(En_STARR) 
  if (!is.null(ResizeEnhancer)) {
    En_STARR_GR <- GenomicRanges::resize(En_STARR_GR, width = ResizeEnhancer, fix = "center")
  }
  
  for (i in names(Region_allLS)){
    cat(">>> ", i, "\n")
    ol <- findOverlaps(En_STARR_GR, Region_allGR[[i]], minoverlap = MinOverlapSize) %>%
      as_tibble() %>%
      left_join(., dplyr::select(En_STARR, CRE_id, queryHits)) %>%
      left_join(., dplyr::mutate(Region_allLS[[i]], subjectHits = 1:nrow(Region_allLS[[i]])) %>%
                  dplyr::select(subjectHits, region_id)) %>%
      dplyr::select(CRE_id, region_id)
    
    # add CRE info into region list
    ol_tmp <- ol %>%
      group_by(region_id) %>%
      summarise(CRE_ls = str_c(CRE_id, collapse = ";"))
    Region_allLS[[i]] <- Region_allLS[[i]] %>%
      left_join(., ol_tmp)
    
    # add region id to CRE list
    ol_tmp <- ol %>%
      group_by(CRE_id) %>%
      summarise(region_ls = str_c(region_id, collapse = ";"))
    En_STARR <- left_join(En_STARR, ol_tmp)
    colnames(En_STARR) <- str_replace_all(colnames(En_STARR), "region_ls", i)
  }
  
  Region_all <- Region_allLS %>%
    map_dfr(., as_tibble)
  
  
## summary
  # Enhancers located in each genomic regions.
    names(En_STARR)
    library(UpSetR)
    png(filename = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "Upset_enhancer_EnResize_ov_GenomicRegion_EnResize_",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.png"),
        width = 16, height = 12, units = "cm", res = 600)
    En_STARR %>%
      dplyr::select(CDS_First:Three_UTR) %>%
      dplyr::mutate(across(.cols = everything(), 
                           .fns = function(x){
                             x[!is.na(x)] <- "1"
                             x[is.na(x)] <- "0"
                             x %>%
                               as.numeric() %>%
                               return()
                           })) %>%
      dplyr::mutate(Other = ifelse(apply(., 1, sum) == 0, 1, 0),
                    Sample = str_c("En", 1:nrow(.))) %>%
      as.data.frame() %>%
      upset(nsets = 11, mainbar.y.label = "Number of enhancers was 4423")
    dev.off()
  
  # Percentage of genomic regions overlapped with enhancer
    Region_all %>%
      dplyr::select(Feature_id, CRE_ls) %>%
      dplyr::mutate(CRE_tmp = ifelse(!is.na(CRE_ls), "Enhancer", "No")) %>%
      group_by(Feature_id, CRE_tmp) %>%
      dplyr::summarise(Num = n()) %>%
      group_by(Feature_id) %>%
      dplyr::mutate(Total = sum(Num),
                    Rate = Num / Total,
                    Label = str_c(Num, " / ", Total)) %>%
      ggplot(aes(x = CRE_tmp, y = Rate, fill = CRE_tmp)) +
        geom_col() + 
        toolkit_tyj$theme_my + 
        facet_wrap("Feature_id", ncol = 4) + 
        scale_fill_npg() + 
        theme(legend.position = "none") + 
        geom_text(aes(y = 0.5, label = scales::percent(Rate, accuracy = 0.01)), size = 6/.pt, angle = 45) + 
        labs(caption = str_c("Min overlap between enhancer and region was ", MinOverlapSize, "bp\n",
                             "Enhancer resize to ", ResizeEnhancer, "bp"))
    
    toolkit_tyj$SavePlot(filename_prefix = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "Percentage_genomic_regions_ov_enhancer_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508"),
                         device = "png", width = 12, height = 12)
    
  saveRDS(En_STARR, file = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "En_list_with_overlapped_UTR_CDS_intron_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.rds"))
  saveRDS(Region_all, file = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "Region_list_with_overlapped_Enhancer_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.rds"))
  NA
```

### .3 Gene enrichment

#### . UTR/CDS/Intron (findoverlap())
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  setwd("D:/SUST/Enhancers_A.T/")
  Region_all <- read_rds("./R4.Distribution_of_enhancers/Exonic_enhancer/Regions_all_tidy_20220507.rds")
  library(org.At.tair.db)
  MinOverlapSize <- 50 # regions whose width less than MinOverlapSize will be discarded.
  ResizeEnhancer <- 200 # resize enhancer to determined size (i.e. the core region of enhancer).
  
  En_STARR <- read_rds(str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "En_list_with_overlapped_UTR_CDS_intron_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.rds"))
  Region_all <- read_rds(str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "Region_list_with_overlapped_Enhancer_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.rds"))
  ExtractGeneID <- function(x){
    x %>%
      str_extract_all(., "AT[^_]+") %>%
      unlist() %>%
      na.omit() %>%
      as.character() %>%
      unique() %>%
      return()
  }
  
  # GO enrichment analysis.
  library(clusterProfiler)
  ego_list <- En_STARR %>%
    dplyr::select(CDS_First:Three_UTR) %>%
    map(., ExtractGeneID) %>%
    map(.x = ., .f = function(y){ # error reported when use future_map() in this step.
                                      enrichGO(gene = y,
                                              keyType = "TAIR",
                                              pvalueCutoff = 0.05,
                                              pAdjustMethod = "BH",
                                              OrgDb = org.At.tair.db,
                                              ont = "All",
                                              readable = T
                                              ) %>% return()
                                     }
      )
  # ego_list %>%
  #   map_dfr(., function(x){
  #     x@result %>%
  #       as_tibble %>%
  #       return()
  #   }, .id = "Features") %>%
  #   saveRDS(., file = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
  #                                                  "GO_enrich_results_findOverlap_Enhancer_EnResize",
  #                                                  ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
  #                                                  "_20220508.rds"))
  
```

#### . all results (annotatePeak())
all genomic features.
```{r}
  rm(list = ls())
  library(org.At.tair.db)
  library(clusterProfiler)
  source("D:/SUST/code/TanYongjun_code.R")
# load enhancer annotation (Distribution of enhancers according to genomic features. generated in 4.1)
  load(file = "./R4.Distribution_of_enhancers/Enhancer_annotation_ChIPseeker_Peak_up500_down0_IntronicEnhancer_20210605.bin")  
  rm(list = ls(pattern = "cis|enhancer|^i$|pas|peak|random|temp|Enhancer|ChIP"))
  
  enhancer_ls <- anno_list$`cSTARR-seq`@anno %>% 
    as_tibble() %>%
    dplyr::mutate(Features = str_replace_all(annotation, "([^\\(+]+) \\(.+", "\\1"))
  
  table(enhancer_ls$Features)
  
  ego_all <- enhancer_ls %>%
    dplyr::select(Features, geneId) %>%
    split(., .$Features) %>%
    map(., function(x){
      x$geneId %>%
        unlist()
    }) %>%
    map(.x = ., .f = function(y){ # error reported when use future_map() in this step.
                                      enrichGO(gene = y,
                                              keyType = "TAIR",
                                              pvalueCutoff = 0.05,
                                              pAdjustMethod = "BH",
                                              OrgDb = org.At.tair.db,
                                              ont = "All",
                                              readable = T
                                              ) %>% return()
                                     }
      )
  # ego_all %>%
  #   map_dfr(., function(x){
  #     x@result %>%
  #       as_tibble %>%
  #       return()
  #   }, .id = "Features") %>%
  #   saveRDS(., file = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
  #                                                  "GO_enrich_results_annotatePeak_Enhancer_20220508.rds"))
  
  
```

```{r}
  ego_all <- read_rds("./R4.Distribution_of_enhancers/Exonic_enhancer/GO_enrich_results_findOverlap_Enhancer_EnResize200bp_MinOverlap50_20220508.rds")
```


### .4 List of genes with CDS hit enhancer.
```{r}
  rm(list = ls())
  
# gene to SYMBOL
  library(org.At.tair.db)
  library(AnnotationDbi)
  library(clusterProfiler)
  uniKeys <- keys(org.At.tair.db, keytype="TAIR")
  cols <- c("TAIR", "SYMBOL")
  TAIR_gene <- AnnotationDbi::select(org.At.tair.db, keys=uniKeys, columns=cols, keytype="TAIR") %>%
    dplyr::rename(gene_id = TAIR) %>%
    group_by(gene_id) %>%
    summarise(SYMBOL_ls = str_c(SYMBOL, collapse = ";"))
  
# phenotype of mutants
  phe_gene <- fread("./refgenome/TAIR/Locus_Germplasm_Phenotype_20210331.txt.gz") %>%
    as_tibble %>%
    dplyr::rename(gene_id = LOCUS_NAME) %>%
    group_by(gene_id) %>%
    dplyr::summarise(mut_phe = str_c(PHENOTYPE, collapse = ";"))

# Number of citation
  num_cit <- fread("./refgenome/tair/Locus_Published_20210331.txt.gz") %>%
    as_tibble() %>%
    dplyr::mutate(gene_id = str_replace_all(name, "^([^\\.]+)\\.\\d+$", "\\1")) %>%
    group_by(gene_id) %>%
    summarise(Num_citation = length(unique(reference_id)))
  
# gene discription
  gene_dis <- fread("./refgenome/TAIR/Araport11_functional_descriptions_20210331.txt.gz") %>%
    as_tibble() %>%
    dplyr::rename(gene_id = name)
  
  
# merge
  TAIR_gene <- left_join(TAIR_gene, phe_gene) %>%
    left_join(., num_cit) %>%
    left_join(., gene_dis)
  
  # fwrite(TAIR_gene, file = "./refgenome/TAIR_genes_info_merge_20220510.csv.gz")
  
# Regions of gene to CRE
  source("D:/SUST/code/TanYongjun_code.R")
  library(org.At.tair.db)
  MinOverlapSize <- 50 # regions whose width less than MinOverlapSize will be discarded.
  ResizeEnhancer <- 200 # resize enhancer to determined size (i.e. the core region of enhancer).
  
  En_STARR <- read_rds(str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "En_list_with_overlapped_UTR_CDS_intron_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.rds"))
  Region_all <- read_rds(str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "Region_list_with_overlapped_Enhancer_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.rds")) %>%
    left_join(., TAIR_gene)
  
  GeneCDShitEn <- Region_all %>%
    dplyr::filter(!is.na(CRE_ls), !is.na(SYMBOL_ls),
                  Features == "CDS") %>%
    arrange(desc(Num_citation))
  
  fwrite(GeneCDShitEn, file = str_c("./R4.Distribution_of_enhancers/Exonic_enhancer/",
                                                 "List_gene_CDS_hit_Enhancer_EnResize",
                                                 ResizeEnhancer, "bp_MinOverlap", MinOverlapSize,
                                                 "_20220508.csv"))
```



# 5. Epigenetic profile of enhancers.

## .0 Info


*How to define which chromatin states used in cluster analysis?*

### .x Data transformation

```{r, Demo, eval=F}
    # Box-Cox transformation
      library(MASS)
      pheno.blup.boxcox <- pheno.blup
      for (i in 2:ncol(pheno.blup.boxcox)) {
        pheno.blup.boxcox[,i][!is.na(pheno.blup.boxcox[,i])] <- 
          toolkit_tyj$powerTransform(na.omit(pheno.blup.boxcox[,i])) - mean(toolkit_tyj$powerTransform(na.omit(pheno.blup.boxcox[,i])))
      }
    
    # Rank-Norm transformation
      library(RNOmni)
      pheno.blup.rank <- pheno.blup
      for (i in 2:ncol(pheno.blup.rank)) {
        pheno.blup.rank[,i][!is.na(pheno.blup.rank[,i])] <- RankNorm(pheno.blup.rank[,i][!is.na(pheno.blup.rank[,i])])
      }
```

## 5.0 CRE hit peak

overlap between CRE and peaks

#### .1 prepare bed.gz files for GAT.

all CREs type in one four-columns .bed file : chr, start, end, cre_ID

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz")

cSTARRseq <- enhancer %>%
  dplyr::filter(method == "cSTARR-seq", enrichment > 1.3)
enhancerK4 <- cSTARRseq %>%
  dplyr::filter(clusterK4 != 0) %>%
  mutate(ID = str_c("KmeansCluster", clusterK4, sep = "")) %>%
  dplyr::select(seqnames, start, end, ID)
enhancerActCluster <- cSTARRseq %>%
  mutate(ID = str_replace_all(ClusterActivity, "Part", "ActivityCluster")) %>%
  dplyr::select(seqnames, start, end, ID)
enhancerWhole <- enhancer %>%
  dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
                !(method == "iSTARR-seq" & enrichment < 1.3),
                method != "PAS") %>%
  dplyr::mutate(ID = method) %>%
  dplyr::select(seqnames, start, end, ID)

df <- rbind(enhancerK4, enhancerActCluster) %>%
  rbind(., enhancerWhole) %>%
  dplyr::filter(seqnames != "Pt", seqnames != "Mt") %>%
  dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""))
table(df$ID)
table(df$seqnames)

fwrite(df, file = "./R5.chromatin_states/CRE_4column_bed_for_GAT_peak_overlap_20210416.bed.gz", sep = "\t", 
       col.names = F, row.names = F)
```

#### .2 run GAT

`/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/5.0_CRE_chrPeaks_overlap_GAT.sh` `/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/5.0_CRE_chrPeaks_overlap_GAT.R`

#### .3 plot

#####. load
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
df <- NULL
for (i in list.files(path = "./R5.chromatin_states/bed_file_chromatin_states/", 
                     pattern = ".+tsv")) {
  df <- fread(paste("./R5.chromatin_states/bed_file_chromatin_states/", 
                    i, sep = "")) %>%
    dplyr::mutate(track = str_replace_all(i, ".tsv", "")) %>%
    rbind(., df)
}

# discard broad peak of H3K4me1 (narrow peak provide by ENCODE)
df <- df %>%
  dplyr::filter(track != "H3K4me1_peaks.broadPeak") %>%
  dplyr::mutate(track = str_replace_all(track, "_.+", ""), 
                track = str_replace_all(track, "DNase", "DHS"),
                track = factor(track, levels = c(
                  "ATAC", "DHS", "MHS", "RNAPOII", 
                  sort(unique(track)[str_detect(unique(track), "(ATAC|DHS|MHS|RNAPOII)", negate = T)])
                )))

```

##### . Peaks hit by CREs
Attention: Different number of CRE and Random regions used in GAT! So, did not compare betwee different CRE!


```{r}
## percent of peaks hit by CREs (Can individual chip-seq peaks be used to predict enhancer activity)
  # all CRE
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "Cluster", negate = T))
  p <- tmp %>%
    ggplot(aes(x = annotation, y = percent_overlap_size_track)) + 
      geom_col(fill = "grey45") + 
      facet_wrap(facets = "track") +
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1)) + 
      ggtitle("Percent of peaks hit by CREs") + 
      ylab("Percent of peaks hit by CREs") +
      geom_text(aes(x = annotation, y = percent_overlap_size_track + 2, 
                    label = paste(round(percent_overlap_size_track, digits = 2), 
                                  "%", sep = "")), size = 1.2) + 
      labs(caption = "length-based")
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Percent_of_peaks_hit_by_CREs_20210416",
  #                      width = 16, height = 12)

  # only cSTARRseq enhancer
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "Cluster", negate = T), 
                  annotation == "cSTARR-seq")
  p <- tmp %>%
    ggplot(aes(x = track, y = percent_overlap_size_track)) + 
      geom_col(fill = "grey45") + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1)) + 
      # ggtitle("Percent of peaks with enhancer activity") + 
      ylab("Percent of peaks with enhancer activity") +
      geom_text(aes(x = track, y = percent_overlap_size_track + 1, 
                    label = paste(round(percent_overlap_size_track, digits = 2), 
                                  "%", sep = "")), size = 1.2) + 
      labs(caption = "length-based")
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Percent_of_peaks_with_enhancer_activity_20210416",
  #                      width = 8, height = 6)
```

##### . CREs hit by Peaks

*Dot plot*

```{r}
## Fold change as y label, enrichment (only STARRseq enhancer)----------------------------------
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "Cluster", negate = T), 
                  annotation == "cSTARR-seq")
  
  # percent of `peaks` with enhancer activity
  tmp %>%
    dplyr::mutate(Coverage = round(percent_overlap_size_track / 100, digits = 2)) %>%
    ggplot(aes(x = track, y = l2fold)) + 
      geom_vline(aes(xintercept = track), linetype = 3, color = "grey55") +
      geom_hline(yintercept = 0, linetype = 2, color = "grey55") +
      geom_point(aes(shape = ifelse(qvalue < 0.05, "qvalue < 0.05", "qvalue >= 0.05"), 
                     size = Coverage), color = "steelblue") + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      scale_shape_manual(values = c(16, 1)) + 
      ylab(label = expression(log[2](Fold~change))) + 
      guides(shape = guide_legend(title = "Significant level"), 
             size = guide_legend(title = "Coverage")) + 
      scale_x_discrete(expand = expansion(mult = c(0.05, 0.05))) + 
      labs(caption = "length-based")
      # scale_y_reverse()
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_peaks_with_enhancer_activity",
  #                      width = 10, height = 6)

  # percent of `enhancer` hit by peaks as ylabel (used in ReSE, Nature Genetics.)
  tmp %>%
    dplyr::mutate(Coverage = round(percent_overlap_size_annotation / 100, digits = 2)) %>%
    ggplot(aes(x = track, y = l2fold)) + 
      geom_vline(aes(xintercept = track), linetype = 3, color = "grey55") +
      geom_hline(yintercept = 0, linetype = 2, color = "grey55") +
      geom_point(aes(color = ifelse(qvalue < 0.05, "qvalue < 0.05", "qvalue >= 0.05"), 
                     size = Coverage),alpha = 0.9) + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      # scale_shape_manual(values = c(16, 1)) + 
      scale_color_npg() +
      ylab(label = expression(log[2](Fold~change))) + 
      guides(shape = guide_legend(title = "Significant level"), 
             color = guide_legend(title = "Significant level"),
             size = guide_legend(title = "Coverage")) + 
      scale_x_discrete(expand = expansion(mult = c(0.05, 0.05))) + 
      scale_size_continuous(range = c(1, 4)) + 
      theme(legend.box.margin = margin(0,0,0,0), legend.spacing.y = unit(0, units = "cm")) + 
      labs(caption = "length-based")
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_enhancer_hit_peaks",
  #                      width = 10, height = 6)
  

  # percent of `enhancer` hit by peaks as ylabel (used in ReSE, Nature Genetics, set depleted histone modification as 0)
  tmp %>%
    dplyr::filter(str_detect(track, "MHS|ATAC", negate = T)) %>%
    dplyr::mutate(Coverage = round(percent_overlap_nsegments_annotation / 100, digits = 2),
                  qvalue = ifelse(l2fold < 0, 1, qvalue),
                  sigLevel = ifelse(qvalue < 0.01, "Sig", "ns"),
                  l2fold = ifelse(l2fold <0, 0, l2fold),
                  track = str_replace_all(as.character(track), "RNAPOII", "RNAPII"),
                  track = factor(track, levels = c("ATAC", "DHS", "MHS", "RNAPII", "H2A", "H2AW",
                                                   "H2AX", "H2AZ", "H3K9ac", 
                                                   "H3K14ac", "H3K23ac", "H3K27ac",
                                                   "H3K36ac", "H3K56ac", "H4K16ac",
                                                   "H3K4me1", "H3K4me2", "H3K4me3", "H3K9me1",
                                                   "H3K9me2", "H3K27me3", "H3K36me3"))
                  ) %>%
    ggplot(aes(x = track, y = l2fold)) + 
      # geom_vline(aes(xintercept = track), linetype = 3, color = "grey55") +
      geom_hline(yintercept = 0, linetype = 2, color = "grey55") +
      geom_segment(aes(x = track, xend = track, y = 0, yend = l2fold), 
                   color = "steelblue", size = 0.3) +
      geom_point(aes(#shape = ifelse(qvalue < 0.05, "qvalue < 0.05", "ns"), 
                     fill = ifelse(qvalue < 0.05, "q<0.05", "ns"), 
                     size = Coverage),alpha = 1, color = "steelblue", shape = 21) + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      scale_fill_manual(values = c("transparent", "steelblue")) +
      ylab(label = expression(log[2](Fold~enrichment))) + 
      guides(fill = guide_legend(title = "Significant level"), 
             shape = guide_legend(title = "Significant level"),
             size = guide_legend(title = "Enhancers\ncoverage\nrate")) + 
      scale_x_discrete(expand = expansion(mult = c(0.05, 0.05))) + 
      scale_size_continuous(range = c(1, 4)) + 
      theme(legend.box.margin = margin(20,0,0,-12),
            legend.spacing.y = unit(0, units = "cm"),
            legend.key.height = unit(0.3, "cm"),
            axis.title.x = element_blank())
  
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_enhancer_hit_peaks",
                       width = 10, height = 4.5)
  
  
  # qvalue as ylabel (used in ReSE, Nature Genetics, set depleted histone modification as 0)
  tmp %>%
    dplyr::mutate(Coverage = round(percent_overlap_size_annotation / 100, digits = 2),
                  qvalue = ifelse(l2fold < 0, 1, qvalue),
                  l2fold = ifelse(l2fold <0, 0, l2fold)) %>%
    ggplot(aes(x = track, y = -log10(qvalue))) + 
      # geom_vline(aes(xintercept = track), linetype = 3, color = "grey55") +
      geom_col(aes(x = track, y = -log10(qvalue)), width = 0.005, color = "steelblue") +
      geom_hline(yintercept = 0, linetype = 2, color = "grey55") +
      geom_point(aes(size = Coverage),alpha = 1, color = "steelblue") + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      # scale_shape_manual(values = c(21, 1)) +
      # scale_fill_manual(values = c("steelblue", "white")) +
      ylab(label = expression(-log[10](Q~value))) + 
      guides(shape = guide_legend(title = "Significant level"), 
             color = guide_legend(title = "Significant level"),
             size = guide_legend(title = "Enhancers coverage rate")) + 
      scale_x_discrete(expand = expansion(mult = c(0.05, 0.05))) + 
      scale_size_continuous(range = c(1, 4)) + 
      theme(legend.box.margin = margin(0,0,0,0), legend.spacing.y = unit(0, units = "cm")) #+ 
      # labs(caption = "length-based")
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_enhancer_hit_peaks_qvalue",
  #                      width = 10, height = 6)
  
  
```

```{r}
## qvalue as y label, enrichment (only STARRseq enhancer)----------------------------------
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "Cluster", negate = T), 
                  annotation == "cSTARR-seq")
  
  # percent of `peaks` with enhancer activity
  tmp %>%
    dplyr::mutate(Coverage = round(percent_overlap_size_track / 100, digits = 2)) %>%
    ggplot(aes(x = track, y = l2fold)) + 
      geom_vline(aes(xintercept = track), linetype = 3, color = "grey55") +
      geom_hline(yintercept = 0, linetype = 2, color = "grey55") +
      geom_point(aes(shape = ifelse(qvalue < 0.05, "qvalue < 0.05", "qvalue >= 0.05"), 
                     size = Coverage), color = "steelblue") + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      scale_shape_manual(values = c(16, 1)) + 
      ylab(label = expression(log[2](Fold~change))) + 
      guides(shape = guide_legend(title = "Significant level"), 
             size = guide_legend(title = "Coverage")) + 
      scale_x_discrete(expand = expansion(mult = c(0.05, 0.05))) + 
      labs(caption = "length-based")
      # scale_y_reverse()
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_peaks_with_enhancer_activity",
  #                      width = 10, height = 6)

  # percent of `enhancer` hit by peaks (used in ReSE, Nature Genetics.)
  tmp %>%
    dplyr::mutate(Coverage = round(percent_overlap_size_annotation / 100, digits = 2)) %>%
    ggplot(aes(x = track, y = l2fold)) + 
      geom_vline(aes(xintercept = track), linetype = 3, color = "grey55") +
      geom_hline(yintercept = 0, linetype = 2, color = "grey55") +
      geom_point(aes(color = ifelse(qvalue < 0.05, "qvalue < 0.05", "qvalue >= 0.05"), 
                     size = Coverage),alpha = 0.9) + 
      toolkit_tyj$theme_my + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      # scale_shape_manual(values = c(16, 1)) + 
      scale_color_npg() +
      ylab(label = expression(log[2](Fold~change))) + 
      guides(shape = guide_legend(title = "Significant level"), 
             color = guide_legend(title = "Significant level"),
             size = guide_legend(title = "Coverage")) + 
      scale_x_discrete(expand = expansion(mult = c(0.05, 0.05))) + 
      scale_size_continuous(range = c(1, 4)) + 
      theme(legend.box.margin = margin(0,0,0,0), legend.spacing.y = unit(0, units = "cm")) + 
      labs(caption = "length-based")
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_enhancer_hit_peaks",
  #                      width = 10, height = 6)
  
```


*Percentage as y label*
```{r}
## percent of `enhancer` hit by peaks (bar plot)
  tmp <- df %>%
    dplyr::filter(str_detect(annotation, "Cluster", negate = T), 
                  annotation == "cSTARR-seq")

# All datasets
  tmpPlot <- tmp %>%
    dplyr::mutate(Observed = observed / annotation_size, 
                  Expected = expected / annotation_size) %>%
    pivot_longer(cols = c(Observed, Expected)) %>%
    dplyr::select(track, qvalue, fold, l2fold, name, value) %>%
    dplyr::mutate(Percent = percent(value, accuracy = 0.01), 
                  sigLabel = toolkit_tyj$returnAsterisk(qvalue), 
                  FC = round(fold, digits = 2),
                  name = factor(name, levels = c("Observed", "Expected")), 
                  text_y_pos = ifelse(value > 0.4, 0.05, value))
  
  tmpSigLabel <- tmpPlot %>%
    dplyr::select(track, FC, sigLabel) %>% 
    distinct() %>%
    dplyr::mutate(start = "Observed", end = "Expected",
                  label = str_c("FC=", FC, "\n", sigLabel, sep = ""))
  
  tmpPlot %>%
    ggplot(aes(x = name, y = value, fill = name)) + 
      geom_col() + 
      geom_signif(data = tmpSigLabel,
                  aes(xmin = start, xmax = end,
                      annotations = label, y_position = 0.6),
                  manual = T, inherit.aes = F,
                  color = "grey35", textsize = 1.5,
                  size = 0.2, vjust = 0.1) +
      facet_wrap(facets = "track", ncol = 6) + 
      toolkit_tyj$theme_my + 
      scale_fill_npg() + 
      scale_y_continuous(labels = scales::percent, 
                         expand = expansion(mult = c(0,0), add = c(0, 0.025))) + 
      theme(legend.key.size = unit(0.3, units = "cm"), 
            legend.position = "none", 
            legend.margin = margin(-5, 0, 0, 0), 
            axis.title.x = element_blank(), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      ylab("Percent overlaped size of enhancer") + 
      geom_text(aes(label = Percent, y = text_y_pos), angle = 90, 
                hjust = -0.1, color = "grey25", size = 1.5) + 
      coord_cartesian(ylim = c(0, 0.8))
 
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_enhancer_hit_peaks_Barplot",
  #                      width = 12, height = 10)

  # tmpPlot %>% 
  #   dplyr::filter(name == "Observed") %>% 
  #   fwrite("./R5.chromatin_states/bed_file_chromatin_states/Summary_ChrStates_enrichment_Enhancers_20210624.csv", 
  #          quote = F)
  
# no ATAC/MHS
  tmpPlot <- tmp %>%
    dplyr::filter(str_detect(track, "MHS|ATAC", negate = T)) %>%
    dplyr::mutate(Observed = observed / annotation_size, 
                  Expected = expected / annotation_size) %>%
    pivot_longer(cols = c(Observed, Expected)) %>%
    dplyr::select(track, qvalue, fold, l2fold, name, value) %>%
    dplyr::mutate(Percent = percent(value, accuracy = 0.01), 
                  sigLabel = toolkit_tyj$returnAsterisk(qvalue), 
                  FC = round(fold, digits = 2),
                  name = factor(name, levels = c("Observed", "Expected")), 
                  text_y_pos = ifelse(value > 0.4, 0.05, value))
  
  tmpSigLabel <- tmpPlot %>%
    dplyr::select(track, FC, sigLabel) %>% 
    distinct() %>%
    dplyr::mutate(start = "Observed", end = "Expected",
                  label = str_c("FC=", FC, "\n", sigLabel, sep = ""))
  
  tmpPlot %>%
    ggplot(aes(x = name, y = value, fill = name)) + 
      geom_col() + 
      geom_signif(data = tmpSigLabel,
                  aes(xmin = start, xmax = end,
                      annotations = label, y_position = 0.6),
                  manual = T, inherit.aes = F,
                  color = "grey35", textsize = 6/.pt,
                  size = 0.2, vjust = 0.1) +
      facet_wrap(facets = "track", ncol = 10) + 
      toolkit_tyj$theme_my + 
      scale_fill_npg() + 
      scale_y_continuous(labels = scales::percent, 
                         expand = expansion(mult = c(0,0), add = c(0, 0.025))) + 
      theme(legend.key.size = unit(0.3, units = "cm"), 
            legend.position = "none", 
            legend.margin = margin(-5, 0, 0, 0), 
            axis.title.x = element_blank(), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      ylab("Percent overlaped size of enhancer") + 
      geom_text(aes(label = Percent, y = text_y_pos), angle = 90, 
                hjust = -0.1, color = "grey15", size = 6/.pt) + 
      coord_cartesian(ylim = c(0, 0.7))
 
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/bed_file_chromatin_states/Enrichment_enhancer_hit_peaks_Barplot_1",
  #                      width = 16, height = 10)

```


## 5.1 Enrichment profile

### 5.1.1 DHS/RNPII/histone

#### .0 prepare .bed files

for generate heatmap and profile in deeptools.

```{r}
# 20210416
  # source("D:/SUST/code/TanYongjun_code.R")
  # enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz")
  # 
  # cSTARRseq <- enhancer %>%
  #   dplyr::filter(method == "cSTARR-seq", enrichment > 1.3)
  # enhancerK4 <- cSTARRseq %>%
  #   dplyr::filter(clusterK4 != 0) %>%
  #   mutate(ID = str_c("KmeansCluster", clusterK4, sep = "")) %>%
  #   dplyr::select(seqnames, start, end, ID, enrichment, strand)
  # enhancerActCluster <- cSTARRseq %>%
  #   mutate(ID = str_replace_all(ClusterActivity, "Part", "ActivityCluster")) %>%
  #   dplyr::select(seqnames, start, end, ID, enrichment, strand)
  # enhancerWhole <- enhancer %>%
  #   dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
  #                 !(method == "iSTARR-seq" & enrichment < 1.3),
  #                 method != "PAS") %>%
  #   dplyr::mutate(ID = method) %>%
  #   dplyr::select(seqnames, start, end, ID, enrichment, strand)
  # 
  # df <- rbind(enhancerK4, enhancerActCluster) %>%
  #   rbind(., enhancerWhole) %>%
  #   dplyr::filter(seqnames != "Pt", seqnames != "Mt") %>%
  #   dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""))
  # table(df$ID)
  # table(df$seqnames)
  # table(df$strand, df$ID)
  # fwrite(df, file = "./R5.chromatin_states/CRE_6column_bed_for_GAT_peak_overlap_20210416.bed.gz", sep = "\t", 
  #        col.names = F, row.names = F)

# 20210608 (Add intronic enhancer, clustered/dispersed enhancer.)
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./R5.chromatin_states/CRE_6column_bed_for_GAT_peak_overlap_20210416.bed.gz")
  
  # add intronic enhancer
  IntronicEnhancer <- fread("./refgenome/Intronic_DHS_enhancer_seedling_20210604.csv.gz") %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
                  ID = "Enhancer_Meng2021") %>%
    dplyr::select(seqnames:end, ID)
  colnames(IntronicEnhancer) <- str_c("V", 1:4)
  dfMerge <- full_join(df, IntronicEnhancer)
  
  # Clustered/dispersed enhancer
  SuperEn <- fread("Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::filter(seqnames != "Mt") %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    dplyr::select(seqnames:end, ClusterType)
  colnames(SuperEn) <- str_c("V", 1:4)
  dfMerge <- full_join(dfMerge, SuperEn)
  
  table(dfMerge$V4)
  dfMerge %>%
    fwrite("./R5.chromatin_states/CRE_6column_bed_for_GAT_peak_overlap_20210608.bed.gz", 
           sep = "\t", col.names = F, row.names = F, na = "NA")
```

#### .1 heatmap profile

Genrate enriched `heatmap` and `profile` of DHS/RNPII/histone in the CREs with deeptools (use `reference-point` mode of `deeptools-computeMatrix`).
`/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/5.1.1_ChrStates_heatmap_profile.lsf` `/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/5.1.1_ChrStates_heatmap_profile.R`

#### .2 plot in R

Files with name of `*plotProfile.tab` generated in previous step were used in plot.

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")

## load
  df <- NULL
  for (i in list.files(path = "./R5.chromatin_states/CRE_heatmap_ChrState/", 
                       pattern = ".+_10kbWidth_10bpBin_plotProfile.tab")) {
    tmp <- fread(paste("./R5.chromatin_states/CRE_heatmap_ChrState/", i, sep = ""),
                skip = 1)
    df <- rbind(df, tmp[-1,])
  }
  
  dim(df)
  colnames(df) <- c("Chromatin_states", "CRE", 1:(ncol(df)-2))
  df <- df %>% 
    pivot_longer(cols = 3:ncol(.), names_to = "Position") %>%
    dplyr::mutate(Position = as.numeric(Position), 
                  CRE = str_replace_all(CRE, "cSTARR-seq", "Enhancer_STARRseq"), 
                  Chromatin_states = str_replace_all(Chromatin_states, "seq", ""),
                  Chromatin_states = factor(Chromatin_states, levels = c(
                      "ATAC", "DHS", "MNase", "FAIRE", "RNAPOII", 
                      sort(unique(Chromatin_states)[str_detect(unique(Chromatin_states), "(ATAC|DHS|MNase|RNAPOII|FAIRE)", negate = T)])
                    ))
                ) %>%
    dplyr::filter(str_detect(Chromatin_states, "(ATAC|FAIRE|MNase)", negate = T))
    

## each CRE type as a whole.
  library(ggformula)
  p <- df %>%
    dplyr::filter(str_detect(CRE, "(Enhancer|Random)", negate = F)) %>%
    dplyr::mutate(CRE = factor(CRE, levels = c("Enhancer_STARRseq", "Enhancer_Zhu2015",
                                               "Enhancer_Wang2019", "Enhancer_Meng2021",
                                               "Random"))) %>%
    ggplot() + 
      facet_wrap(facets = "Chromatin_states", scales = "free_y", ncol = 7) + 
      geom_spline(aes(x = Position, y = value, color = CRE, linetype = CRE), size = 0.3) +
      # geom_point(size = 0.05) +
      scale_color_manual(values = c(pal_npg()(9)[c(1, 2, 3, 4)], 
                                    "grey40")) +
      scale_linetype_manual(values = c(1,1,1,1,3)) +
      toolkit_tyj$theme_my + 
      theme(legend.position = "bottom", legend.margin = margin(-12, 0,0,0), legend.title = element_blank()) + 
      guides(color=guide_legend(nrow = 1)) + 
      ylab("score") +
      scale_x_continuous(breaks = c(1, 500, 1000), labels = c("-5kb", "Center", "5kb"))
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_heatmap_ChrState/Profile_each_CRE_as_whole_20210609",
  #                      width = 15, height = 9, plot = p)

## each ActivityCluster
  library(ggformula)
  unique(df$CRE)
  p <- df %>%
    dplyr::filter(str_detect(CRE, "(Activity|Random)", negate = F), 
                  # str_detect(Chromatin_states, "(ATAC|FAIRE|MNase)", negate = T)
                  ) %>%
    dplyr::mutate(CRE = str_replace_all(CRE, "ActivityCluster", "Act")) %>%
    ggplot() + 
      facet_wrap(facets = "Chromatin_states", scales = "free_y", ncol = 5) + 
      geom_spline(aes(x = Position, y = value, color = CRE, linetype = CRE), size = 0.3) +
      # geom_point(size = 0.05) +
      scale_color_manual(values = c(carto_pal(5, "Temps"),
                                    "grey40")) +
      scale_linetype_manual(values = c(1,1,1,1,1,3)) +
      toolkit_tyj$theme_my + 
      theme(legend.position = "bottom", legend.margin = margin(-12, 0,0,0), legend.title = element_blank()) + 
      guides(color=guide_legend(nrow = 1)) + 
      ylab("score") +
      scale_x_continuous(breaks = c(1, 500, 1000), labels = c("-5kb", "Center", "5kb"))
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_heatmap_ChrState/Profile_CRE_ActivityCluster_20210417",
  #                      width = 10, height = 12, plot = p)
  
## each KmeansCluster
  library(ggformula)
  unique(df$CRE)
  p <- df %>%
    dplyr::filter(str_detect(CRE, "(Kmeans|Random)", negate = F), 
                  # str_detect(Chromatin_states, "(ATAC|FAIRE|MNase)", negate = T)
                  ) %>%
    dplyr::mutate(CRE = str_replace_all(CRE, "KmeansCluster", "C")) %>%
    ggplot() + 
      facet_wrap(facets = "Chromatin_states", scales = "free_y", ncol = 5) + 
      geom_spline(aes(x = Position, y = value, color = CRE, linetype = CRE), size = 0.3) +
      # geom_point(size = 0.05) +
      scale_color_manual(values = c(toolkit_tyj$AtEn_KmeanC_color, "grey40")) +
      scale_linetype_manual(values = c(1,1,1,1,3)) +
      toolkit_tyj$theme_my + 
      theme(legend.position = "bottom", legend.margin = margin(-12, 0,0,0), legend.title = element_blank()) + 
      guides(color=guide_legend(nrow = 1)) + 
      ylab("score") +
      scale_x_continuous(breaks = c(1, 500, 1000), labels = c("-5kb", "Center", "5kb"))
  p
  
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_heatmap_ChrState/Profile_CRE_KmeansCluster_20210417", 
                       width = 10, height = 12, plot = p)
  
## Clustered/Dispersed 
  p <- df %>%
    dplyr::filter(str_detect(CRE, "(Dispersed|Clustered|Random)", negate = F), 
                  # str_detect(Chromatin_states, "(ATAC|FAIRE|MNase)", negate = T)
                  ) %>%
    dplyr::mutate(CRE = str_replace_all(CRE, "_enhancer", "")) %>%
    ggplot() + 
      facet_wrap(facets = "Chromatin_states", scales = "free_y", ncol = 5) + 
      geom_spline(aes(x = Position, y = value, color = CRE, linetype = CRE), size = 0.3) +
      # geom_point(size = 0.05) +
      scale_color_manual(values = c(pal_npg()(5)[c(3, 1)],
                                    "grey40")) +
      scale_linetype_manual(values = c(1,1,3)) +
      toolkit_tyj$theme_my + 
      theme(legend.position = "bottom", legend.margin = margin(-12, 0,0,0), legend.title = element_blank()) + 
      guides(color=guide_legend(nrow = 1)) + 
      ylab("score") +
      scale_x_continuous(breaks = c(1, 500, 1000), labels = c("-5kb", "Center", "5kb"))
  p
  
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_heatmap_ChrState/Profile_CRE_Clustered_Dispersed_20210417", 
                       width = 10, height = 12, plot = p)
```

### 5.1.2 TSS

A total of four related studies were reported.
1.
  GRO-seq, GRO-cap, RNA-seq; by Hetzel et.al, 2016, PNAS.
  6-days-old seedlings.
  *GSE83108(tsv)* (only a tsv files contain regions of `transcripts` and related infos).
2.
  GRO-seq, pNET-seq; by Zhu et.al, 2018, NatPlant.
  12-days-old seedling (varied ROPII isoforms).
  *GSE109974(tsv, bedgraph), GSE117014(tsv)* 
3.
  pNET-seq; by Kindgren et.al, 2019, NAR, 10-days-old seedling (22/4 degree).
  *GSE131733(NET-seq, bw)* 
4.
  CAGE, smallRNA, RNA-seq; by Thieffry et.al, 2020, PC(and BioRxiv), 14-days-old seedlings (GEO, 12-days-old?) (wt and two RNA decay pathway mutants).
  *GSE136356*(CAGE, seedling, bw), `GSE136362 (RNA-seq, seedling, bw), and GSE142555(smallRNA, leaf and flower bud)` 
  Only 1 and 4 can used in this study.

*Hypothesis:* 
1.
  Identified TSS (nocoding, bidirectional, and 3' located) enriched in enhancers.
  datasets: CAGE (bw) and GRO-cap(bed).
2.
  RNAPOLII paused near enhancers located in gene body?
  datasets: pNET (bw).
  `heatmap of enrichment` 
3.
  Higher antisense transcription around enhancer than other gene region?
  datasets: GRO-seq/RNA-seq bw file (+ and -).`heatmap of enrichment`

#### .1 Number of TSS in CREs.

*Results did not suppurt hypothesis 1 !!* Number of TSS in STARRseq enhancer was NOT higher than random!.

```{r, TSS from transcript region of GSE83108}
# load
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("Enahncers_DHS_STARRseq_H3K4_Pro_Random_20210313.csv.gz") %>%
  dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3))
table(enhancer$method)

# transcripts
regions <- fread("./refgenome/nacentRNAseq/GSE83108_transcripts.txt") %>%
  mutate(width = `end (TAIR10)` - `start (TAIR10)`)

tmp_label <- regions %>%
  group_by(`HOMER Detailed Transcript Type`) %>%
  dplyr::summarise(Num = n(), Median = median(width)) %>%
  mutate(label = paste("N=", Num, sep = ""))
regions %>%
  ggplot(aes(x = `HOMER Detailed Transcript Type`, y = width)) + 
    geom_violin() + 
    toolkit_tyj$theme_my + 
    scale_y_log10() + 
    coord_flip() + 
    ggtitle("Width of regions") + 
    geom_text(data = tmp_label, aes(x = `HOMER Detailed Transcript Type`, y = Median, label = label))


table(regions$`HOMER Detailed Transcript Type`)
table(regions$`HOMER Transcript Type`)

# TSS locate in enhancers identified by different methods
tss_all <- regions %>%
  dplyr::select(2:5, 9:11)
names(tss_all) <- c("seqnames", "start", "end", "strand", "Type", "Gene_NearestGeneName", "Alias")
tss_gr <- makeGRangesFromDataFrame(tss_all, keep.extra.columns = T) %>%
  GenomicRanges::resize(width = 1, fix = "start") %>%
  GenomicRanges::resize(width = 100, fix = "center")

# # write to bed for GAT.(four columns)
# tss_gr %>%
#   GenomicRanges::resize(width = 50, fix = "center") %>%
#   as.data.frame() %>%
#   dplyr::select(seqnames, start, end, Type) %>% 
#   dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
#                 Type = str_c("GSE83108_", Type, sep = "")) %>%
#   fwrite("./refgenome/nacentRNAseq/TSS_regions_for_GAT_all_50bp.bed", 
#          sep = "\t", quote = F, col.names = F, row.names = F)
# tss_gr %>%
#   GenomicRanges::resize(width = 50, fix = "center") %>%
#   as.data.frame() %>%
#   dplyr::select(seqnames, start, end, Type) %>% 
#   dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
#                 Type = "GSE83108_All") %>%
#   fwrite("./refgenome/nacentRNAseq/TSS_regions_for_GAT_all_50bp.bed", 
#          sep = "\t", quote = F, col.names = F, row.names = F, append = T)


df_tss <- NULL
df_cre <- NULL
for (i in unique(enhancer$method)) {
  en_starr <- enhancer %>%
    # dplyr::filter(str_detect(method, "STARR"))
    dplyr::filter(method == i)
  
  ol <- findOverlaps(tss_gr, 
                     makeGRangesFromDataFrame(en_starr))
  
  # tss
  tss_all$CRE <- "No"
  tss_all$CRE[unique(ol@from)] <- "Yes"
  df_tss <- table(tss_all$CRE, tss_all$Type) %>%
    as.data.frame() %>%
    pivot_wider(id_cols = "Var2", names_from = "Var1", values_from = "Freq") %>%
    mutate(Ratio = percent(Yes / (Yes + No)), 
           method = i) %>%
    rbind(., df_tss)
  
  # CRE
  en_starr$TSS <- "No"
  en_starr$TSS[unique(ol@to)] <- "Yes"
  df_cre <- table(en_starr$TSS) %>%
    as.data.frame() %>%
    mutate(method = i) %>%
    rbind(., df_cre)
}

df_tss %>%
  mutate(Ratio = str_c("(", Yes, "/", Yes+No, "); ", Ratio, sep = "")) %>%
  dplyr::select(-No, -Yes) %>%
  pivot_wider(id_cols = Var2, names_from = "method", values_from = "Ratio") %T>%
  fwrite("./R5.chromatin_states/nacentRNA_TSS/Summary_of_TSS_located_in_enhancers_20210508.csv")

df_cre %>%
  pivot_wider(id_cols = method, names_from = "Var1", values_from = "Freq") %>%
  mutate(Ratio = percent(Yes / (Yes + No))) %T>%
  fwrite("./R5.chromatin_states/nacentRNA_TSS/Summary_of_enhancers_with_TSS_20210508.csv")
  
```

```{r, dataset from Axel_Thieffry_2019_Biorxiv_Supp_data1}
# load
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("Enahncers_DHS_STARRseq_H3K4_Pro_Random_20210313.csv.gz") %>%
  dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3))
regions <- fread("./refgenome/nacentRNAseq/Axel_Thieffry_2019_Biorxiv_Supp_data1_media-2.tsv") %>%
  mutate(seqnames = str_replace_all(seqnames, "Chr", ""))
table(regions$type, regions$annotation)
regions$ID <- paste(regions$type, regions$annotation, sep = "-")
regionsGR <- regions %>%
  makeGRangesFromDataFrame()

# ## write to bed for GAT
# regions %>%
#   dplyr::select(seqnames, start, end, type) %>% 
#   dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
#                 type = "Axel_Thieffry_2019_all") %>%
#   fwrite("./refgenome/nacentRNAseq/TSS_regions_for_GAT_all_50bp.bed", 
#          sep = "\t", quote = F, col.names = F, row.names = F, append = T)
# regions %>%
#   dplyr::select(seqnames, start, end, type) %>% 
#   dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""), 
#                 type = str_c("Axel_", type, sep = "")) %>%
#   fwrite("./refgenome/nacentRNAseq/TSS_regions_for_GAT_all_50bp.bed", 
#          sep = "\t", quote = F, col.names = F, row.names = F, append = T)

# calculate overlap
df_tss <- NULL
df_cre <- NULL

for (i in unique(enhancer$method)) {
  cre_tmp <- enhancer %>%
    dplyr::filter(method == i)
  cre_tmpgr <- makeGRangesFromDataFrame(cre_tmp)
  
  ol <- findOverlaps(cre_tmpgr, regionsGR)
  
  # Regions/TSS
  cre_tmp$TSS <- "No"
  cre_tmp$TSS[ol@from] <- "Yes"
  df_tss <- table(cre_tmp$TSS) %>%
    as.data.frame() %>%
    mutate(method = i) %>%
    rbind(., df_tss)
  
  # CRE
  cre_tmp$TSS <- "No"
  cre_tmp$TSS[ol@from] <- regions$ID[ol@to]
  df_cre <- table(cre_tmp$TSS) %>%
    as.data.frame() %>%
    mutate(method = i) %>%
    rbind(., df_cre)
  
}

df_tss %>%
  pivot_wider(id_cols = method, names_from = "Var1", values_from = "Freq") %>%
  mutate(Ratio = percent(Yes / (No + Yes))) %T>%
  fwrite("./R5.chromatin_states/nacentRNA_TSS/Summary_of_Total_TSS_located_in_enhancers_Axel_Thieffry_2019_Biorxiv_20210508.csv")

  

df_cre %>%
  group_by(method) %>%
  mutate(Sum = sum(Freq), Ratio = percent(Freq / Sum, accuracy = 0.01)) %>%
  # mutate(label = str_c(Freq, "/", Sum, "; ", Ratio)) %>%
  mutate(label = Ratio) %>%
  dplyr::select(Var1, method, label) %>%
  pivot_wider(id_cols = "Var1", names_from = "method", values_from = label, values_fill = NA) %T>%
  fwrite("./R5.chromatin_states/nacentRNA_TSS/Summary_of_TSS_located_in_enhancers_Axel_Thieffry_2019_Biorxiv_20210508.csv")
  View()
```

#### .2 Plot profile with deeptools

`/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/5.1.2_ChrStates_heatmap_profile.lsf` `/scratch/2021-04-10/bioTanyj/At_enhancer/scripts/5.1.2_ChrStates_heatmap_profile.R`

#### .3 GAT
bed files generated in `.1`.
`/scratch/2021-05-25/bioTanyj/At_enhancer/scripts/5.1.2_nacentRNA_TSS_GAT.sh`

*No significantly difference was observed when using 50bp or 100bp.*
*Collect results and plot (set 100bp width of TSS listed in GSE83108)*
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
df <- fread("./R5.chromatin_states/nacentRNA_TSS/Overlap_GAT_100000times_results.tsv") %>%
  dplyr::filter(str_detect(annotation, "(Cluster|Enhancer|Random|enhancer)")) %>%
  arrange(track, annotation)
unique(df$annotation) %>% sort()
unique(df$track)

# # table (fold + qvalue)
# df %>%
#   dplyr::mutate(label = str_c(round(fold, digits = 2), 
#                               "(",
#                               scientific(qvalue, digits = 4), 
#                               ")", 
#                               sep = "")) %>%
#   pivot_wider(id_cols = annotation, names_from = track, values_from = label) %>%
#   View()

# # Number of regions in each annotation track
# df %>%
#   dplyr::select(track, track_nsegments) %>%
#   unique()

# heatmap with significant label all
df %>%
  dplyr::filter(str_detect(annotation, "Random_", negate = T)) %>%
  dplyr::mutate(sigLable = toolkit_tyj$returnAsterisk(qvalue)) %>%
  dplyr::mutate(label = ifelse(sigLable == "ns", "ns",
                               str_c(format(l2fold, digits = 2), "\n", sigLable, sep = ""))) %>%
  ggplot(aes(x = track, y = annotation, fill = l2fold)) + 
    geom_tile() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    geom_text(aes(label = label), size = 1.5, lineheight = 0.8) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    labs(caption = "100bp") +
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/nacentRNA_TSS/noRandom_GAT_results_TSS_ol_enhancers_100bp_20210526", 
                     width = 12, height = 8)

# heatmap with significant label (selected)
df %>%
  dplyr::filter(str_detect(annotation, "(Random_|sEnhancer|pEnhancer)", negate = T), 
                track_nsegments > 100) %>%
  dplyr::mutate(sigLable = toolkit_tyj$returnAsterisk(qvalue)) %>%
  dplyr::mutate(label = ifelse(sigLable == "ns", "ns",
                               str_c(format(l2fold, digits = 2), "\n", sigLable, sep = ""))) %>%
  ggplot(aes(x = track, y = annotation, fill = l2fold)) + 
    geom_tile() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    geom_text(aes(label = label), size = 1.5, lineheight = 0.8) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    labs(caption = "100bp") +
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/nacentRNA_TSS/selected_GAT_results_TSS_ol_enhancers_100bp_20210526", 
                     width = 12, height = 8)

# filter
df %>%
  dplyr::filter(str_detect(track, "(all|All)"),
                str_detect(annotation, "^Enhancer", negate = F))

```

*Collect results and plot (set 50bp width of TSS listed in GSE83108)*
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
df <- fread("./R5.chromatin_states/nacentRNA_TSS/Overlap_GAT_100000times_50bp_results.tsv") %>%
  dplyr::filter(str_detect(annotation, "(Cluster|Enhancer|Random|enhancer)")) %>%
  arrange(track, annotation)
unique(df$annotation) %>% sort()
unique(df$track)

# # table (fold + qvalue)
# df %>%
#   dplyr::mutate(label = str_c(round(fold, digits = 2), 
#                               "(",
#                               scientific(qvalue, digits = 4), 
#                               ")", 
#                               sep = "")) %>%
#   pivot_wider(id_cols = annotation, names_from = track, values_from = label) %>%
#   View()

# # Number of regions in each annotation track
# df %>%
#   dplyr::select(track, track_nsegments) %>%
#   unique()

# heatmap with significant label all
df %>%
  dplyr::filter(str_detect(annotation, "Random_", negate = T)) %>%
  dplyr::mutate(sigLable = toolkit_tyj$returnAsterisk(qvalue)) %>%
  dplyr::mutate(label = ifelse(sigLable == "ns", "ns",
                               str_c(format(l2fold, digits = 2), "\n", sigLable, sep = ""))) %>%
  ggplot(aes(x = track, y = annotation, fill = l2fold)) + 
    geom_tile() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    geom_text(aes(label = label), size = 1.5, lineheight = 0.8) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    labs(caption = "50bp") +
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/nacentRNA_TSS/noRandom_GAT_results_TSS_ol_enhancers_50bp_20210526", 
                     width = 12, height = 8)

# heatmap with significant label (selected)
df %>%
  dplyr::filter(str_detect(annotation, "(Random_|sEnhancer|pEnhancer)", negate = T), 
                track_nsegments > 100) %>%
  dplyr::mutate(sigLable = toolkit_tyj$returnAsterisk(qvalue)) %>%
  dplyr::mutate(label = ifelse(sigLable == "ns", "ns",
                               str_c(format(l2fold, digits = 2), "\n", sigLable, sep = ""))) %>%
  ggplot(aes(x = track, y = annotation, fill = l2fold)) + 
    geom_tile() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    geom_text(aes(label = label), size = 1.5, lineheight = 0.8) + 
      scale_fill_gradient2(high = brewer.pal(11, "RdYlGn")[1], 
                           mid = brewer.pal(11, "RdYlGn")[6], 
                           low = brewer.pal(11, "RdYlGn")[9], midpoint = 0) + 
    toolkit_tyj$theme_my + 
    labs(caption = "50bp") +
    theme(legend.key.width = unit(0.4, units = "cm"), 
          legend.margin = margin(-1, -1, -1, -6))
toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/nacentRNA_TSS/selected_GAT_results_TSS_ol_enhancers_50bp_20210526", 
                     width = 12, height = 8)

```


### 5.1.3 lncRNA ol enhancer

#### .0 Datasets
1. lncRNA from gff3 (ensembl v51)
2. 37238 lncNATs listed in `Supplemental Table 1a` of `Wang2014_GenomeRes`.
3. 6480 intergenic lncRNA listed in `Supplemental Dataset 6` of `Li et al., 2012, Plant Cell`.

*extract from TxDb*
Only 394 ncRNA were contained in `TxDb.Athaliana.BioMart.plantsmart28`.
```{r}
source("D:/SUST/code/TanYongjun_code.R")
library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28
txdb51 <- makeTxDbFromGFF(file = "./refgenome/Arabidopsis_thaliana.TAIR10.51.gff3" )

genesAll <- genes(txdb, columns = columns(txdb)) %>% as.data.frame()
table(unlist(genesAll$TXTYPE))

genesAll51 <- genes(txdb51, columns = columns(txdb51)) %>% as.data.frame()
table(unlist(genesAll51$TXTYPE))
```

*extract info from GFF3*
```{r}
source("D:/SUST/code/TanYongjun_code.R")
At_ensembl51 <- read.gff("./refgenome/Arabidopsis_thaliana.TAIR10.51.gff3.gz")
table(At_ensembl51$type)
lncRNA <- At_ensembl51 %>%
  dplyr::filter(type == "lnc_RNA")

# lncRNA %>%
#   dplyr::select(1, 4, 5) %>%
#   fwrite(file = "./refgenome/lncRNA_ensembl51_extracted_20210508.bed.gz", sep = "\t", 
#          row.names = F, col.names = F)

```

*Generate .bed file for GAT*
Four columns: chr, start, end, annotation.
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  lncRNA <- fread("./refgenome/lncRNA/lncRNA_ensembl51_extracted_20210508.bed.gz") %>%
    dplyr::mutate(V4 = "ensembl_v51_gff3")
  tmp <- fread("./refgenome/lncRNA/Intergenic_lncRNA_Li2012_PlantCell_Supplemental Dataset 6. Classification of the TUs identifed by RepTAS.tsv") %>%
    dplyr::select(2, 4, 5) %>%
    dplyr::mutate(V4 = "Intergenic_lncRNA_Li2012")
  names(tmp) <- str_c("V", 1:4, sep = "")
  
  tmp1 <- fread("./refgenome/lncRNA/lncNAT_Wang2014_GenomeRes_Supplemental_Tables_1a.tsv") %>%
    dplyr::select(2, 4, 5) %>%
    dplyr::mutate(V4 = "lncNATs_Wang2014")
  names(tmp1) <- str_c("V", 1:4, sep = "")
  
  lncRNA <- rbind(lncRNA, tmp, tmp1) %>%
    dplyr::mutate(V1 = str_replace_all(V1, "chr", "")) %>%
    dplyr::mutate(V1 = str_c("chr", V1, sep = ""))
  table(lncRNA$V1, lncRNA$V4)
  
  lncRNA %>%
    dplyr::mutate(width = V3 - V2) %>%
    ggplot(aes(x = width, color = V4, fill = V4)) + 
      geom_density() + 
      scale_x_log10() + 
      toolkit_tyj$theme_my + 
      scale_color_npg() + 
      scale_fill_npg(alpha = 1/2) + 
      theme(legend.position = c(0.8, 0.85), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.title = element_blank()) +
      ggtitle("Length of lncRNA from three sources")
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/lncRNA/Length_of_lncRNA_from_three_sources", 
                       width = 8, height = 6)
  
  
 # write to bed file
  for (i in unique(lncRNA$V4)) {
    lncRNA %>%
      dplyr::filter(V4 == i) %>%
      fwrite(file = paste("./refgenome/lncRNA/lncRNA_", i,"_for_GAT_20210508.bed.gz", sep = ""),
       sep = "\t", col.names = F)
  }

```

#### .1 GAT
`/scratch/2021-05-03/bioTanyj/At_enhancer/scripts/5.1.3.lncRNA_overlap_GAT.sh`
`/scratch/2021-05-03/bioTanyj/At_enhancer/scripts/5.1.3.lncRNA_overlap_GAT.R`

#### .2 collelct and plot
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")

# # in seperate file (20210508)
# df <- NULL
# for (i in list.files(path = "./R5.chromatin_states/lncRNA/", 
#                      pattern = ".+20210508.tsv")) {
#   df <- fread(paste("./R5.chromatin_states/lncRNA/", i, sep = "")) %>%
#     dplyr::mutate(track = str_replace_all(i, "GAT_results_(.+)_2021.+", "\\1")) %>%
#     rbind(., df)
# }

# 20210514 (no difference was found between results calculated in 20210508 and 20210514)
df <- fread("./R5.chromatin_states/lncRNA/GAT_results_all_20210514.tsv")

df %>%
  dplyr::mutate(siglabel = toolkit_tyj$returnAsterisk(p.adjust(pvalue, method = "BH"))) %>%
  ggplot(aes(y = annotation, x = l2fold, fill = siglabel)) + 
    geom_col() + 
    facet_wrap("track", nrow = 1, scales = "free_x") + 
    toolkit_tyj$theme_my + 
    theme(legend.position = "bottom", legend.key.size = unit(0.3, units = "cm"), legend.margin = margin(-5, 0,0,0)) + 
    scale_fill_manual(values = c("grey55", "grey45", "grey40", "grey79")) + 
    xlab(expression(log[2](fold~change)))
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/lncRNA/Enrichment_of_CRE_all_in_lncRNA",
                       width = 12, height = 8)

## each CRE as a whole
tmp <- df %>%
  dplyr::mutate(siglabel = toolkit_tyj$returnAsterisk(p.adjust(pvalue, method = "BH")),
                annotation = str_replace_all(annotation, "^Enhancer$", "Enhancer_STARRseq")) %>%
  dplyr::filter(str_detect(annotation, "(Enhancer_)"), 
                str_detect(track, "ensembl")) 
tmp %>%
  ggplot(aes(y = annotation, x = l2fold, fill = siglabel)) + 
    geom_col() + 
    # facet_wrap("track", nrow = 1, scales = "free_x") + 
    toolkit_tyj$theme_my + 
    theme(legend.position = c(0.82, 0.2), 
          legend.key.size = unit(0.3, units = "cm"), 
          legend.margin = margin(-5, 0,0,0), 
          legend.title = element_blank()) + 
    scale_fill_manual(values = c("grey40", "grey79")) + 
    xlab(expression(log[2](fold~change)))
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/lncRNA/Enrichment_of_CREwhole_in_lncRNA",
                       width = 6, height = 4)
```

### 5.1.4 ChrState on TSS-TTS

#### .1 prepare bed for genes
All gene, active genes, silence genes
```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  gene_tab <- fread("./R6.expression/Gene_list_with_CREs_500TSS100_300TTS300_20210322.csv.gz") %>%
    dplyr::select(seqnames, start, end, gene_id, strand, RPKM) %>%
    dplyr::mutate(RPKM = ifelse(is.na(RPKM), 0, RPKM))
    
# CorePro of active/silent genes
  genesAT <- genes(txdb) %>% 
    as.data.frame() %>%
    left_join(., gene_tab %>% dplyr::select(gene_id, RPKM))
  
# write to bed file (Active vs Silent)
  genesAT$Type <- ifelse(genesAT$RPKM > 1, "ActiveGene", "SilentGene")
  table(genesAT$Type)
  
  for (i in unique(genesAT$Type)) {
    genesAT %>%
      dplyr::filter(Type == i) %>%
      dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
      dplyr::select(seqnames, start, end, gene_id, RPKM, strand) %>%
      fwrite(str_c("./R5.chromatin_states/TSS2TTS_heatmap_ChrState/", 
                   i, sep = ""), 
             col.names = F, row.names = F, sep = "\t")
  }
  
# Activity level
  quantile(genesAT$RPKM, probs = seq(0, 1, 0.1))
  table(cut(genesAT$RPKM, breaks = c(-1, 0.0001, 1, 6, 14, 35, Inf)))
  genesAT$Type <- cut(genesAT$RPKM, breaks = c(-1, 0.0001, 1, 6, 14, 35, Inf)) %>%
    as.numeric() %>%
    paste("ActivityPart", ., sep = "-")
  
  for (i in unique(genesAT$Type)) {
    genesAT %>%
      dplyr::filter(Type == i) %>%
      dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
      dplyr::select(seqnames, start, end, gene_id, RPKM, strand) %>%
      fwrite(str_c("./R5.chromatin_states/TSS2TTS_heatmap_ChrState/", 
                   i, sep = ""), 
             col.names = F, row.names = F, sep = "\t")
  }
  
  
```

#### .2 calculate enrichment profile with deeptools.


#### .3 plot in R
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")

## load
  df <- NULL
  for (i in list.files(path = "./R5.chromatin_states/TSS2TTS_heatmap_ChrState/", 
                       pattern = ".+_plotProfile.tab.gz")) {
    cat(i, "\n")
    tmp <- fread(paste("./R5.chromatin_states/TSS2TTS_heatmap_ChrState/", i, sep = ""),
                skip = 1)
    df <- rbind(df, tmp[-1,])
  }
  
  dim(df)
  colnames(df) <- c("Chromatin_states", "CRE", 1:(ncol(df)-2))
  df <- df %>% 
    pivot_longer(cols = 3:ncol(.), names_to = "Position") %>%
    dplyr::mutate(Position = as.numeric(Position), 
                  Chromatin_states = str_replace_all(Chromatin_states, "seq", ""))
    

## each CRE type as a whole.
  library(ggformula)
  p <- df %>%
    ggplot() + 
      facet_wrap(facets = "Chromatin_states", scales = "free_y") + 
      geom_vline(xintercept = c(250, 500), linetype = 3, color = "grey35", size = 0.2) +
      geom_spline(aes(x = Position, y = value, color = CRE, linetype = CRE), size = 0.3) +
      # geom_point(size = 0.05) +
      scale_color_manual(values = carto_pal(6, "Temps")) +
      scale_linetype_manual(values = c(1,1,1,1,1,1)) +
      toolkit_tyj$theme_my + 
      theme(legend.key.height = unit(0.3, units = "cm"),
            legend.position = c(0.92, 0.06), legend.title = element_blank()) + 
      guides(color=guide_legend(ncol = 1)) + 
      scale_x_continuous(breaks = c(1, 250, 500, 750), labels = c("-5kb", "TSS", "TTS", "5kb")) + 
      ylab("Epigenetic score")
  
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/TSS2TTS_heatmap_ChrState/All_chromatin_states_profile_20210624", 
                       width = 14, height = 10, plot = p)

```


## 5.2 Kmeans

Steps: 1.
Compute score matrix (scale-region mode).
2.
PCA 3.
Kmeans cluter.

### 5.2.1 Matrix of mean score

Calculate score matrix using `sclae-region mode` of `deeptools-compute matrix` *Attention: 1. CREs located on chrC and chrM were not used in analysis. 2. Random in cre_score and enhancer were generated in different times. 3. Promoters of some transcript from the same gene locus were the same.* Profile of the whole genome.

```{bash}
cd /scratch/2021-03-11/bioTanyj/At_enhancer/R5.chromatin_states/CRE_score_matrix
# make bins of the whole genome.
bedtools makewindows -b TAIR10_chr_size_noCM.tsv -w 600 >  bins_100bp_TAIR10.bed
```

*Calculate the score matrix of CREs and the whole genome* `/scratch/2021-03-11/bioTanyj/At_enhancer/scripts/5.1.1_ChrStates_heatmap_profile.R`

*Collect the score matrix* `/scratch/2021-03-11/bioTanyj/At_enhancer/scripts/5.2.1_collect_CRE_score_matrix.R`

### 5.2.2 PCA and Kmeans

*Target* 1.
Illustrate the chromatin states of all CREs. 2.
Classify enhancer/promoters into different clusters: 2.1 Relationship between enhancer cluster and expression.
=\> `Activity of enhancer affected by chromatin states.` 2.2 TFs enriched in each enhancer cluster/promoter, and enhancer_DHS/H3K4me1 (can be seem as only one state).

```{r, summary of ChrState score}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
genome_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_whole_genome_100bpbins_20210314.csv.gz") %>%
  dplyr::mutate(method = "WholeGenome") %>%
  pivot_wider(id_cols = c(seqnames:end), names_from = "chrState", values_from = "score")
cre_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210314.csv.gz") %>%
  unique() %>%
  pivot_wider(id_cols = c(seqnames, start, end, method), names_from = "chrState", values_from = "score") # CRE located on chrC and chrM were discarded in `ComputeMatrix`.

## Heatmap of all chromatin states.
# png(filename = "./R5.chromatin_states/CRE_score_matrix/Heatmap_correlation_among_ChrStates_WholeGenome.png",
    # width = 18, height = 16, units = "cm", bg = "transparent", res = 600)
# ComplexHeatmap::Heatmap(cor(scale(genome_score[,4:ncol(genome_score)]),
#                             use = "pairwise.complete.obs", method = "pearson"))
# dev.off()

pheatmap(cor(scale(genome_score[,4:ncol(genome_score)]),
                            use = "pairwise.complete.obs", method = "pearson"),
         display_numbers = T,
         filename = "./R5.chromatin_states/CRE_score_matrix/Heatmap_correlation_among_ChrStates_WholeGenome.png")

## compare score of CRE to the whole genome.
p <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_whole_genome_100bpbins_20210314.csv.gz") %>%
  
  mutate(method = "Whole Genome") %>%
  rbind(., unique(fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210314.csv.gz"))) %>%
  dplyr::filter(method != "iSTARR-seq", method != "Random", method != "PAS") %>%
  dplyr::mutate(method = str_replace_all(method, "cSTARR-seq", "Enhancer_STARRseq"),
                method = factor(method, levels = c("Promoter", "Whole Genome", 
                                                   "Enhancer_STARRseq", "Enhancer_Zhu2015",
                                                   "Enhancer_Wang2019"))) %>%
  ggplot(aes(x = method, y = score, color = method)) +
    geom_boxplot(outlier.alpha = 0, fill = "transparent") +
    scale_fill_npg() +
    scale_color_npg() +
    toolkit_tyj$theme_my +
    scale_y_log10() +
    facet_wrap(facets = "chrState", scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), legend.position = "none")
# toolkit_tyj$SavePlot("./R5.chromatin_states/CRE_score_matrix/Distribution_of_chromatin_score3",
#                      width = 16, height = 12, plot = p)
rm(p)

```

#### .1 XXX  use all chrStates (Not used)

```{r PCA}
source("D:/SUST/code/TanYongjun_code.R")
cre_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210314.csv.gz") %>%
  unique() %>%
  pivot_wider(id_cols = c(seqnames, start, end, method), names_from = "chrState", values_from = "score") # CRE located on chrC and chrM were discarded in `ComputeMatrix`.

### PCA
    res.pca <- FactoMineR::PCA(cre_score[,5:ncol(cre_score)], scale.unit = T, graph = T)
    pca_score <- res.pca$ind$coord %>% as.data.frame()
    
    cre_score <- cbind(cre_score, pca_score)
    rm(res.pca, p, pca_score)

## merge with metainfo
    cre_all <- fread("./R5.chromatin_states/CRE_score_matrix/Enahncers_DHS_STARRseq_H3K4_Pro_Random_resized_20210313.csv.gz") %>%
      mutate(seqnames = str_c("chr", seqnames, "")) %>%
      mutate(startBK = ifelse(is.na(startBK), start, startBK), 
             endBK = ifelse(is.na(endBK), end, endBK))
    
    df <- full_join(cre_all, cre_score) %>%
      mutate(start = startBK, 
             end = endBK)

    # hist(df$endBK - df$startBK)
    # fwrite(df, file = "./Enhancers_DHS_STARRseq_H3K4_Pro_Random_PCA_score_20210315.csv.gz")
    # fwrite(df, file = "./Enhancers_DHS_STARRseq_H3K4_Pro_Random_PCA_score_20210320.csv.gz")
    # rm(cre_score)
```

```{r optimal K value}
## determined optimal K value based on elbow method.

  source("D:/SUST/code/TanYongjun_code.R")
  # df <- fread("./Enhancers_DHS_STARRseq_H3K4_Pro_Random_PCA_score_20210315.csv.gz") %>%
  df <- fread("./Enhancers_DHS_STARRseq_H3K4_Pro_Random_PCA_score_20210320.csv.gz") %>%
    dplyr::filter(str_detect(method, "(cSTARR|iSTARR)")) %>%
    dplyr::filter(method != "Promoter", 
                  method != "PAS",
                  !is.na(Dim.1))
  table(df$method)
  
# run
  k_ls <- 2:20
  N_iter <- 100 # number of iteration in kmeans()

  ss_ratio <- NULL
  for (j in k_ls) {
    cat(j, " ")
    x <- kmeans(as.matrix(df[,which(str_detect(colnames(df), "Dim"))]), iter.max = N_iter, centers = j)
    ss_ratio <- c(ss_ratio, x$tot.withinss)
  }

  p <- ggplot(data = data.frame(k = k_ls, SS_ratio = ss_ratio), aes(x = k, y = SS_ratio)) + 
    geom_vline(aes(xintercept = k), linetype = 2, color = "grey80") +
    # geom_vline(xintercept = 10, color = "steelblue", linetype = 2) +
    geom_point(color = "grey20") +
    geom_line() + 
    toolkit_tyj$theme_my + 
    xlab(expression(K[kmeans])) + 
    ylab(label = "tot.withinss") + 
    labs(caption = "PCA score") +
    scale_x_continuous(breaks = k_ls) +
    ggtitle("Determination of ptimal K values for Kmeans cluster") 
  p
  toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_score_matrix/Optiam_K_value_SS-tot_withness_kmeans_cSTARRseq1.3_pca_20210320",
    device = "png", width = 12, height = 8, dpi = 600)

## Kmeans cluster
  k <- 4
  N_iter <- 100

  set.seed(12345678)
  cre_kmeans <- kmeans(as.matrix(df[,which(str_detect(colnames(df), "Dim"))]), iter.max = N_iter, centers = k)
  df$cluster <- cre_kmeans$cluster
  
  # Number of CRE in each cluster
  df %>%
    group_by(method, cluster) %>%
    dplyr::summarise(Num = n()) %>%
    mutate(cluster = as.character(cluster)) %>%
    ggplot(aes(x = method, y = Num, color = cluster, fill = cluster)) + 
      geom_col(position = "dodge2") +
      toolkit_tyj$theme_my +
      scale_fill_npg() +
      scale_color_npg() + 
      # scale_y_log10() + 
      ggtitle("Number of CREs in each cluster") + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  toolkit_tyj$SavePlot(filename_prefix = paste("./R5.chromatin_states/CRE_score_matrix/Number_of_elements_in_each_clusters_K=",
                                               k, "20210320.png", sep = ""))
  
  fwrite(df, paste("./Enhancers_DHS_STARRseq_H3K4_Random_PCA_score_KmeansCluster_K=", 
                   k, "_20210320.csv.gz", sep = ""))
  
```

##### .XXX plot

```{r, Heatmap of chrStates of each CRE cluster}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
# df <- fread("./Enhancers_DHS_STARRseq_H3K4_Random_PCA_score_KmeansCluster6_20210320.csv.gz")
df$name <- str_c(df$method, df$cluster, df$seqnames, df$start, df$end, sep = "_")


p <- df %>%
  dplyr::select(cluster, method, ATACseq:RNAPOII) %>%
  pivot_longer(cols = ATACseq:RNAPOII) %>%
  mutate(cluster = as.factor(cluster)) %>%
  ggplot(aes(x = method, y = value, color = cluster)) +
    geom_violin(fill = "transparent") + 
    toolkit_tyj$theme_my + 
    scale_y_log10() + 
    facet_wrap("name") + 
    theme(legend.position = "bottom", axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    scale_color_npg() + 
    guides(guide_legend(nrow = 1))
toolkit_tyj$SavePlot(filename_prefix = paste("./R5.chromatin_states/CRE_score_matrix/ViolinPlot_Chromatin_score_CRE_cluters_K=",
                                             k, "_20210320", sep = ""),
                     width = 25, height = 16, plot = p)

## scale chromatin score
df_scale <- df %>%
  dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
  pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
               values_to = "signal") %>%
  mutate(signal = ifelse(is.na(signal), 0, signal)) %>%
  group_by(method, histone) %>%
  dplyr::mutate(bins = as.numeric(cut(signal, 
                               breaks = unique(c(-1, 
                                          quantile(signal,probs = seq(0, 1, 0.01))[2:100],
                                          1000)))
                                  )
         ) %>%
  dplyr::mutate(bins = (bins - min(bins)) / (max(bins) - min(bins))) %>%
  dplyr::mutate(bins = bins * 2 - 1)

df_scale %>%
  ggplot(aes(x = histone, y = bins)) + 
    geom_violin(fill = "transparent") + 
    facet_grid(method ~ .) + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))

# to wider format
df_scale <- df_scale %>%
  dplyr::select(seqnames:end, cluster, method, histone, bins) %>% 
  pivot_wider(id_cols = seqnames:method, 
              names_from = "histone",
              values_from = "bins")

# if not scale to quantile
# if (T) {
#   df_scale <- df %>%
#     dplyr::select(seqnames, start, end, cluster, method, ATACseq:RNAPOII)
#   for (i in 6:ncol(df_scale)) {
#     df_scale[,i] <- (df_scale[,i] - mean(df_scale[,i])) / sd(df_scale[,i])
#   }
# }

## plot heatmap
for (i in unique(df_scale$method)) {
  # subset data
  tmp <- df_scale %>%
    dplyr::filter(method == i) %>%
    arrange(cluster) %>%
    mutate(cluster =  paste("Cluster-", cluster, sep = "")) %>%
    dplyr::rename(Cluster = cluster)
  
  # prepare
  mat_df <- as.matrix(tmp[,6:ncol(tmp)])
  anno_temp <- data.frame(Cluster = tmp$Cluster)
  row.names(anno_temp) <- row.names(mat_df) <- paste(1:nrow(mat_df))
  # show_col(pal_npg("nrc")(9))
  anno_color <- list(Cluster = c(pal_npg("nrc")(9)[c(1,2,4,3,5,6)]))
  names(anno_color$Cluster) <- paste("Cluster-", 1:6, sep = "")
  
  # plot
  pheatmap(mat = mat_df, 
         kmeans_k = NA, 
         scale = "none", 
         cluster_rows = F, 
         show_rownames = F, 
         cluster_cols = F,
         annotation_row = anno_temp,
         annotation_colors = anno_color,
         width = 4.8, height = 4, 
         main = i,
         filename = paste("./R5.chromatin_states/CRE_score_matrix/Heatmap_scaled_",
                          i, "_K=", k,"_20210320-1.png", sep = ""),
         color = colorRampPalette(c("steelblue", "white", "red"), space = "Lab",
                                    interpolate = "linear")(50))
}

```

#### .2 exclude ATAC/FAIRE/MNase

```{r PCA}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
cre_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210314.csv.gz") %>%
  unique() %>%
  pivot_wider(id_cols = c(seqnames, start, end, method), names_from = "chrState", values_from = "score") # CRE located on chrC and chrM were discarded in `ComputeMatrix`.

### PCA
  columID  <- c(6:7, 9:19, 21)
  names(cre_score)[columID]
  set.seed(12345)
  res.pca <- FactoMineR::PCA(cre_score[, columID], scale.unit = T, graph = F)
  pca_score <- res.pca$ind$coord %>% as.data.frame()
  
  cre_score <- cbind(cre_score, pca_score)
  rm(res.pca, p, pca_score)

## merge with CRE info
  cre_all <- fread("./R5.chromatin_states/CRE_score_matrix/Enahncers_DHS_STARRseq_H3K4_Pro_Random_resized_20210313.csv.gz") %>%
    mutate(seqnames = str_c("chr", seqnames, ""))
  cre_all$startBK[is.na(cre_all$startBK)] <- cre_all$start[is.na(cre_all$startBK)]
  cre_all$endBK[is.na(cre_all$endBK)] <- cre_all$end[is.na(cre_all$endBK)]
  
  df <- full_join(cre_all, cre_score) %>%
    mutate(start = startBK, end = endBK)

  hist(df$endBK - df$startBK)
  # fwrite(df, file = "./Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210320.csv.gz")
  # rm(cre_score)
```

```{r optimal K value}
## determined optimal K value based on elbow method.
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210320.csv.gz") %>%
    dplyr::filter(str_detect(method, "(iSTARR|cSTARR)"),
                  !is.na(Dim.1))
  # keep enhancer_cSTARRseq > 1.3 and promoter
  df <- df[!(df$method == "cSTARR-seq" & df$enrichment < 1.3),]
  df <- df[!(df$method == "iSTARR-seq" & df$enrichment < 1.3),]
  table(df$method)
  
# run
  k_ls <- 2:20
  N_iter <- 100000 # number of iteration in kmeans()

  ss_ratio <- NULL
  for (j in k_ls) {
    cat(j, " ")
    set.seed(12345)
    x <- kmeans(as.matrix(df[,which(str_detect(colnames(df), "Dim"))]), iter.max = N_iter, centers = j)
    ss_ratio <- c(ss_ratio, x$tot.withinss)
  }

  p <- ggplot(data = data.frame(k = k_ls, SS_ratio = ss_ratio), aes(x = k, y = SS_ratio)) + 
    geom_vline(aes(xintercept = k), linetype = 2, color = "grey80") +
    # geom_vline(xintercept = 10, color = "steelblue", linetype = 2) +
    geom_point(color = "grey20") +
    geom_line() + 
    xlab(expression(K[kmeans])) + 
    ylab(label = "tot.withinss") + 
    # labs(caption = "PCA score with subseted ChrStates") +
    scale_x_continuous(breaks = k_ls) +
    # ggtitle("Determination of ptimal K values for Kmeans cluster") +
    toolkit_tyj$theme_my
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_score_matrix/Optiam_K_value_SS-tot_withnss_K_kmeans_subsetChrState_PCA_cSTARRseq1.3_DHS_H3K4_20210415",
  #                      width = 7, height = 5, dpi = 600)

```

##### .plot

*2021*
14 epigenetic scores were used in heatmap.
```{r, Heatmap of chrStates of each CRE cluster}
source("D:/SUST/code/TanYongjun_code.R")
## Kmeans cluster
for (k in 4:8) {
  ## kmeans
      # k <- 4
      N_iter <- 100000
    
      set.seed(12345)
      cre_kmeans <- kmeans(as.matrix(df[,which(str_detect(colnames(df), "Dim"))]), iter.max = N_iter, centers = k)
      df$cluster <- cre_kmeans$cluster
      
      # Number of CRE in each cluster
      df %>%
        group_by(method, cluster) %>%
        dplyr::summarise(Num = n()) %>%
        mutate(cluster = as.character(cluster)) %>%
        ggplot(aes(x = method, y = Num, color = cluster, fill = cluster)) + 
          geom_col(position = "dodge2") +
          toolkit_tyj$theme_my +
          scale_fill_npg() +
          scale_color_npg() + 
          # scale_y_log10() + 
          ggtitle("Number of CREs in each cluster")
      
      # fwrite(df, paste("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_K=",
      #                  k, "_20210315.csv.gz", sep = ""))
    df <- fread(paste("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_K=",
                       k, "_20210315.csv.gz", sep = ""))
      
    df$name <- str_c(df$method, df$cluster, df$seqnames, df$start, df$end, sep = "_")

  ## Violin plot of each cluster
    # p <- df %>%
    #   dplyr::select(cluster, method, ATACseq:RNAPOII) %>%
    #   pivot_longer(cols = ATACseq:RNAPOII) %>%
    #   mutate(cluster = as.factor(cluster)) %>%
    #   ggplot(aes(x = method, y = value, color = cluster)) +
    #     geom_violin() + 
    #     toolkit_tyj$theme_my + 
    #     scale_y_log10() + 
    #     facet_wrap("name") + 
    #     theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    #     scale_color_npg()
    # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_score_matrix/Chromatin_score_CRE_subsetChrState_cluters_20210315",
    #                      width = 16, height = 16, plot = p)

  ## scale chromatin score
    df_scale <- df %>%
      dplyr::select(seqnames, start, end, method, cluster, DHS, DNAmethy, H2A:H3K9me2, RNAPOII) %>%
      pivot_longer(cols = DHS:RNAPOII, names_to = "histone", 
                   values_to = "signal") %>%
      group_by(method, histone) %>%
      mutate(signal = ifelse(is.na(signal), 0, signal)) %>%
      dplyr::mutate(bins = as.numeric(cut(signal, 
                                   breaks = unique(c(-1, 
                                              quantile(signal,probs = seq(0, 1, 0.01))[2:100],
                                              1000)))
                                      )
             ) %>%
      dplyr::mutate(bins = (bins - min(bins)) / (max(bins) - min(bins))) %>%
      dplyr::mutate(bins = bins * 2 - 1)
    
    # df_scale %>%
    #   ggplot(aes(x = histone, y = bins)) + 
    #     geom_violin(fill = "transparent") + 
    #     facet_grid(method ~ .)
    
    df_scale <- df_scale %>%
      dplyr::select(seqnames:end, cluster, method, histone, bins) %>% 
      pivot_wider(id_cols = seqnames:method, 
                  names_from = "histone",
                  values_from = "bins")

  ## plot heatmap
    for (i in unique(df_scale$method)) {
      # subset data
      tmp <- df_scale %>%
        dplyr::filter(method == i) %>%
        arrange(cluster) %>%
        mutate(cluster =  paste("Cluster-", cluster, sep = "")) %>%
        dplyr::rename(Cluster = cluster)
      
      # prepare
      mat_df <- as.matrix(tmp[,6:ncol(tmp)])
      anno_temp <- data.frame(Cluster = tmp$Cluster)
      row.names(anno_temp) <- row.names(mat_df) <- paste(1:nrow(mat_df))
      # show_col(pal_npg("nrc")(9))
      anno_color <- list(Cluster = c(pal_npg("nrc")(9)[1:k]))
      names(anno_color$Cluster) <- paste("Cluster-", 1:k, sep = "")
      
      # plot
      pheatmap(mat = mat_df, 
             kmeans_k = NA, 
             scale = "none", 
             cluster_rows = F, 
             show_rownames = F, 
             cluster_cols = F,
             annotation_row = anno_temp,
             annotation_colors = anno_color,
             width = 8 * 0.393700787, height = 8 * 0.393700787, 
             main = paste(i, " K=",k, sep = ""),
             filename = paste("./R5.chromatin_states/CRE_score_matrix/Heatmap_scaled_subsetChrStates_K=",
                              k, "_", i, "_20210329_1.png", sep = ""),
             # color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
             color = colorRampPalette(colors = rev(c("firebrick3", "white", "dodgerblue3")))(100),
             fontsize = 6, fontsize_col = 6, fontsize_row = 6,
             legend_labels = T
             )
    }
} # k
```


*20220422*
with Kmeans cluster info generated in 2021, but add other epigenetic scores (N=21).
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")

## load kmeans info
  df <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_K=4_20210315.csv.gz") %>%
    dplyr::filter(str_detect(method, "cSTARR|Promoter|Random"),
                  !is.na(Dim.1)) %>%
    dplyr::select(seqnames:end, enrichment, method, ID, cluster)
  # keep enhancer_cSTARRseq > 1.3
  df <- df[!(df$method == "cSTARR-seq" & df$enrichment < 1.3),]
  table(df$method)

## add epigenetic scores
  cre_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210416.csv.gz") %>%
    unique() %>%
    pivot_wider(id_cols = c(seqnames, start, end, method), names_from = "chrState", values_from = "score") # CRE located on chrC and chrM were discarded in `ComputeMatrix`.

  # merge with CRE info (the real start/end of each CRE)
  cre_all <- fread("./R5.chromatin_states/CRE_score_matrix/Enahncers_DHS_STARRseq_H3K4_Pro_Random_resized_20210313.csv.gz") %>%
    mutate(seqnames = str_c("chr", seqnames, ""))
  cre_all$startBK[is.na(cre_all$startBK)] <- cre_all$start[is.na(cre_all$startBK)]
  cre_all$endBK[is.na(cre_all$endBK)] <- cre_all$end[is.na(cre_all$endBK)]
  
  df1 <- full_join(cre_all, cre_score) %>%
    mutate(start = startBK, end = endBK)
  
  df <- left_join(df, df1) %>%
    dplyr::select(-c(ATACseq, FAIREseq, MNaseseq))
  rm(cre_all, df1, cre_score)
  
## scale chromatin score
  df_scale <- df %>%
    dplyr::select(seqnames, start, end, method, cluster, DHS:RNAPOII) %>%
    pivot_longer(cols = DHS:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    group_by(method, histone) %>%
    mutate(signal = ifelse(is.na(signal), 0, signal)) %>%
    dplyr::mutate(bins = as.numeric(cut(signal, 
                                 breaks = unique(c(-1, 
                                            quantile(signal,probs = seq(0, 1, 0.01))[2:100],
                                            1000)))
                                    )
           ) %>%
    dplyr::mutate(bins = (bins - min(bins)) / (max(bins) - min(bins))) %>%
    dplyr::mutate(bins = bins * 2 - 1)
  
  # df_scale %>%
  #   ggplot(aes(x = histone, y = bins)) +
  #     geom_violin(fill = "transparent") +
  #     facet_grid(method ~ .)
  
  df_scale <- df_scale %>%
    dplyr::select(seqnames:end, cluster, method, histone, bins) %>% 
    pivot_wider(id_cols = seqnames:method, 
                names_from = "histone",
                values_from = "bins")

  ## plot heatmap
  k <- 4
  i <- unique(df_scale$method)[1]
  for (i in unique(df_scale$method)) {
    # i <- unique(df_scale$method)[1]
    # subset data
    tmp <- df_scale %>%
      dplyr::filter(method == i) %>%
      arrange(cluster) %>%
      mutate(cluster =  paste("Cluster-", cluster, sep = "")) %>%
      dplyr::rename(Cluster = cluster)
    
    # prepare
    Epi_order <- c("DHS", "RNAPOII", "H3K27ac", "H3K36me3", 
                   "H3K4me2", "H3K4me3", "H3K9ac", "H3K14ac", "H4K16ac", "H3K23ac", "H3K36ac","H3K56ac", "H2AZ",
                   
                   "H3K9me1", "H3K9me2", "DNAmethy", "H2AW", 
                   
                   "H3K4me1", "H3K27me3", "H2A", "H2AX")
    mat_df <- tmp %>%
      ungroup() %>%
      # dplyr::relocate(DHS, RNAPOII, H2A, , , , , , , , , , , , , ) %>%
      dplyr::select(DHS:RNAPOII) %>%
      map_dfc(., as.numeric)
    mat_df <- mat_df[Epi_order] %>%
      as.matrix()
    colnames(mat_df) <- colnames(mat_df) %>%
      str_replace_all(., "H2A", "H2A.") %>%
      str_replace_all(., "^H2A.$", "H2A")
    
    anno_temp <- data.frame(Cluster = tmp$Cluster)
    row.names(anno_temp) <- row.names(mat_df) <- paste(1:nrow(mat_df))
    # show_col(pal_npg("nrc")(9))
    anno_color <- list(Cluster = toolkit_tyj$AtEn_KmeanC_color)
    names(anno_color$Cluster) <- paste("Cluster-", 1:k, sep = "")
    
    # plot (order of histone f)
    pheatmap::pheatmap(mat = mat_df, 
           kmeans_k = NA, 
           scale = "none", 
           cluster_rows = F, 
           show_rownames = F, 
           cluster_cols = F,
           annotation_row = anno_temp,
           annotation_colors = anno_color,
           gaps_row = table(anno_temp) %>% cumsum() %>% head(3), 
           treeheight_col = 0,
           # main = paste(i, " K=",k, sep = ""),
           filename = paste("./R5.chromatin_states/CRE_score_matrix/Heatmap_scaled_AllChrStates_",i, "_K=4_20220420.png", sep = ""),
           # color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
           color = colorRampPalette(colors = rev(c("firebrick3", "white", "dodgerblue3")))(100),
           fontsize = 6, fontsize_col = 6, fontsize_row = 6,
           legend_labels = T, width = 8/2.54, height = 6/2.54
           )
  }

  NA
```


#### .generate CRE table with cluster info

*Merge CREs used in Kmeans cluster and others(random, Promoter, PAS)*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer <- fread("Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210320.csv.gz")
dfK4 <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_K=4_20210315.csv.gz") %>%
  dplyr::select(seqnames:strand, method, cluster) %>%
  dplyr::rename(clusterK4 = cluster)
dfK7 <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_K=7_20210315.csv.gz") %>%
  dplyr::select(seqnames:strand, method, cluster) %>%
  dplyr::rename(clusterK7 = cluster)
enhancer <- full_join(enhancer, dfK4) %>%
  full_join(., dfK7) %>%
  dplyr::mutate(clusterK4 = ifelse(is.na(clusterK4), 0, clusterK4)) %>%
  dplyr::mutate(clusterK7 = ifelse(is.na(clusterK7), 0, clusterK7))

# fwrite(enhancer, "Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_20210321.csv.gz")

##  correlation between enhancer activity and PCs?
df <- enhancer %>%
  dplyr::filter(!is.na(enrichment), 
                !is.na(Dim.1)) %>%
  dplyr::select(c(1:5, 9, 10, 34:38))
cor(df[, c(6, 8:12)]) %>%
  pheatmap(kmeans_k = NA, 
           scale = "none", 
           cluster_rows = F, 
           show_rownames = T, 
           cluster_cols = F,
           display_numbers = T,
           main = "Pearson correlation among PCs and STARRseq enhancer activity",
           color = colorRampPalette(rev(brewer.pal(n = 7, name =
                                                   "RdBu")))(100))

## correlaiton between enhancer activity and ChrStates.
enhancer %>%
  dplyr::filter(!is.na(enrichment),
                method == "cSTARR-seq", 
                # enrichment > 1.3,
                !is.na(Dim.1)) %>%
  dplyr::select(enrichment, ATACseq:RNAPOII) %>%
  cor(use = "pairwise.complete.obs") %>%
  pheatmap(kmeans_k = NA, 
           scale = "none", 
           cluster_rows = F, 
           show_rownames = T, 
           cluster_cols = F,
           display_numbers = T,
           main = "Pearson correlation among Chromatin states and enhancer activity",
           color = colorRampPalette(rev(brewer.pal(n = 7, name ="RdBu")))(100), 
           width = 7, height = 7, 
           filename = "./R5.chromatin_states/CRE_score_matrix/Correlation_among_enhancer_enrichment_ChrState_20210321.png")
```

#### .3 plot histone score (20210608)
Just plot heatmap/violinPlot/DensityPlot of chromatin states (Enhancers identified using STARR-seq.)

##### .KmeansCluster
Kmeans cluster was performed in 20210320 with only 17 Chr Score matrix. K = 4 was used in this part.
```{r PCA}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
cre_score <- fread("./R5.chromatin_states/CRE_score_matrix/Chromatin_score_matrix_all_CRE_resized100bp_20210416.csv.gz") %>%
  unique() %>%
  pivot_wider(id_cols = c(seqnames, start, end, method), names_from = "chrState", values_from = "score") # CRE located on chrC and chrM were discarded in `ComputeMatrix`.

### PCA
  columID  <- c(6:7, 9:26, 28)
  names(cre_score)[columID]
  set.seed(12345)
  res.pca <- FactoMineR::PCA(cre_score[, columID], scale.unit = T, graph = F)
  pca_score <- res.pca$ind$coord %>% as.data.frame()
  
  cre_score <- cbind(cre_score, pca_score)
  rm(res.pca, p, pca_score)

## merge with CRE info
  cre_all <- fread("./R5.chromatin_states/CRE_score_matrix/Enahncers_DHS_STARRseq_H3K4_Pro_Random_resized_20210313.csv.gz") %>%
    mutate(seqnames = str_c("chr", seqnames, ""))
  cre_all$startBK[is.na(cre_all$startBK)] <- cre_all$start[is.na(cre_all$startBK)]
  cre_all$endBK[is.na(cre_all$endBK)] <- cre_all$end[is.na(cre_all$endBK)]
  
  df <- full_join(cre_all, cre_score) %>%
    mutate(start = startBK, end = endBK)

  # hist(df$endBK - df$startBK)
  # fwrite(df, file = "./R5.chromatin_states/CRE_score_matrix/Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210608.csv.gz")
  # rm(cre_score)
```

```{r optimal K value}
## determined optimal K value based on elbow method (Only cSTARRseq, enrichment > 1.3).
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./R5.chromatin_states/CRE_score_matrix/Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210608.csv.gz") %>%
    dplyr::filter(str_detect(method, "(cSTARR)"),
                  !is.na(Dim.1))
  # keep enhancer_cSTARRseq > 1.3 and promoter
  df <- df[!(df$method == "cSTARR-seq" & df$enrichment < 1.3),]
  df <- df[!(df$method == "iSTARR-seq" & df$enrichment < 1.3),]
  table(df$method)
  
# run
  k_ls <- 2:20
  N_iter <- 100000 # number of iteration in kmeans()

  ss_ratio <- NULL
  for (j in k_ls) {
    cat(j, " ")
    set.seed(12345)
    x <- kmeans(as.matrix(df[,which(str_detect(colnames(df), "Dim"))]), iter.max = N_iter, centers = j)
    ss_ratio <- c(ss_ratio, x$tot.withinss)
  }

  p <- ggplot(data = data.frame(k = k_ls, SS_ratio = ss_ratio), aes(x = k, y = SS_ratio)) + 
    geom_vline(aes(xintercept = k), linetype = 2, color = "grey80") +
    # geom_vline(xintercept = 10, color = "steelblue", linetype = 2) +
    geom_point(color = "grey20") +
    geom_line() + 
    toolkit_tyj$theme_my + 
    xlab(expression(K[kmeans])) + 
    ylab(label = "tot.withinss") + 
    labs(caption = "PCA score with subseted ChrStates") +
    scale_x_continuous(breaks = k_ls) +
    ggtitle("Determination of ptimal K values for Kmeans cluster") 
  # p
  # toolkit_tyj$SavePlot(filename_prefix = "./R5.chromatin_states/CRE_score_matrix/Optiam_K_value_SS-tot_withnss_K_kmeans_subsetChrState_PCA_cSTARRseq1.3_20210608",
  #                       device = "png", width = 8, height = 6, dpi = 600)

```

*Kmeans cluster results was load from 20210315*
```{r, Heatmap of chrStates of each CRE cluster}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    dplyr::select(seqnames:end, clusterK4) %>%
    dplyr::rename(cluster = clusterK4)
  df <- fread("./R5.chromatin_states/CRE_score_matrix/Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210608.csv.gz") %>%
    dplyr::filter(str_detect(method, "(cSTARR)"),
                  !is.na(Dim.1), 
                  !(method == "cSTARR-seq" & enrichment < 1.3)) %>%
    left_join(., enhancer)

## distribution of CRE's histone modification score.
  df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(x = signal)) + 
      geom_density() + 
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10()
    

## scale chromatin score
  # scale into even distribution
  df_scale <- df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    group_by(method, histone) %>%
    mutate(signal = ifelse(is.na(signal), 0, signal)) %>%
    dplyr::mutate(bins = as.numeric(cut(signal, 
                                 breaks = unique(c(-1, 
                                            quantile(signal,probs = seq(0, 1, 0.01))[2:100],
                                            1000)))
                                    )
           ) %>%
    dplyr::mutate(bins = (bins - min(bins)) / (max(bins) - min(bins))) %>%
    dplyr::mutate(bins = bins * 2 - 1)
  
  # df_scale %>%
  #   ggplot(aes(x = histone, y = bins)) +
  #     geom_violin(fill = "transparent") +
  #     facet_grid(method ~ .)
  
  df_scale <- df_scale %>%
    dplyr::select(seqnames:end, cluster, method, histone, bins) %>% 
    pivot_wider(id_cols = seqnames:method, 
                names_from = "histone",
                values_from = "bins")

## plot heatmap
  k <- 4
  for (i in unique(df_scale$method)) {
    # subset data
    tmp <- df_scale %>%
      dplyr::filter(method == i) %>%
      arrange(cluster) %>%
      dplyr::rename(Cluster = cluster)
    
    # prepare
    mat_df <- as.matrix(tmp[,6:ncol(tmp)])
    anno_temp <- data.frame(Cluster = paste("Cluster-", tmp$Cluster, sep = ""))
    row.names(anno_temp) <- row.names(mat_df) <- paste(1:nrow(mat_df))
    # show_col(pal_npg("nrc")(9))
    anno_color <- list(Cluster = c(pal_npg("nrc")(9)[c(2,1,3:9)][1:k]))
    names(anno_color$Cluster) <- paste("Cluster-", 1:k, sep = "")
    
    # plot
    pheatmap::pheatmap(mat = mat_df, 
           kmeans_k = NA, 
           scale = "none", 
           cluster_rows = F, 
           show_rownames = F, 
           cluster_cols = F,
           annotation_row = anno_temp,
           annotation_colors = anno_color,
           width = 4.8, height = 4, 
           main = paste(i, " K=",k, sep = ""),
           filename = paste("./R5.chromatin_states/CRE_score_matrix/Heatmap_scaled_subsetChrStates_K=",
                            k, "_", i, "_20210608.png", sep = ""),
           # color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100)
           color = colorRampPalette(colors = rev(c("firebrick3", "white", "dodgerblue3")))(100)
           )
  }

```

##### .ActivityCluster
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    dplyr::select(seqnames:end, V4) %>%
    dplyr::rename(cluster = V4)
  df <- fread("./R5.chromatin_states/CRE_score_matrix/Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210608.csv.gz") %>%
    dplyr::filter(str_detect(method, "(cSTARR)"),
                  !is.na(Dim.1), 
                  !(method == "cSTARR-seq" & enrichment < 1.3)) %>%
    left_join(., enhancer)

## distribution of CRE's histone modification score.
  df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(x = signal)) + 
      geom_density() + 
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10()
    
## Heatmap ---------------------
  
## scale chromatin score
  # scale into even distribution
  df_scale <- df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    group_by(method, histone) %>%
    mutate(signal = ifelse(is.na(signal), 0, signal)) %>%
    dplyr::mutate(bins = as.numeric(cut(signal, 
                                 breaks = unique(c(-1, 
                                            quantile(signal,probs = seq(0, 1, 0.01))[2:100],
                                            1000)))
                                    )
           ) %>%
    dplyr::mutate(bins = (bins - min(bins)) / (max(bins) - min(bins))) %>%
    dplyr::mutate(bins = bins * 2 - 1)
  
  
  df_scale %>%
    ggplot(aes(x = histone, y = bins)) +
      geom_violin(fill = "transparent") +
      facet_grid(method ~ .)
  
  df_scale <- df_scale %>%
    dplyr::select(seqnames:end, cluster, method, histone, bins) %>% 
    pivot_wider(id_cols = seqnames:method, 
                names_from = "histone",
                values_from = "bins")

## plot heatmap
  k <- 5
  for (i in unique(df_scale$method)) {
    # subset data
    tmp <- df_scale %>%
      dplyr::filter(method == i) %>%
      arrange(cluster) %>%
      dplyr::rename(Cluster = cluster)
    
    # prepare
    mat_df <- as.matrix(tmp[,6:ncol(tmp)])
    anno_temp <- data.frame(Cluster = tmp$Cluster)
    row.names(anno_temp) <- row.names(mat_df) <- paste(1:nrow(mat_df))
    # show_col(pal_npg("nrc")(9))
    anno_color <- list(Cluster = c(pal_npg("nrc")(9)[c(2,1,3:9)][1:k]))
    names(anno_color$Cluster) <- paste("ActivityCluster", 1:k, sep = "")
    
    # plot
    pheatmap(mat = mat_df, 
           kmeans_k = NA, 
           scale = "none", 
           cluster_rows = F, 
           show_rownames = F, 
           cluster_cols = F,
           annotation_row = anno_temp,
           annotation_colors = anno_color,
           width = 4.8, height = 4, 
           main = paste(i, " K=",k, sep = ""),
           filename = paste("./R5.chromatin_states/CRE_score_matrix/Heatmap_scaled_subsetChrStates_ActivityCluster",
                            k, "_", i, "_20210608.png", sep = ""),
           # color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100)
           color = colorRampPalette(colors = rev(c("firebrick3", "white", "dodgerblue3")))(100)
           )
  }

## Violin plot -----------------
  df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(y = cluster, x = signal, color = cluster)) + 
      # geom_density() + 
      geom_violin(fill = "transparent") +
      geom_boxplot(fill = "transparent", width = 0.2, outlier.alpha = 0) +
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10() + 
      scale_color_manual(values = c(carto_pal(5, "Temps"),
                              "grey40")) + 
      theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  toolkit_tyj$SavePlot("./R5.chromatin_states/CRE_score_matrix/ViolinPlot_ActivityCluster_Enhancer_STARRseq1.3_20210609", width = 12, height = 12, bg = "white")
  
## Density plot -----------------
  p <- df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(x = signal, color = cluster)) + 
      geom_density(size = 0.3) +
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10() + 
      scale_color_manual(values = c(carto_pal(5, "Temps"),
                              "grey40")) + 
      theme(legend.position = c(0.9,0.07), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.title = element_blank(), 
            legend.margin = margin(0,0,0,-5))
  toolkit_tyj$SavePlot("./R5.chromatin_states/CRE_score_matrix/DensityPlot_ActivityCluster_Enhancer_STARRseq1.3_20210609", width = 14, height = 12, plot = p)
```

##### .Clustered/Dispersed
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = "")) %>%
    dplyr::select(seqnames:end, ClusterType) %>%
    dplyr::rename(cluster = ClusterType)
  df <- fread("./R5.chromatin_states/CRE_score_matrix/Enhancers_DHS_STARRseq_H3K4_Pro_Random_subsetChrState_PCA_score_20210608.csv.gz") %>%
    dplyr::filter(str_detect(method, "(cSTARR)"),
                  !is.na(Dim.1), 
                  !(method == "cSTARR-seq" & enrichment < 1.3)) %>%
    left_join(., enhancer)

## distribution of CRE's histone modification score.
  df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(x = signal)) + 
      geom_density() + 
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10() + 
      ggtitle("Histone modification signal of all enhancers")
    
## Violin plot -----------------
  df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(y = cluster, x = signal, color = cluster)) + 
      # geom_density() + 
      geom_violin(fill = "transparent") +
      geom_boxplot(fill = "transparent", width = 0.2, outlier.alpha = 0) +
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10() + 
      scale_color_manual(values = c(carto_pal(5, "Temps"),
                              "grey40")) + 
      theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  toolkit_tyj$SavePlot("./R5.chromatin_states/CRE_score_matrix/ViolinPlot_ActivityCluster_Enhancer_STARRseq1.3_20210609", width = 12, height = 12, bg = "white")
  
## Density plot -----------------
  p <- df %>%
    dplyr::select(seqnames, start, end, method, cluster, ATACseq:RNAPOII) %>%
    pivot_longer(cols = ATACseq:RNAPOII, names_to = "histone", 
                 values_to = "signal") %>%
    ggplot(aes(x = signal, color = cluster)) + 
      geom_density(size = 0.3) +
      facet_wrap(facets = "histone") + 
      toolkit_tyj$theme_my + 
      scale_x_log10() + 
      scale_color_manual(values = c(carto_pal(5, "Temps"),
                              "grey40")) + 
      theme(legend.position = c(0.9,0.07), 
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.title = element_blank(), 
            legend.margin = margin(0,0,0,-5))
  toolkit_tyj$SavePlot("./R5.chromatin_states/CRE_score_matrix/DensityPlot_ActivityCluster_Enhancer_STARRseq1.3_20210609", width = 14, height = 12, plot = p)
```


## 5.3 ActivityCluster Mean



# 6. Expression

## 6.0 background info

*Distribution of genes on chromosome* Length of gene; Distance between flanking-located genes.

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28
tair10Size <- fread("./refgenome/TAIR10_chr_size_noCM.tsv") %>%
  mutate(V1 = str_replace_all(V1, "chr", ""))
names(tair10Size) <- c("seqnames", "start", "end")
TssAll <- promoters(txdb) %>%
  GenomicRanges::resize(width = 1, fix = "start", use.names = T)

## distance between TSS of nearest genes
  DisBetweenNearestTSS <- GenomicRanges::setdiff(GenomicRanges::makeGRangesFromDataFrame(tair10Size), 
                                   TssAll, ignore.strand = T) %>%
    as.data.frame() %>%
    arrange(seqnames, start)
  quantile(DisBetweenNearestTSS$width, probs = seq(0, 1, 0.05))
  DisBetweenNearestTSS %>%
    ggplot(aes(x = width)) +
      toolkit_tyj$theme_my +
      geom_vline(xintercept = median(DisBetweenNearestTSS$width), linetype = 2) +
      geom_histogram(color = "white", fill = "grey50", alpha = 1/2) + 
      scale_x_log10() +
      xlab("Distance") +
      # scale_x_continuous(breaks = c(0, 100, 500, 1000, 20000, 40000, 60000, 80000)) +
      ggtitle("Distance between TSS of each two nearest genes", 
              subtitle = paste("Median = ", median(DisBetweenNearestTSS$width), sep = ""))
  # toolkit_tyj$SavePlot(filename_prefix = "./refgenome/Distance_between_TSS_of_two_nearest_genes_20210321",
  #                      width = 12, height = 8)
  quantile(DisBetweenNearestTSS$width, probs = seq(0, 1, 0.05))
  
## distance between TTS and the following genes's TSS.
  geneAll <- genes(txdb) %>% 
    GenomicRanges::reduce()
  DisBeweenTTS2TSS <- GenomicRanges::setdiff(GenomicRanges::makeGRangesFromDataFrame(tair10Size), 
                                   geneAll, ignore.strand = T) %>%
    as.data.frame() %>%
    arrange(seqnames, start)
  DisBeweenTTS2TSS %>%
    ggplot(aes(x = width)) +
      toolkit_tyj$theme_my +
      geom_vline(xintercept = median(DisBeweenTTS2TSS$width), linetype = 2) +
      geom_histogram(color = "white", fill = "grey50", alpha = 1/2) + 
      scale_x_log10() +
      xlab("Distance") +
      # scale_x_continuous(breaks = c(0, 100, 500, 1000, 20000, 40000, 60000, 80000)) +
      ggtitle("Distance between TTS and TSS of two nearest genes", 
              subtitle = paste("Median = ", median(DisBeweenTTS2TSS$width), sep = ""))
  # toolkit_tyj$SavePlot(filename_prefix = "./refgenome/Distance_between_TTS_and_TSS_of_two_nearest_genes_20210321",
  #                      width = 12, height = 8)
  quantile(DisBeweenTTS2TSS$width, probs = seq(0, 1, 0.05))
  
## width of gene (TSS to TTS)
  df <- data.frame(width = width(geneAll)) 
  df %>%
      ggplot(aes(x = width)) +
      toolkit_tyj$theme_my +
      geom_vline(xintercept = median(df$width), linetype = 2) +
      geom_histogram(color = "white", fill = "grey50", alpha = 1/2) + 
      scale_x_log10() +
      xlab("Distance") +
      # scale_x_continuous(breaks = c(0, 100, 500, 1000, 20000, 40000, 60000, 80000)) +
      ggtitle("Distribution of ene length (from TSS to TTS)", 
              subtitle = paste("Median = ", median(df$width), 
                               "; The totoal number of non-overlapped was 32378", sep = ""))
  # toolkit_tyj$SavePlot(filename_prefix = "./refgenome/Distribution_of_gene_length_20210321",
  #                      width = 12, height = 8)
  quantile(df$width, probs = seq(0, 1, 0.05))

```

*RPKM of gene in seedling* `RNAseq datasets:` 1.
GSE114950; Zhang et.al, 2019, Nucleic Acid Res. 10-days after `vertilization.` *NOT USED* 2.
GSE34318; Zhang et.al, 2012, (Jiang jiming lab).
14-days old.

## 6.0 Annotate
```{r}
source("D:/SUST/code/TanYongjun_code.R")
RPKM_tab1 <- fread("./refgenome/RNA/GSE34318_fpkm_exp.txt.gz")

RPKM_tab2 <- fread("./refgenome/RNA/GSE114950_col_4col_RNA-expression.txt.gz") %>%
  dplyr::select(chrom:col_fpkm) %>%
  dplyr::rename(id = gene)

df <- full_join(RPKM_tab2, RPKM_tab1)

df %>%
  dplyr::select(6:ncol(df)) %>%
  cor(use = "pairwise.complete.obs") %>%
  pheatmap::pheatmap(display_numbers = T)

df$RPKM <- (df$cufflinks_ath_leaf_rep1 + df$cufflinks_ath_leaf_rep3) / 2

# fwrite(df, file = "./refgenome/GSE34318_leaf_mean_FPKM_20210321.tsv.gz", sep = "\t")
```

### .1 annotatePeak(), NOT USED.
*Annotation (CRE-centered)* This method was ChIPseeker-based, only the nearest gene of each CRE were reported in the result table!

```{r, annotation}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
GeneExp <- fread("./refgenome/RNA/GSE34318_leaf_mean_FPKM_20210321.tsv.gz") %>%
  dplyr::select(id, RPKM) %>%
  dplyr::rename(geneId = id)
enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_20210321.csv.gz") %>%
  mutate(seqnames = str_replace_all(seqnames, "chr", ""))
library(TxDb.Athaliana.BioMart.plantsmart28)

# annotation to gene
df <- enhancer %>%
  makeGRangesFromDataFrame(keep.extra.columns = T) %>%
  GenomicRanges::resize(width = 1, fix = "center") %>%
  ChIPseeker::annotatePeak(., tssRegion = c(-3000, 500), 
                           TxDb = TxDb.Athaliana.BioMart.plantsmart28, 
                           level = "gene",
                           overlap = "all") %>%
  as.data.frame() %>%
  dplyr::mutate(start = startBK, end = endBK) 

# add RPKM
df <- left_join(df, GeneExp)
sum(is.na(df$RPKM))
df$RPKM[is.na(df$RPKM)] <- 0

# fwrite(df, file = "Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_RPKM_20210321.csv.gz")
```

### .2 FindOverlaps()
*Annotation (Gene-centered)* This method was fondOverlap-based, all CRE located within the determined region were reported in the result table.
Classify genes into groups based on it with or without flanking one kind of enhancers.
Steps: 
1. prepare GR object of gene TSS, gene TTS, and the left gene body regions based on provided distance.
(priority: TSS \> TTS \> GeneBody) 
2. mark CRE in each genes based on findoverlap.
3. collect results and plot.

*each kind of Enhancer*
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_20210321.csv.gz") %>%
    mutate(seqnames = str_replace_all(seqnames, "chr", "")) %>%
    dplyr::filter(str_detect(method, "(cSTARR|Enhancer)")) %>%
    dplyr::mutate(name = str_c(method, seqnames, start, "K4=", clusterK4, "K7=", clusterK7, sep = "_")) %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment <= 1.3 ))

## parameters
  TSS_region <- c(-500, 100) # c(upstream, downstream)
  TTS_region <- c(-300, 300)
  
  
## prepare regions.
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  genes_hg <- genes(txdb) # TSS to TTS
  
  # TSS
  TSS_gene <- resize(genes_hg, fix = "start", width = 1, use.names = T)
  TSS_gene <- TSS_gene %>%
    shift(shift = ifelse(strand(TSS_gene) == "+", sum(TSS_region) / 2, (0 - sum(TSS_region) / 2))) %>%
    GenomicRanges::resize(width = sum(abs(TSS_region)), fix = "center")
  
  # TTS
  TTS_gene <- resize(genes_hg, width = 1, fix = "end", use.names = T)
  TTS_gene <- TTS_gene %>%
    shift(shift = ifelse(strand(TTS_gene) == "+", sum(TTS_region) / 2, (0 - sum(TTS_region) / 2))) %>%
    GenomicRanges::resize(width = sum(abs(TTS_region)), fix = "center")
  
## mark CRE (peak) located in different gene region
  gene_tab <- as.data.frame(genes_hg)
  cre_tab <- enhancer
  
  # gene body
  ol <- findOverlaps(genes_hg, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$name[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_GeneBody = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # TSS region
  ol <- findOverlaps(TSS_gene, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$name[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_TSS = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # TTS region
  ol <- findOverlaps(TTS_gene, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$name[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_TTS = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # Discard duplicated marked CREs in the same gene. (priority: TSS > TTS > GeneBody)
  gene_tab$tmp <- paste(gene_tab$cre_GeneBody, gene_tab$cre_TSS, gene_tab$cre_TTS, sep = "#") %>%
    str_split(., "#", n = 20) %>%
    map(., function(x){x[x!="NA"]})
  gene_tab$tmp_Num <- map_int(gene_tab$tmp, length)
  gene_tab$tmp_NumUniq <- map_int(gene_tab$tmp, function(x){length(unique(x))})
  row_id <- which(gene_tab$tmp_Num > gene_tab$tmp_NumUniq)
  for (i in row_id) {
    tmp_GB <- gene_tab$cre_GeneBody[i] %>% str_split(., "#") %>% unlist()
    tmp_TSS <- gene_tab$cre_TSS[i] %>% str_split(., "#") %>% unlist()
    tmp_TTS <- gene_tab$cre_TTS[i] %>% str_split(., "#") %>% unlist()
    tmp_TTS <- base::setdiff(tmp_TTS, tmp_TSS)
    tmp_GB <- base::setdiff(tmp_GB, tmp_TSS) %>%
      base::setdiff(., tmp_TTS)
    
    gene_tab$cre_GeneBody[i] <- str_c(unlist(tmp_GB), collapse = "#")
    gene_tab$cre_TSS[i] <- str_c(unlist(tmp_TSS), collapse = "#")
    gene_tab$cre_TTS[i] <- str_c(unlist(tmp_TTS), collapse = "#")
  }
  gene_tab$tmp <- gene_tab$tmp_Num <- gene_tab$tmp_NumUniq <- NULL
  
## merge with gene expression
  GeneExp <- fread("./refgenome/RNA/GSE34318_leaf_mean_FPKM_20210321.tsv.gz") %>%
    dplyr::select(id, RPKM) %>%
    dplyr::rename(gene_id = id)
  
  gene_tab <- left_join(gene_tab, GeneExp)
  

## Count and info of CRE (total Number of cre, cluster ID based on K=4 or K=7)
  for (i in c("cre_GeneBody", "cre_TSS", "cre_TTS")) {
    cat("\n>>>>>> ", i, "\n")
    for (j in unique(cre_tab$method)) {
      cat(" ", j)
      tmp <- gene_tab[[i]] %>%
        # sort(decreasing = T) %>%
        # head(n = 20) %>%
        str_split(., pattern = "#", n = 20) %>%
        map(., function(x){
          return(x[str_detect(x, j)])
        })
      df <- data.frame(Number = map_int(tmp, function(x){sum(str_detect(x, j))}), 
                              clusterK4 = map_chr(tmp, function(x){str_c(str_replace_all(x, ".+K4=_(\\d).*", "\\1"), collapse = ";")}), 
                              clusterK7 = map_chr(tmp, function(x){str_c(str_replace_all(x, ".+K7=_(\\d).*", "\\1"), collapse = ";")}))
      colnames(df) <- str_c(str_replace(i, "cre_", ""), 
                            j, 
                            colnames(df), sep = "_")
      
      gene_tab <- cbind(gene_tab, df)
    }
  }

## Gene types: House-keeping or Tissue-specific
  HKgenes <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv") %>%
    dplyr::mutate(gene_id = str_replace_all(`Transcript ID`,"(AT.+)\\.\\d+", "\\1"))
  gene_tab$Gene_type <- "Tissue-specific"
  gene_tab$Gene_type[gene_tab$gene_id %in% HKgenes$gene_id] <- "House-keeping"

  # fwrite(gene_tab, file = "./R6.expression/Gene_list_with_CREs_500TSS100_300TTS300_20210322.csv.gz")
```

*Enhancer_STARRseq, no TE*
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_IntronicEn_TE_20220424.csv.gz") %>%
    mutate(seqnames = str_replace_all(seqnames, "chr", "")) %>%
    dplyr::filter(str_detect(method, "(cSTARR)")) %>%
    dplyr::mutate(name = str_c(method, seqnames, start, "K4=", clusterK4, "K7=", clusterK7, sep = "_")) %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment <= 1.3 ))
  table(enhancer$TE_SuperFamily)
  
  # no TE-derived enhancers
  enhancer <- enhancer %>%
    dplyr::filter(str_length(TE) < 2)
  
## parameters
  TSS_region <- c(-500, 100) # c(upstream, downstream)
  TTS_region <- c(-300, 300)
  
  
## prepare regions.
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  genes_hg <- genes(txdb) # TSS to TTS
  
  # TSS
  TSS_gene <- resize(genes_hg, fix = "start", width = 1, use.names = T)
  TSS_gene <- TSS_gene %>%
    shift(shift = ifelse(strand(TSS_gene) == "+", sum(TSS_region) / 2, (0 - sum(TSS_region) / 2))) %>%
    GenomicRanges::resize(width = sum(abs(TSS_region)), fix = "center")
  
  # TTS
  TTS_gene <- resize(genes_hg, width = 1, fix = "end", use.names = T)
  TTS_gene <- TTS_gene %>%
    shift(shift = ifelse(strand(TTS_gene) == "+", sum(TTS_region) / 2, (0 - sum(TTS_region) / 2))) %>%
    GenomicRanges::resize(width = sum(abs(TTS_region)), fix = "center")
  
## mark CRE (peak) located in different gene region
  gene_tab <- as.data.frame(genes_hg)
  cre_tab <- enhancer
  
  # gene body
  ol <- findOverlaps(genes_hg, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$name[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_GeneBody = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # TSS region
  ol <- findOverlaps(TSS_gene, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$name[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_TSS = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # TTS region
  ol <- findOverlaps(TTS_gene, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$name[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_TTS = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # Discard duplicated marked CREs in the same gene. (priority: TSS > TTS > GeneBody)
  gene_tab$tmp <- paste(gene_tab$cre_GeneBody, gene_tab$cre_TSS, gene_tab$cre_TTS, sep = "#") %>%
    str_split(., "#", n = 20) %>%
    map(., function(x){x[x!="NA"]})
  gene_tab$tmp_Num <- map_int(gene_tab$tmp, length)
  gene_tab$tmp_NumUniq <- map_int(gene_tab$tmp, function(x){length(unique(x))})
  row_id <- which(gene_tab$tmp_Num > gene_tab$tmp_NumUniq)
  for (i in row_id) {
    tmp_GB <- gene_tab$cre_GeneBody[i] %>% str_split(., "#") %>% unlist()
    tmp_TSS <- gene_tab$cre_TSS[i] %>% str_split(., "#") %>% unlist()
    tmp_TTS <- gene_tab$cre_TTS[i] %>% str_split(., "#") %>% unlist()
    tmp_TTS <- base::setdiff(tmp_TTS, tmp_TSS)
    tmp_GB <- base::setdiff(tmp_GB, tmp_TSS) %>%
      base::setdiff(., tmp_TTS)
    
    gene_tab$cre_GeneBody[i] <- str_c(unlist(tmp_GB), collapse = "#")
    gene_tab$cre_TSS[i] <- str_c(unlist(tmp_TSS), collapse = "#")
    gene_tab$cre_TTS[i] <- str_c(unlist(tmp_TTS), collapse = "#")
  }
  gene_tab$tmp <- gene_tab$tmp_Num <- gene_tab$tmp_NumUniq <- NULL
  
## merge with gene expression
  GeneExp <- fread("./refgenome/RNA/GSE34318_leaf_mean_FPKM_20210321.tsv.gz") %>%
    dplyr::select(id, RPKM) %>%
    dplyr::rename(gene_id = id)
  
  gene_tab <- left_join(gene_tab, GeneExp)
  

## Count and info of CRE (total Number of cre, cluster ID based on K=4 or K=7)
  for (i in c("cre_GeneBody", "cre_TSS", "cre_TTS")) {
    cat("\n>>>>>> ", i, "\n")
    for (j in unique(cre_tab$method)) {
      cat(" ", j)
      tmp <- gene_tab[[i]] %>%
        # sort(decreasing = T) %>%
        # head(n = 20) %>%
        str_split(., pattern = "#", n = 20) %>%
        map(., function(x){
          return(x[str_detect(x, j)])
        })
      df <- data.frame(Number = map_int(tmp, function(x){sum(str_detect(x, j))}), 
                              clusterK4 = map_chr(tmp, function(x){str_c(str_replace_all(x, ".+K4=_(\\d).*", "\\1"), collapse = ";")}), 
                              clusterK7 = map_chr(tmp, function(x){str_c(str_replace_all(x, ".+K7=_(\\d).*", "\\1"), collapse = ";")}))
      colnames(df) <- str_c(str_replace(i, "cre_", ""), 
                            j, 
                            colnames(df), sep = "_")
      
      gene_tab <- cbind(gene_tab, df)
    }
  }

## Gene types: House-keeping or Tissue-specific
  HKgenes <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv") %>%
    dplyr::mutate(gene_id = str_replace_all(`Transcript ID`,"(AT.+)\\.\\d+", "\\1"))
  gene_tab$Gene_type <- "Tissue-specific"
  gene_tab$Gene_type[gene_tab$gene_id %in% HKgenes$gene_id] <- "House-keeping"

  # fwrite(gene_tab, file = "./R6.expression/Gene_list_with_Enhancer_STARRseq_noTE_500TSS100_300TTS300_20220516.csv.gz")
```

## 6.1 Enhancers as a whole

### .1 EXP X With/without

#### . Count
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  noTE_enhancer <- F
  if (noTE_enhancer) {
    gene_tab <- fread("./R6.expression/Gene_list_with_Enhancer_STARRseq_noTE_500TSS100_300TTS300_20220516.csv.gz")
  }else{
    gene_tab <- fread("./R6.expression/Gene_list_with_CREs_500TSS100_300TTS300_20210322.csv.gz")
  }
  gene_tab <- gene_tab[,str_detect(colnames(gene_tab), "(cre_|cluster)", negate = T)]

# fill NA as 0
  for (i in str_detect(colnames(gene_tab), "Number")) {
    gene_tab[,i][is.na(gene_tab[,i])] <- 0
  }

# Count enhancers located near genes
  gene_tab$Num_cSTARRseq <- gene_tab$`TSS_cSTARR-seq_Number` + gene_tab$`TTS_cSTARR-seq_Number` + gene_tab$`GeneBody_cSTARR-seq_Number`
  gene_tab$cSTARRseq <- ifelse(gene_tab$Num_cSTARRseq > 0, "With enhancer", "Without enhancer")
    
  if (!noTE_enhancer) {
    gene_tab$Num_Enhancer_Zhu2015 <- gene_tab$TSS_Enhancer_Zhu2015_Number + gene_tab$TTS_Enhancer_Zhu2015_Number + gene_tab$GeneBody_Enhancer_Zhu2015_Number
    gene_tab$Num_Enhancer_Wang2019 <- gene_tab$TSS_Enhancer_Wang2019_Number + gene_tab$TTS_Enhancer_Wang2019_Number + gene_tab$GeneBody_Enhancer_Wang2019_Number
    gene_tab$Enhancer_Zhu2015 <- ifelse(gene_tab$Num_Enhancer_Zhu2015 > 0, "With enhancer", "Without enhancer")
    gene_tab$Enhancer_Wang2019 <- ifelse(gene_tab$Num_Enhancer_Wang2019 > 0, "With enhancer", "Without enhancer")
  }
```

#### .  Exp x En_Num
```{r}
# number of genes with varied flanking enhancers
  p <- gene_tab %>%
    dplyr::select(gene_id, Num_Enhancer_Wang2019, Num_cSTARRseq, Num_Enhancer_Zhu2015) %>%
    pivot_longer(cols = Num_Enhancer_Wang2019:Num_Enhancer_Zhu2015, names_to = "Enhancer", values_to = "Number") %>%
    dplyr::filter(Number > 0) %>%
    group_by(Enhancer, Number) %>%
    dplyr::summarise(Count = n()) %>%
    pivot_wider(id_cols = Enhancer, names_from = Number, values_from = Count, values_fill = 0) %>%
    pivot_longer(cols = 2:9, names_to = "Number", values_to = "Count") %>%
    dplyr::mutate(Number = factor(Number),
                  Count = ifelse(Count == 0, NA, Count)) %>%
    ggplot(aes(x = Number, color = Enhancer, fill = Enhancer, y = Count)) + 
      geom_col(position = position_dodge(), width = 0.7) + 
      scale_color_npg() + 
      scale_fill_npg() + 
      toolkit_tyj$theme_my +
      theme(legend.position = c(0.7, 0.9), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.text = element_text(size = 7), 
            legend.title = element_blank()) +
      # scale_y_log10() + 
      geom_text(aes(x = Number, y = (Count + 1)/2, label = Count), 
                color = "grey30",
                size = 1.5,
                position = position_dodge(width = 0.8)) + 
      ggtitle("Genes with flanking enhancers") + 
      labs(caption = "cSTARR-seq enhancer, enrichment > 1.3")
  # toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Number_genes_with_enhancers_STARRseq1.3_20210322", 
  #                      width = 10, height = 8, plot = p)

# relationship between RPKM and enhancer number
  df <- gene_tab %>%
    as_tibble() %>%
    dplyr::mutate(type = ifelse(Num_cSTARRseq == 1, "1", "None"),
                  type = ifelse(Num_cSTARRseq == 2, "2", type),
                  type = ifelse(Num_cSTARRseq >2 , ">=3", type),
                  type = factor(type, levels = c("None", "1", "2", ">=3"))) 
  
  # all gene
  df_label <- df %>%
    toolkit_tyj$MultiComparison2(., Name_variable1 = NA, Name_variable2 = "type", 
                                 Name_value = "RPKM", test.name = "kruskal", OnlyDF = T) %>%
    dplyr::mutate(label_n = str_c(sigLabel, "\n", Num_var2_level, sep = ""),
                  type = factor(type, levels = c("None", "1", "2", ">=3"))) %>%
    arrange(type)
  df %>%
    ggplot(aes(x = type, y = log10(RPKM + 1)))  + 
      toolkit_tyj$theme_my + 
      geom_boxplot(width = 0.1, outlier.shape = NA, fill = "transparent") + 
      geom_violin(fill = "transparent", scale = "width") +
      geom_text(data = df_label, aes(x = type, y = 4, label = sigLabel), 
                size = 6/.pt, ) + 
      xlab("Number of enhancers") + 
      scale_x_discrete(labels = str_c(df_label$type, "\n(", df_label$Num_var2_level, ")"))
  
  FileName <- ifelse(noTE_enhancer, "Expression_genes_with_varied_flanking_noTEenhancers",
                     "Expression_genes_with_varied_flanking_enhancers")
  toolkit_tyj$SavePlot(filename_prefix = str_c("./R6.expression/", FileName),
                       width = 4, height = 6, device = c("png", "pdf"))
  
  # Tissue-specific and House-keeping
  df_label <- df %>%
    toolkit_tyj$MultiComparison2(., Name_variable1 = "Gene_type", Name_variable2 = "type", 
                                 Name_value = "RPKM", test.name = "kruskal", OnlyDF = T) %>%
    dplyr::mutate(label_n = str_c(sigLabel, "\n", Num_var2_level, sep = ""))
  df %>%
    ggplot(aes(x = type, y = log10(RPKM + 1)))  + 
      toolkit_tyj$theme_my + 
      geom_boxplot(width = 0.1, outlier.shape = NA, fill = "transparent") + 
      geom_violin(fill = "transparent", scale = "width") +
      geom_text(data = df_label, aes(x = type, y = 4, label = label_n), 
                size = 6/.pt, ) + 
      facet_grid(. ~ Gene_type) +
      xlab("Number of enhancers")
  FileName <- ifelse(noTE_enhancer, "Expression_TSHKgenes_with_varied_flanking_noTEenhancers",
                     "Expression_TSHKgenes_with_varied_flanking_enhancers")
  toolkit_tyj$SavePlot(filename_prefix = str_c("./R6.expression/", FileName),
                       width = 8, height = 6, device = "png")

```

#### . Exp X with/without En
```{r}
# RPKM ~ enhancer
  gene_tab %>%
    dplyr::select(gene_id, RPKM, cSTARRseq:Enhancer_Wang2019) %>%
    dplyr::mutate(Random = "Random") %>%
    pivot_longer(cols = cSTARRseq:Random, names_to = "Enhancer", values_to = "States") %>%
    mutate(RPKM = log10(RPKM + 1)) %>%
    ggplot(aes(x = Enhancer, y = RPKM, color = States)) + 
      geom_hline(yintercept = median(log10(gene_tab$RPKM + 1)), linetype = 2, color = "grey40") +
      geom_violin(fill = "transparent", position = position_dodge(width = 1)) + 
      geom_boxplot(fill = "transparent", outlier.alpha = 0, width = 0.05, 
                   position = position_dodge(width = 1)) +
      toolkit_tyj$theme_my + 
      # scale_color_npg() + 
      scale_color_manual(values = c("grey50", "#E64B35FF", "#4DBBD5FF")) +
      theme(legend.title = element_blank(), 
            legend.position = "right", 
            legend.key.size = unit(0.3, units = "cm"),
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Expression_of_gene_with_without_enhancer_STARRseq1.3_20210322",
  #                      width = 14, height = 10)
  
  df <- gene_tab %>%
    dplyr::select(gene_id, RPKM, cSTARRseq) %>%
    dplyr::mutate(Random = "Random") %>%
    pivot_longer(cols = cSTARRseq:Random, names_to = "Enhancer", values_to = "States") %>%
    mutate(States = factor(States, levels = c("With enhancer", "Without enhancer", "Random"))) %>%
    dplyr::filter(Enhancer == "cSTARRseq" |Enhancer == "Random") 
  df_label <- df %>%
    toolkit_tyj$MultiComparison2(df = ., Name_variable1 = NA, Name_variable2 = "States",
                                 Name_value = "RPKM", test.name = "kruskal", OnlyDF = T) %>%
    dplyr::mutate(States = factor(States, c("With enhancer", "Without enhancer", "Random")),
                  xlabel = str_c(States, "\n(", Num_var2_level, ")")) %>%
    arrange(States)
  
  df %>%
    ggplot(aes(x = States, y = log10(RPKM + 1), color = States)) + 
      geom_hline(yintercept = median(log10(gene_tab$RPKM + 1)), linetype = 2, color = "grey40") +
      geom_violin(fill = "transparent", position = position_dodge(width = 1), scale = "width") + 
      geom_boxplot(fill = "transparent", outlier.alpha = 0, width = 0.05, 
                   position = position_dodge(width = 1)) +
      toolkit_tyj$theme_my + 
      # scale_color_npg() + 
      # geom_signif(comparisons = list(c("With enhancer", "Random"), 
      #                                c("With enhancer", "Without enhancer"),
      #                                c("Random", "Without enhancer")), 
      #             y_position = c(4.5, 4.1, 3.7), textsize = 6/.pt, 
      #             color = "grey30", test = "wilcox.test", size = 0.2) +
      geom_text(aes(x = States, y = 4, label = sigLabel), size = 6/.pt, data = df_label,
                color = "grey15") +
      scale_color_manual(values = toolkit_tyj$AtEn_YNRandom) +
      theme(legend.title = element_blank(), 
            legend.position = "none", 
            legend.key.size = unit(0.3, units = "cm"),
            # axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            axis.title.x = element_blank()) +
      ylab(expression(log[10](RPKM+1))) + 
      # coord_cartesian(ylim = c(0, 4.7)) + 
      scale_x_discrete(labels = str_replace_all(df_label$xlabel, " ", "\n"))
  
  toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Expression_of_gene_with_without_STARRseq1.3_noTE_20210322",
                       width = 4, height = 6, units = "cm")
```

### . 2 Exp X En_act

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_IntronicEn_TE_20220424.csv.gz") %>%
    mutate(seqnames = str_replace_all(seqnames, "chr", "")) %>%
    dplyr::filter(str_detect(method, "(cSTARR)")) %>%
    dplyr::mutate(name = str_c(method, seqnames, start, "K4=", clusterK4, "K7=", clusterK7, sep = "_")) %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment <= 1.3 ))
  table(enhancer$TE_SuperFamily)
  
  # no TE-derived enhancers
  enhancer <- enhancer %>%
    dplyr::filter(str_length(TE) < 2,
                  clusterK4  == 1 | clusterK4 == 4)
  
## parameters
  TSS_region <- c(-500, 100) # c(upstream, downstream)
  TTS_region <- c(-300, 300)
  
  
## prepare regions.
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  genes_hg <- genes(txdb) # TSS to TTS
  
  # TSS
  TSS_gene <- resize(genes_hg, fix = "start", width = 1, use.names = T)
  TSS_gene <- TSS_gene %>%
    shift(shift = ifelse(strand(TSS_gene) == "+", sum(TSS_region) / 2, (0 - sum(TSS_region) / 2))) %>%
    GenomicRanges::resize(width = sum(abs(TSS_region)), fix = "center")
  
  # TTS
  TTS_gene <- resize(genes_hg, width = 1, fix = "end", use.names = T)
  TTS_gene <- TTS_gene %>%
    shift(shift = ifelse(strand(TTS_gene) == "+", sum(TTS_region) / 2, (0 - sum(TTS_region) / 2))) %>%
    GenomicRanges::resize(width = sum(abs(TTS_region)), fix = "center")
  
## mark CRE (peak) located in different gene region
  gene_tab <- as.data.frame(genes_hg)
  cre_tab <- enhancer
  
  # gene body
  ol <- findOverlaps(genes_hg, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$enrichment[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_GeneBody = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # TSS region
  ol <- findOverlaps(TSS_gene, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$enrichment[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_TSS = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
  # TTS region
  ol <- findOverlaps(TTS_gene, 
                     GenomicRanges::resize(makeGRangesFromDataFrame(cre_tab), 
                                           width = 1, fix = "center")) %>%
    as.data.frame()
  ol$gene_id <- gene_tab$gene_id[ol$queryHits]
  ol$cre_name <- cre_tab$enrichment[ol$subjectHits]
  ol <- ol %>%
    dplyr::select(gene_id, cre_name) %>%
    group_by(gene_id) %>%
    summarise(cre_TTS = str_c(cre_name, collapse = "#"))
  gene_tab <- full_join(gene_tab, ol)
  
 ## merge with gene expression
  GeneExp <- fread("./refgenome/RNA/GSE34318_leaf_mean_FPKM_20210321.tsv.gz") %>%
    dplyr::select(id, RPKM) %>%
    dplyr::rename(gene_id = id)
  
  gene_tab <- left_join(gene_tab, GeneExp)
  

## Gene types: House-keeping or Tissue-specific
  HKgenes <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv") %>%
    dplyr::mutate(gene_id = str_replace_all(`Transcript ID`,"(AT.+)\\.\\d+", "\\1"))
  gene_tab$Gene_type <- "Tissue-specific"
  gene_tab$Gene_type[gene_tab$gene_id %in% HKgenes$gene_id] <- "House-keeping"

## Calculate Activity of Enhancer.
  gene_tab$Max_Act <- apply(gene_tab %>% dplyr::select(contains("cre")), 1, function(x){
    x %>%
      na.omit() %>%
      str_c(., collapse = "#")
  }) %>%
    map(., function(y){
      strsplit(x = y, split = "#", fixed = T) %>%
        unlist() %>%
        as.numeric() %>%
        max(na.rm = T) %>%
        return()
    }) %>%
    unlist()
  
  # scatterplot
  library(ggpmisc)
  # my.formula <- y ~ x -1
  df <-  gene_tab %>%
    dplyr::filter(Max_Act > 0) %>%
    as_tibble 
  library(ggrastr)
  df %>%
    ggplot(aes(x = Max_Act, y = log10(RPKM + 1))) +
        # geom_point(shape = 1, alpha = 1/3) +
        geom_point_rast(shape = 1, alpha = 1/3, dpi = 600) +
        # geom_smooth(method = "lm", se= T, color="steelblue", formula = my.formula, size = 0.5) +
        # stat_poly_eq(formula = my.formula,
        #              aes(label = paste(..eq.label.., ..rr.label.., ..p.value.label.., sep = "~~~")),
        #              parse = TRUE, size = 2) +
        geom_text(x = 2, y = 3.2, label = str_c("p=", 
                                                round(cor.test(df$RPKM, df$Max_Act, use = "pairwise.complete.obs")$p.value, 4),
                                                "  r=",
                                                round(cor(df$RPKM, df$Max_Act, use = "pairwise.complete.obs"), 4)),
                  size = 6/.pt) +
        toolkit_tyj$theme_my + 
      xlab("Enhancet activity") + 
      ylab(expression(log[10]^(RPKM+1)))
  toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Scatterplot_Exp_vs_EnAct",
                       width = 6, height = 6)
  
  
  
```



## 6.2 cluster enhancer X RPKM

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  gene_tab <- fread("./R6.expression/Gene_list_with_CREs_500TSS100_300TTS300_20210322.csv.gz")
  gene_tab <- gene_tab[,str_detect(colnames(gene_tab), "(Enhancer)", negate = T)]

# fill NA as 0
  for (i in which(str_detect(colnames(gene_tab), "Number"))) {
    gene_tab[,i][is.na(gene_tab[,i])] <- 0
  }

# Count enhancers located near genes
  splitAndSort <- function(x){
    str_split(x, ";") %>%
      map_chr(., function(y){
        na.omit(y) %>% 
          sort() %>% 
          unique() %>% 
          str_c(collapse = ";")
        }) %>%
      str_replace_all(c("^;" = "",
                        ";NA" = ""))%>%
      return() 
  }
    
  gene_tab$clusterK4 <- paste(gene_tab$`TSS_cSTARR-seq_clusterK4`, 
                              gene_tab$`TTS_cSTARR-seq_clusterK4`, 
                              gene_tab$`GeneBody_cSTARR-seq_clusterK4`, 
                              sep = ";") %>%
    splitAndSort()
  gene_tab$clusterK7 <- paste(gene_tab$`TSS_cSTARR-seq_clusterK7`,
                              gene_tab$`TTS_cSTARR-seq_clusterK7`,
                              gene_tab$`GeneBody_cSTARR-seq_clusterK7`,
                              sep = ";") %>%
    splitAndSort()
  table(gene_tab$clusterK4)
  table(gene_tab$clusterK7)
  
## plot Cluster K=4
  df <- gene_tab %>%
    dplyr::filter(str_detect(clusterK4, ";", negate = T), 
                  clusterK4 != 0) %>%
    dplyr::mutate(clusterK4 = ifelse(clusterK4 == "NA", "Without\nenhancer", paste("C", clusterK4, sep = ""))) %>%
    full_join(., mutate(., clusterK4 = "All genes")) %>%
    dplyr::mutate(clusterK4 = factor(clusterK4, levels = c(paste("C", 1:4, sep = ""), "Without\nenhancer", "All genes")))
  df_label <- df%>%
    group_by(clusterK4) %>%
    dplyr::summarise(Num = n(), Median = round(median(log10(RPKM + 1), na.rm = T), digits = 2)) %>%
    dplyr::mutate(label = paste("N=", Num, "\nMedian=", Median, sep = ""),
                  xlabel = str_c(clusterK4, "\n(", Num, ")"))
  df_label <- df %>%
    toolkit_tyj$MultiComparison2(., Name_variable1 = NA, Name_variable2 = "clusterK4",
                                 Name_value = "RPKM",test.name = "kruskal")
  df_label <- df_label$significant.label %>%
    dplyr::mutate(xlabel = str_c(clusterK4, "\n(", Num_var2_level, ")"),
                  clusterK4 = factor(clusterK4, levels = c(paste("C", 1:4, sep = ""), "Without\nenhancer", "All genes"))) %>%
    arrange(clusterK4)
  
  df %>%
    ggplot(aes(x = clusterK4, y = log10(RPKM + 1), color = clusterK4)) +
      geom_hline(yintercept = 0.63, linetype = 2, color = "grey40") +
      geom_violin(fill = "transparent", scale = "width") +
      geom_boxplot(outlier.alpha = 0, width = 0.1, fill = "transparent") + 
      # scale_color_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.position = "none") +
      xlab("") +
      ylab(expression(log[10](RPKM+1))) +
      # geom_label(data = df_label, 
      #           aes(x = clusterK4, y = 3.4, label = label), 
      #           fill = "white", size = 6/.pt, alpha = 1/2) + 
      # coord_flip() + 
      geom_text(data = df_label, aes(x = clusterK4, y = 4, label = sigLabel), 
                size = 6/.pt, color = "grey15") +
      scale_color_manual(values = c(toolkit_tyj$AtEn_KmeanC_color, toolkit_tyj$AtEn_YNRandom[c(2,3)])) + 
      scale_x_discrete(label = df_label$xlabel)

  toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Expression_of_gene_with_STARRseq_enhancerClusterK4_20210322",
                       width = 6, height = 4.5)
  
## plot Cluster K=4 (exclude "ALL genes")
  df <- gene_tab %>%
    dplyr::filter(str_detect(clusterK4, ";", negate = T), 
                  clusterK4 != 0) %>%
    dplyr::mutate(clusterK4 = ifelse(clusterK4 == "NA", "Without\nenhancer", paste("C", clusterK4, sep = ""))) %>%
    # full_join(., mutate(., clusterK4 = "All genes")) %>%
    dplyr::mutate(clusterK4 = factor(clusterK4, levels = c(paste("C", 1:4, sep = ""), "Without\nenhancer", "All genes")))
  df_label <- df%>%
    group_by(clusterK4) %>%
    dplyr::summarise(Num = n(), Median = round(median(log10(RPKM + 1), na.rm = T), digits = 2)) %>%
    dplyr::mutate(label = paste("N=", Num, "\nMedian=", Median, sep = ""),
                  xlabel = str_c(clusterK4, "\n(", Num, ")"))
  df_label <- df %>%
    toolkit_tyj$MultiComparison2(., Name_variable1 = NA, Name_variable2 = "clusterK4",
                                 Name_value = "RPKM",test.name = "kruskal")
  df_label <- df_label$significant.label %>%
    dplyr::mutate(xlabel = str_c(clusterK4, "\n(", Num_var2_level, ")"),
                  clusterK4 = factor(clusterK4, levels = c(paste("C", 1:4, sep = ""), "Without\nenhancer", "All genes"))) %>%
    arrange(clusterK4)
  
  df %>%
    ggplot(aes(x = clusterK4, y = log10(RPKM + 1), color = clusterK4)) +
      geom_hline(yintercept = 0.63, linetype = 2, color = "grey40") +
      geom_violin(fill = "transparent", scale = "width") +
      geom_boxplot(outlier.alpha = 0, width = 0.1, fill = "transparent") + 
      # scale_color_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.position = "none") +
      xlab("") +
      ylab(expression(log[10](RPKM+1))) +
      # geom_label(data = df_label, 
      #           aes(x = clusterK4, y = 3.4, label = label), 
      #           fill = "white", size = 6/.pt, alpha = 1/2) + 
      # coord_flip() + 
      geom_text(data = df_label, aes(x = clusterK4, y = 4, label = sigLabel), 
                size = 6/.pt, color = "grey15") +
      scale_color_manual(values = c(toolkit_tyj$AtEn_KmeanC_color, toolkit_tyj$AtEn_YNRandom[c(2,3)])) + 
      scale_x_discrete(label = df_label$xlabel)

  toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Expression_of_gene_with_STARRseq_enhancerClusterK4_noAllGene_20210322",
                       width = 5, height = 4.5)
  
## plot Cluster K=7
  df <- gene_tab %>%
    dplyr::filter(str_detect(clusterK7, ";", negate = T), 
                  clusterK7 != 0) %>%
    dplyr::mutate(clusterK7 = ifelse(clusterK7 == "NA", "Without enhancer", paste("C", clusterK7, sep = ""))) %>%
    full_join(., mutate(., clusterK7 = "All genes")) %>%
    dplyr::mutate(clusterK7 = factor(clusterK7, levels = rev(c(paste("C", 1:7, sep = ""), "Without enhancer", "All genes"))))
  df_label <- df%>%
    group_by(clusterK7) %>%
    dplyr::summarise(Num = n(), Median = round(median(log10(RPKM + 1), na.rm = T), digits = 2)) %>%
    dplyr::mutate(label = paste("N=", Num, "\nMedian=", Median, sep = ""))
  
  df %>%
    ggplot(aes(x = clusterK7, y = log10(RPKM + 1), color = clusterK7)) +
      geom_hline(yintercept = 0.63, linetype = 2, color = "grey40") +
      geom_violin(fill = "transparent", scale = "width") +
      geom_boxplot(outlier.alpha = 0, width = 0.1, fill = "transparent") + 
      # scale_color_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.position = "none") +
      xlab("") +
      ylab(expression(log[10](RPKM+1))) +
      geom_label(data = df_label, 
                aes(x = clusterK7, y = 3.5, label = label), 
                fill = "white", size = 2.5, alpha = 1/2) + 
      coord_flip() + 
      scale_color_manual(values = rev(c(pal_npg("nrc")(7), "grey35", "grey40")))

  # toolkit_tyj$SavePlot(filename_prefix = "./R6.expression/Expression_of_gene_with_STARRseq_enhancerclusterK7_20210322", 
  #                      width = 12, height = 10)
```

# 7. Varied TFs enriched in enhancer clusters.

Two levels: All enhancer as a whole and each enhancer clusters, promoters were also used in analysis.
Two strategy:         
1. overlap between DAPseq/Chip-seq peaks and CREs. 
2. PWMs enriched in CREs and control.(This strategy also contain two kinds of methods: 1. region based; and 2. affinity based.)

## 7.0 related datasets:

Peaks and motifs(DRMEME output): DAPSeq, Motifs: MotifDb, Info and regNet: AGRIS.
### .1 DAPseq DAP-seq: O'Malley et.al, 2016, Cell.
`GSE60143` Based on input DNA, two kind of results generated: After discard Chip-seq and protein expressed in ecoli, a total of 811 bed files were used in following analysis.
Col, and Col_amplified denotes DNA modification were not discarded or discarded by PCR.

```{r, prepare table for DAP-seq datasets}
source("/work/bio-tanyj/soft/TanYongjun_code.R")
FileLs <- list.files(path = "./refgenome/GSE60143_RAW_DAPSeq/", pattern = "narrowPeak")
FileLs <- data.frame(file = FileLs,
                     DAPSeq = ifelse(str_detect(FileLs, "DAPSeq"), "Yes", "No"),
                     Methy = str_replace_all(FileLs, ".+(col_|colamp_).+", "\\1"), 
                     Family = str_replace_all(FileLs, ".+DAPSeq-(.+?)_tnt.+", "\\1"),
                     Protein = str_replace_all(FileLs, ".+DAPSeq-.+_tnt-(.+?)_.+", "\\1")
                     ) %>%
  dplyr::filter(str_detect(file, "ChIPSeq", negate = T), 
                str_detect(file, "(ecoli|xlsx)", negate = T)) %>%
  dplyr::mutate(Methy = str_replace_all(Methy, c("col_" = "Col", "colamp_" = "Col_amplified")), 
                Methy = ifelse(str_detect(Methy, "GSM"), "Col", Methy))

FileLs$NumRegions <- NA
FileLs$MedianWidthRegion <- NA
for (i in seq_len(nrow(FileLs))) {
  cat(i, " ")
  df <- fread(paste("./refgenome/GSE60143_RAW_DAPSeq/", 
                                      FileLs$file[i], sep = ""), header = F)
  FileLs$NumRegions[i] <- nrow(df)
  FileLs$MedianWidthRegion[i] <- median(df$V3 - df$V2, na.rm = T)
  
}

# fwrite(FileLs, file = "./refgenome/GSE60143_RAW_DAPSeq/meta_info_byTYJ_20210324.tsv",
#        sep = "\t")
```

### .2 MotifDb

Also provide *PWM* and meta infos, 803 PWMs of 499 proteins which belong to 40 tfFamilies were provided.

```{r}
library(MotifDb)
sort (table (values (MotifDb)$organism), decreasing=TRUE)
sort (table (values (MotifDb)$dataSource), decreasing=TRUE)
hs_motif <- query (MotifDb, 'thaliana')
hs_motif <- query (MotifDb, 'Osativa') ## Only five motifs for O.s !
# query (MotifDb, 'thaliana')
# query (MotifDb, 'Osativa')


MetaInfo <- as.tibble(hs_motif@elementMetadata)

table(MetaInfo$geneSymbol) %>% hist(main = "Number of motifs for each TF")
table(MetaInfo$dataSource)
unique(MetaInfo$geneSymbol) %>% length()

```

### .3 AGRIS

the `AGRIS: Arabidopsis gene regulatory information server`: <https://agris-knowledgebase.org/>: *PWMs* were not provided!
`AtTFDB`: meta info of TFs family.
`AtcisDB`: Promoter and TFBS of genes (upstream of CDS).
`AtRegNet`: relationship between TF and target genes.

### .4 PlantTFDB (maintained by Gao-lab)

provide information of TFs in 163 plant secies.
A total of 290 motifs for A.t were provided (in meme motif format).

### .x motifs used in recent study

*Study 1*
Data from `ReMap 2020` and the some other source (also used in Study 2) were used to identify TFBS (using FIMO) located in the determined enhancer. TFs related data in `ReMap` was also included in the `JASPAR 2020`.
"Enhancer-mediated reporter gene expression in Arabidopsis thaliana: a forward genetic screen", Yuan Lin, 2021, the plant journal.

*Study 2*
a total of 1721 motifs belong to 763 TFs were used (FIMO).  "Dynamic control of enhancer activity drives stage-specific gene expression during flower morphogenesis", Yan et al., 2019, Nature Communication.
```{r, out.width='50%'}
knitr::include_graphics("./refgenome/List_of_motifs_A.t_Yan_2019_NC.png")
```

### .x PPI DB

*BioGRID*: a database contain `protein-protein interactions` (PPI) information.
          Abundance informations were provided in tab format for model animals, A.t.
          (less for Oryza sativa.)        
*PlantRegMap*: a database contain `TFs-CRE/gene interaction` (Not appropriate for comparison between different TF groups). they also develop a database for A.t "ATRM". However, only 1431 interactions between TFs and target genes were listed. It is insufficient to perform analysis.

```{r}
# load
  source("D:/SUST/code/TanYongjun_code.R")
  # some PPI were detected between proteins from A.t and other species.
  PPIbioGRID <- fread("./refgenome/BIOGRID-ORGANISM-Arabidopsis_thaliana_Columbia-4.3.196.tab3.txt.gz") %>%
    dplyr::filter(`Organism Name Interactor A` == "Arabidopsis thaliana (Columbia)", 
                  `Organism Name Interactor B` == "Arabidopsis thaliana (Columbia)")

# summary
  table(PPIbioGRID$`Experimental System`, PPIbioGRID$`Experimental System Type`)
  table(PPIbioGRID$`Organism Name Interactor A`, PPIbioGRID$`Organism Name Interactor B`)
  
# Count number of interaction for each protein.
  # Interaction between two proteins may identified by different methods. Discard duplicated interactions.
  df <- NULL
  df <- apply(PPIbioGRID, 1, function(x){
    sort(x[c(6,7)])
  })
  df <- df %>%
    t %>%
    as.data.frame()
  nrow(df)
  nrow(unique(df))
  df <- unique(df)
  
  Num_PPI <- c(df$V1, df$V2) %>%
    table() %>%
    as.data.frame()
  colnames(Num_PPI) <- c("geneID", "Number_PPI")
  
  quantile(Num_PPI$Number_PPI, probs = seq(0, 1, 0.05))
  
# add gene info
  # GeneInfo <- data.frame(geneID = c(PPIbioGRID$`Systematic Name Interactor A`, PPIbioGRID$`Systematic Name Interactor B`), 
  #                        geneSymble = c(PPIbioGRID$`Official Symbol Interactor A`, PPIbioGRID$`Official Symbol Interactor B`)) %>%
  #   unique()
  metaInfo <- fread("./refgenome/GSE60143_RAW_DAPSeq/TableS1D.tsv") %>%
    dplyr::rename(geneID = AT_ID)
  Num_PPI <- Num_PPI %>%
    right_join(., metaInfo)
  
  # fwrite(Num_PPI, file = "./refgenome/Number_Protein_protein_interaction_BioGRID_DAPseq_tidy_2021617.csv.gz")

```

## 7.1 DAP Peaks vs CREs.

Promoter, each kind of enhancer, enhancer cluster, Random selected regions.
perform enrichment analysis with GAT.
Two steps: 
1. prepare bed.gz files for CREs and TF peaks.
2. run GAT.

List of CRE data sets:
1. each kind of CRE: Enhancer_STARRseq, three DHS-related enhancers, CorePro, Random, TSS2TTS, FirstExon, and FirstIntron.
2. Enhancer_STARRseq groups: Kmeans cluster, Act cluster, TS/HK/Distal enhancers(located +-500bp of two kind of genes or not).
3. sCorePro (not overlapped with En), sEnhancer, pEnhancer, eCorePro. identified based on 1bp overlap between CorePro (100bp, from -50 to 50bp around TSS) and Enhancer(600bp).
4. sEnhancer300CorePro, sEnhancer600CorePro.(similar to 3, but the CorePro were resize to 700bp and 1300bp, respectively.)
5. TS/HK CorePro. corepromoter of TS and HK genes.

### .1 prepare bed.gz files for GAT.

*CREs 202103*

```{r}
### CREs (all CREs in one bed files with different id in the 4th column. col1-col3: seqname, start, end)
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_20210321.csv.gz") %>%
    dplyr::filter(method != "PAS") %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
                  !(method == "iSTARR-seq" & enrichment < 1.3),
                  str_detect(seqnames, "(chrMt|chrPt)", negate = T))
  table(enhancer$method)
  table(enhancer$seqnames)
  addmargins(table(enhancer$clusterK4, enhancer$method))
  
  CREs_all <- NULL
  # each CRE
  for (i in unique(enhancer$method)) {
    CREs_all <- enhancer %>%
      dplyr::filter(method == i) %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(id = paste(i, "_all", sep = "")) %>%
      rbind(CREs_all, .)
  }
  
  # each cluster of CRE (Enhancer_Wang2019-cluster1, Enhancer_Zhu2015-cluster1 were discarded for number of regions less than 10)
  for (i in unique(enhancer$method)) {
    df <- enhancer %>% dplyr::filter(method == i)
    for (j in unique(df$clusterK4)) {
        CREs_all <- df %>%
          dplyr::filter(clusterK4 == j) %>%
          dplyr::select(1:3) %>%
          dplyr::mutate(id = paste(i, "_", j, sep = "")) %>%
          rbind(CREs_all, .)
    }
  }
  
  # varied number of randomly slected regions
  for (i in c(50, 100, 200, 300, 400, 1000, 2000, 5000, 10000)) {
      CREs_all <- enhancer %>%
        dplyr::filter(method == "Random") %>%
        dplyr::select(1:3) %>%
        slice_sample(n = i, replace = F) %>%
        dplyr::mutate(id = paste("Random_N", i, sep = "")) %>%
        rbind(CREs_all, .)
  }
  
  CREs_all %>%
    dplyr::filter(id != "Enhancer_Wang2019_1", 
                  id != "Enhancer_Zhu2015_1", 
                  id != "Random_0",
                  id != "Promoter_0") %>%
    dplyr::mutate(start = format(start, scientific = F), 
                  end = format(end, scientific = F)) %>%
    fwrite(file = "./R7.TFs/bed_files/at_CREs_all_20210324.bed.gz", 
           sep = "\t", col.names = F)

```

*CREs 20210416*
20210416: ActivityClusters, 1'st exon, 1'st intron, CorePro, eCorePro, sCorePro, sEnhancer, pEnhancer, HSCorePro, TSCorePro.
CorPro (-50,50) were classified into two kinds, eCorePro and sCorePro, based on them overlap 1bp with enhancer (600 bp width) or not, respectively.


```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
## load CREs generate in 20210324 and filtration
  CREs_all <- fread("./R7.TFs/bed_files/at_CREs_all_20210324.bed.gz") %>%
    dplyr::filter(str_detect(V4, "(iSTARR|Enhancer_Zhu2015_\\d|Enhancer_Wang2019_\\d|Promoter)", negate = T)) %>%
    dplyr::mutate(V1 = str_replace_all(V1, "chr", ""),
                  V4 = str_replace_all(V4, c("cSTARR-seq_1" = "KmeansCluster1", 
                                             "cSTARR-seq_2" = "KmeansCluster2", 
                                             "cSTARR-seq_3" = "KmeansCluster3", 
                                             "cSTARR-seq_4" = "KmeansCluster4",
                                             "cSTARR-seq_all" = "Enhancer", 
                                             "Enhancer_Wang2019_all" = "Enhancer_Wang2019",
                                             "Enhancer_Zhu2015_all" = "Enhancer_Zhu2015")))
  table(CREs_all$V4)
  unique(CREs_all$V4) %>% sort
  
## Activity clusters
  activeCluster <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_20210415.csv.gz") %>%
    dplyr::filter(str_detect(ClusterActivity, "Part"), 
                  method == "cSTARR-seq", 
                  seqnames != "Mt",
                  seqnames != "Pt") %>%
    dplyr::select(seqnames, start, end, ClusterActivity) %>%
    dplyr::mutate(V4 = str_replace_all(ClusterActivity, "Part", "ActivityCluster")) %>%
    dplyr::select(seqnames, start, end, V4)
  
# ## enhancers located near HK and TS genes? (less than 50 enhancer located near 705 HK genes.)
#   MaxDistance <- 2000
#   hk_gene <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv")
#   library(TxDb.Athaliana.BioMart.plantsmart28)
#   txdb <- TxDb.Athaliana.BioMart.plantsmart28
#   enhancer <- CREs_all %>%
#   dplyr::filter(V4 == "Enhancer")
#   colnames(enhancer) <- c("seqnames", "start", "end", "V4")
#   df <- makeGRangesFromDataFrame(enhancer) %>%
#     ChIPseeker::annotatePeak(TxDb = txdb) %>%
#     as.data.frame() %>%
#     dplyr::filter(distanceToTSS < MaxDistance)
#   df$Type <- "TissueSpecific"
#   df$Type[df$transcriptId %in% hk_gene$`Transcript ID`] <- "HouseKeeping"
#   table(df$Type)
  
## Add other types of CRE: (epromoter, Random in gene body, HKpromoter, TSpromoter, 1st exon, 1st intron)
  # epromoter
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  CorePro <- promoters(txdb, upstream = 50, downstream = 50, use.names = T)
  enhancer <- CREs_all %>%
    dplyr::filter(V4 == "Enhancer")
  colnames(enhancer) <- c("seqnames", "start", "end", "V4")
  enhancerGR <- makeGRangesFromDataFrame(enhancer)
  ol <- findOverlaps(CorePro, enhancerGR) # 1bp?
  eCorePro <- as.data.frame(CorePro)[unique(ol@from),] %>%
    dplyr::mutate(V4 = "eCorePro") %>%
    dplyr::select(seqnames, start, end, V4)
  sCorePro <- as.data.frame(CorePro)[-unique(ol@from),] %>%
    dplyr::mutate(V4 = "sCorePro") %>%
    dplyr::select(seqnames, start, end, V4)
  sEnhancer <- enhancer[-unique(ol@to),] %>%
    dplyr::mutate(V4 = "sEnhancer")
  pEnhancer <- enhancer[unique(ol@to),] %>%
    dplyr::mutate(V4 = "pEnhancer")
  
  # TS HK promoter
  hk_gene <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv")
  CorePro <- as.data.frame(CorePro)
  HKCorePro <- CorePro[which(CorePro$tx_name %in% hk_gene$`Transcript ID`),] %>%
    dplyr::mutate(V4 = "HKCorePro") %>%
    dplyr::select(seqnames, start, end, V4)
  TSCorePro <- CorePro[-which(CorePro$tx_name %in% hk_gene$`Transcript ID`),] %>%
    dplyr::mutate(V4 = "TSCorePro") %>%
    dplyr::select(seqnames, start, end, V4)
  CorePro <- CorePro %>%
    dplyr::mutate(V4 = "CorePro") %>%
    dplyr::select(seqnames, start, end, V4)
  
  # Random regions in gene body
  geneAt <- genes(txdb)
  randomGR <- CREs_all %>% 
    dplyr::filter(str_detect(V4, "Random")) %>% 
    makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3") %>%
    GenomicRanges::resize(width = 600, fix = "center")
  ol <- findOverlaps(geneAt, randomGR, minoverlap = 600)
  RandomGeneBody <- as.data.frame(randomGR)[unique(ol@to),] %>%
    dplyr::mutate(V4 = "RandomGeneBody") %>%
    dplyr::select(seqnames, start, end, V4)
  
  # 1'st exon, 1'st intron.
    # first intron
    intron_all <- intronsByTranscript(txdb, use.names = T) %>% as.data.frame()
    first_intron1 <- intron_all %>%
      dplyr::filter(strand == "+") %>%
      dplyr::arrange(group_name, start)
    first_intron1 <- first_intron1[match(unique(first_intron1$group_name), first_intron1$group_name),]
    
    first_intron2 <- intron_all %>%
      dplyr::filter(strand == "-") %>%
      dplyr::arrange(group_name, desc(start))
    first_intron2 <- first_intron2[match(unique(first_intron2$group_name), first_intron2$group_name),]
    
    first_intron <- rbind(first_intron1, first_intron2) %>%
      dplyr::arrange(seqnames, group_name, start) %>%
      na.omit() %>%
      makeGRangesFromDataFrame() %>%
      GenomicRanges::reduce() %>%
      as.data.frame() %>%
      dplyr::mutate(V4 = "FirstIntron") %>%
      dplyr::select(seqnames, start, end, V4)
    rm(first_intron1, first_intron2, intron_all)  
    
    # first exon
    exon_all <- exonsBy(txdb, use.names = T) %>% as.data.frame()
    first_exon1 <- exon_all %>%
      dplyr::filter(strand == "+") %>%
      dplyr::arrange(group_name, start)
    first_exon1 <- first_exon1[match(unique(first_exon1$group_name), first_exon1$group_name),]
    
    first_exon2 <- exon_all %>%
      dplyr::filter(strand == "-") %>%
      dplyr::arrange(group_name, desc(start))
    first_exon2 <- first_exon2[match(unique(first_exon2$group_name), first_exon2$group_name),]
    
    first_exon <- rbind(first_exon1, first_exon2) %>%
      dplyr::arrange(seqnames, group_name, start) %>%
      na.omit() %>%
      makeGRangesFromDataFrame() %>%
      GenomicRanges::reduce() %>%
      as.data.frame() %>%
      dplyr::mutate(V4 = "FirstExon") %>%
      dplyr::select(seqnames, start, end, V4)
    rm(first_intron1, first_exon2, exon_all) 
  
  # merge
    df <- rbind(first_intron, first_exon, RandomGeneBody, CorePro, 
                eCorePro, sCorePro, HKCorePro, TSCorePro, pEnhancer, sEnhancer)
    names(CREs_all) <- names(df)
    table(CREs_all$V4)
    table(df$V4)
    ALL <- rbind(CREs_all, df, activeCluster) 
    table(ALL$V4)
    
  # write to file
    # all
    ALL %>%
      dplyr::mutate(start = format(start, scientific = F), 
                  end = format(end, scientific = F), 
                  seqnames = str_c("chr", seqnames, "")) %>%
      dplyr::filter(seqnames != "chrMt", seqnames != "chrPt") %>%
      fwrite(file = "./R7.TFs/bed_files/at_CREs_all_20210416.bed.gz", 
             sep = "\t", col.names = F)
    
    # only ActiveCluster & CorePro
    tmp <- ALL %>%
      dplyr::mutate(start = format(start, scientific = F), 
                  end = format(end, scientific = F), 
                  seqnames = str_c("chr", seqnames, "")) %>%
      dplyr::filter(seqnames != "chrMt", seqnames != "chrPt", 
                    str_detect(V4, "(Activity|^CorePro$)")) %>%
      fwrite(file = "./R7.TFs/bed_files/at_CREs_20210420_add.bed.gz", 
             sep = "\t", col.names = F)
      
```

*20210617*
20210508: CorePro of active/silence gene, distal intergenic, gene body
```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  ALL <- fread("./R7.TFs/bed_files/at_CREs_all_20210416.bed.gz")
  
# CorePro of active/silent genes
  CorePro <- promoters(txdb, upstream = 50, downstream = 50, use.names = T) %>% 
    as.data.frame() %>%
    dplyr::mutate(gene_id = str_replace_all(tx_name, "(.+)\\..+", "\\1")) %>%
    dplyr::select(seqnames, start, end, gene_id)
  gene_tab <- fread("./R6.expression/Gene_list_with_CREs_500TSS100_300TTS300_20210322.csv.gz") %>%
    dplyr::select(seqnames, start, end, gene_id, strand, RPKM) %>%
    dplyr::mutate(RPKM = ifelse(is.na(RPKM), 0, RPKM))
  hist(gene_tab$RPKM, breaks = 2000, xlim = c(0, 200))
  quantile(gene_tab$RPKM, probs = seq(0, 1, 0.05), na.rm = T)
  table(cut(gene_tab$RPKM, breaks = c(-1, 1, Inf)))
  
  CorePro <- gene_tab %>%
    dplyr::select(gene_id, RPKM) %>%
    left_join(CorePro, .)
  CorePro$V4 <- ifelse(CorePro$RPKM > 1, "ActiveCorePro", "SilentCorePro")
  CorePro <- CorePro %>% 
    dplyr::select(seqnames, start, end, V4) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""))
  names(CorePro) <- str_c("V", 1:4)

# distal intergenic and gene body
  cre_all <- fread("./R4.Distribution_of_enhancers/GAT_overlap/GenomicFeatures_20210509.bed.gz")
  table(cre_all$V4)
  regions <- cre_all %>%
    dplyr::filter(str_detect(V4, "(DistalIntergenic|TSS2TTS)"))
  
# merge
  ALLmerge <- ALL %>%
    rbind(., CorePro) %>%
    rbind(., regions)
  CreSummary <- ALLmerge %>%
    dplyr::mutate(Width = V3 - V2) %>%
    group_by(V4) %>%
    dplyr::summarise(MedianWidth = median(Width), 
              Num = n())
  table(ALLmerge$V4)
  table(ALLmerge$V1)
  
  fwrite(ALLmerge, file = "./R7.TFs/bed_files/at_CREs_all_20210509.bed.gz",
         sep = "\t", col.names = F, row.names = F)
  
  # just keep CREs which did not run in GAT to save time.
  tmp <- ALLmerge %>%
    dplyr::filter(str_detect(V4, "(ActiveCorePro|SilentCorePro|FirstIntron|FirstExon|DistalIntergenic|TSS2TTS)"))
  unique(tmp$V4)
  fwrite(tmp, file = "./R7.TFs/bed_files/at_CREs_PartReRun_20210509.bed.gz",
         sep = "\t", col.names = F, row.names = F)
  
  # just keep CREs which did not run in GAT to save time.
  tmp <- ALLmerge %>%
    dplyr::filter(str_detect(V4, "(ActiveCorePro|SilentCorePro)"))
  table(tmp$V4)
  fwrite(tmp, file = "./R7.TFs/bed_files/at_CREs_PartReRun_20210617.bed.gz",
         sep = "\t", col.names = F, row.names = F)
```

*20210517*
20210517: Clustered and dispersed Enhancers
```{r}
# load
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  ALL <- fread("./R7.TFs/bed_files/at_CREs_all_20210509.bed.gz")

# clustered dispersed enhancers.
  cre_all <- fread("Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    dplyr::select(seqnames, start, end, ClusterType) %>%
    dplyr::filter(seqnames != "Mt")
  table(cre_all$seqnames)
  table(cre_all$ClusterType)
  colnames(cre_all) <- paste("V", 1:4, sep = "")
  cre_all$V1 <- str_c("chr", cre_all$V1, sep = "")
  
# merge
  ALLmerge <- ALL %>%
    rbind(., cre_all)
  table(ALLmerge$V4)
  unique(cre_all$V1)
  
  # summary
  CreSummary <- ALLmerge %>%
    dplyr::mutate(Width = V3 - V2) %>%
    group_by(V4) %>%
    dplyr::summarise(MedianWidth = median(Width), 
              Num = n()) %>%
    dplyr::filter(str_detect(V4, "Random", negate = T))
  fwrite(ALLmerge, file = "./R7.TFs/bed_files/at_CREs_all_20210517.bed.gz",
         sep = "\t", col.names = F, row.names = F)
  
# only clustered/dispersed enhancers
  fwrite(cre_all, file = "./R7.TFs/bed_files/at_clusteredDispersed_enhancers_20210517.bed.gz",
         sep = "\t", col.names = F, row.names = F)
  
```

*20210630*
20210630: Enhancer_Meng2021
```{r}
source("/work/bio-tanyj/soft/TanYongjun_code.R")
fread("./R4.Distribution_of_enhancers/TEs/CREs_forGAT_WholeKmeanActivityClustered_20210629.bed.gz") %>%
  dplyr::filter(V4 == "Enhancer_Meng2021") %>%
  fwrite(file = "./R7.TFs/bed_files/Enhancers_Meng2021_20210630.bed.gz",
         sep = "\t", col.names = F, row.names = F)

```

*20210702*
enhancers (border of 600bp region) located 300 or 600 bp away from CorePro(-50, 50)
```{r}
  source("D:/SUST/code/TanYongjun_code.R")
  cre_all <- fread("./R7.TFs/bed_files/at_CREs_all_20210509.bed.gz") %>%
    dplyr::filter(str_detect(V4, "(sCorePro|sEnhancer)"))
  table(cre_all$V4)
  
  sCorePro <- fread("./R7.TFs/bed_files/at_CREs_all_20210509.bed.gz") %>%
    dplyr::filter(V4 == "sCorePro")
  sEnhancer <- fread("./R7.TFs/bed_files/at_CREs_all_20210509.bed.gz") %>%
    dplyr::filter(V4 == "sEnhancer")
  
  colnames(sCorePro) <- colnames(sEnhancer) <- c("seqnames", "start", "end", "type")
  sCoreProGR300 <- makeGRangesFromDataFrame(sCorePro) %>%
    GenomicRanges::resize(width = 700, fix = "center")
  sCoreProGR600 <- makeGRangesFromDataFrame(sCorePro) %>%
    GenomicRanges::resize(width = 1300, fix = "center")
  sEnhancerGR <- makeGRangesFromDataFrame(sEnhancer)
  
  ol300 <- findOverlaps(sCoreProGR300, sEnhancerGR)
  ol600 <- findOverlaps(sCoreProGR600, sEnhancerGR)
  
  sEnhancer300 <- sEnhancer[-c(unique(ol300@to)),] %>%
    dplyr::mutate(type = "sEnhancer300CorePro")
  sEnhancer600 <- sEnhancer[-c(unique(ol600@to)),] %>%
    dplyr::mutate(type = "sEnhancer600CorePro")
  rbind(sEnhancer300, sEnhancer600) %>%
    fwrite("./R7.TFs/bed_files/at_CREs_enhancer_located_xxbp_from_CorePro.bed.gz", 
           col.names = F, row.names = F, sep = "\t")
```

*20220517*
enhancers located near tissue-specific/house-keeping genes.

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  En_STARR <- fread("./R7.TFs/bed_files/at_CREs_all_20210509.bed.gz") %>%
    dplyr::filter(str_detect(V4, "^Enhancer$")) %>%
    dplyr::mutate(V1 = str_replace_all(V1, "chr", ""))
  En_STARRGR <- makeGRangesFromDataFrame(En_STARR, seqnames.field = "V1", start.field = "V2", end.field = "V3")
  
  # TS and HK gene (-300bp, TSS-TTS, 300bp)
  hk_gene <- fread("./refgenome/Arabidopsis_thaliana_HouseKeeping_genes_10.1111.tpj.13415.tsv") %>%
    dplyr::mutate(gene_id = str_replace_all(`Transcript ID`, "^(.+)\\.\\d+$", "\\1"))
  library(TxDb.Athaliana.BioMart.plantsmart28)
  TSS2TTS <- genes(TxDb.Athaliana.BioMart.plantsmart28) %>%
    as.data.frame() %>%
    dplyr::mutate(Type = ifelse(gene_id %in% hk_gene$gene_id, "HK", "TS"),
                  start = start - 500,
                  end = end + 500) %>%
    split(., .$Type) %>%
    map(., makeGRangesFromDataFrame)
  
  ol <- findOverlaps(En_STARRGR, TSS2TTS$HK)
  HKenhancer <- En_STARR[unique(ol@from), ] %>%
    dplyr::mutate(V4 = "HKenhancer")
  ol <- findOverlaps(En_STARRGR, TSS2TTS$TS)
  TSenhancer <- En_STARR[unique(ol@from), ] %>%
    dplyr::mutate(V4 = "TSenhancer")
  
  full_join(HKenhancer, TSenhancer) %>%
    dplyr::mutate(V1 = str_c("chr", V1)) %>%
    fwrite("./R7.TFs/bed_files/Enhancer_TSHK_gene_up500_down500_20220517.bed.gz", 
           col.names = F, row.names = F, sep = "\t")

```


*DAP peaks*

```{r}
## DAP peaks (run on Taiyi, each TFs in a bed file)
  source("/work/bio-tanyj/soft/TanYongjun_code.R")
  setwd("/scratch/2021-06-16/bioTanyj/At_enhancer/")
  FileLs <- fread(file = "./refgenome/GSE60143_RAW_DAPSeq/meta_info_byTYJ_20210324.tsv")
  for (i in seq_len(nrow(FileLs))) {
    cat(i, " ")
    fread(paste("./refgenome/GSE60143_RAW_DAPSeq/", 
                                        FileLs$file[i], sep = ""), header = F) %>%
      dplyr::select(1:3) %>%
      arrange(V1, V2, V3) %>%
      dplyr::mutate(V2 = ifelse(V2 < 1, 1, V2)) %>% # some negative value found in bed file!
      dplyr::mutate(V2 = format(V2, scientific = F), 
                    V3 = format(V3, scientific = F)) %>%
      fwrite(file = paste("/scratch/2021-06-16/bioTanyj/At_enhancer/R7.TFs/bed_DAPSeq/",
                          FileLs$Family[i], "_",FileLs$Protein[i], 
                          "_", FileLs$Methy[i], ".bed.gz", sep = ""), 
             sep = "\t", col.names = F)
  }
```

### .2 Run GAT

`/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/7.1_CRE_TFs_overlap_GAT.R` `/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/7.1_CRE_TFs_overlap_GAT.sh`


## ---------- 20210702 ------
Enrichment analyais (GAT) of DAP-seq peaks in all these types of CREs were performed in several times separately.
Somr CREs may calculate in different runs, and duplicate results may contained in results.
collected GAT overlap results calculated in 20210616 (Act, Kmeans, Pro, En_Zhu, En_Wang, etc.), 20210617 (silent/active CorePro), 20210630 (Intronic En), 20210702(sEnhancer300CorePro), and 20220517 (TSenhancer, HKenhancer).

### .3 collect results and adjust

#### . collect on Taiyi.

```{r, collecte results on Taiyi, eval=F}
setwd("/scratch/2021-06-16/bioTanyj/At_enhancer/R7.TFs/calculation_GAT_20210616")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

# load results of each DAP peaks (all CRE in the same file)
results_all <- NULL
for(i in list.files(pattern = ".+.tsv$")){
  temp <- fread(i, header = T) %>%
    dplyr::filter(str_detect(annotation, "(ActiveCorePro|SilentCorePro)", negate = T)) %>%
    dplyr::mutate(track = i)
  results_all <- rbind(results_all, temp)
}
fwrite(results_all, file = "All_GAT_100000_results_20210616.csv.gz")

```

```{r, collecte results on Taiyi, eval=F}
setwd("/scratch/2021-07-02/bioTanyj/At_enhancer/R7.TFs/calculation_GAT_20210617")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

# load results of each DAP peaks (all CRE in the same file)
results_all <- NULL
for(i in list.files(pattern = ".+.tsv$")){
  temp <- fread(i, header = T) %>%
    mutate(track = i)
  results_all <- rbind(results_all, temp)
}
fwrite(results_all, file = "All_GAT_100000_results_20210617.csv.gz")

```

```{r, collecte results on Taiyi, eval=F}
setwd("/scratch/2021-06-16/bioTanyj/At_enhancer/R7.TFs/calculation_GAT_20210630")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

# load results of each DAP peaks (all CRE in the same file)
results_all <- NULL
for(i in list.files(pattern = ".+.tsv$")){
  temp <- fread(i, header = T) %>%
    dplyr::filter(str_detect(annotation, "(ActiveCorePro|SilentCorePro)", negate = T)) %>%
    dplyr::mutate(track = i)
  results_all <- rbind(results_all, temp)
}
fwrite(results_all, file = "All_GAT_100000_results_20210630.csv.gz")

```

```{r, collecte results on Taiyi, eval=F}
setwd("/scratch/2021-07-02/bioTanyj/At_enhancer/R7.TFs/calculation_GAT_20210702")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

# load results of each DAP peaks (all CRE in the same file)
results_all <- NULL
for(i in list.files(pattern = ".+.tsv$")){
  temp <- fread(i, header = T) %>%
    dplyr::filter(str_detect(annotation, "(ActiveCorePro|SilentCorePro)", negate = T)) %>%
    dplyr::mutate(track = i)
  results_all <- rbind(results_all, temp)
}
fwrite(results_all, file = "All_GAT_100000_results_20210702.csv.gz")

```

```{r, collecte results on Taiyi, eval=F}
setwd("/scratch/2022-05-11/bioTanyj/At_enhancer/R7.TFs/calculation_GAT_20220517")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

# load results of each DAP peaks (all CRE in the same file)
results_all <- NULL
for(i in list.files(pattern = ".+.tsv$")){
  temp <- fread(i, header = T) %>%
    dplyr::mutate(track = i)
  results_all <- rbind(results_all, temp)
}
fwrite(results_all, file = "All_GAT_100000_results_20220517.csv.gz")

```


####. local

```{r}
  rm(list=ls())
  source("D:/SUST/code/TanYongjun_code.R")
# table of DAPseq peaks meta info
  FileLs <- fread("./refgenome/GSE60143_RAW_DAPSeq/meta_info_byTYJ_20210324.tsv") %>%
    mutate(track = paste(Family, "_", Protein, "_", Methy, ".tsv", sep = ""))
  
# results  of GAT.
  r1 <- fread("./R7.TFs/All_GAT_100000_results_20210616.csv.gz") %>%
    dplyr::mutate(Data = "20210616") 
  r2 <- fread("./R7.TFs/All_GAT_100000_results_20210617.csv.gz") %>%
    dplyr::mutate(Data = "20210617")
  r3 <- fread("./R7.TFs/All_GAT_100000_results_20210630.csv.gz") %>%
    dplyr::mutate(Data = "20210630")
  r4 <- fread("./R7.TFs/All_GAT_100000_results_20210702.csv.gz") %>%
    dplyr::mutate(Data = "20210702")
  r5 <- fread("./R7.TFs/All_GAT_100000_results_20220517.csv.gz") %>%
    dplyr::mutate(Data = "20220517")
  results_all <- rbind(r1, r2, r3, r4, r5) %>%
    as.data.frame() %>%
    dplyr::mutate(track = str_replace_all(track, "_add", ""))
  rm(r1, r2, r3, r4, r5)
  
  # match(results_all$track, FileLs$track)
  results_all <- right_join(FileLs, results_all)
  unique(results_all$annotation)
  table(results_all$annotation, results_all$Data)

## summary of CREs
  CREsAll <- fread("./R7.TFs/bed_files/at_CREs_all_20210517.bed.gz") %>%
    rbind(., fread("./R7.TFs/bed_files/Enhancers_Meng2021_20210630.bed.gz")) %>%
    rbind(., fread("./R7.TFs/bed_files/at_CREs_enhancer_located_xxbp_from_CorePro.bed.gz")) %>%
    rbind(., fread("./R7.TFs/bed_files/Enhancer_TSHK_gene_up500_down500_20220517.bed.gz"))
  table(CREsAll$V4)
  CREs_summary <- CREsAll %>%
    dplyr::mutate(width = V3 - V2) %>%
    group_by(V4) %>%
    dplyr::summarise(MedianWidth = median(width), NumRegions = n()) %>%
    dplyr::rename(annotation = V4)

# rename CRE/CRE clusters
  unique(results_all$annotation) %>% sort()
  unique(CREsAll$V4) %>% sort()
  results_all$annotation <- str_replace_all(results_all$annotation, 
                                            c("^Enhancer$" = "Enhancer_STARRseq"))
  
# adjust pvalue
  results_sig <- results_all %>%
    group_by(annotation) %>%
    mutate(Num = length(pvalue), 
           padj_BH = p.adjust(pvalue, method = "BH")) %>%
    mutate(padj_BH_lab = toolkit_tyj$returnAsterisk(padj_BH)) %>%
    # dplyr::filter(abs(l2fold) > 1) %>%
    mutate(fold_type = ifelse(fold > 1, "Enriched", "Depleted")) %>%
    mutate(Type = str_c(fold_type, padj_BH_lab, sep = " ")) %>%
    mutate(Type = str_replace_all(Type, ".+ns", "ns"))

  # fwrite(results_sig, file = "./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz")
  
```

```{r, summary}

  hist(log2(results_sig$fold), breaks = 20)
  table(cut(results_sig$padj_BH, breaks = c(0, 0.0001, 0.001, 0.0079, 0.01, 0.05, 1)))

# summary table of significant
  # Number of DAP peaks enriched, depleted in CREs.
  for (i in unique(results_sig$Methy)) {
      results_sig %>%
        # dplyr::mutate(padj_BH_lab = ifelse(padj_BH >= 0.0040550, "ns", padj_BH_lab)) %>% # use most significant p value generated with Random regions as threshold.
        dplyr::filter(Methy == i) %>%
        group_by(annotation, Type) %>%
        dplyr::summarise(Num = n()) %>%
        pivot_wider(id_cols = annotation, names_from = Type, values_from = Num) %>%
        left_join(., CREs_summary) %>%
        fwrite(paste("./R7.TFs/Summary_GAT_DAPseqPeaks_",i, "_hit_CRE_20210609.csv", sep = ""), na = 0)
  }
  
  # plot of enrichment level in GAT.
  results_sig %>%
    dplyr::select(track, annotation, fold, padj_BH) %>%
    mutate(padj_BH = toolkit_tyj$returnAsterisk(padj_BH),
           TF = str_replace_all(track, ".+_(.+?).tsv", "\\1")) %>%
    # pivot_wider(id_cols = track, names_from = annotation, values_from = qvalue) %>%
    ggplot(aes(x = fold, color = padj_BH, fill = padj_BH)) +
      geom_histogram(position = position_dodge()) +
      scale_x_continuous(trans = "log10") + 
      ggtitle("Density plot of enrichment") + 
      theme_bw()ee
```


### .4 Col vs ColAmp

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  
# load
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz")
  
  # Protein with Col and Col_amp.
  TF_ls <- results_sig %>%
    dplyr::select(Protein, Methy) %>%
    distinct() %>%
    group_by(Protein) %>%
    dplyr::summarise(Num = n()) %>%
    dplyr::filter(Num > 1) %>%
    dplyr::select(Protein) %>%
    unlist() %>%
    as.vector()
  results_comp <- results_sig %>%
    dplyr::filter(Protein %in% TF_ls)
  
  # relationship between Col and Col_amp
  CompareSummary <- results_comp %>%
    dplyr::mutate(Type = str_replace_all(Type, "[\\*\\s]+", "")) %>%
    pivot_wider(id_cols = c(Protein, annotation), names_from = Methy, values_from = Type)
  
  p <- CompareSummary %>%
    dplyr::mutate(value = str_c("Amp ", Col_amplified, " - Col ", Col, sep = "")) %>%
    group_by(value) %>%
    dplyr::summarise(Num = n()) %>%
    arrange(Num) %>% 
    dplyr::mutate(value = factor(value, levels = value)) %>%
    dplyr::mutate(Percent = Num / sum(Num), 
                  label = percent(Percent, accuracy = 0.01)) %>%
    ggplot(aes(x = Percent, y = value)) + 
      geom_col(fill = "grey55") + 
      toolkit_tyj$theme_my + 
      geom_text(aes(label = label), nudge_x = 0.02, size = 2.5) + 
    ggtitle("Compare GAT reuslts of DAP-seq peaks \nuse Col and Col_amp DNA")
  p
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Compare_Col_ColAmp_DAPpeaks_enriched_in_all_CRE_datasets",
  #                      width = 12, height = 8, device = "png", plot = p)
```


### .5 Num of TFs (Tower plot)

#### .Supp talbe

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz")
  dep_ls <- fread("./refgenome/GSE60143_RAW_DAPSeq/meta_info_byTYJ_20210324.tsv")
  table(dep_ls$Family, dep_ls$Methy)
  table(dep_ls$Methy)
  
  table(results_sig$annotation)
  tmp <- results_sig %>%
    dplyr::filter(str_detect(annotation, "ActivityCluster|Kmeans|HKCorePro|TSCorePro|sCorePro|^sEnhancer$|Enhancer_")) %>%
    dplyr::mutate(annotation = str_replace_all(annotation, c("ActivityCluster" = "Act", 
                                "KmeansCluster" = "C",
                                "HKCorePro" = "Promoter_House-keeping",
                                "TSCorePro" = "Promoter_Tissue-specific",
                                "sEnhancer" = "sEnhancer_STARRseq",
                                "sCorePro" = "sPromoter"))) %>%
    dplyr::select(-c(file, DAPSeq, Methy)) %>%
    dplyr::mutate(across(.cols = matches("percent"), .fns = function(x){x/100}))
  unique(tmp$annotation) %>%sort()
  head(tmp)
  
  # tmp %>%
    # fwrite("./R7.TFs/GAT_results_for_Supp_Table.csv")
  
```


#### .Count

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz")
  dep_ls <- fread("./refgenome/GSE60143_RAW_DAPSeq/meta_info_byTYJ_20210324.tsv")
  table(dep_ls$Family, dep_ls$Methy)
  table(dep_ls$Methy)

## summary of CREs
  CREsAll <- fread("./R7.TFs/bed_files/at_CREs_all_20210517.bed.gz") %>%
    rbind(., fread("./R7.TFs/bed_files/Enhancers_Meng2021_20210630.bed.gz")) %>%
    rbind(., fread("./R7.TFs/bed_files/at_CREs_enhancer_located_xxbp_from_CorePro.bed.gz")) %>%
    rbind(., fread("./R7.TFs/bed_files/Enhancer_TSHK_gene_up500_down500_20220517.bed.gz")) %>%
    dplyr::mutate(V4 = str_replace_all(V4, "^Enhancer$", "Enhancer_STARRseq"))
  table(CREsAll$V4, CREsAll$V1)
  CREs_summary <- CREsAll %>%
    dplyr::mutate(width = V3 - V2) %>%
    group_by(V4) %>%
    dplyr::summarise(MedianWidth = median(width), NumRegions = n()) %>%
    dplyr::rename(annotation = V4)
  
  setdiff(unique(results_sig$annotation), unique(CREsAll$V4))
  setdiff(unique(CREsAll$V4), unique(results_sig$annotation))

## Table of TFs enriched in CREs/CREs cluster
  df <- results_sig %>%
    dplyr::filter(padj_BH < 0.05) %>%
    dplyr::mutate(tmp = str_c(round(l2fold, digits = 2), "(", format(padj_BH, scientific = T, digits = 2), ")", sep = "")) %>%
    pivot_wider(id_cols = c(Family, Protein, Methy), names_from = annotation, values_from = tmp) %>%
    arrange(Family, Protein, Methy)
  # fwrite(df, "./R7.TFs/Table_of_TFs_enriched_in_CREs_cluster_20210616.csv")

## Number of TFs and TFfamilys enriched in each CRE or CRE clusters.
  dap_summary <- dep_ls %>%
    group_by(Methy, Family) %>%
    dplyr::count(name = "Total")
  

### Percent of enriched TFs------
  table(dep_ls$Methy)
  
  # Count All TFs
  Num_TFs <- results_sig %>%
    dplyr::filter(padj_BH < 0.05) %>%
    group_by(Methy, annotation, fold_type) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = c(annotation, fold_type), names_from = Methy, values_from = Num, values_fill = NA) %>%
    dplyr::mutate(Col = Col / 470, 
                  Col_amplified = Col_amplified / 341) %>%
    pivot_longer(cols = Col:Col_amplified, names_to = "Methy") %>%
    pivot_wider(id_cols = fold_type:Methy, names_from = annotation, values_from = value, values_fill = 0) %>%
    dplyr::mutate(Family = "Total") 
  
  # Count at each TFfamily
  Num_Family <- results_sig %>%
    dplyr::filter(padj_BH < 0.05) %>%
    group_by(annotation, Methy, Family, fold_type) %>%
    dplyr::count(name = "Num") %>%
    left_join(., dap_summary) %>%
    dplyr::mutate(Ratio = Num / Total) %>%
    pivot_wider(id_cols = c(Family, Methy, fold_type), names_from = annotation, values_from = Ratio, values_fill = 0) %>%
    arrange(Family, Methy)
  
  # merge and classify TFs (based on enriched in CREs or not.)
  df <- full_join(Num_TFs, Num_Family) %>%
    dplyr::select(ncol(.), 1:(ncol(.) - 1)) %>%
    left_join(., dap_summary)
  
   # label and classify
    df <- df %>%
      dplyr::mutate(sCorePro = ifelse(is.na(sCorePro), 0.00001, sCorePro),
                    sEnhancer = ifelse(is.na(sEnhancer), 0.00001, sEnhancer),
                    EPratio = sEnhancer/sCorePro,
                    log2EPratio = log2(sEnhancer/sCorePro)) %>%
      dplyr::mutate(VariedFamily = ifelse(abs(log2EPratio) > 1 & Total > 5, "Yes", "No"))
    
   ## write
    # df %>%
    #   fwrite("./R7.TFs/Percent_of_Methy_Amp_TFs_TFFamily_members_enriched_depleted_in_CREs_20210616.csv", quote = T)

### Number of enriched TFs------
  
  # each TFs level
  Num_TFs <- results_sig %>%
    dplyr::filter(padj_BH < 0.05) %>%
    group_by(Methy, annotation, fold_type) %>%
    dplyr::summarise(Num = n()) %>%
    pivot_wider(id_cols = c(annotation, fold_type), names_from = Methy, values_from = Num, values_fill = NA) %>%
    pivot_longer(cols = Col:Col_amplified, names_to = "Methy") %>%
    pivot_wider(id_cols = fold_type:Methy, names_from = annotation, values_from = value, values_fill = 0) %>%
    dplyr::mutate(Family = "Total") 
  
  # each TFfamily level
  Num_Family <- results_sig %>%
    dplyr::filter(padj_BH < 0.05) %>%
    group_by(annotation, Methy, Family, fold_type) %>%
    dplyr::count(name = "Num") %>%
    left_join(., dap_summary) %>%
    dplyr::mutate(Ratio = Num / Total) %>%
    pivot_wider(id_cols = c(Family, Methy, fold_type), names_from = annotation, values_from = Num, values_fill = 0) %>%
    arrange(Family, Methy)
  
  # merge and classify TFs (based on enriched in CREs or not.)
  df <- full_join(Num_TFs, Num_Family) %>%
    dplyr::select(ncol(.), 1:(ncol(.) - 1)) %>%
    left_join(., dap_summary)
  
   # label and classify
  df <- df %>%
    dplyr::mutate(sCorePro = ifelse(is.na(sCorePro), 0.01, sCorePro),
                  sEnhancer = ifelse(is.na(sEnhancer), 0.01, sEnhancer),
                  EPratio = sEnhancer/sCorePro,
                  log2EPratio = log2(sEnhancer/sCorePro),
                  EPdiff = round(sEnhancer - sCorePro, digits = 0)) %>%
    dplyr::mutate(VariedFamily = ifelse(abs(log2EPratio) > 1 & Total > 5, "Yes", "No"))
   
   # # write
   # df %>%
   #   fwrite("./R7.TFs/Number_of_Methy_Amp_TFs_TFFamily_members_enriched_depleted_in_CREs_20210616.csv", quote = T)

```

#### .prepare
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
df <- read_csv("./R7.TFs/Number_of_Methy_Amp_TFs_TFFamily_members_enriched_depleted_in_CREs_20210616.csv") %>%
  dplyr::filter(Methy == "Col") # discard "Col_amplified"

# prepare
FamilyOder <- df %>%
  dplyr::select(Family, Total) %>%
  na.omit() %>%
  unique() %>%
  arrange(desc(Total))
sigVaryTFfamily <- df %>%
  dplyr::filter(abs(log2EPratio) > 1 & Total > 5) %>%
  arrange(desc(abs(log2EPratio))) %>%
  dplyr::select(Family) %>%
  unlist() %>%
  unique()

tmp <- df %>%
  dplyr::select(matches("(^Family|fold_type|Total|CorePro|Enhancer|enhancer|Activity|GeneBody|Kmeans|First|TSS2TTS|Exon|Intron)")) %>%
  dplyr::filter(Family != "Total") %>%
  # dplyr::filter(Family %in% sigVaryTFfamily) %>%
  pivot_longer(cols = matches("CorePro|Enhancer|enhancer|Activity|GeneBody|Kmeans|First|TSS2TTS|Exon|Intron"), values_to = "Num_sig") %>%
  dplyr::mutate(Total = if_else(fold_type == "Depleted", -Total, Total), 
                Num_sig = if_else(fold_type == "Depleted", -Num_sig, Num_sig), 
                Family = factor(Family, levels = FamilyOder$Family), 
                fold_type = factor(fold_type, levels = c("Enriched", "Depleted")), 
                Num_sig = ifelse(abs(Num_sig) < 0.5, 0, Num_sig))

# summary of each family level
df %>%
  dplyr::select(matches("(^Family|fold_type|Total|CorePro|Enhancer|enhancer|Activity|GeneBody|Kmeans|First|TSS2TTS|Exon|Intron)")) %>%
  dplyr::mutate(Total = ifelse(is.na(Total), 470, Total)) %>%
  dplyr::mutate(across(.cols = matches("(CorePro|Enhancer|enhancer|Activity|GeneBody|Kmeans|First|TSS2TTS|Exon|Intron)"), 
         function(x){
           x[is.na(x)] <- 0
           return(paste(percent(x / .$Total, accuracy = 0.01), 
                        "(", x, "/", .$Total, ")", sep = ""))
         })) %T>%
  # fwrite("./R7.TFs/Percent_of_TF_family_members_enriched_depleted_in_CREs_summary_20210704.csv", row.names = T)
  View()
```

#### .plot All CorePro/ Enhancer
*Two styles*
```{r}
### plot ---------------------
### style 1 (Not used)-------------------
 unique(tmp$name)
 tmp %>%
  dplyr::filter(str_detect(name, "sCorePro|sEnhancer")) %>%
  ggplot(aes(x = Family, y = Num_sig)) +
    geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
    geom_col(aes(fill = fold_type)) +
    toolkit_tyj$theme_my + 
    facet_grid(name ~ .) + 
    scale_fill_npg()+
    # coord_flip() +
    scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                       labels = c(40, 20, 0, 20, 40)) +
    xlab("Protein family") + ylab("Number of protein family members") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.position = "bottom", 
          legend.key.size = unit(0.2, units = "cm")) + 
    guides(fill = guide_legend(title = NULL, nrow = 1))
 
  # toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Promoter_Enhancer_style1_20210616",
  #                      width = 16, height = 16, device = "png")
  
### style 2 (prefer) --------------------
tmp %>%
  dplyr::filter(str_detect(name, "sCorePro|sEnhancer")) %>%
  ggplot(aes(x = Family, y = Num_sig)) +
    geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
    geom_col(aes(fill = fold_type)) +
    toolkit_tyj$theme_my + 
    facet_grid(. ~ name) + 
    scale_fill_npg()+
    coord_flip() +
    scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                       labels = c(40, 20, 0, 20, 40)) +
    xlab("Protein family") + ylab("Number of protein family members") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.position = "bottom", 
          legend.key.size = unit(0.2, units = "cm")) + 
    guides(fill = guide_legend(title = NULL, nrow = 1))
  
  # toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Promoter_Enhancer_style2_20210616",
  #                      width = 12, height = 12, device = "png")
```


#### .Pro/En, Kmeans, Activity, Clustered/Dispersed.

*Enhancers identified using different methods*

```{r}
  unique(tmp$name)
  # Enhancers identified by different methods
  tmp %>%
    dplyr::filter(str_detect(name, "Enhancer_.+")) %>%
    # dplyr::mutate(name =str_replace_all(name, "s", "")) %>%
    dplyr::mutate(name = factor(name, levels = c("Enhancer_STARRseq", 
                                                 "Enhancer_Zhu2015", 
                                                 "Enhancer_Wang2019", 
                                                 "Enhancer_Meng2021"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      guides(fill = guide_legend(title = NULL, nrow = 1))
  
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Four_kinds_of_enhancers_style2_20210616",
                       width = 10.5, height = 10)  
```

*Promoter / enhancer*
```{r}
  unique(tmp$name)
  # promoter vs enhancer
  tmp %>%
    dplyr::filter(str_detect(name, "(^sCorePro$|^sEnhancer600)")) %>%
    dplyr::mutate(name =str_replace_all(name, c("s" = "",
                                                "\\d00.+" = "",
                                                "CorePro" = "Promoter"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      # labs(caption = "sCorePro, sEnhancer (600bp away from CorePro)") +
      guides(fill = guide_legend(title = NULL, nrow = 1))
      
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_sPromoter_sEnhancer600bpCorePro_style2_20210704",
                       width = 6, height = 10)  
```

*Kmeans Cluster*
```{r}
  # KmeansCluster
  unique(tmp$name) %>% as.character() %>% sort
  tmp %>%
    dplyr::filter(str_detect(name, "(^Kmeans|^Enhancer_STARRseq$)")) %>%
    dplyr::mutate(name = str_replace_all(name, c("KmeansCluster" = "C",
                                                 "Enhancer_STARRseq" = "All")),
                  name = factor(name, levels = c(str_c("C", 1:4, sep = ""), "All"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      guides(fill = guide_legend(title = NULL, nrow = 1))
  
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_KmeansEnhancer_style2_20210616",
                       width = 11, height = 10, device = "png")  

```

*Kmeans Cluster & promoter*
```{r}
  # KmeansCluster
  unique(tmp$name) %>% as.character() %>% sort
  tmp %>%
    dplyr::filter(str_detect(name, "(^Kmeans|^Enhancer_STARRseq$|^sCorePro$)")) %>%
    dplyr::mutate(name = str_replace_all(name, c("KmeansCluster" = "C",
                                                 "Enhancer_STARRseq" = "Enhancer",
                                                 "sCorePro" = "Promoter")),
                  name = factor(name, levels = c(str_c("C", 1:4, sep = ""), "Enhancer", "Promoter"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      guides(fill = guide_legend(title = NULL, nrow = 1))
  
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_KmeansEnhancer_Promoter_style2_20210616",
                       width = 9, height = 10)  

```


*Activity Cluster*
```{r}
  # ActivityCluster
  unique(tmp$name) %>% as.character() %>% sort
  tmp %>%
    dplyr::filter(str_detect(name, "(^Activity|^Enhancer_STARRseq$)")) %>%
    dplyr::mutate(name = str_replace_all(name, c("ActivityCluster" = "Act",
                                                 "Enhancer_STARRseq" = "All"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      guides(fill = guide_legend(title = NULL, nrow = 1))
  
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_ActivityEnhancer_style2_20210616",
                       width = 12, height = 10)
  
```

*Core promoter of Tissue Specific vs House Keeping genes*
```{r}
  # Tissue-Specific/House-Keeping 
  unique(tmp$name) %>% as.character() %>% sort
  tmp %>%
    dplyr::filter(str_detect(name, "TSCorePro|HKCorePro")) %>%
    dplyr::mutate(name = str_replace_all(name, c("^CorePro$" = "All",
                                                 "HKCorePro" = "House-keeping",
                                                 "TSCorePro" = "Tissue-specific")),
                  name = factor(name, levels = c("House-keeping", "Tissue-specific", "All"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      guides(fill = guide_legend(title = NULL, nrow = 1))
  
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_TS-HK_CorePro_style2_20210616",
                       width = 6, height = 10, device = "png")  
  
```

*Clustered vs Dispersed*
```{r}
  unique(tmp$name) %>% as.character() %>% sort
  tmp %>%
    dplyr::filter(str_detect(name, "Clustered_enhancer|Dispersed_enhancer|^Enhancer_STARRseq$")) %>%
    dplyr::mutate(name = str_replace_all(name, c("_enhancer" = "", 
                                                 "Enhancer_STARRseq" = "All")),
                  name = factor(name, levels = c("Clustered", "Dispersed", "All"))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      guides(fill = guide_legend(title = NULL, nrow = 1))
  
  toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Clustered_Dispersed_Enhancer_style2_20210616",
                       width = 8, height = 10)  
  
```

### .6 Num of TFs (barplot, pitplot, and upset plot)
#### . Four types of Enhancers.
Enhancers identified or predicted by different methods.

```{r}
  rm(list = ls())
  library(UpSetR)
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz") %>%
    dplyr::filter(Methy == "Col", 
                  str_detect(annotation, "(Enhancer_)"))
  # unique(results_sig$annotation)
  metaInfo <- fread("./refgenome/GSE60143_RAW_DAPSeq/TableS1D.tsv") %>%
    dplyr::rename(geneID = AT_ID)
    
# prepare table for Upset
  df <- results_sig %>%
    dplyr::mutate(Type = str_replace(Type, "\\*+", ""), 
                  fold_type = ifelse(Type == "ns", "ns", fold_type), 
                  value = str_c(annotation, fold_type, sep = "-")) %>%
    pivot_wider(id_cols = c(Family, Protein), names_from = annotation, values_from = value) %>%
    pivot_longer(cols = matches("Enhancer"), names_to = "CRE", values_to = "enrich_type") %>%
    dplyr::mutate(value = 1) %>%
    dplyr::select(-CRE) %>%
    pivot_wider(names_from = enrich_type, values_from = value, values_fill = 0)

  
# plot
  # bar plot (Number)
  SetSize <- map_dbl(df[,3:ncol(df)], function(x){sum(x)}) %>% sort(decreasing = T)
  data.frame(Enhancer = names(SetSize), 
             Number = SetSize) %>%
    dplyr::mutate(Method = str_replace_all(Enhancer, "(Enhancer_.+)-.+", "\\1"), 
                  Type = str_replace_all(Enhancer, "(Enhancer_.+)-(.+)", "\\2"),
                  Type = factor(Type, levels = c("Enriched", "ns", "Depleted"))) %>%
    dplyr::group_by(Method) %>%
    dplyr::mutate(Total = sum(Number),
                  Percent = percent(Number / Total, accuracy = 0.01),
                  label = str_c(Number, "\n", Percent))  %>%
    ggplot(aes(y = Number, x = Type, fill = Type)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ Method) + 
      scale_fill_npg() + 
      theme(# legend.position = c(0.92, 0.88),
            legend.position = "none",
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            legend.key.size = unit(0.3, units = "cm"),
            axis.title.x = element_blank(),
            panel.spacing.x = unit(0.08, "cm")) + 
      geom_text(aes(label = Number), nudge_y = 8, size = 5/.pt)
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Percent_of_TFfamily_members_enriched_depleted_in_Four_kinds_of_enhancers_total_20210703",
  #                      width = 10, height = 6)

## Upset
  # prepare parameters
  SetColor <- rep("grey60", length(SetSize))
  SetColor[str_detect(names(SetSize), "Enhancer_STARR")] <- pal_npg()(9)[1]
  SetColor[str_detect(names(SetSize), "Enhancer_Zhu")] <- pal_npg()(9)[2]
  SetColor[str_detect(names(SetSize), "Enhancer_Wang")] <- pal_npg()(9)[5]
  SetColor[str_detect(names(SetSize), "Enhancer_Meng")] <- pal_npg()(9)[3]

  # png("./R7.TFs/Upset_TFs_enriched_depleted_Four_Types_enhancers_20210703.png",
  #     width = 16, height = 12, units = "cm", res = 600, bg = "transparent")
    UpSetR::upset(data = as.data.frame(df),
                  nsets = NA,
                  sets = colnames(df)[3:ncol(df)],
                  order.by = c("freq"),
                  decreasing = c(TRUE,T),
                  # queries = list(list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Enriched"),
                  #             color = pal_npg()(9)[1], active = T),
                  #                list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Depleted"),
                  #             color = pal_npg()(9)[5], active = T),
                  #                list(query = intersects, params = list("Enhancer_Enriched", "Promoter_ns"),
                  #             color = pal_npg()(9)[5], active = T),
                  #                list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Enriched"),
                  #             color = "gold", active = T),
                  #                list(query = intersects, params = list("Enhancer_ns", "Promoter_Enriched"),
                  #             color = "gold", active = T),
                  #               list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Depleted"),
                  #             color = pal_npg()(9)[4], active = T),
                  #                list(query = intersects, params = list("Enhancer_ns", "Promoter_ns"),
                  #             color = pal_npg()(9)[6], active = T)
                  #             ),
                  main.bar.color = "grey60",
                  sets.bar.color = SetColor
                  )
  # dev.off()

## Table of TFs only enriched or depleted in STARR-seq.
  df$Only_EnSTARR_enriched <- "N"
  df$Only_EnSTARR_enriched[rowSums(df[, str_detect(colnames(df), "Enhancer.+Enriched")]) == 1 &
                           df$`Enhancer_STARRseq-Enriched` == 1] <- "Y"
  df$Only_EnSTARR_depleted <- "N"
  df$Only_EnSTARR_depleted[rowSums(df[, str_detect(colnames(df), "Enhancer.+Depleted")]) == 1 &
                           df$`Enhancer_STARRseq-Depleted` == 1] <- "Y"
  table(df$Only_EnSTARR_enriched)
  table(df$Only_EnSTARR_depleted)

  tmp <- df %>%
    dplyr::mutate(Type = str_c(Only_EnSTARR_enriched, Only_EnSTARR_depleted, sep = "-"),
                  Type = str_replace_all(Type, c("Y-N" = "Enhancer_STARRseq_enriched",
                                                 "N-Y" = "Enhancer_STARRseq_depleted",
                                                 "N-N" = "Other"))) %>%
    group_by(Family, Type) %>%
    count() %>%
    group_by(Family) %>%
    dplyr::mutate(Total = sum(n)) %>%
    arrange(desc(Total)) %>%
    dplyr::mutate(Family = factor(Family, levels = unique(.$Family)))
  tmp %>%
    ggplot(aes(x = n, y = Family, fill = Type)) +
      geom_col() +
      toolkit_tyj$theme_my +
      scale_fill_manual(values = c(pal_npg()(9)[c(3,1)], "grey65")) +
      theme(legend.position = c(0.69, 0.94),
          legend.key.size = unit(0.3, units = "cm"),
          legend.title = element_blank()) + 
      xlab("Number of TFs") + ylab("Protein family")

  # toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Percent_TFs_groups_enriched_Enhancer_STARRseq_20210704",
  #      width = 7.5, height = 10)

# write to files
  # fwrite(df, file = "./R7.TFs/TFs_enriched_depleted_Four_Types_enhancers_20210703.csv")
  # 
  # tmp %>%
  #   pivot_wider(names_from = Type, values_from = n, values_fill = 0) %>%
  #   arrange(desc(Total)) %>%
  #   fwrite(., file = "./R7.TFs/TFs_enriched_depleted_Four_Types_enhancers_Summary_20210703.csv")

```


#### . KmeansCluster
GO of TFs enriched in each Kmeans clusters.

```{r}
  rm(list = ls())
  library(UpSetR)
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz") %>%
    dplyr::filter(Methy == "Col", 
                  str_detect(annotation, "(Kmeans)"))
  unique(results_sig$annotation)
  metaInfo <- fread("./refgenome/GSE60143_RAW_DAPSeq/TableS1D.tsv") %>%
    dplyr::rename(geneID = AT_ID)
    
# prepare table for Upset
  df <- results_sig %>%
    dplyr::mutate(Type = str_replace(Type, "\\*+", ""), 
                  fold_type = ifelse(Type == "ns", "ns", fold_type), 
                  value = str_c(annotation, fold_type, sep = "-")) %>%
    pivot_wider(id_cols = c(Family, Protein), names_from = annotation, values_from = value) %>%
    pivot_longer(cols = matches("Kmeans"), names_to = "CRE", values_to = "enrich_type") %>%
    dplyr::mutate(value = 1) %>%
    dplyr::select(-CRE) %>%
    pivot_wider(names_from = enrich_type, values_from = value, values_fill = 0)

  
# plot
  # bar plot
  SetSize <- map_dbl(df[,3:ncol(df)], function(x){sum(x)}) %>% sort(decreasing = T)
  
  data.frame(Enhancer = names(SetSize), 
             Number = SetSize) %>%
    dplyr::mutate(Method = str_replace_all(Enhancer, "(Kmeans.+)-.+", "\\1"), 
                  Method = str_replace_all(Method, "KmeansCluster", "C"),
                  Type = str_replace_all(Enhancer, "(Kmeans.+)-(.+)", "\\2"),
                  Type = factor(Type, levels = c("Enriched", "ns", "Depleted"))) %>%
    ggplot(aes(y = Number, x = Type, fill = Type)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ Method) + 
      scale_fill_npg() + 
      ylab("Number of TFs") +
      theme(
            # legend.position = c(0.92, 0.88),
            legend.position = "none",
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            legend.key.size = unit(0.3, units = "cm"),
            axis.title.x = element_blank()) + 
      geom_text(aes(label = Number), nudge_y = 8, size = 6/.pt)
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Four_KmeansClusters_total_20210703.png", 
  #                      width = 8, height = 8, device = "png")
  
  ## Upset
  # prepare parameters
  SetColor <- rep("grey60", length(SetSize))
  show_col(pal_npg()(9))
  SetColor[str_detect(names(SetSize), "(Cluster1)")] <- pal_npg()(9)[1]
  SetColor[str_detect(names(SetSize), "(Cluster2)")] <- pal_npg()(9)[5]
  SetColor[str_detect(names(SetSize), "(Cluster3)")] <- pal_npg()(9)[3]
  SetColor[str_detect(names(SetSize), "(Cluster4)")] <- pal_npg()(9)[4]
  
  # png("./R7.TFs/Upset_TFs_enriched_depleted_Four_KmeansClusters_20210703.png",
  #     width = 16, height = 12, units = "cm", res = 600, bg = "transparent")
    UpSetR::upset(data = as.data.frame(df), 
                  nsets = NA,
                  sets = colnames(df)[3:ncol(df)], 
                  order.by = c("freq"), 
                  decreasing = c(TRUE,T), 
                  # queries = list(list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Enriched"), 
                  #             color = pal_npg()(9)[1], active = T),
                  #                list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Depleted"), 
                  #             color = pal_npg()(9)[5], active = T),
                  #                list(query = intersects, params = list("Enhancer_Enriched", "Promoter_ns"), 
                  #             color = pal_npg()(9)[5], active = T),
                  #                list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Enriched"), 
                  #             color = "gold", active = T),
                  #                list(query = intersects, params = list("Enhancer_ns", "Promoter_Enriched"), 
                  #             color = "gold", active = T),
                  #               list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Depleted"), 
                  #             color = pal_npg()(9)[4], active = T),
                  #                list(query = intersects, params = list("Enhancer_ns", "Promoter_ns"), 
                  #             color = pal_npg()(9)[6], active = T)
                  #             ), 
                  main.bar.color = "grey60", 
                  sets.bar.color = SetColor
                  )
  # dev.off()
  
# TFs enriched in Cluster2 and cluster3.
  tmp <- df %>%
    dplyr::group_by(Family) %>%
    dplyr::mutate(TotalFamilySize = n()) %>%
    ungroup() %>%
    dplyr::filter(.$`KmeansCluster2-Enriched` == 1 | .$`KmeansCluster3-Enriched` == 1) %>%
    group_by(Family) %>%
    dplyr::mutate(NumberFamily = n(), 
                  Percent = NumberFamily / TotalFamilySize) %>%
    ungroup() %>%
    dplyr::mutate(Type = "Cluster2",
                  Type = ifelse(.$`KmeansCluster3-Enriched` == 1, "Cluster3", Type),
                  Type = ifelse(.$`KmeansCluster2-Enriched` == 1 & .$`KmeansCluster3-Enriched` == 1, "Both", Type),
                  )
  # fwrite(tmp, file = "./R7.TFs/TFs_enriched_in_KmeansCluster2_KmeansCluster3_20210704.csv", 
  #        col.names = T, row.names = T)

```

#### . KmeansCluster & promoter (Pie plot)
```{r}
 rm(list = ls())
  library(UpSetR)
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz") %>%
    dplyr::filter(Methy == "Col", 
                  str_detect(annotation, "Kmeans|^sCorePro$|Enhancer_STARRseq"))
  unique(results_sig$annotation)
  metaInfo <- fread("./refgenome/GSE60143_RAW_DAPSeq/TableS1D.tsv") %>%
    dplyr::rename(geneID = AT_ID)
    
# prepare table for Upset
  df <- results_sig %>%
    dplyr::mutate(Type = str_replace(Type, "\\*+", ""), 
                  fold_type = ifelse(Type == "ns", "ns", fold_type), 
                  value = str_c(annotation, fold_type, sep = "-")) %>%
    pivot_wider(id_cols = c(Family, Protein), names_from = annotation, values_from = value) %>%
    pivot_longer(cols = matches("Kmeans|STARR|CorePro"), names_to = "CRE", values_to = "enrich_type") %>%
    dplyr::mutate(value = 1) %>%
    dplyr::select(-CRE) %>%
    pivot_wider(names_from = enrich_type, values_from = value, values_fill = 0)

  
# plot
  # bar plot
  SetSize <- map_dbl(df[,3:ncol(df)], function(x){sum(x)}) %>% sort(decreasing = T)
  
  tmp <- data.frame(Enhancer = names(SetSize), 
             Number = SetSize) %>%
    dplyr::mutate(Method = str_replace_all(Enhancer, c("(Kmeans.+)-.+" = "\\1",
                                                       "KmeansCluster" = "C",
                                                       "sCorePro.+" = "Promoter",
                                                       "Enhancer.+" = "Enhancer")), 
                  Type = str_replace_all(Enhancer, ".+-(.+?)", "\\1"),
                  Type = factor(Type, levels = c("Enriched", "ns", "Depleted"))) %>%
    group_by(Method) %>%
    dplyr::mutate(Total = sum(Number),
                  Percent = scales::percent(Number/Total, accuracy = 0.01),
                  label = str_c(Number, "\n", Percent)) 
  
  # bar plot
  tmp %>%
    ggplot(aes(y = Number, x = Type, fill = Type)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ Method) + 
      scale_fill_manual(values = toolkit_tyj$cre_color_fill[c(2,4,1)]) +
      ylab("Number of TFs") +
      theme(# legend.position = c(0.92, 0.88),
            legend.position = "none",
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            legend.key.size = unit(0.3, units = "cm"),
            axis.title.x = element_blank(),
            panel.spacing.x = unit(0.08, "cm")) + 
      geom_text(aes(label = Number), nudge_y = 8, size = 5/.pt) + 
      scale_y_continuous(expand = expansion(mult = c(0, 0.08)))
  
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Percent_of_TFfamily_members_enriched_depleted_in_KmeansCluster_Promoter_20210703",
                       width = 8, height = 6)

  tmp %>%
    group_by(Method) %>%
    dplyr::arrange(Method, desc(Type)) %>%
    dplyr::mutate(PercentNum = Number/Total,
                  y_pos = cumsum(PercentNum) - PercentNum/2) %>%
    ggplot(aes(y = PercentNum, x = 0, fill = Type)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ Method) + 
      scale_fill_manual(values = toolkit_tyj$cre_color_fill[c(2,4,1)]) +
      ylab("Number of TFs") +
      theme(# legend.position = c(0.92, 0.88),
            legend.position = "bottom",
            legend.title = element_blank(),
            legend.key.size = unit(0.3, units = "cm"),
            axis.title = element_blank(),
            axis.text = element_blank(),
            axis.ticks = element_blank(),
            panel.spacing.x = unit(0.08, "cm"),
            legend.margin = margin(-10, 0, 0, 0)) + 
      geom_text(aes(label = Number, y = y_pos), size = 6/.pt) +
      # scale_y_continuous(expand = expansion(mult = c(0, 0.08))) +
      coord_polar(theta = "y")
  
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Percent_of_TFfamily_members_Pie_enriched_depleted_in_KmeansCluster_Promoter_20210703",
                       width = 16, height = 4)
```



#### . ActivityClusters


```{r}
  rm(list = ls())
  library(UpSetR)
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz") %>%
    dplyr::filter(Methy == "Col", 
                  str_detect(annotation, "(Enhancer_STARRseq|Activity)"))
  unique(results_sig$annotation)
  metaInfo <- fread("./refgenome/GSE60143_RAW_DAPSeq/TableS1D.tsv") %>%
    dplyr::rename(geneID = AT_ID)
    
# prepare table for Upset
  df <- results_sig %>%
    dplyr::mutate(Type = str_replace(Type, "\\*+", ""), 
                  fold_type = ifelse(Type == "ns", "ns", fold_type), 
                  value = str_c(annotation, fold_type, sep = "-")) %>%
    pivot_wider(id_cols = c(Family, Protein), names_from = annotation, values_from = value) %>%
    pivot_longer(cols = matches("(Enhancer|Activity)"), names_to = "CRE", values_to = "enrich_type") %>%
    dplyr::mutate(value = 1) %>%
    dplyr::select(-CRE) %>%
    pivot_wider(names_from = enrich_type, values_from = value, values_fill = 0)

  
# plot
  # bar plot (Number of TFs enriched/depleted in each groups)
  SetSize <- map_dbl(df[,3:ncol(df)], function(x){sum(x)}) %>% sort(decreasing = T)
  data.frame(Enhancer = names(SetSize), 
             Number = SetSize) %>%
    dplyr::mutate(Method = str_replace_all(Enhancer, "(.+)-.+", "\\1"), 
                  Type = str_replace_all(Enhancer, "(.+)-(.+)", "\\2"),
                  Type = factor(Type, levels = c("Enriched", "ns", "Depleted"))) %>%
    ggplot(aes(y = Number, x = Type, fill = Type)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      facet_grid(. ~ Method) + 
      scale_fill_npg() + 
      theme(legend.position = "none", 
            legend.key.size = unit(0.3, units = "cm")) + 
      geom_text(aes(label = Number), nudge_y = 8, size = 2.5)
  
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Enhancer_STARRseq_ActivityCluster_total_20210703.png", 
                       width = 14, height = 8, device = "png")
  
  ## Upset
  # prepare parameters
  SetColor <- rep("grey60", length(SetSize))
  # SetColor[str_detect(names(SetSize), "Enhancer_STARR")] <- pal_npg()(9)[1]
  # SetColor[str_detect(names(SetSize), "Enhancer_Zhu")] <- pal_npg()(9)[2]
  # SetColor[str_detect(names(SetSize), "Enhancer_Wang")] <- pal_npg()(9)[5]
  # SetColor[str_detect(names(SetSize), "Enhancer_Meng")] <- pal_npg()(9)[3]
  
  # png("./R7.TFs/Upset_TFs_enriched_depleted_Enhancer_STARRseq_ActivityCluster_20210703.png",
  #     width = 16, height = 12, units = "cm", res = 600, bg = "transparent")
    UpSetR::upset(data = as.data.frame(df), 
                  nsets = NA,
                  sets = colnames(df)[3:ncol(df)], 
                  order.by = c("freq"), 
                  decreasing = c(TRUE,T), 
                  # queries = list(list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Enriched"), 
                  #             color = pal_npg()(9)[1], active = T),
                  #                list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Depleted"), 
                  #             color = pal_npg()(9)[5], active = T),
                  #                list(query = intersects, params = list("Enhancer_Enriched", "Promoter_ns"), 
                  #             color = pal_npg()(9)[5], active = T),
                  #                list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Enriched"), 
                  #             color = "gold", active = T),
                  #                list(query = intersects, params = list("Enhancer_ns", "Promoter_Enriched"), 
                  #             color = "gold", active = T),
                  #               list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Depleted"), 
                  #             color = pal_npg()(9)[4], active = T),
                  #                list(query = intersects, params = list("Enhancer_ns", "Promoter_ns"), 
                  #             color = pal_npg()(9)[6], active = T)
                  #             ), 
                  main.bar.color = "grey60", 
                  sets.bar.color = SetColor
                  )
  # dev.off()
  
## Table of TFs only enriched or depleted in ActivityCluster5 (compare to other four ActivityClusteres)
  df$Only_A5_enriched <- "N"
  df$Only_A5_enriched[rowSums(df[, str_detect(colnames(df), "(Cluster).+Enriched")]) == 1 & 
                           df$`ActivityCluster5-Enriched` == 1] <- "Y"
  df$Only_A5_depleted <- "N"
  df$Only_A5_depleted[rowSums(df[, str_detect(colnames(df), "(Cluster).+Depleted")]) == 1 & 
                           df$`Only_A5_depleted-Depleted` == 1] <- "Y"
  table(df$Only_A5_enriched)
  table(df$Only_A5_depleted)
  
  df %>%
    dplyr::filter(Only_A5_enriched == "Y" | Only_A5_depleted == "Y")
    
  
fwrite(df, file = "./R7.TFs/TFs_enriched_depleted_Enhancer_STARRseq_ActivityCluster_20210703.csv")

## Number of TFs enriched/depledted in each activityCluster at family level.
tmp <- results_sig %>%
  dplyr::filter(padj_BH < 0.05) %>%
  group_by(Family, fold_type, annotation) %>%
  count() %>%
  pivot_wider(values_from = n, names_from = annotation)

tmp %>%
  fwrite("./R7.TFs/Number_of_TFs_enriched_dpleted_ActivityCluster_family_level_20210706.csv")

```




### .7 Pro vs En 
> only Col (did not amplification to discard DNA modification) DAPseq peaks
sEnhancer vs sCorePro (overlaped enhancers and promoters were not included in analysis.)

*sCorePro* vs *sEnhancer600bpCorePro*

#### . TOP diff families
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- read_csv("./R7.TFs/Number_of_Methy_unMethy_TFs_TFFamily_members_enriched_depleted_in_CREs_20210616.csv") %>%
    dplyr::filter(Methy == "Col")
  
  # order of TF families in plot
  FamilyOder <- df %>%
    dplyr::select(Family, Total) %>%
    na.omit() %>%
    unique() %>%
    arrange(desc(Total))
  
  # varied families (sEnhancer vs sCorePro)
  sigVaryTFfamily <- df %>%
    dplyr::mutate(log2EPratio = log2(sEnhancer600CorePro / sCorePro)) %>%
    dplyr::filter(abs(log2EPratio) > 1 & Total > 5) %>%
    arrange(desc(abs(log2EPratio))) %>%
    dplyr::select(Family) %>%
    unlist() %>%
    unique()
  
  # Top different familes En/Pro
  df %>%
    dplyr::select(which(str_detect(colnames(.), "(^Family|fold_type|sEnhancer600|Total|sCorePro)"))) %>%
    dplyr::filter(Family != "Total") %>%
    dplyr::filter(Family %in% sigVaryTFfamily) %>%
    pivot_longer(cols = matches("sEnhancer|sCorePro"), values_to = "Num_sig") %>%
    dplyr::mutate(Total = if_else(fold_type == "Depleted", -Total, Total), 
                  Num_sig = if_else(fold_type == "Depleted", -Num_sig, Num_sig), 
                  Family = factor(Family, levels = FamilyOder$Family), 
                  fold_type = factor(fold_type, levels = c("Enriched", "Depleted")), 
                  Num_sig = ifelse(abs(Num_sig) < 0.5, 0, Num_sig)) %>%
    dplyr::mutate(name = str_replace_all(name, c("s" = "", 
                                                 "\\d+.+" = ""))) %>%
    ggplot(aes(x = Family, y = Num_sig)) +
      geom_col(aes(x = Family, y = Total), fill = "grey60", alpha = 0.5) +
      geom_col(aes(fill = fold_type)) +
      toolkit_tyj$theme_my + 
      facet_grid(. ~ name) + 
      scale_fill_npg()+
      coord_flip() +
      scale_y_continuous(breaks = c(-40, -20, 0, 20, 40), 
                         labels = c(40, 20, 0, 20, 40)) +
      xlab("Protein family") + ylab("Number of protein family members") +
      theme(legend.position = "bottom", 
            legend.key.size = unit(0.2, units = "cm")) + 
      guides(fill = guide_legend(title = NULL, nrow = 1)) + 
      labs(caption = "sCorePro vs sEnhancer600bap away from CorePro")
  
    toolkit_tyj$SavePlot("./R7.TFs/Number_of_TFfamily_members_enriched_depleted_in_Promoter_Enhancer_TOP_style2",
                         width = 8, height = 8, device = "png")

```

#### . TFs fold ratio

```{r, eval=F, NOT used}
  # Caculate ratio between fold change of TFs overlap with enhancer and promoter is misleading!
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz") %>%
    dplyr::filter(Methy == "Col", 
                  str_detect(annotation, "(sEnhancer600CorePro|sCorePro)")) %>%
    dplyr::mutate(annotation = str_replace_all(annotation, c("^s" = "",
                                                             "(Enhancer).+" = "\\1")))
  TFls <- results_sig %>%
    dplyr::filter(padj_BH < 0.05, abs(l2fold) > 1) %>%
    dplyr::select(Protein) %>%
    unlist() %>%
    unique()  
    
  df <- results_sig %>%
    dplyr::filter(Protein %in% TFls, 
                  NumRegions > 200) %>%
    pivot_wider(id_cols = c(Family, Protein), names_from = annotation, values_from = fold) %>%
    dplyr::mutate(EPratio = Enhancer / CorePro,
                  l2EPratio = log2(Enhancer / CorePro), 
                  Type = ifelse(l2EPratio > 0, "Enhancer prefer", "Promoter prefer")) %>% 
    arrange(desc(abs(l2EPratio)))
    
  # TOP xx TFs which enriched in enhancer or promoter
  df %>%
    dplyr::mutate(Protein = factor(Protein, levels = .$Protein)) %>%
    head(200) %>%
    ggplot(aes(y = Protein, x = l2EPratio, fill = Type)) + 
      geom_vline(xintercept = c(-8, -4, 0, 4), color = "grey50", linetype  =2) +
      scale_x_continuous(breaks = c(-8, -4, 0, 4)) +
      geom_col() + 
      scale_fill_npg() + 
      # facet_grid(Family ~ ., space = "free", scales = "free_y") +
      toolkit_tyj$theme_my +
      theme(legend.position = c(0.22, 0.97), 
            legend.key.size = unit(0.3, units = "cm")) + 
      guides(fill = guide_legend(title = NULL)) + 
      labs(caption = "sCorePro vs sEnhancer (600bp away from CorePro)")
      
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Top_TFs_EnPro_fold_ratio_20210617", 
                       width = 10, height = 12, device = "png")

  # Top xxx TFs prefer enriched
  df %>%
    head(200) %>%
    pivot_longer(cols = c(Enhancer, CorePro), names_to = "CRE", values_to = "fold") %>%
    dplyr::mutate(fold = ifelse(CRE == "CorePro", 0 - fold, fold)) %>%
    ggplot(aes(x = fold, y = Protein, fill = CRE)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      scale_fill_npg() +
      theme(legend.position = c(0.8, 0.05), 
            legend.key.size = unit(0.3, units = "cm"), 
            axis.text.y = element_text(size = 3.5)) + 
      xlab("Enrichment") + 
      # facet_wrap(facets = "Family", scales = "free_y") +
      geom_vline(xintercept = c(-1, 1), color = "grey50", linetype = 2) + 
      labs(caption = "sCorePro vs sEnhancer (600bp away from CorePro)")
  
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Top_all_TFs_fold_Pro_En_20210616", 
                       width = 12, height = 18, device = "png")
```


#### . Upset plot and barplot.
Classify TFs into different types based on enriched in pro/enhancer states.(Y+Y, Y+N, N+Y, N+N).
Upsetplot?
TFs and TFfamily level.

```{r}
  rm(list = ls())
  library(UpSetR)
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R7.TFs/All_GAT_100000_results_padjBH_siglable_20210609.csv.gz") %>%
    dplyr::filter(Methy == "Col", 
                  str_detect(annotation, "(sEnhancer600|sCorePro)"))
  metaInfo <- fread("./refgenome/GSE60143_RAW_DAPSeq/TableS1D.tsv") %>%
    dplyr::rename(geneID = AT_ID)
    
  # prepare table for Upset
  df <- results_sig %>%
    dplyr::mutate(Type = str_replace(Type, "\\*+", ""), 
                  annotation = str_replace_all(annotation, c("sCorePro" = "Promoter",
                                                             "sEnhancer.*" = "Enhancer"))) %>%
    pivot_wider(id_cols = c(Family, Protein), names_from = annotation, values_from = Type) %>%
    dplyr::mutate(Promoter = str_c("Promoter", Promoter, sep = "_"), 
                  Enhancer = str_c("Enhancer", Enhancer, sep = "_")) %>%
    pivot_longer(cols = Promoter:Enhancer, names_to = "CRE", values_to = "enrich_type") %>%
    dplyr::mutate(value = 1, 
                  enrich_type = str_replace(enrich_type, " ", "")) %>%
    dplyr::select(-CRE) %>%
    pivot_wider(names_from = enrich_type, values_from = value, values_fill = 0)
  
  # plot
  # png("./R7.TFs/Upset_TFs_enriched_depleted_CorePro_Enhancer600CorePro_20210703.png", 
  #     width = 12, height = 8, units = "cm", res = 600, bg = "transparent")
    UpSetR::upset(data = as.data.frame(df), 
                  nsets = NA,
                  sets = colnames(df)[3:ncol(df)], 
                  order.by = c("freq"), 
                  decreasing = c(TRUE,T), 
                  queries = list(list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Enriched"), 
                              color = pal_npg()(9)[1], active = T),
                                 list(query = intersects, params = list("Enhancer_Enriched", "Promoter_Depleted"), 
                              color = pal_npg()(9)[5], active = T),
                                 list(query = intersects, params = list("Enhancer_Enriched", "Promoter_ns"), 
                              color = pal_npg()(9)[5], active = T),
                                 list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Enriched"), 
                              color = "gold", active = T),
                                 list(query = intersects, params = list("Enhancer_ns", "Promoter_Enriched"), 
                              color = "gold", active = T),
                                list(query = intersects, params = list("Enhancer_Depleted", "Promoter_Depleted"), 
                              color = pal_npg()(9)[4], active = T),
                                 list(query = intersects, params = list("Enhancer_ns", "Promoter_ns"), 
                              color = pal_npg()(9)[6], active = T)
                              ), 
                  main.bar.color = "grey60", 
                  sets.bar.color = pal_npg()(9)[c(1,6,3,6,1,3)])
  # dev.off()
  
  ## summary of TFs
    results_sig %>%
      group_by(annotation, fold_type, Type) %>%
      count() %>%
      arrange(annotation, fold_type, Type) %>%
      View()
    
    df %>%
      dplyr::select(matches("(Promoter|Enhancer)")) %>%
      map_dbl(., sum) %>%
      data.frame(id = names(.), number = .) %>%
      dplyr::mutate(CRE = str_replace_all(id, "(.+)_.+", "\\1"),
                    Type = str_replace_all(id, "(.+)_(.+)", "\\2"), 
                    Type = factor(Type, levels = c("Enriched", "ns", "Depleted"))) %>%
      ggplot(aes(x = Type, y = number, fill = Type)) + 
        geom_col() +
        toolkit_tyj$theme_my + 
        facet_grid(. ~ CRE) + 
        scale_fill_npg() + 
        geom_text(aes(label = number), nudge_y = 8, size = 2) + 
        theme(legend.position = "right", legend.key.size = unit(0.3, units = "cm"))
    
    toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Number_of_All_TFs_enriched_depleted_in_Promoter_Enhancer_20210704",
                         width = 8, height = 6, device = "png")
    
  # fwrite(df, file = "./R7.TFs/Upset_TFs_enriched_depleted_CorePro_Enhancer300bpCorePro_20210703.csv")

```

##### . Count in each family

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./R7.TFs/Upset_TFs_enriched_depleted_CorePro_Enhancer300bpCorePro_20210703.csv")
  
## four types
  tmp <- df %>%
    group_by(Family, Type1) %>%
    count() %>%
    group_by(Family) %>%
    dplyr::mutate(Type1 = str_replace_all(Type1, c("_enriched" = "",
                                                   "Only_enhancer" = "Enhancer",
                                                   "Only_promoter" = "Promoter")),
                  Sum = sum(n), 
                  Percent = n / Sum) %>%
    arrange(desc(Sum)) %>%
    dplyr::mutate(Family = factor(Family, levels = unique(.$Family)), 
                  Type1 = factor(Type1, levels = c("Both", 
                                                   "Enhancer", 
                                                   "Promoter",
                                                   "None")))
  tmp %>%
    ggplot(aes(y = Family, x = n, fill = Type1)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      # scale_fill_npg() + 
      # scale_fill_manual(values = c(pal_npg()(9)[c(8, 5, 2)], "grey60")) +
      scale_fill_manual(values = c(brewer.pal(9, "Set1")[c(4,1,5)], "grey60")) +
      theme(#axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
            legend.position = "bottom", 
            legend.box.just = "left",
            legend.title = element_blank(),
            legend.key.size = unit(0.2, units = "cm"),
            panel.spacing.x = unit(0.1, "cm"),
            legend.margin = margin(-10, 0, 0, -30)) + 
      scale_x_continuous(breaks = seq(0, 70, by = 10)) +
      xlab("Protein family") + ylab("Number of protein family members")
  
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Number_TFs_4groups_enriched_CorePro_sEnhancer600bpCorePro",
                       width = 5, height = 9.4)
  
  # table
  tmp %>%
    dplyr::mutate(label = str_c(n, "/", Sum, "(", round(Percent, digits = 4), ")", sep = "")) %>%
    pivot_wider(id_cols = c(Family, Sum), names_from = Type1, values_from = Percent, values_fill = 0) %>%
    fwrite(., file = "./R7.TFs/Number_TFs_4groups_enriched_CorePro_sEnhancer600bpCorePro_20210703.csv", 
           quote = F)

## eight types (too complicated)
  tmp <- df %>%
    dplyr::mutate(Type = str_replace_all(Type, "\n", "-")) %>%
    group_by(Family, Type) %>%
    count() %>%
    group_by(Family) %>%
    dplyr::mutate(Sum = sum(n), 
                  Percent = n / Sum) %>%
    arrange(desc(Sum)) %>%
    dplyr::mutate(Family = factor(Family, levels = unique(.$Family)))
  
  tmp %>%
    ggplot(aes(y = Family, x = n, fill = Type)) + 
      geom_col() + 
      toolkit_tyj$theme_my + 
      # scale_fill_npg() +
      scale_fill_manual(values = c(pal_npg()(9)[c(4, 2, 3, 1, 8, 5, 6, 7)], "grey60")) +
      theme(legend.position = c(0.75, 0.83), 
            legend.key.size = unit(0.3, "cm"), 
            legend.title = element_blank()) + 
      xlab("Number") + 
      ylab("Family")
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/Number_TFs_9groups_enriched_CorePro_sEnhancer600bpCorePro", 
  #                      width = 10, height = 10, device = "png")

  # table
  tmp %>%
    dplyr::mutate(label = str_c(n, "/", Sum, "(", round(Percent, digits = 4), ")", sep = "")) %>%
    pivot_wider(id_cols = c(Family, Sum), names_from = Type, values_from = Percent, values_fill = 0) %>%
    fwrite(., file = "./R7.TFs/Number_TFs_9groups_enriched_CorePro_sEnhancer600bpCorePro_20210703.csv", 
           quote = F)
```


#### . PPI of each kind of TFs

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("./R7.TFs/Upset_TFs_enriched_depleted_CorePro_Enhancer600bpCorePro_20210703.csv")
  PPiNum <- fread("./refgenome/Number_Protein_protein_interaction_BioGRID_DAPseq_tidy_2021617.csv.gz") %>%
    dplyr::select(geneID, Number_PPI, Family, Protein)

# classify Proteins based on enriched/depleted profiles in sEnhancer and sCorePro.
  names(df)
  namesTMP <- colnames(df)
  df$Type <- apply(df, 1, function(x){
    namesTMP[which(x == 1)] %>%
      sort() %>%
      str_c(collapse = "\n")
  })
  table(df$Type) %>% sort(decreasing = T)
  rm(namesTMP)
  df$geneID <- df$Protein
  
# merge based on geneID and Protein name.
  df <- df %>%
    left_join(., dplyr::select(PPiNum, geneID, Number_PPI)) %>%
    left_join(., dplyr::select(PPiNum, Protein, Number_PPI) %>%
                dplyr::rename(Number_PPI1 = Number_PPI)) %>%
    dplyr::mutate(Number_PPI =  ifelse(is.na(Number_PPI), Number_PPI1, Number_PPI)) %>%
    dplyr::select(-Number_PPI1)
  
### plot
  # all nine types
  dfLab <- df %>%
    dplyr::filter(!is.na(Number_PPI)) %>%
    group_by(Type) %>%
    dplyr::summarise(Num = n(),
                     Median = median(Number_PPI), 
                     Mean =round( mean(Number_PPI), 2)) %>%
    arrange(desc(Num)) %>%
    dplyr::mutate(label = str_c("N=", Num, "\nMedian=", Median, "\nMean=", Mean))
    
  df %>%
    dplyr::mutate(Type = factor(Type, levels = dfLab$Type)) %>%
    ggplot(aes(x = Type, y = Number_PPI)) + 
      geom_beeswarm(size = 0.5, color = "grey45") +
      geom_boxplot(width = 0.2, fill = "transparent", outlier.alpha = 0) + 
      geom_violin(fill = "transparent") +
      toolkit_tyj$theme_my + 
      scale_y_log10() + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      geom_text(data = dfLab, aes(x = Type, y = 280, label = label),
                color = "grey35", size = 2) + 
      coord_cartesian(ylim = c(1, 350)) + 
      ggtitle("Number of PPI for TFs enriched or depleted in enhancers and CorePro")
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/PPI_of_TFs_enriched_depleted_in_CorePro_sEnhancer_Nine_types",
  #                      width = 16, height = 8, device = "png")
  
  # Two types: Enhancer_enriched(Pro_depleted/ns), Pro_enriched (Enhancer_depleted/ns).
  df$Type1 <- df$Type %>%
    str_replace_all(., c("Enhancer_Enriched\nPromoter_Enriched" = "Both_enriched",
                         "Enhancer_Enriched\nPromoter_Depleted" = "Only_enhancer_enriched", 
                         "Enhancer_Enriched\nPromoter_ns" = "Only_enhancer_enriched", 
                         "Enhancer_Depleted\nPromoter_Enriched" = "Only_promoter_enriched",
                         "Enhancer_ns\nPromoter_Enriched" = "Only_promoter_enriched", 
                         ".+\n.+" = "None_enriched"))
  
  dfLab <- df %>%
    dplyr::filter(!is.na(Number_PPI)) %>%
    group_by(Type1) %>%
    dplyr::summarise(Num = n(),
                     Median = median(Number_PPI), 
                     Mean =round( mean(Number_PPI), 2)) %>%
    arrange(desc(Num)) %>%
    dplyr::mutate(label = str_c("N=", Num, "\nMedian=", Median, "\nMean=", Mean))
  
  df %>%
    dplyr::mutate(Type1 = factor(Type1, levels = dfLab$Type1)) %>%
    ggplot(aes(x = Type1, y = Number_PPI)) + 
      geom_beeswarm(size = 0.5, color = "grey45") +
      geom_boxplot(width = 0.2, fill = "transparent", outlier.alpha = 0) + 
      geom_violin(fill = "transparent") +
      toolkit_tyj$theme_my + 
      scale_y_log10() + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      geom_text(data = dfLab, aes(x = Type1, y = 280, label = label),
                color = "grey35", size = 2) + 
      coord_cartesian(ylim = c(1, 450)) + 
      ggtitle("Number of PPI for TFs enriched in enhancers and CorePro") + 
      geom_signif(comparisons = list(c("Only_Enhancer_Enriched", "Only_Promoter_Enriched")), 
                  test = "wilcox.test", textsize = 2, color = "grey35")
  
  toolkit_tyj$SavePlot(filename_prefix = "./R7.TFs/PPI_of_TFs_enriched_Enhancer_CorePro_20210703",
                       width = 12, height = 8, device = "png")
  
# write to file
  # fwrite(df, file = "./R7.TFs/Upset_TFs_enriched_depleted_CorePro_Enhancer600bpCorePro_labeled_20210703.csv")
```

### .x Top TFs enriched in enhancers.

Illustrate top enriched or depleted TFs of each CRE with loliplot.
How to subset TFs in illustration?

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
results_sig <- fread("./R7.TFs/All_GAT_200000_results_padjBH_siglable_20210325.csv.gz")
results_sig %>%
  ggplot(aes(x = l2fold)) + 
    geom_histogram(fill = "grey45", color = "white") + 
    toolkit_tyj$theme_my + 
    ggtitle("log2(FC) of overlap between DAP peaks and CREs") + 
    xlab(expression(log[2](FC)))

## only cSTARR-seq
df <- results_sig %>%
  dplyr::filter(str_detect(annotation, "cSTARR-seq"), 
                Methy == "Col", ) %>%
  group_by(annotation) %>%
  arrange(annotation, Family)
  
```


## 7.2 Motifs enriched in CREs.

### .1 collect Motifs



## 7.3 cooccurance of some motifs? Grammer of TFBS? in Pro and En?




# 8. Conservation.

## 8.0 Informations

1.  *macroevolution*: `CNS: constrained noncoding sequences`; two index: `PhastCons`: PhastCons score is the score from 0 to 1 to show the `conservation level`.
    but it is not a linear system, most part of genome are 0 score or even not have score at all.
    Higher value denote higher conservation level.
    `PhyloP`: abs(phyloP) is the `-log(p-value)` under a null hypothesis of neutral evolution, and a negative sign indicates faster-than expected evolution, while positive values imply conservation.
    phyloP scores measure evolutionary conservation at individual alignment sites.
    Interpretations of the scores are compared to the evolution that is expected under neutral drift.Positive scores --- Measure conservation, which is slower evolution than expected, at sites that are predicted to be conserved.Negative scores --- Measure acceleration, which is faster evolution than expected, at sites that are predicted to be fast-evolving.
    PhastCons and PhyloP score of A.t (and other 62 plants) can be download from `PlantRegMap` which was maintained by Gao-lab of Peking university (<http://plantregmap.gao-lab.org/>).

2.  *microevolution*: allele frequency variation in population of same species.
    Genotype and meta info of 1135 Arabidopsis.thaliana accessions were download from database (<https://1001genomes.org/data/GMI-MPI/releases/v3.1/>).
    Index: Fst, pi, CLR, omega, TajimaD.

## 8.1 PhastCons

Two kinds of dataset were provided: 
  1. Conserved elements:`ConservedElements_CE_Ath.gtf.gz` (predicted based on PhastCons) and
  2. PhaseCons score, `Ath_PhastCons.temp.sort.bedGraph`.

### 8.1.1 GAT Conserved elements

##### .1 prepare .bed
```{r}
###  Conserved Elements
  source("D:/SUST/code/TanYongjun_code.R")
  ceAt <- fread("./refgenome/PhastCons_PhyloP_GaoLab/ConservedElements_CE_Ath.gtf.gz")
  hist(ceAt$V5 - ceAt$V4 + 1, main = "Width of Conserved elements")
  quantile(ceAt$V5 - ceAt$V4 + 1, probs = seq(0, 1, 0.1))
  sum(ceAt$V5 - ceAt$V4) / 1e6
  ceAt <- ceAt  %>%
    dplyr::select(V1, V4, V5) %>%
    dplyr::mutate(V1 = str_replace_all(V1, "Chr", "chr")) 
  
  # # files for GAT
  # ceAt %>%
  #   fwrite("./refgenome/PhastCons_PhyloP_GaoLab/ConservedElements_CE_Ath.bed.gz", 
  #          row.names = F, col.names = F, sep = "\t")

### CREs
  cre_all <- fread("./R7.TFs/bed_files/at_CREs_all_20210517.bed.gz") %>%
    rbind(., fread("./R7.TFs/bed_files/Enhancers_Meng2021_20210630.bed.gz")) %>%
    dplyr::filter(str_detect(V4, "^Enhancer.*")) %>%
    dplyr::mutate(V4 = str_replace_all(V4, "^Enhancer$", "Enhancer_STARRseq"))
  
  table(cre_all$V4, cre_all$V1)
  cre_all %>%
    split(., .$V4) %>%
    map_dfr(., function(x){quantile(x$V3 - x$V2, probs = seq(0, 1, 0.1))}, .id = "id")
  quantile(cre_all$V3 - cre_all$V2 + 1, probs = seq(0, 1, 0.1))
  # cre_all %>%
  #   fwrite("./R8.conservation/Enhancers_four_methods_for_GAT_20210630.bed.gz", 
  #          row.names = F, col.names = F, sep = "\t")

```

##### .2 run GAT
`/scratch/2021-06-25/bioTanyj/At_enhancer/scripts/8.1.8.conservedElements_overlap_GAT.R`
`/scratch/2021-06-25/bioTanyj/At_enhancer/scripts/8.1.8.conservedElements_overlap_GAT.sh`

##### .3 collect results

```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  results_sig <- fread("./R8.conservation/ConservedElements_ol_GAT_results_all_20210630.tsv")
  
  results_sig %>%
    dplyr::mutate(percent_overlap_size_annotation = percent_overlap_size_annotation / 100,
                  annotation = factor(annotation, levels = c("Enhancer_STARRseq", 
                                                             "Enhancer_Zhu2015", 
                                                             "Enhancer_Wang2019",
                                                             "Enhancer_Meng2021")),
                  qvalue = p.adjust(pvalue, method = "BH"),
                  label = str_c(scales::percent(percent_overlap_size_annotation, accuracy = 0.01),
                                "\nFC=", format(fold, digits = 2), "\n", toolkit_tyj$returnAsterisk(qvalue))) %>%
    ggplot(aes(x = annotation, y = percent_overlap_size_annotation)) + 
      geom_col(fill = "grey65") + 
      toolkit_tyj$theme_my + 
      theme(axis.title.x = element_blank(),
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      scale_y_continuous(labels = scales::percent) + 
      geom_text(aes(y = percent_overlap_size_annotation / 2, label = label), size = 5/.pt) + 
      ylab("Enhancers overlaped with conserved elements")
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/ConservedElements_ol_GAT_results_BarPlot_20210514",
  #                      width = 6, height = 6)
  NA
```




### 8.1.2 score
### .Step1, calculate PhastCons in beds

1.  convert bedgraph file to .bw file.

2.  generate bed files (all features were cut into regions near 600bp using `tile()`)

3.  `with enhancer`: feature regions hit by enhancer peak; without enhancer: feature regions not hit by enhancer peak.


```{r prepare bed files for each genomic regions, eval=F}

  source("D:/SUST/code/TanYongjun_code.R")

# enhancers
  enhancer <- fread("Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans_4&7_RPKM_20210321.csv.gz") %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment <= 1.3))
  table(enhancer$method)
  
# genome regions (cut into ~ 600 bins)
  library(TxDb.Athaliana.BioMart.plantsmart28)
  txdb <- TxDb.Athaliana.BioMart.plantsmart28
  
  intron_all <- intronsByTranscript(txdb, use.names = T) %>% unlist() %>% IRanges::tile(width = 600) %>% unlist()
  exon_all <- cdsBy(txdb, by = "gene") %>% unlist() %>% IRanges::tile(width = 600) %>% unlist()
  five_utr <- fiveUTRsByTranscript(txdb, use.names = T) %>% unlist() %>% IRanges::tile(width = 600) %>% unlist()
  three_utr <- threeUTRsByTranscript(txdb, use.names = T) %>% unlist() %>% IRanges::tile(width = 600) %>% unlist()
  pro_core <- promoters(txdb, upstream = 50, downstream = 50, use.names = T) %>% IRanges::tile(width = 600) %>% unlist()
  pro_enlarge <- promoters(txdb, upstream = 500, downstream = 100, use.names = T) %>% IRanges::tile(width = 600) %>% unlist()
  promoter <- promoters(txdb, upstream = 200, downstream = 50, use.names = T) %>% IRanges::tile(width = 600) %>% unlist()
  gene_all <- genes(txdb) %>% IRanges::tile(width = 600) %>% unlist()
  intergenic <- IRanges::gaps(gene_all, start = 1L, end = seqlengths(gene_all)) %>%
    GenomicRanges::reduce() %>%
    GenomicRanges::setdiff(., intron_all, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., exon_all, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., five_utr, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., three_utr, ignore.strand=TRUE) %>%
    GenomicRanges::setdiff(., pro_enlarge, ignore.strand=TRUE) %>% IRanges::tile(width = 600) %>% unlist()
  
  genomic_regions <- list(Intron = intron_all,
                          Exon = exon_all,
                          FiveUTR = five_utr,
                          ThreeUTR = three_utr,
                          CorePro = pro_core,
                          ProEnlarge = pro_enlarge,
                          Promoter = promoter,
                          Interneric = intergenic,
                          Gene = gene_all)
  
### two set of files containing genomic features overlapped by the center of STARR-seq enhancer.
  cSTARR <- enhancer %>%
    dplyr::filter(method == "cSTARR-seq")
  cSTARRgr <- makeGRangesFromDataFrame(cSTARR) %>%
    GenomicRanges::resize(width = 1, fix = "start")
  
  genomic_regions_STARR <- list()
  genomic_regions_noSTARR <- list()
  for (i in 1:length(genomic_regions)) {
    id <- names(genomic_regions)[i]
    region_tmp <- genomic_regions[[id]]
    ol <- findOverlaps(region_tmp, cSTARRgr)
    STARR_tmp <- region_tmp[unique(ol@from)]
    noSTARR_tmp <- region_tmp[-unique(ol@from)]
    
    genomic_regions_STARR <- c(genomic_regions_STARR, list(STARR_tmp))
    names(genomic_regions_STARR)[length(genomic_regions_STARR)] <- paste(id, "_STARRseq", sep = "")
    genomic_regions_noSTARR <- c(genomic_regions_noSTARR, list(noSTARR_tmp))
    names(genomic_regions_noSTARR)[length(genomic_regions_noSTARR)] <- paste(id, "_noSTARRseq", sep = "")
  }

#### write to bed file
  # genomic regions
  for (i in 1:length(genomic_regions)) {
    genomic_regions[[i]] %>%
      as.data.frame(row.names = NULL) %>%
      dplyr::mutate(seqnames = str_c("Chr", as.character(seqnames), sep = "")) %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(start = format(start, scientific = F), 
                    end = format(end, scientific = F), 
                    ) %>%
      fwrite(file = paste("./R8.conservation/bed_files/",
             names(genomic_regions)[i], ".bed", sep = ""),
             quote = F, row.names = F, col.names = F, sep = "\t")
    
    genomic_regions_STARR[[i]] %>%
      as.data.frame(row.names = NULL) %>%
      dplyr::mutate(seqnames = str_c("Chr", as.character(seqnames), sep = "")) %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(start = format(start, scientific = F), 
                    end = format(end, scientific = F)) %>%
      fwrite(file = paste("./R8.conservation/bed_files/",
             names(genomic_regions_STARR)[i], ".bed", sep = ""),
             quote = F, row.names = F, col.names = F, sep = "\t")
    
    genomic_regions_noSTARR[[i]] %>%
      as.data.frame(row.names = NULL) %>%
      dplyr::mutate(seqnames = str_c("Chr", as.character(seqnames), sep = "")) %>%
      dplyr::select(1:3) %>%
      dplyr::mutate(start = format(start, scientific = F), 
                    end = format(end, scientific = F)) %>%
      fwrite(file = paste("./R8.conservation/bed_files/",
             names(genomic_regions_noSTARR)[i], ".bed", sep = ""),
             quote = F, row.names = F, col.names = F, sep = "\t")
  }

  # CREs (discard Promoters (core promoter) in CRE table.)
  table(enhancer$method)
  for (i in unique(enhancer$method)) {
    if (i != "Promoter") {
        enhancer %>%
          dplyr::filter(method == i) %>%
          dplyr::mutate(start = format(startBK, scientific = F), 
                end = format(endBK, scientific = F)) %>%
          dplyr::select(seqnames, start, end) %>%
          dplyr::mutate(seqnames = str_c("Chr", as.character(seqnames), sep = "")) %>%
          fwrite(file = paste("./R8.conservation/bed_files/",
                 i, ".bed", sep = ""),
                 quote = F, row.names = F, col.names = F, sep = "\t")
    }
  }
  
```

3.  calculate mean score of each CRE following code used in 5.2.1.        
`/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/8.1.1_PhastConst.R`         
`/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/8.1.1_PhastConst.sh`        

### .Step2, Comparison and plot
Promoter: c(-200, 50)
CorePro: c(-50, 50)
ProEnlarge: c(-500, 100)


*202105* all 7000 cSTARRseq enhancer
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(tidyverse)

# laod PhaseCons score and get the true region width. (CREs located on chrC/M were discared in Step1)
phastConsAt <- fread("./R8.conservation/PhastCons_score_matrix_all_CRE_20210326.csv.gz") %>%
  dplyr::rename(PhastCons = score) %>%
  dplyr::filter(str_detect(id, "(Promoter|Random|iSTARR-seq|PAS)", negate = T)) %>%
  dplyr::mutate(id = str_replace_all(id, "cSTARR-seq", "Enhancer")) %>%
  dplyr::mutate(Features = str_replace_all(id, "(_noSTARRseq|_STARRseq)", ""), 
                tmp = str_replace_all(string = id, pattern = ".+_(noSTARRseq|STARRseq)", replacement = "\\1"),
                Type = ifelse(str_detect(tmp, "noSTARRseq"), tmp, "All"),
                Type = ifelse(str_detect(tmp, "STARRseq"), tmp, "All"))
x <- unique(phastConsAt$id) %>% sort()
table(phastConsAt$Type, phastConsAt$Features)
table(phastConsAt$Type)
table(phastConsAt$Features)

## ALL
df <- phastConsAt %>%
  dplyr::group_by(id, Type, Features) %>%
  dplyr::summarise(Num = n(), Median = median(PhastCons, na.rm = T)) %>%
  dplyr::mutate(label = str_c("Median=", round(Median, digits = 4), "\nN=", Num, sep = ""))

p <- phastConsAt %>%
  ggplot(aes(x = Type, y = PhastCons)) +
    geom_violin(fill = "transparent", scale = "width", width = 0.75) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0) +
    toolkit_tyj$theme_my + 
    coord_flip() +
    geom_hline(yintercept = median(phastConsAt$PhastCons[phastConsAt$id == "WholeGenome600"], na.rm = T), 
               linetype = 2) +
    geom_point(data = df, aes(x = Type, y = Median), color = "red") + 
    facet_grid(rows = "Features", space = "free", scales = "free") + 
    geom_label(data = df, aes(x = Type, y = 0.92, label = label), size = 1.5, alpha = 1/2)

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/PhastCons_all_tile", 
                     width = 16, height = 22, plot = p)

## each CRE as a whole
df <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  dplyr::group_by(id, Type, Features) %>%
  dplyr::summarise(Num = n(), Median = median(PhastCons, na.rm = T)) %>%
  dplyr::mutate(label = str_c("Median=", round(Median, digits = 4), "\nN=", Num, sep = ""))


p <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  ggplot(aes(x = Features, y = PhastCons)) +
    geom_violin(fill = "transparent", scale = "width", width = 0.75) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0) +
    toolkit_tyj$theme_my + 
    coord_flip() +
    geom_hline(yintercept = median(phastConsAt$PhastCons[phastConsAt$id == "WholeGenome600"], na.rm = T), 
               linetype = 2) +
    geom_point(data = df, 
               aes(x = Features, y = Median), color = "red") + 
    geom_label(data = df, aes(x = Features, y = 0.95, label = label), size = 1.8, alpha = 1/2)

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/PhastCons_each_Features_tile", 
                     width = 16, height = 12, plot = p)
```

*20210630*  all 7000 cSTARRseq enhancer
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(tidyverse)

# laod PhaseCons score and get the true region width. (CREs located on chrC/M were discared in Step1)
phastConsAt <- fread("./R8.conservation/PhastCons_score_matrix_all_CRE_20210326.csv.gz") %>%
  dplyr::rename(PhastCons = score) %>%
  dplyr::filter(str_detect(id, "(Promoter|ProEnlarge|iSTARR-seq|PAS)", negate = T)) %>%
  dplyr::mutate(id = str_replace_all(id, "cSTARR-seq", "Enhancer")) %>%
  dplyr::mutate(Features = str_replace_all(id, "(_noSTARRseq|_STARRseq)", ""), 
                tmp = str_replace_all(string = id, pattern = ".+_(noSTARRseq|STARRseq)", replacement = "\\1"),
                Type = ifelse(str_detect(tmp, "noSTARRseq"), tmp, "All"),
                Type = ifelse(str_detect(tmp, "STARRseq"), tmp, "All"), 
                Features  = str_replace_all(Features, "^Enhancer$", "Enhancer_STARRseq")) %>%
  dplyr::filter(str_detect(Features, "(Enhancer_W.+|Enhancer_Z.+)", negate = T)) %>%
  dplyr::mutate(Features = factor(Features, levels = rev(c("CorePro", "FiveUTR", "Exon", "Intron",
                                                         "ThreeUTR",  "Gene", 
                                                         "Interneric","Enhancer_STARRseq", "Random"))),
                Type = str_replace_all(Type, c("^STARRseq$" = "With enhancer",
                                               "noSTARRseq" = "Without enhancer")))
x <- unique(phastConsAt$id) %>% sort()
table(phastConsAt$Type, phastConsAt$Features)
table(phastConsAt$Type)
table(phastConsAt$Features)

## plot --------------------------------------------------------------------

## each CRE as a whole
sigLabel <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  dplyr::select(Type, Features, PhastCons) %>%
  dplyr::rename(value = PhastCons) %>%
  toolkit_tyj$Multicomparison(., Name_variable1 = "Type", Name_variable2 = "Features", 
                              test.name = "kruskal", label.sigLevel = 0.05)

df <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  dplyr::group_by(id, Type, Features) %>%
  dplyr::summarise(Num = n(), Median = median(PhastCons, na.rm = T)) %>%
  dplyr::mutate(label = str_c("Median=", round(Median, digits = 4), "\nN=", Num, sep = "")) %>%
  dplyr::filter(Type == "All")

df <- sigLabel$significant.label %>%
  dplyr::rename(Features = variable2) %>%
  full_join(., df)

p <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  ggplot(aes(x = Features, y = PhastCons)) +
    geom_violin(fill = "transparent", scale = "width", width = 0.75) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0) +
    toolkit_tyj$theme_my + 
    coord_flip() +
    geom_hline(yintercept = median(phastConsAt$PhastCons[phastConsAt$id == "Random"], na.rm = T), 
               linetype = 2) +
    geom_point(data = df, 
               aes(x = Features, y = Median), color = "red") + 
    geom_label(data = df, aes(x = Features, y = 1.1, label = label), 
               size = 1.5, alpha = 0, label.padding = unit(0.03, units = "cm")) + 
    scale_y_continuous(expand = expansion(c(0.02, 0.05), 0)) + 
    geom_text(data = df, aes(x = Features, y = 1.2, label = value), 
              color = "red", size = 2.5)

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/PhastCons_each_Features_tile_20210630", 
                     width = 8, height = 6, plot = p)



## Features x (STARRseq, noSTARRseq, All)
df <- phastConsAt %>%
  dplyr::filter(str_detect(Features, "(Random|Enhancer_STARRseq)", negate = T))%>%
  dplyr::group_by(id, Type, Features) %>%
  dplyr::summarise(Num = n(), Median = median(PhastCons, na.rm = T)) %>%
  dplyr::mutate(label = str_c("Median=", round(Median, digits = 4), "\nN=", Num, sep = ""))

sigLabel <- phastConsAt %>%
  dplyr::filter(str_detect(Features, "(Random|Enhancer_STARRseq)", negate = T)) %>%
  dplyr::select(Features, Type, PhastCons) %>%
  dplyr::rename(value = PhastCons) %>%
  toolkit_tyj$Multicomparison(., Name_variable1 = "Features", 
                              Name_variable2 = "Type", test.name = "kruskal",
                              label.sigLevel = 0.05)
df <- sigLabel$significant.label %>%
  dplyr::rename(Features = variable1, Type = variable2) %>%
  full_join(df, .) %>%
  dplyr::mutate(Features = factor(Features, levels = c("CorePro", "FiveUTR", "Exon", "Intron",
                                                         "ThreeUTR",  "Gene", 
                                                         "Interneric","Enhancer_STARRseq", "Random")))

p <- phastConsAt %>%
  dplyr::filter(str_detect(Features, "(Random|Enhancer_STARRseq)", negate = T))%>%
  dplyr::mutate(Features = factor(Features, levels = c("CorePro", "FiveUTR", "Exon", "Intron",
                                                         "ThreeUTR",  "Gene", 
                                                         "Interneric","Enhancer_STARRseq", "Random"))) %>%
  ggplot(aes(x = Type, y = PhastCons)) +
    geom_violin(fill = "transparent", scale = "width", width = 0.75) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0) +
    toolkit_tyj$theme_my + 
    coord_flip() +
    geom_hline(yintercept = median(phastConsAt$PhastCons[phastConsAt$id == "Random"], na.rm = T), 
               linetype = 2) +
    geom_point(data = df, aes(x = Type, y = Median), color = "red") + 
    facet_grid(rows = "Features", space = "free", scales = "free") + 
    geom_label(data = df, aes(x = Type, y = 1.1, label = label), 
               size = 1.5, alpha = 0, label.padding = unit(0.03, units = "cm")) + 
    scale_y_continuous(expand = expansion(c(0.02, 0.05), 0), 
                       breaks = c(0, 0.25, 0.5, 0.75, 1)) + 
    geom_text(data = df, aes(x = Type, y = 1.2, label = value), 
              color = "red", size = 2.5)

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/PhastCons_all_tile_20210701", 
                     width = 12, height = 12, plot = p)

```

*20220420* 4423 cSTARRseq enhancer, > 1.3
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
library(tidyverse)

Feature_level <- c("Promoter", "FiveUTR", "Exon",
                  "Intron", "ThreeUTR", "Gene","Intergenic",
                  "Enhancer_STARRseq", "GenomeWide", "Random")

# laod PhaseCons score and get the true region width. (CREs located on chrC/M were discared in Step1)
phastConsAt <- fread("./R8.conservation/PhastCons_score_matrix_all_CRE_20220420.csv.gz") %>%
  dplyr::rename(PhastCons = score) %>%
  dplyr::filter(str_detect(id, "(CorePro|^Promoter$|Promoter_|iSTARR-seq|PAS)", negate = T)) %>%
  dplyr::mutate(id = str_replace_all(id, c("cSTARR-seq" = "Enhancer",
                                           "ProEnlarge" = "Promoter",
                                           "Interneric" = "Intergenic"))) %>%
  dplyr::mutate(Features = str_replace_all(id, "(_noSTARRseq|_STARRseq)", ""), 
                tmp = str_replace_all(string = id, pattern = ".+_(noSTARRseq|STARRseq)", replacement = "\\1"),
                Type = ifelse(str_detect(tmp, "noSTARRseq"), tmp, "All"),
                Type = ifelse(str_detect(tmp, "STARRseq"), tmp, "All"), 
                Features  = str_replace_all(Features, "^Enhancer$", "Enhancer_STARRseq")) %>%
  dplyr::filter(str_detect(Features, "(Enhancer_W.+|Enhancer_Z.+)", negate = T)) %>%
  dplyr::mutate(Features = factor(Features, levels = rev(Feature_level)),
                Type = str_replace_all(Type, c("^STARRseq$" = "With enhancer",
                                               "noSTARRseq" = "Without enhancer")))
x <- unique(phastConsAt$id) %>% sort()
table(phastConsAt$Type, phastConsAt$Features)
table(phastConsAt$Type)
table(phastConsAt$Features)
table(phastConsAt$id)

## plot --------------------------------------------------------------------

## each CRE as a whole
sigLabel <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  dplyr::select(Type, Features, PhastCons) %>%
  dplyr::rename(value = PhastCons) %>%
  toolkit_tyj$Multicomparison(., Name_variable1 = "Type", Name_variable2 = "Features", 
                              test.name = "kruskal", label.sigLevel = 0.05)

df <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  dplyr::group_by(id, Type, Features) %>%
  dplyr::summarise(Num = n(), Median = median(PhastCons, na.rm = T)) %>%
  dplyr::mutate(label = str_c("Median=", round(Median, digits = 2), "\nN=", Num, sep = "")) %>%
  dplyr::filter(Type == "All")

df <- sigLabel$significant.label %>%
  dplyr::rename(Features = variable2) %>%
  full_join(., df)

p <- phastConsAt %>%
  dplyr::filter(Type == "All") %>%
  ggplot(aes(x = Features, y = PhastCons)) +
    geom_violin(fill = "transparent", scale = "width", width = 0.75, size = 0.3) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0) +
    toolkit_tyj$theme_my + 
    coord_flip() +
    geom_hline(yintercept = median(phastConsAt$PhastCons[phastConsAt$id == "Random"], na.rm = T), 
               linetype = 2, size = 0.5, color = "grey35") +
    geom_point(data = df, 
               aes(x = Features, y = Median), color = "red", size=0.5) + 
    geom_label(data = df, aes(x = Features, y = 1.2, label = label), color = "grey35", 
               size = 5/.pt, alpha = 0, label.padding = unit(0.03, units = "cm")) + 
    scale_y_continuous(expand = expansion(c(0.02, 0.08), 0), breaks = seq(0, 1, 0.25)) + 
    geom_text(data = df, aes(x = Features, y = 1.45, label = value), 
              color = "red", size = 6/.pt)
p

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/PhastCons_each_Features_tile_20220420", 
                     width = 8, height = 6, plot = p)


## Features x (STARRseq, noSTARRseq, All)
df <- phastConsAt %>%
  dplyr::filter(str_detect(Features, "(Random|Enhancer_STARRseq)", negate = T))%>%
  dplyr::group_by(id, Type, Features) %>%
  dplyr::summarise(Num = n(), Median = median(PhastCons, na.rm = T)) %>%
  dplyr::mutate(label = str_c("Median=", round(Median, digits = 2), "\nN=", Num, sep = ""))

sigLabel <- phastConsAt %>%
  dplyr::filter(str_detect(Features, "(Random|Enhancer_STARRseq)", negate = T)) %>%
  dplyr::select(Features, Type, PhastCons) %>%
  dplyr::rename(value = PhastCons) %>%
  toolkit_tyj$Multicomparison(., Name_variable1 = "Features", 
                              Name_variable2 = "Type", test.name = "kruskal",
                              label.sigLevel = 0.05)
df <- sigLabel$significant.label %>%
  dplyr::rename(Features = variable1, Type = variable2) %>%
  full_join(df, .) %>%
  dplyr::mutate(Features = factor(Features, levels = Feature_level))

p <- phastConsAt %>%
  dplyr::filter(str_detect(Features, "(Random|Enhancer_STARRseq)", negate = T))%>%
  dplyr::mutate(Features = factor(Features, levels = Feature_level)) %>%
  ggplot(aes(x = Type, y = PhastCons)) +
    geom_violin(fill = "transparent", scale = "width", width = 0.75, size = 0.3) + 
    geom_boxplot(width = 0.05, outlier.alpha = 0) +
    toolkit_tyj$theme_my + 
    coord_flip() +
    geom_hline(yintercept = median(phastConsAt$PhastCons[phastConsAt$id == "Random"], na.rm = T), 
               linetype = 2, size = 0.5, color = "grey35") +
    geom_point(data = df, aes(x = Type, y = Median), color = "red", size = 0.5) + 
    facet_grid(rows = "Features", space = "free", scales = "free") + 
    geom_label(data = df, aes(x = Type, y = 1.2, label = label), color = "grey35",
               size = 4/.pt, alpha = 0, label.padding = unit(0.03, units = "cm")) + 
    scale_y_continuous(expand = expansion(c(0.02, 0.08), 0), 
                       breaks = c(0, 0.25, 0.5, 0.75, 1)) + 
    geom_text(data = df, aes(x = Type, y = 1.45, label = value), 
              color = "red", size = 6/.pt) + 
    theme(panel.spacing = unit(0.04, "cm"))

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/PhastCons_all_tile_20220420", 
                     width = 8, height = 10, plot = p)

```


## 8.2 Microevolution

Related index can be classified into two types:             
1. per-base: pi, fst        
2. bin-based: pi, fst, tajimad (200bp window, 200bp step), CLR, Omega.       
`Window/bin size must less than width of the majority CREs.`       

bed files of CREs and genomic features were generated in 8.1.

### .1 calculate index

*Width of CREs and genomic features*

```{r}
cat(">>> Five numbers of regions in each bed file:\n")
for (i in list.files(path = "./R8.conservation/bed_files/", pattern = ".bed")) {
  tmp <- fread(paste("./R8.conservation/bed_files/", i, sep = "")) %>%
    makeGRangesFromDataFrame(seqnames.field = "V1", start.field = "V2", end.field = "V3")
  cat(i, ": ", fivenum(width(tmp)), "\n")
}
```

*Prepare table of accession ID for following calculation*

```{r}
source("D:/SUST/code/TanYongjun_code.R")
metaInfo <- fread("./refgenome/At1135_accessions.csv")
# accession id of each group
table(metaInfo$`Admixture Group`)
for (i in unique(metaInfo$`Admixture Group`)) {
  metaInfo %>%
    dplyr::filter(`Admixture Group` == i) %>%
    dplyr::select(accession) %>%
    fwrite(file = paste("./refgenome/Accession_id_of_Group_", i, ".list", sep = ""),
           col.names = F, quote = F, row.names = F)
}
dim(metaInfo)
```

*Calculation* Pi, Tajima'sD, Fst, Omega, CLR.
`/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/8.2.1_microevolution_fil_pi_tajimaD.sh` `/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/8.2.2_microevolution_Fst.R` `/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/8.2.3_microevolution_OmegaCLR.sh` `/scratch/2021-03-21/bioTanyj/At_enhancer/scripts/8.2.4_collect_pi_TajimaD_Fst.R`

### .2 Backup (not used)

##### \~2 Load and summary (Not used! because very large window was used.)

```{r}
source("D:/SUST/code/TanYongjun_code.R")

# load data and convert to non-overlapped regions
df_pi <- fread("./R8.conservation/index/Pi_10kwindow_2kstep_maf0.05_beagle.csv.gz") %>%
  dplyr::mutate(position = (BIN_END + BIN_START -1 ) / 2) %>%
  dplyr::select(group, chr, position, PI) %>%
  dplyr::rename(value = PI) %>%
  dplyr::mutate(start = position - 1000, 
                end = position + 1000)

df_tajima <- fread("./R8.conservation/index/TajimaD_5kwindow_maf0.05_beagle.csv.gz") %>%
  dplyr::mutate(position = BIN_START + 2500) %>%
  dplyr::filter(N_SNPS > 30) %>%
  dplyr::select(group, chr, position, TajimaD) %>%
  dplyr::rename(value = TajimaD) %>%
  dplyr::mutate(start = position - 2500, 
                end = position + 2500)

df_fst <- fread("./R8.conservation/index/Fst_10kwindow_5kstep_maf0.05_beagle.csv.gz") %>%
  dplyr::filter(N_VARIANTS > 30) %>%
  dplyr::mutate(position = (BIN_END + BIN_START - 1) / 2,
                chr = str_c("chr", CHROM, sep = ""), 
                WEIGHTED_FST = ifelse(WEIGHTED_FST >0, WEIGHTED_FST, 0), 
                MEAN_FST = ifelse(MEAN_FST > 0, MEAN_FST, 0)) %>%
  dplyr::rename(value = WEIGHTED_FST) %>%
  dplyr::select(group, chr, position, value) %>%
  dplyr::mutate(start = position - 2500, 
                end = position + 2500)

df_omega <- fread("./R8.conservation/index/OmegaPlus_all_maf0.05_beagle.csv.gz") %>%
  dplyr::rename(position = Position, value = Omega) %>%
  dplyr::select(group, chr, position, value) %>%
  dplyr::mutate(chr = str_replace_all(chr, "Chr", "chr"), 
                start = round(position - 15210/2, digits = 0), 
                end = round(position + 15210/2, digits = 0))

df_clr <- fread("./R8.conservation/index/Sweed_CLR_all_maf0.05_beagle.csv.gz") %>%
  dplyr::mutate(chr = str_replace_all(chr, "Chr", "chr"), 
                position = Position, 
                value = Likelihood) %>%
  dplyr::select(group, chr, position, value) %>%
  dplyr::mutate(start = round(position - 15210/2, digits = 0), 
                end = round(position + 15210/2, digits = 0))

index_all <- list(CLR = df_clr, 
                  Omega = df_omega, 
                  PI = df_pi, 
                  TajimaD = df_tajima, 
                  FST = df_fst)

# add column in each df of list.
for (i in 1:length(index_all)) {
  index_all[[i]]$index <- names(index_all)[i]
}

## summary of all index
summary_all <- map(index_all, .f = function(x){
  x %>%
    group_by(group, index) %>%
    summarise(Min = min(value), Median = median(value), Max = max(value))
})
summary_all <- purrr::reduce(summary_all, rbind)

index_all %>% 
  purrr::reduce(., rbind) %>%
  ggplot(aes(x = group, y = value)) +
    geom_violin(fill = "transparent") + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
    facet_wrap(facets = "index", scales = "free") + 
    toolkit_tyj$theme_my
# toolkit_tyj$SavePlot("./R8.conservation/index/ViolinPlot_index_genowise_20210407", 
#                      width = 25, height = 20)

# save(index_all, file = "./R8.conservation/index/All_index_list_object_20210407.bin")
```

##### \~3 values of bed

```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
load("./R8.conservation/index/All_index_list_object_20210407.bin")

## convert index into GR object
index_all <- index_all[c("PI")] %>%
  purrr::reduce(., rbind) %>%
  dplyr::mutate(seqnames = str_replace_all(chr, "chr", "Chr")) %>%
  dplyr::select(-position)
indexAllGR <- index_all %>%
  GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "seqnames", 
                                          start.field = "start", 
                                          end.field = "end", 
                                          keep.extra.columns = T)

## subset scores in bed file.
# max region width less than 600 except Enhancer_Wang/zhu, and random.
score_all <- NULL
for (i in list.files("./R8.conservation/bed_files/", pattern = ".+bed$")) {
  # i <- list.files("./R8.conservation/bed_files/", pattern = ".+bed$")[1]
  cat(i, "\n")
  # load bed
  tmp <- fread(paste("./R8.conservation/bed_files/", i, sep = "")) %>%
    GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "V1", 
                                            start.field = "V2", 
                                            end.field = "V3")
  
  # findoverlap
  ol <- findOverlaps(indexAllGR, tmp)
  
  # subset score index (group x index x bed)
  df <- index_all
  df$ID <- NA
  df$ID[unique(ol@from)] <- str_replace_all(i, ".bed", "")
  df <- df %>% dplyr::filter(!is.na(df$ID))
  score_all <- rbind(score_all, df)
  rm(df)
}

```

##### \~4 Plot

```{r}
# summary and plot
df <- score_all %>%
  group_by(ID, group) %>%
  dplyr::summarise(Median = median(value))
p <- score_all %>%
  ggplot(aes(x = ID, y = value)) +
    geom_boxplot(fill = "transparent", width = 0.2, outlier.alpha = 0) +
    geom_violin(fill = "transparent") + 
    toolkit_tyj$theme_my + 
    facet_wrap("group", nrow = 2) +
    coord_flip() +
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.key.size = unit(0.4, units = "cm")) + 
    geom_point(data = df, aes(x = ID, y = Median), color = "red", size = 0.5)

toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/index/PI_compare_20210407", 
                     width = 25, height = 18, plot = p)
```

### .5 collect and plot

#### \~ 1 subset population index in CRE.

```{r}
##################################################################################
#  subset population index in CRE; evenly subset from whole genome for plot on PC.
#                         Tanyongjun 20210418 modified 20210416
##################################################################################

rm(list = ls())
setwd("/scratch/2022-04-15/bioTanyj/At_enhancer/R8.conservation/population_genetics")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

#### merge -------------------------------------------------
# pi
    pi_group <- fread("Pi_200window_50step_maf0.05_beagle.csv.gz")
    pi_1001genomes <- fread("All_1001genomes_Pi_200window_200step_maf0.05_beagle.csv.gz")
    pi_all <- rbind(pi_group, pi_1001genomes) %>%
      dplyr::mutate(start = BIN_START, end = BIN_END, 
                    value = PI) %>%
      dplyr::select(group, chr, start, end, value)
    cat("\n>>>>>>> load pi completed!\n")
    
# TajimaD
    tajima_group <- fread("TajimaD_200window_maf0.05_beagle.csv.gz")
    tajima_1001genomes <- fread("All_1001genomes_TajimaD_200window_maf0.05_beagle.csv.gz")
    tajima_all <- rbind(tajima_1001genomes, tajima_group) %>%
      dplyr::mutate(start = BIN_START, end = BIN_START + 200, 
                    value = TajimaD) %>%
      dplyr::select(group, chr, start, end, value)
    cat(">>>>>>> load TajimaD completed!\n")
    
# fst
    fst_all <- fread("Fst_200bpwindow_200bpstep_maf0.05_beagle.csv.gz") %>%
      dplyr::mutate(start = BIN_START, end = BIN_END,
                    value = ifelse(WEIGHTED_FST > 0, WEIGHTED_FST, 0)) %>%
      dplyr::select(group, chr, start, end, value)
    cat(">>>>>>> load FST completed!\n")
    
# merge and to GR
    indexAll <- list(FST = fst_all, PI = pi_all, TajimaD = tajima_all)
    indexAllGR <- purrr::map(indexAll, GenomicRanges::makeGRangesFromDataFrame, 
                             seqnames.field = "chr", 
                             start.field = "start", 
                             end.field = "end")
    cat(">>>> Generate GR completed!\n")
    
#### CRE score -------------------
for(i in list.files(path = "../bed_files",
                    pattern = ".+\\.bed$")){
      cat("\n>>>> Calculate score of ", i, ":  ")
      tmp <- fread(paste("../bed_files/", i, sep = ""), header = F, sep = "\t") %>%
        dplyr::mutate(id = str_replace_all(i, ".bed", ""), 
        V1 = str_replace_all(V1, "Chr", "chr"))
      tmpGR <- tmp %>%
        GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "V1", 
                                                start.field = "V2", end.field = "V3")

      # findOverlap
      hit_ls <- purrr::map(indexAllGR, function(x){
        findOverlaps(x, tmpGR)
      })
      cat(" Findoverlap.... ")

      # extract score
      for (j in names(indexAllGR)) {
        cat(j, ":  ")
        df <- cbind(value = indexAll[[j]]$value[hit_ls[[j]]@from],
                    tmp[hit_ls[[j]]@to,])
        # cat(" cbind ")
        df <- df %>%
          group_by(V1, V2, V3) %>%
          dplyr::summarise(index_tmp = median(value))
        # cat(" group/summarise ")
        names(df) <- str_replace_all(names(df), "index_tmp", j)
        
        # merge to origin CRE df
        tmp <- full_join(tmp, df)
        # cat(" full_join ")
        rm(df)
      } # for loop
      
      # write to file
      tmp %>%
        fwrite(paste(str_replace_all(i, "\\.bed", ""), "_population_index.tsv", sep = ""), 
               sep = "\t")
    } # for loop

### subset population index of the whole genome as background --------------------
    thin <- c(5, 10, 20, 50, 100) # select one in thin.
    cat("<==========================================>\n
    >>>>>>> Thin the populatin index of the whole genome:\n")
    for (i in thin) {
        cat("\n    ", i, ":  ")
      for (j in names(indexAll)) {
        cat(j)
        indexAll[[j]] %>%
        #   dplyr::slice_head(n = 10000) %>%
          dplyr::arrange(group, start, end) %>%
          dplyr::slice(seq(i, nrow(.), i)) %>%
          fwrite(., paste(j, "_GenomeWide_thin", i, "_20210418.tsv.gz", sep = ""), 
                 sep = "\t")
      }
    }
    
```
`/scratch/2022-04-15/bioTanyj/At_enhancer/scripts/old_scripts_20210908/8.2.5_subset_population_index_for_plot.R`


#### \~ 2 plot PI

```{r}
  rm(list = ls())
  setwd("D:/SUST/Enhancers_A.T/")
  
  Feature_level <- c("Promoter", "FiveUTR", "Exon",
                  "Intron", "ThreeUTR", "Gene","Intergenic",
                  "Enhancer_STARRseq", "GenomeWide", "Random")
  
## load
  source("D:/SUST/code/TanYongjun_code.R")
  cre_score <- fread("./R8.conservation/index/PI_CRE_all_20220420.tsv.gz") %>%
    dplyr::filter(str_detect(id, "(iSTARR-seq|PAS)", negate = T)) %>%
    dplyr::mutate(Features = str_replace_all(id, "(.+)_(.+seq)", "\\1"), 
                  Type = str_replace_all(id, "(.+)_(.+seq)", "\\2"),
                  Type = ifelse(Features == Type, "All", Type),
                  id = str_replace_all(id, "cSTARR-seq", "Enhancer_STARRseq"),
                  Features = str_replace_all(Features, "cSTARR-seq", "Enhancer_STARRseq"))
  table(cre_score$Features, cre_score$Type)
  
  geno_score <- fread("./R8.conservation/index/PI_GenomeWide_thin20_20210418.tsv.gz") %>%
    dplyr::select(chr:end, group, value) %>%
    dplyr::rename(PI = value) %>%
    dplyr::mutate(id = "GenomeWide", Features = "GenomeWide", Type = "All")
  colnames(cre_score)[1:3] <- c("chr", "start", "end")
  
  cre_score <- full_join(cre_score, geno_score)
  
## Genomic features -------------------
## plot of each group
  for (i in unique(cre_score$group)) {
    # i <- unique(cre_score$group)[1]
    cat(i, ";  ")
    ## each CRE/genomic features
    tmp <- cre_score %>% 
      dplyr::filter(group == i, 
                    Type == "All", 
                    str_detect(Features, "iSTARR|Random|CorePro|Promoter|PAS|Enhancer_W.+|Enhancer_Z.+", negate = T)) %>%
      dplyr::mutate(PI = as.numeric(PI) * 1000, 
                    Features = str_replace_all(Features, c("Interneric" = "Intergenic",
                                                           "ProEnlarge" = "Promoter")), 
                    Features = factor(Features, levels = rev(Feature_level)))
    
    # summary
    tmp_label <- tmp %>% 
      dplyr::group_by(group, Features) %>%
      dplyr::summarise(Median = round(median(PI, na.rm = T), digits = 2),
                Num = n()) %>%
      dplyr::mutate(label = str_c("Median=", Median, "\nN=", Num, sep = ""))
    
    # sig label
    sig_label <- toolkit_tyj$Multicomparison(Data_comapre = dplyr::select(tmp, group, Features, PI) %>% dplyr::rename(value = PI), 
                                Name_variable1 = "group", 
                                Name_variable2 = "Features", 
                                test.name = "kruskal")
    names(sig_label$significant.label) <- str_replace_all(names(sig_label$significant.label), 
                                                          c("variable1" = "group", 
                                                            "variable2" = "Features"))
    
    # plot
    p <- tmp %>%
      ggplot(aes(x = PI, y = Features)) +
      geom_vline(xintercept = tmp_label$Median[tmp_label$Features == "GenomeWide"], 
                 linetype = 2, size = 0.5, color = "grey35") +
      geom_violin(fill = "transparent", scale = "width", width = 0.75, size = 0.3) + 
      geom_boxplot(width = 0.05, outlier.alpha = 0) +
      # facet_grid(region ~ ., scales = "free_y") +
      # ggtitle(i) +
      coord_cartesian(xlim = c(0.01, 50)) +
      xlab(expression(nucleotide~diversity~(x10^-3))) +
      ylab("Features") +
      # labs(caption = "Kruskal-Wallis test, p < 0.05") +
      scale_x_log10() +
      theme() + 
      toolkit_tyj$theme_my +
      geom_point(data = tmp_label, aes(x = Median, y = Features), color = "red", size = 0.5) + 
      geom_label(data = tmp_label, aes(x = 0.02, y = Features, label = label), 
                 color = "grey35", size = 5/.pt, 
                 label.padding = unit(0.03, units = "cm"), 
                 fill = "transparent") + 
      geom_text(data = sig_label$significant.label, 
                aes(x = 50, y = Features, label = value), 
                color = "red", size = 6/.pt)
    
    toolkit_tyj$SavePlot(filename_prefix = paste("./R8.conservation/index/PI_", i, "_all_regions_1", sep = ""),
                           width = 8, height = 6, plot = p)
    
  }
  

  
## all group in a figure
    tmp <- cre_score %>% 
      dplyr::filter(group != "All_1001genomes", 
                    Type == "All", 
                    str_detect(Features, "iSTARR|Random|CorePro|Promoter|PAS|Enhancer_W.+|Enhancer_Z.+", negate = T)) %>%
      dplyr::mutate(PI = as.numeric(PI) * 1000, 
                    Features = str_replace_all(Features, c("Interneric" = "Intergenic",
                                                           "ProEnlarge" = "Promoter")), 
                    Features = factor(Features, levels = rev(Feature_level)))
    # summary
    tmp_label <- tmp %>% 
      dplyr::group_by(group, Features) %>%
      dplyr::summarise(Median = round(median(PI, na.rm = T), digits = 3),
                Num = n()) %>%
      dplyr::mutate(label = str_c("Median=", Median, "\nN=", Num, sep = ""))
    
    # sig label
    sig_label <- toolkit_tyj$Multicomparison(Data_comapre = dplyr::select(tmp, group, Features, PI) %>% dplyr::rename(value = PI), 
                                Name_variable1 = "group", 
                                Name_variable2 = "Features", 
                                test.name = "kruskal")
    names(sig_label$significant.label) <- str_replace_all(names(sig_label$significant.label), 
                                                          c("variable1" = "group", 
                                                            "variable2" = "Features"))
    
    p <- tmp %>%
      ggplot(aes(x = PI, y = Features)) +
      geom_vline(data = dplyr::filter(tmp_label, Features == "GenomeWide"),
                 aes(xintercept = Median), 
                 color = "grey45", linetype = 3) +
      geom_boxplot(outlier.alpha = 0, width = 0.1, fill = "transparent", size = 0.2) +
      geom_violin(fill = "transparent", size = 0.2, scale = "width", width = 0.8) +
      facet_wrap(facets = "group", nrow = 2) +
      # ggtitle("Nucleotide diversity of enhancer and genomic features") +
      # coord_cartesian(xlim = c(0.01, 80)) + 
      xlab(expression(nucleotide~diversity~(x10^-3))) +
      ylab("Enhancers and genomic features") +
      # labs(caption = "Kruskal-Wallis test, p < 0.05") +
      scale_x_log10(breaks = c(0.01, 0.1, 1, 10)) +
      theme() + 
      toolkit_tyj$theme_my +
      geom_point(data = tmp_label, aes(x = Median, y = Features), color = "red", size = 0.5) + 
      geom_label(data = tmp_label, aes(x = 0.01, y = Features, label = label), 
                 color = "grey25", size = 4/.pt, 
                 label.padding = unit(0.03, units = "cm"), 
                 fill = "transparent") + 
      geom_text(data = sig_label$significant.label, 
                aes(x = 50, y = Features, label = value), 
                color = "red", size = 5/.pt)
    
    toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/index/PI_all_groups_1",
                           width = 16, height = 10, plot = p)

    
## GenomicFeatures X Enhancer
    
#plot of each group
  for (i in unique(cre_score$group)) {
    # i <- unique(cre_score$group)[2]
    cat(i, ";  ")
    ## each CRE/genomic features
    tmp <- cre_score %>% 
      dplyr::filter(group == i, 
                    str_detect(id, "iSTARR|Random|CorePro|Promoter|PAS|Enhancer_W.+|Enhancer_Z.+|Enhancer_STARRseq|GenomeWide", 
                               negate = T)) %>%
      dplyr::mutate(PI = as.numeric(PI) * 1000, 
                    Features = str_replace_all(Features, c("Interneric" = "Intergenic",
                                                           "ProEnlarge" = "Promoter")), 
                    Type = str_replace_all(Type, c("noSTARRseq" = "Without enhancer", 
                                                   "^STARRseq$" = "With enhancer")),
                    Features = factor(Features, levels = Feature_level))
    
    # summary
    tmp_label <- tmp %>% 
      group_by(Type, group, Features) %>%
      summarise(Median = round(median(PI, na.rm = T), digits = 2),
                Num = n()) %>%
      dplyr::mutate(label = str_c("Median=", Median, "\nN=", Num, sep = ""))
    
    # sig label
    sig_label <- toolkit_tyj$Multicomparison(Data_comapre = dplyr::select(tmp, Features, Type, PI) %>% 
                                               dplyr::rename(value = PI), 
                                            Name_variable1 = "Features", 
                                            Name_variable2 = "Type", 
                                            test.name = "kruskal")
    names(sig_label$significant.label) <- str_replace_all(names(sig_label$significant.label), 
                                                          c("variable1" = "Features", 
                                                            "variable2" = "Type"))
   
    sig_label$significant.label$Features <- factor(sig_label$significant.label$Features, 
                                                 levels = Feature_level)

    # plot
    p <- tmp %>%
      ggplot(aes(x = PI, y = Type)) +
      # geom_vline(xintercept = tmp_label$Median[tmp_label$Features == "Random"], 
                 # color = "grey45", linetype = 3) +
      geom_violin(fill = "transparent", scale = "width", width = 0.75, size = 0.3) + 
      geom_boxplot(width = 0.05, outlier.alpha = 0) +
      facet_grid(Features ~ ., scales = "free_y", space = "free_y") +
      # ggtitle(i) +
      coord_cartesian(xlim = c(0.01, 50)) +
      xlab(expression(nucleotide~diversity~(x10^-3))) +
      ylab("Enhancers and genomic features") +
      # labs(caption = "Kruskal-Wallis test, p < 0.05") +
      scale_x_log10() +
      theme(panel.spacing = unit(0.05, "cm")) + 
      toolkit_tyj$theme_my +
      geom_point(data = tmp_label, aes(x = Median, y = Type), color = "red", size = 0.5) + 
      geom_label(data = tmp_label, aes(x = 0.02, y = Type, label = label), 
                 color = "grey35", size = 4/.pt, 
                 label.padding = unit(0.03, units = "cm"), 
                 fill = "transparent") + 
      geom_text(data = sig_label$significant.label, 
                aes(x = 50, y = Type, label = value), 
                color = "red", size = 6/.pt)
    
    toolkit_tyj$SavePlot(filename_prefix = paste("./R8.conservation/index/PI_", i, "_region_X_En_1", sep = ""),
                           width = 8, height = 10, plot = p)
    
  }
  
```

#### \~ 3 FST

the data file `FST_CRE_all_20210418.tsv.gz` was 477 Mb, run on TaiYi.

```{r}
##################################################################################
#  plot Fst of CRE for each group pairs.
#                         Tanyongjun 20210420
##################################################################################

rm(list = ls())
setwd("/scratch/2021-04-10/bioTanyj/At_enhancer/R8.conservation/population_genetics")
source("/work/bio-tanyj/soft/TanYongjun_code.R")

# load file
  cre_score <- fread("./FST_CRE_all_20210418.tsv.gz")
  geno_score <- fread("./FST_GenomeWide_thin10_20210418.tsv.gz") %>%
    dplyr::mutate(id = "GenomeWide") %>%
    dplyr::select(chr:end, group, id, value)
  names(geno_score) <- names(cre_score)
  rm(geno_score)  

## plot of each group pairs
  for (i in unique(cre_score$group)) {
    # i <- unique(cre_score$group)[1]
    cat(i, ";  ")
    ## each CRE/genomic features
    tmp <- cre_score %>% 
      dplyr::filter(group == i, 
                    # str_detect(id, "_", negate = T),
                    str_detect(id, "iSTARR|GenomeWide|ProEnlarge|CorePro|PAS", negate = T)) %>%
      dplyr::mutate(region = str_replace_all(id, "(.+)_.+", "\\1"), 
                    id = str_replace_all(id, c("cSTARR-seq" = "Enhancer", 
                                               "Interneric" = "Intergenic")), 
                    id = factor(id, levels = rev(c("Enhancer", "Promoter", "FiveUTR", "Exon", 
                                               "Intron", "ThreeUTR", "Intergenic", 
                                               "Gene", "Random"))))
    
    # summary
    tmp_label <- tmp %>% 
      group_by(id, group) %>%
      summarise(Median = round(median(FST, na.rm = T), digits = 2),
                Num = n()) %>%
      dplyr::mutate(label = str_c("Median=", Median, "\nN=", Num, sep = ""))
    
    # sig label
    sig_label <- toolkit_tyj$Multicomparison(Data_comapre = dplyr::select(tmp, group, id, FST) %>% 
                                                 dplyr::rename(value = FST), 
                                             Name_variable1 = "group", 
                                             Name_variable2 = "id", 
                                             test.name = "kruskal")
    
    names(sig_label$significant.label) <- str_replace_all(names(sig_label$significant.label), 
                                                          c("variable1" = "group", 
                                                            "variable2" = "id"))
    
    # plot
    p <- tmp %>%
      ggplot(aes(x = FST, y = id)) +
      geom_vline(xintercept = tmp_label$Median[tmp_label$id == "Random"], 
                 color = "grey45", linetype = 3) +
      geom_boxplot(outlier.alpha = 0, width = 0.1, fill = "transparent", size = 0.2) +
      geom_violin(fill = "transparent", size = 0.2) +
      # facet_grid(region ~ ., scales = "free_y") +
      ggtitle(i) +
      coord_cartesian(xlim = c(0.01, 50)) +
      xlab(expression(nucleotide~diversity~(x10^-3))) +
      ylab("Enhancers and genomic features") +
      labs(caption = "Kruskal-Wallis test, p < 0.05") +
      scale_x_log10() +
      theme() + 
      toolkit_tyj$theme_my +
      geom_point(data = tmp_label, aes(x = Median, y = id), color = "red", size = 0.5) + 
      geom_label(data = tmp_label, aes(x = 0.02, y = id, label = label), 
                 color = "grey35", size = 1.2, 
                 label.padding = unit(0.03, units = "cm"), 
                 fill = "transparent") + 
      geom_text(data = sig_label$significant.label, 
                aes(x = 50, y = id, label = value), 
                color = "red", size = 1.5)
    
    toolkit_tyj$SavePlot(filename_prefix = paste("FST_", i, "_all_regions", sep = ""),
                           width = 8, height = 6, plot = p, device = "png")

  }
  
```


## 8.3 SNP density according to CRE
Relative dentisity of SNP in CREs

### .1 each enhancer as a whole
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")

# snp
  genoAll <- fread("./R11.GWAS/1001genomes_snp-short-indel_only_ACGTN_loci_position.tsv.gz")
  genoAllGR <- genoAll %>%
    dplyr::mutate(start = V2 - 1) %>%
    dplyr::rename(seqnames = V1, end = V2) %>%
    makeGRangesFromDataFrame()
  
# CRE (Number of CREs used must be the same!)

  cre_all <- fread("Enhancer_Four_types_Random_PAS_all_20220420.csv.gz") %>%
    dplyr::mutate(strand = "*",
                  method = str_replace_all(method, c("cSTARR-seq" = "Enhancer_STARRseq",
                                                     "IntronicEnhancer" = "Enhancer_Meng2021")),
                  method = factor(method, c("Enhancer_STARRseq", "Enhancer_Zhu2015", 
                                            "Enhancer_Wang2019", "Enhancer_Meng2021",
                                            "Random", "iSTARR-seq", "PAS")))
  table(cre_all$method)
  table(cre_all$strand, cre_all$ID)

  NumCRE <- 1919 # min number of CREs
  
  # # add ten random 
  for (i in 1:10) {
    cre_all <- cre_all %>%
      dplyr::filter(method == "Random") %>%
      slice_sample(n = NumCRE) %>%
      dplyr::mutate(method = str_c(method, i, sep = "")) %>%
      rbind(cre_all, .)
  }
  
  
  # use same number of regions for each CRE in analysis.
  set.seed(12345)
  crePeakGR <- cre_all %>%
    dplyr::filter(str_detect(method, "(PAS|iSTARR-seq)", negate = T)) %>%
    split(., .$method) %>%
    map(., slice_sample, n = NumCRE) %>%
    map(., makeGRangesFromDataFrame) %>%
    map(., resize, width = 1, fix = "center")
  
# calculate distance
  df <- NULL
  for (i in names(crePeakGR)) {
    Sys.Date()
    cat(">>> ", i, "\n")
    tmp <- distanceToNearest(genoAllGR, crePeakGR[[i]]) %>%
      as.data.frame() %>%
      dplyr::filter(distance < 5000)
    tmp <- cbind(Pos_SNP = as.data.frame(genoAllGR)$end[tmp$queryHits], 
                  tmp, 
                  Pos_Peak = as.data.frame(crePeakGR[[i]])$end[tmp$subjectHits])
    tmp$Distance <- tmp$Pos_SNP - tmp$Pos_Peak
    tmp$method <- i
    df <- rbind(df, 
                tmp[,6:7])
  }
  
  
# Density plot
  df %>%
    dplyr::filter(str_detect(method, "(PAS)", negate = T)) %>%
    ggplot(aes(x = Distance, color = method)) + 
      geom_density() + 
      toolkit_tyj$theme_my + 
      scale_color_npg()
  
# deduct Random.
  tmp <- df %>%
    dplyr::mutate(Bins = cut(Distance, breaks = seq(-5000, 5000, 100))) %>%
    group_by(Bins, method) %>%
    summarise(Num = n()) %>%
    group_by(method) %>%
    dplyr::mutate(Sum = sum(Num)) %>%
    dplyr::mutate(Rate = Num / Sum) %>%
    pivot_wider(id_cols = Bins, names_from = method, values_from = Rate)
  
  # deduct random
  tmp$Random = apply(tmp[,which(str_detect(colnames(tmp), "Random"))], 
                     1, mean)
  for (i in which(str_detect(colnames(tmp), "(cSTARR|Enhancer)"))) {
    tmp[,i] <- tmp[,i] - tmp$Random
  }
  tmp <- tmp[,str_detect(colnames(tmp), "Random", negate = T)]
  
  tmp <- tmp %>%
    pivot_longer(cols = -Bins, names_to = "method", values_to = "density")
  
  # fwrite(tmp, file = "./R8.conservation/SNP_Density_according_to_enhancers_20210517.csv")
  
# plot
  library(ggformula)
  
  # each enhancer as a whole
  tmp %>%
    dplyr::mutate(x = as.numeric(Bins)) %>%
    ggplot(aes(x = x, y = density, color = method)) + 
      geom_point()+
      geom_spline(spar = 0.35) +
      toolkit_tyj$theme_my + 
      guides(color = guide_legend(title = "Type")) + 
      theme(legend.position = c(0.19, 0.8), 
            legend.title = element_blank(), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.key.width = unit(0.2, units = "cm")) + 
      xlab("Posotion") + 
      ylab("Relative density of SNPs") + 
      scale_x_continuous(breaks = c(0, 100, 200), 
                         labels = c("-5kb", "Center", "5kb"))
  
  # toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/Density_of_SNP_small_Indels_Enhancers_deductRandom", 
  #                      width = 8, height = 6)
  
  # Enhancer_STARRseq
  tmp %>%
    dplyr::filter(str_detect(method, "STARR")) %>%
    dplyr::mutate(x = as.numeric(Bins)) %>%
    ggplot(aes(x = x, y = density)) + 
      geom_point(color = "grey55")+
      geom_spline(color = "grey35", spar = 0.5, cv = F) +
      # geom_line() +
      toolkit_tyj$theme_my + 
      guides(color = guide_legend(title = "Type")) + 
      theme(legend.position = c(0.19, 0.8), 
            legend.title = element_blank(), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.key.width = unit(0.2, units = "cm")) + 
      xlab("Posotion") + 
      ylab("Relative density of SNPs") + 
      scale_x_continuous(breaks = c(0, 100, 200), 
                         labels = c("-5kb", "Center", "5kb"))
  
  toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/Density_of_SNP_small_Indels_Enhancers_STARRseq_deductRandom",
                       width = 8, height = 6)
```

### .2 Cluster
```{r}
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")

# snp
  genoAll <- fread("./R11.GWAS/1001genomes_snp-short-indel_only_ACGTN_loci_position.tsv.gz")
  genoAllGR <- genoAll %>%
    dplyr::mutate(start = V2 - 1) %>%
    dplyr::rename(seqnames = V1, end = V2) %>%
    makeGRangesFromDataFrame()
  
# CRE (Number of CREs used must be the same!)
  cre_all <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3), 
                  seqnames != "Mt", seqnames != "Pt") %>%
    dplyr::filter(str_detect(method, "(cSTARR|Random)"))
  table(cre_all$method)
  table(cre_all$clusterK4)
  table(cre_all$ClusterActivity)
  
  cre_merge <- cre_all %>%
    dplyr::select(seqnames, start, end, method) %>%
    dplyr::filter(method == "cSTARR-seq") %>%
    dplyr::mutate(method = str_replace_all(method, "cSTARR-seq", "Enhancer_STARRseq")) %>%
    slice_sample(n = 1919)
    
  cre_merge <- cre_all %>%
    dplyr::select(seqnames, start, end, clusterK4) %>%
    dplyr::filter(clusterK4 != 0) %>%
    dplyr::mutate(clusterK4 = str_c("KmeansCluster-", clusterK4, sep = "")) %>%
    dplyr::rename(method = clusterK4) %>%
    rbind(., cre_merge)
  cre_merge <- cre_all %>%
    dplyr::select(seqnames, start, end, ClusterActivity) %>%
    dplyr::mutate(ClusterActivity = str_replace_all(ClusterActivity, "Part", "ActivityCluster-")) %>%
    dplyr::rename(method = ClusterActivity) %>%
    rbind(., cre_merge)
  cre_merge <- cre_merge %>%
    dplyr::filter(str_length(method) > 0, method != 0)
  table(cre_merge$method)
  
  # add according random (same number of regions)
  random <- cre_all %>%
    dplyr::filter(method == "Random") %>%
    dplyr::select(seqnames, start, end, method)
  cre_all <- cre_merge %>%
    dplyr::filter(method != "Random") 
  CREcount <- table(cre_all$method) %>% as.data.frame
  
  for (i in 1:10) { # number of random
    for (j in 1:nrow(CREcount)) {
      cre_all <- random %>%
        slice_sample(n = CREcount$Freq[j]) %>%
        dplyr::mutate(method = str_c(CREcount$Var1[j], "Random", i, sep = "")) %>%
        rbind(cre_all, .)
    }
  }
  table(cre_all$method)
  
  # to GR object
  crePeakGR <- cre_all %>%
    split(., .$method) %>%
    map(., makeGRangesFromDataFrame) %>%
    map(., resize, width = 1, fix = "center")
  
# calculate distance between SNP and CRE center
  df <- NULL
  for (i in names(crePeakGR)) {
    timestamp()
    cat(">>> ", i, "\n")
    tmp <- distanceToNearest(genoAllGR, crePeakGR[[i]]) %>%
      as.data.frame() %>%
      dplyr::filter(distance < 5000)
    tmp <- cbind(Pos_SNP = as.data.frame(genoAllGR)$end[tmp$queryHits], 
                  tmp, 
                  Pos_Peak = as.data.frame(crePeakGR[[i]])$end[tmp$subjectHits])
    tmp$Distance <- tmp$Pos_SNP - tmp$Pos_Peak
    tmp$method <- i
    df <- rbind(df, 
                tmp[,6:7])
  }
  
# deduct mean value of the corresponding Random regions.
  # reshape
  tmp <- df %>%
    dplyr::mutate(Bins = cut(Distance, breaks = seq(-5000, 5000, 100))) %>%
    group_by(Bins, method) %>%
    summarise(Num = n()) %>%
    group_by(method) %>%
    dplyr::mutate(Sum = sum(Num)) %>%
    dplyr::mutate(Rate = Num / Sum) %>%
    pivot_wider(id_cols = Bins, names_from = method, values_from = Rate)
  
  # deduct
  for (i in unique(cre_all$method)[str_detect(unique(cre_all$method), "Random", negate = T)]) {
    random_tmp <- apply(tmp[,which(str_detect(colnames(tmp), str_c(i, "Random", sep = "")))], 1, mean)
    tmp[,str_detect(colnames(tmp), str_c("^", i, "$", sep = ""))] <- 
      tmp[,str_detect(colnames(tmp), str_c("^", i, "$", sep = ""))] - random_tmp
  }
  tmp <- tmp[, str_detect(colnames(tmp), "Random", negate = T)]

  # to longer format
  tmp <- tmp %>%
    pivot_longer(cols = -Bins, names_to = "method", values_to = "density")
  
  # fwrite(tmp, file = "./SNP_Density_according_to_enhancers_clusters_20210517.csv")
  
  # plot
  library(ggformula)
  
  # Activity clusters
  tmp %>%
    dplyr::filter(str_detect(method, "(Activity|Enhancer)")) %>%
    dplyr::mutate(x = as.numeric(Bins), 
                  method = str_replace_all(method, "cSTARR-seq", "Enhancer_STARRseq")) %>%
    ggplot(aes(x = x, y = density, color = method)) + 
      # geom_point()+
      geom_spline() +
      toolkit_tyj$theme_my + 
      guides(color = guide_legend(title = "Type")) + 
      theme(legend.position = c(0.19, 0.8), 
            legend.title = element_blank(), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.key.width = unit(0.2, units = "cm")) + 
      xlab("Posotion") + 
      ylab("Relative density of SNPs") + 
      scale_x_continuous(breaks = c(0, 100, 200), 
                         labels = c("-5kb", "Center", "5kb")) + 
      scale_color_npg()
  toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/Density_of_SNP_small_Indels_ActivityCluster_deductRandom-1", 
                       width = 8, height = 6)
  
  
  # Kmeans clusters
  tmp %>%
    dplyr::filter(str_detect(method, "(Kmeans|Cluster-0)")) %>%
    dplyr::mutate(x = as.numeric(Bins), 
                  method = str_replace_all(method, "cSTARR-seq", "Enhancer_STARRseq")) %>%
    ggplot(aes(x = x, y = density, color = method)) + 
      # geom_point()+
      geom_spline() +
      toolkit_tyj$theme_my + 
      guides(color = guide_legend(title = "Type")) + 
      theme(legend.position = c(0.19, 0.8), 
            legend.title = element_blank(), 
            legend.key.size = unit(0.3, units = "cm"), 
            legend.key.width = unit(0.2, units = "cm")) + 
      xlab("Posotion") + 
      ylab("Relative density of SNPs") + 
      scale_x_continuous(breaks = c(0, 100, 200), 
                         labels = c("-5kb", "Center", "5kb")) + 
      scale_color_npg()
  toolkit_tyj$SavePlot(filename_prefix = "./R8.conservation/Density_of_SNP_small_Indels_KmeansCluster_deductRandom-1", 
                       width = 8, height = 6)
  NA
  
```

##-------------Calling------------------------------

Identification of enhancers with STARR-seq/iSTARR-seq data.
Run enhancer identification pipeline with modified duplication-discarding method: Fragments with the same start (5') and clustered end (3') were seem as transcript from the same fragments in plasmid library.
(generate bed file)
## .x

> *The poly(A) signal.* The characteristic position distribution peaking at \~15-20 nt upstream of the cleavage sites, and the actual cleavage position typically *fluctuates* by \~10nt with respect to a major cleavage position.
> Elkon et.al, 2013, Nature review genetics.

## 9.0 bam files generate by Wanjing.

The following codes were copy from Wanjing:

*Alignment and file conversion*

```{bash, eval=F}
#!/bin/bash

sample_name="RAtEnWC1"

echo "================================== 比对 ================================"

#比对
bowtie2 -p 16 --no-discordant -X 2000 -x /home/wanjing/data3/ref_genomes/TAIR10/TAIR10_bowtie2/chr_num_modified/TAIR10 -1 ../"$sample_name"_R1.fq.gz -2 ../"$sample_name"_R2.fq.gz | samtools view -bh -@ 8 | samtools sort -@ 8 -o "$sample_name"_raw.bam 

echo "================================== 统计duplicate ================================"

#picard标记duplicate
java -jar /home/wanjing/data3/tools/picard/picard.jar MarkDuplicates I="$sample_name"_raw.bam O="$sample_name"_marked_duplicates.bam M="$sample_name"_marked_dup_metrics.txt  
#samtools统计duplicate
samtools flagstat -@ 8 "$sample_name"_marked_duplicates.bam >"$sample_name"_marked_dup_flagstat.txt

echo "================================== 过滤，转成bw ================================"

#过滤
# -bh : bam file with header, -f 2, match two position, -q 5: skip MAPQ less 5. @ 8: threads
samtools view -bh -f 2 -q 5 -@ 8 "$sample_name"_raw.bam | samtools sort -@ 8 -o "$sample_name".bam 
samtools index -@ 8 "$sample_name".bam
#转成bw
bamCoverage -b "$sample_name".bam -o "$sample_name".bw

echo "================================== 统计片段大小 ================================"

#统计片段大小
java -jar /home/wanjing/data3/tools/picard/picard.jar \
CollectInsertSizeMetrics \
I="$sample_name".bam \
O="$sample_name"_insert_size.txt \
H="$sample_name"_insert_size_histogram.pdf 

echo "================================== 计算覆盖度 ================================"

#计算覆盖度
# `CollectWgsMetricsWithNonZeroCoverage` was designed for estimate the coverage and performance (quality) of whole genome sequencing. The coverage region was only region hit by reads. It may not suit for estimate the coverage of STARR-seq for paired reads were not overlapped. __TYJ.
java -jar /home/wanjing/data3/tools/picard/picard.jar \
CollectWgsMetricsWithNonZeroCoverage \
I="$sample_name".bam \
O="$sample_name"_coverage.txt \
CHART="$sample_name"_coverage.pdf \
R=/home/wanjing/data3/ref_genomes/TAIR10/TAIR10_bowtie2/chr_num_modified/TAIR10.fa

echo "================================== complete! ================================="


```

*Remove duplicates from bam file*

```{bash, eval=F}
#!/bin/bash

sample_name="RAtEnWC1"

# remove duplicates with `MarkDuplicates`
java -jar /home/wanjing/data3/tools/picard/picard.jar MarkDuplicates REMOVE_DUPLICATES=true I="$sample_name"_raw.bam O="$sample_name"_rmdup.bam M="$sample_name"_rmdup_metrics.txt

samtools view -bh -f 2 -q 5 -@ 8 -o "$sample_name"_filter.bam "$sample_name"_rmdup.bam

```

## 9.1 cSTARR-seq (classical STARR-seq)

A total of 4 fastq files were generated, 2 for plasmid and 2 for cDNA.
`/scratch/2021-03-02/bioTanyj/At_enhancer/data_wj/Cleandata` Use xxx_raw.bam files generated by Wanjing.
(xxx_raw.bam was generated by bowtie2, samtools view and samtools sort.)

### 9.1.1 bam to bed

```{r, eval=F}
#!/bin/bash
#BSUB -J CBam2Bed                  ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 8                  ### ask for number of cores (default: 1)
#BSUB -W 40:00               ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

##-------------------------------------------------------------------------------------##
# convert xxx_raw.bam files (from WanJing) to bed files of cSTARR-seq.
#----------------------------------------------------------------------------------------

source /work/bio-tanyj/scripts/path_and_dir.sh

cd /scratch/2021-03-02/bioTanyj/At_enhancer/enhancer_calling/cSTARR-seq
for i in `find /scratch/2021-03-02/bioTanyj/At_enhancer/data_wj/Cleandata/  -name "*raw.bam"`;
do
	id=`basename ${i}`
  # sort
	samtools sort -n -@ 8 ${i} -o ${id%_raw.bam}_sort.bam
  # Fill in mate coordinates, ISIZE (inferred insert size) and mate related flags from a name-sorted alignment.
	samtools fixmate -@ 8 ${id%_raw.bam}_sort.bam ${id%_raw.bam}_fixed.bam
  # to ped
	bedtools bamtobed -i ${id%_raw.bam}_fixed.bam -mate1 -bedpe > ${id%_raw.bam}.bedpe # contain ten columns: chr-R1 start-R1 end-R1 chr-R2 start-R2 end-R2 readsID length strand-R1 strand-R2
  # subset paired reads which mapped on different strand of the same chromosome, and save the first read in new bed file (OPS: the variable of delimiter in awk)
	cat ${id%_raw.bam}.bedpe | awk '{if($1==$4&&$1!="."&&$2!=0&&$5!=0){if($2>=$5&&$9=="-"&&$10=="+"){start=$5;end=$3;strand="-"}else if($2<$5&&$9=="+"&&$10=="-"){start=$2;end=$6;strand="+"};print $1,start,end,strand}}' OFS='\t' \
	| awk '{print $1,$2,$3,$4}' OFS='\t' > ${id%_raw.bam}_fragment_paired.bed # the second awd seen not needed!
  # sort and keep the coordinance information of unique PE reads (the start/end of region covered by PE read)
	cat ${id%_raw.bam}_fragment_paired.bed | sort -T ./ -k1,1 -k2,2n > ${id%_raw.bam}_fragment_paired_sorted.bed
	uniq ${id%_raw.bam}_fragment_paired_sorted.bed | awk '{print $1,$2,$3,$4}' OFS='\t' > ${id%_raw.bam}_fragment_paired_sorted_unique.bed
done
```

### 9.1.2 Unique fragments (as control)

#### . Count

Count unique fragments in each bed files.
The unique fragment was defined based on have the varied `start` (5'), `end` (3') or `both` (`chromosome` and `strand` were also contained).
All code in `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.1.2_unique_fragments.R` and `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.1.2_unique_fragments.sh`.

#### . plot

```{r}
source("D:/SUST/code/TanYongjun_code.R")
df <- fread("./enhancer_calling/cSTARR-seq/Unique_fragments_definition/Summary_of_unique_fragments_based_on_different_criterial.tsv") %>%
  mutate(ID = str_replace_all(ID, "_fragment_paired_sorted_unique", ""))
df %>%
  ggplot(aes(y = Number_uniq_fragments, x = Name, fill = Name, color = Name)) +
    geom_col(position = "dodge") + 
    toolkit_tyj$theme_my + 
    scale_color_npg() + 
    scale_fill_npg() + 
    # facet_grid(. ~ ID) + 
    facet_wrap("ID") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.position = "none", 
          axis.title.x = element_blank()) + 
    geom_text(aes(label = Number_uniq_fragments, y = Number_uniq_fragments/2), color = "grey20", size = 2) + 
    ggtitle("Number of unique fragments filtered based on different criterion\n cSTARR-seq")

# toolkit_tyj$SavePlot("./enhancer_calling/Summary_of_unique_fragments_based_on_different_criterial_cSTARR_seq",
#                      width = 12, height = 8)
```

*The total number of unique fragments defined based on one of the two fragment end was about 60% of unique fragments defined based on both end of fragments.* *No difference was observed on the number of unique fragments based on the start or end of fragments*

### 9.1.4 Fragment length stregth, and filtration.

1.  Add 40bp to each fragment, because only 21-120 bp of each read were used in alignment.
2.  Remove fragments which show length larger than 2k.
3.  Generate files contain only unique fragments (for plasmid library use six index). `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.1.4_stregth_filtration_uniq.R` `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.1.4_stregth_filtration_uniq.sh`

### 9.1.5 merge and calling

#### .1 merge

Fragments were length filtered, stregth (+40bp), and generate unique fragments based on two strategy: Strategy 1.
chr + start + end + strand for `plasmid`, and `cDNA_nodT`; chr + start + strand for `cDNA_dT.` Strategy 2 (startOnly).
chr + start + stand.
(all fragments in plasmid, cDNA_dT, and cDNA_nodT use the same criterion)

```{bash}
# strategy 1
cat RAtEnWP1_add40_filtered_uniq_strategy1.bed RAtEnWP2_add40_filtered_uniq_strategy1.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEnWP_add40_filtered_uniq_strategy1.bed
    
cat RAtEnWC1_add40_filtered_uniq_strategy1.bed RAtEnWC2_add40_filtered_uniq_strategy1.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEnWC_add40_filtered_uniq_strategy1.bed
    
# strategy 2 (startOnly)
cat RAtEnWP1_add40_filtered_uniq_startOnly.bed RAtEnWP2_add40_filtered_uniq_startOnly.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEnWP_add40_filtered_uniq_startOnly.bed
    
cat RAtEnWC1_add40_filtered_uniq_startOnly.bed RAtEnWC2_add40_filtered_uniq_startOnly.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEnWC_add40_filtered_uniq_startOnly.bed

```

#### .2 calling

Calling enhancers: Fragments (two strategies) X 3 sample (rep1, rep2, and merged)

```{bash, eval = F}
#!/bin/bash
#BSUB -J cEnCalling            ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 2                  ### ask for number of cores (default: 1)
#BSUB -W 5:00                ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

###########################################################################
#       Scrips for calling enahncer using R package "BasicSTARRseq".
#                                                TanYongjun 20210306
###########################################################################

targetpath="/scratch/2021-03-02/bioTanyj/At_enhancer/enhancer_calling/cSTARR-seq"
cd ${targetpath}

## Strategy 1.
    # merge
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEnWP_add40_filtered_uniq_strategy1.bed \
        RAtEnWC_add40_filtered_uniq_strategy1.bed \
        ${targetpath} \
        cEnhancer_merge_strategy1_raw.tsv

    # rep1
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEnWP1_add40_filtered_uniq_strategy1.bed \
        RAtEnWC1_add40_filtered_uniq_strategy1.bed \
        ${targetpath} \
        cEnhancer_rep1_strategy1_raw.tsv

    # rep2
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEnWP2_add40_filtered_uniq_strategy1.bed \
        RAtEnWC2_add40_filtered_uniq_strategy1.bed \
        ${targetpath} \
        cEnhancer_rep2_strategy1_raw.tsv


# ## Strategy 2: startOnly
    # merged
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEnWP_add40_filtered_uniq_startOnly.bed \
        RAtEnWC_add40_filtered_uniq_startOnly.bed \
        ${targetpath} \
        cEnhancer_merge_strategy2_raw.tsv

    #
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEnWP1_add40_filtered_uniq_startOnly.bed \
        RAtEnWC1_add40_filtered_uniq_startOnly.bed \
        ${targetpath} \
        cEnhancer_rep1_strategy2_raw.tsv

    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEnWP2_add40_filtered_uniq_startOnly.bed \
        RAtEnWC2_add40_filtered_uniq_startOnly.bed \
        ${targetpath} \
        cEnhancer_rep2_strategy2_raw.tsv

```

## 9.2 iSTARR-seq (improved STARR-seq)

A total of 14 fastq files were generated, 12 for plasmic (6 index in each of 2 replicates)and 2 for cDNA.
`/scratch/2021-03-02/bioTanyj/At_enhancer/data_wj/ANCGD180597_PM-GD180597-14_2019-08-18` Use xxx_raw.bam files generated by Wanjing.

### 9.2.1 bam to bed

```{bash}
#!/bin/bash
#BSUB -J iBam2Bed                  ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 8                  ### ask for number of cores (default: 1)
#BSUB -W 40:00               ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

##-------------------------------------------------------------------------------------##
# convert xxx_raw.bam files (from WanJing) to bed files of iSTARR-seq.
#----------------------------------------------------------------------------------------

source /work/bio-tanyj/scripts/path_and_dir.sh

cd /scratch/2021-03-02/bioTanyj/At_enhancer/enhancer_calling/iSTARR-seq
for i in `find /scratch/2021-03-02/bioTanyj/At_enhancer/data_wj/ANCGD180597_PM-GD180597-14_2019-08-18/Cleandata/  -name "*raw.bam"`;
do
	id=`basename ${i}`
  # sort
	samtools sort -n -@ 8 ${i} -o ${id%_raw.bam}_sort.bam
  # Fill in mate coordinates, ISIZE (inferred insert size) and mate related flags from a name-sorted alignment.
	samtools fixmate -@ 8 ${id%_raw.bam}_sort.bam ${id%_raw.bam}_fixed.bam
  # to ped
	bedtools bamtobed -i ${id%_raw.bam}_fixed.bam -mate1 -bedpe > ${id%_raw.bam}.bedpe # contain ten columns: chr-R1 start-R1 end-R1 chr-R2 start-R2 end-R2 readsID length strand-R1 strand-R2
  # subset paired reads which mapped on different strand of the same chromosome, and save the first read in new bed file (OPS: the variable of delimiter in awk)
	cat ${id%_raw.bam}.bedpe | awk '{if($1==$4&&$1!="."&&$2!=0&&$5!=0){if($2>=$5&&$9=="-"&&$10=="+"){start=$5;end=$3;strand="-"}else if($2<$5&&$9=="+"&&$10=="-"){start=$2;end=$6;strand="+"};print $1,start,end,strand}}' OFS='\t' \
	| awk '{print $1,$2,$3,$4}' OFS='\t' > ${id%_raw.bam}_fragment_paired.bed # the second awd seen not needed!
  # sort and keep the coordinance information of unique PE reads (the start/end of region covered by PE read)
	cat ${id%_raw.bam}_fragment_paired.bed | sort -T ./ -k1,1 -k2,2n > ${id%_raw.bam}_fragment_paired_sorted.bed
	uniq ${id%_raw.bam}_fragment_paired_sorted.bed | awk '{print $1,$2,$3,$4}' OFS='\t' > ${id%_raw.bam}_fragment_paired_sorted_unique.bed
done

```

### 9.2.2 Unique fragments count

Count unique fragments in each bed files.
The unique fragment was defined based on have the varied `start` (5'), `end` (3') or `both` (`chromosome` and `strand` were also contained).

#### .1 Count

All code in `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.1.2_unique_fragments.R` and `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.2.2_unique_fragments.sh`.

#### .2 pot

```{r}
source("D:/SUST/code/TanYongjun_code.R")
df <- fread("./enhancer_calling/iSTARR-seq/Unique_fragments_definition/Summary_of_unique_fragments_based_on_different_criterial.tsv") %>%
  mutate(ID = str_replace_all(ID, "_fragment_paired_sorted_unique", ""))
df %>%
  ggplot(aes(y = Number_uniq_fragments, x = Name, fill = Name, color = Name)) +
    geom_col(position = "dodge") + 
    toolkit_tyj$theme_my + 
    scale_color_npg() + 
    scale_fill_npg() + 
    # facet_grid(. ~ ID) +
    facet_wrap("ID") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
          legend.position = "none", 
          axis.title.x = element_blank()) + 
    geom_text(aes(label = Number_uniq_fragments, y = Number_uniq_fragments/2), color = "grey20", size = 2) +
    ggtitle("Number of unique fragments filtered based on different criterion\n iSTARR-seq")

# toolkit_tyj$SavePlot("./enhancer_calling/Summary_of_unique_fragments_based_on_different_criterial_iSTARR_seq",
#                      width = 16, height = 12)
```

### - Extend fragments with dT (not used, just extend all fragments based on the median length)

#### .Not used for only \~40-50% fragments identified both in plasmid and cDNA.

Extend fragments with dT (PAS) based on fragments in plasmid library.
Steps: 1.
Merge unique fragments with different index/barcode in Illumina seequencing library prepration.
2.
Subset fragments from plasmid library which can not be found in nodT cDNA library.
3.
Find the corresponding full length fragments of dT in subseted plasmid fragments based on the start position (chromosome and strand were both included), extend the length of fragment

*Only 5.4% fragments of plasmid can be found in the nodT_cDNA library. Thus, extend dT fragments based on the corredponding fragments in the plasmid is not possible.* *Only 32.3% fragments in dT_cDNA hit the same start of fragments in the plasmid library*

> Step1: Merge fragments of plasmid with different barcode.

```{bash Merge bed files, eval = F}
# plasmid replicate 1
for i in `ls | grep -E "RAtEP1.+unique.bed"`
do
cat ${i} >> RAtEP1_fragment_paired_sorted_unique.bed
done

for i in `ls | grep -E "RAtEP2.+unique.bed"`
do
cat ${i} >> RAtEP2_fragment_paired_sorted_unique.bed
done
```

> Step2: Subset fragments from plasmid which can not be found in nodT cDNA library.
> - the length variation of fragements in dt with the same start position.
> - how many fragments in cDNA nodT can be found in plasmid.

```{r}

```

### 9.2.4 Fragment length stregth, and filtration.

1.  Add 40bp to each fragment, because only 21-120 bp of each read were used in alignment.
2.  Remove fragments which show length larger than 2k. `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.2.4_stregth_filtration_uniq.R` `/scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.2.4_stregth_filtration_uniq.sh`

### 9.2.5 merge and calling

#### .1 merge cDNA_dT and cDNA_nodT, and merge replicates.

Fragments were length filtered, stregth (+40bp), and generate unique fragments based on two strategy: 1.
Mixs mode: chr + start + end + strand (all fragments in plasmid, cDNA_dT, and cDNA_nodT use the same criterion).
2.
StartOnly mode: chr + start + stand (all fragments in plasmid, cDNA_dT, and cDNA_nodT use the same criterion).
`Fragments of dT which stregthed to median size of nodT were not used in enhancer calling.`

*Not extend dT fragments*

```{bash}
cd /scratch/2021-03-02/bioTanyj/At_enhancer/enhancer_calling/iSTARR-seq/

# strategy 1 (Mixed mode: both start and end + startOnly)
cat RAtEP1_add40_filtered_uniq_strategy1.bed RAtEP2_add40_filtered_uniq_strategy1.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEP_add40_filtered_uniq_strategy1.bed
    
cat RAtEC1_nodT_add40_filtered_uniq_strategy1.bed RAtEC1_dT_add40_filtered_uniq_startOnly.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC1_add40_filtered_uniq_strategy1.bed
    
cat RAtEC2_nodT_add40_filtered_uniq_strategy1.bed RAtEC2_dT_add40_filtered_uniq_startOnly.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC2_add40_filtered_uniq_strategy1.bed
    
cat RAtEC1_add40_filtered_uniq_strategy1.bed RAtEC2_add40_filtered_uniq_strategy1.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC_add40_filtered_uniq_strategy1.bed

# strategy 2 (startOnly)
cat RAtEP1_add40_filtered_uniq_startOnly.bed RAtEP2_add40_filtered_uniq_startOnly.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEP_add40_filtered_uniq_startOnly.bed
    
cat RAtEC1_nodT_add40_filtered_uniq_startOnly.bed RAtEC1_dT_add40_filtered_uniq_startOnly.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC1_add40_filtered_uniq_startOnly.bed
    
cat RAtEC2_nodT_add40_filtered_uniq_startOnly.bed RAtEC2_dT_add40_filtered_uniq_startOnly.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC2_add40_filtered_uniq_startOnly.bed
    
cat RAtEC1_add40_filtered_uniq_startOnly.bed RAtEC2_add40_filtered_uniq_startOnly.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC_add40_filtered_uniq_startOnly.bed

```

*Extended*

```{bash}
cd /scratch/2021-03-02/bioTanyj/At_enhancer/enhancer_calling/iSTARR-seq/

# strategy 1 (Mixed mode: both start and end + startOnly)

cat RAtEC1_nodT_add40_filtered_uniq_strategy1.bed \
    RAtEC1_dT_add40_filtered_uniq_startOnly_extendToMedianLength.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC1_add40_filtered_uniq_strategy1_extendToMedianLength.bed
    
cat RAtEC2_nodT_add40_filtered_uniq_strategy1.bed \
    RAtEC2_dT_add40_filtered_uniq_startOnly_extendToMedianLength.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC2_add40_filtered_uniq_strategy1_extendToMedianLength.bed
    
cat RAtEC1_add40_filtered_uniq_strategy1_extendToMedianLength.bed \
    RAtEC2_add40_filtered_uniq_strategy1_extendToMedianLength.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC_add40_filtered_uniq_strategy1_extendToMedianLength.bed

# strategy 2 (startOnly)

    
cat RAtEC1_nodT_add40_filtered_uniq_startOnly.bed \
    RAtEC1_dT_add40_filtered_uniq_startOnly_extendToMedianLength.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC1_add40_filtered_uniq_startOnly_extendToMedianLength.bed
    
cat RAtEC2_nodT_add40_filtered_uniq_startOnly.bed \
    RAtEC2_dT_add40_filtered_uniq_startOnly_extendToMedianLength.bed  | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC2_add40_filtered_uniq_startOnly_extendToMedianLength.bed
    
cat RAtEC1_add40_filtered_uniq_startOnly_extendToMedianLength.bed \
    RAtEC2_add40_filtered_uniq_startOnly_extendToMedianLength.bed | \
    sort -T ./ -k1,1 -k2,2n > \
    RAtEC_add40_filtered_uniq_startOnly_extendToMedianLength.bed

```

#### .2 calling

```{bash, eval = F}
#!/bin/bash
#BSUB -J iEnCalling            ### set the job Name
#BSUB -q ser               ### specify queue, medium/short/debug/ser
#BSUB -n 2                  ### ask for number of cores (default: 1)
#BSUB -W 5:00                ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

###########################################################################
#       Scrips for calling enahncer using R package "BasicSTARRseq".
#                                                TanYongjun 20210306
###########################################################################

targetpath="/scratch/2021-03-02/bioTanyj/At_enhancer/enhancer_calling/iSTARR-seq"
cd ${targetpath}

## Strategy 1.
    # merge
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEP_add40_filtered_uniq_strategy1.bed \
        RAtEC_add40_filtered_uniq_strategy1.bed \
        ${targetpath} \
        iEnhancer_merge_strategy1_raw.tsv

    # rep1
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEP1_add40_filtered_uniq_strategy1.bed \
        RAtEC1_add40_filtered_uniq_strategy1.bed \
        ${targetpath} \
        iEnhancer_rep1_strategy1_raw.tsv

    # rep2
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEP2_add40_filtered_uniq_strategy1.bed \
        RAtEC2_add40_filtered_uniq_strategy1.bed \
        ${targetpath} \
        iEnhancer_rep2_strategy1_raw.tsv


# ## Strategy 2: startOnly
    # merged
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEP_add40_filtered_uniq_startOnly.bed \
        RAtEC_add40_filtered_uniq_startOnly.bed \
        ${targetpath} \
        iEnhancer_merge_strategy2_raw.tsv

    # rep1
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEP1_add40_filtered_uniq_startOnly.bed \
        RAtEC1_add40_filtered_uniq_startOnly.bed \
        ${targetpath} \
        iEnhancer_rep1_strategy2_raw.tsv

    # rep2
    Rscript /scratch/2021-03-02/bioTanyj/At_enhancer/cSTARR_seq/scripts/9.x_enhancer_calling.R \
        RAtEP2_add40_filtered_uniq_startOnly.bed \
        RAtEC2_add40_filtered_uniq_startOnly.bed \
        ${targetpath} \
        iEnhancer_rep2_strategy2_raw.tsv


```

## 9.3 downsample
Did library coverage enough?

### .1 subset and call enhancer
`/scratch/2021-05-13/bioTanyj/At_enhancer/scripts/9.3_downsample_Saturation_analysis_20210512.R`
`/scratch/2021-05-13/bioTanyj/At_enhancer/scripts/9.3_downsample_Saturation_analysis_20210512.sh`

`/scratch/2021-05-13/bioTanyj/At_enhancer/scripts/9.3.2_enhancer_calling.sh`

### .2 Saturation analysis
```{r}
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
if (F) {
  df <- NULL
  for (i in list.files("./R2.enhancer_calling/downsample/", 
                       pattern = "*.tsv.gz$")) {
    cat(i, "\n")
    tmp <- fread(paste("./R2.enhancer_calling/downsample/", i, sep = "")) %>%
      dplyr::filter(enrichment > 1.3) %>%
      nrow()
    df <- rbind(df, 
                c(str_replace(i, "Enhancers_(.+)_.+Rate(.+)_rep(\\d+).tsv.gz", "\\1"),
                  str_replace(i, "Enhancers_(.+)_.+Rate(.+)_rep(\\d+).tsv.gz", "\\2"),
                  str_replace(i, "Enhancers_(.+)_.+Rate(.+)_rep(\\d+).tsv.gz", "\\3"), 
                  tmp))
  }
  colnames(df) <- c("Sample", "downSampleRate", "rep", "NumEnhancer")
  # fwrite(df, file = "./R2.enhancer_calling/downsample/Number_of_enhancers_1.3.csv")
}else{
  df <- fread("./R2.enhancer_calling/downsample/Number_of_enhancers_1.3.csv")
}

# plot
df <- as.data.frame(df)
for (i in 2:4) {
  df[,i] <- as.numeric(df[,i])
}

df_mean <- df %>%
  group_by(downSampleRate, Sample) %>%
  summarise(Mean = mean(NumEnhancer))

df %>%
  ggplot(aes(x = downSampleRate, y = NumEnhancer, color = Sample)) + 
    geom_beeswarm(alpha = 1/2, size = 0.5) + 
    toolkit_tyj$theme_my + 
    geom_line(data = df_mean, aes(x = downSampleRate, y = Mean, color = Sample)) + 
    theme(legend.position = c(0.15, 0.8), legend.title = element_blank()) + 
    scale_color_npg() + 
    scale_x_continuous(breaks = seq(0, 1, 0.1)) + 
    ylab("Number of identified enhancers") + 
    xlab("Percentage of down sampled data")

toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/downsample/Number_of_identified_enhancers_with_downsampled_data",
                     width = 8, height = 6)

df %>%
  dplyr::filter(Sample == "merge") %>%
  ggplot(aes(x = downSampleRate, y = NumEnhancer)) + 
    geom_beeswarm(alpha = 2/3, size = 0.5, color = "grey30") + 
    toolkit_tyj$theme_my + 
    geom_line(data = dplyr::filter(df_mean, Sample == "merge"), 
              aes(x = downSampleRate, y = Mean), color = "grey30") + 
    theme(legend.position = c(0.15, 0.8), legend.title = element_blank()) + 
    scale_color_npg() + 
    scale_x_continuous(breaks = seq(0, 1, 0.1)) + 
    ylab("Number of identified enhancers") + 
    xlab("Percentage of down sampled data")

toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/downsample/Number_of_identified_enhancers_with_downsampled_data_onlymerge",
                     width = 8, height = 6)
```

### .3 repeatbility of top enhancers
Percent of top enhancers (not subseted) repeately identified in subsampled data.
```{r}

## TOP 2200 enhancers
rm(list = ls())
source("D:/SUST/code/TanYongjun_code.R")
topEn <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
  arrange(desc(enrichment)) %>%
  slice_head(n = 2200) %>%
  dplyr::select(seqnames, start, end) %>%
  dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""))
topEnGR <- makeGRangesFromDataFrame(topEn)  

df <- NULL
for (i in list.files("./R2.enhancer_calling/downsample/", pattern = ".+merge.+.tsv.gz$")) {
  cat(i, "\n")
  tmp <- fread(paste("./R2.enhancer_calling/downsample/", i, sep = "")) %>%
    dplyr::filter(enrichment > 1.3)
  tmpGR <- makeGRangesFromDataFrame(tmp)
  ol <- findOverlaps(topEnGR, tmpGR)
  df <- rbind(df, 
              c(i, length(unique(ol@from)), length(unique(ol@to))))
}

df <- as.data.frame(df)
colnames(df) <- c("id", "NumFrom", "NumTo")

df <- df %>%
  dplyr::mutate(Sample = str_replace_all(id, "Enhancers_(.+)_subRate(.+)_rep(.+).tsv.gz", "\\1"), 
                Rate = str_replace_all(id, "Enhancers_(.+)_subRate(.+)_rep(.+).tsv.gz", "\\2"),
                Rep = str_replace_all(id, "Enhancers_(.+)_subRate(.+)_rep(.+).tsv.gz", "\\3")) %>%
  dplyr::select(-id)

# fwrite(df, file = "./R2.enhancer_calling/downsample/TopEnhancer_repeatly_identified.csv",
#        row.names = F, col.names = T)

fread("./R2.enhancer_calling/downsample/TopEnhancer_repeatly_identified.csv") %>%
  dplyr::mutate(Percent = as.numeric(NumFrom) / 2200) %>%
  ggplot(aes(x = Rate, y = Percent)) + 
    geom_beeswarm(color = "grey40") + 
    toolkit_tyj$theme_my + 
    xlab("down sample rate") + 
    ylab("Percent of repeatly identified top enhancers")

## TOP enhancers
  rm(list = ls())
  source("D:/SUST/code/TanYongjun_code.R")
  allEn <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
    arrange(desc(enrichment)) %>%
    dplyr::select(seqnames, start, end) %>%
    dplyr::mutate(seqnames = str_c("chr", seqnames, sep = ""))
  allEnGR <- makeGRangesFromDataFrame(allEn)  
  
  df <- NULL
  for (i in list.files("./R2.enhancer_calling/downsample/", pattern = ".+merge.+.tsv.gz$")) {
    cat("\n>>>> ", i, "\n    ")
    tmp <- fread(paste("./R2.enhancer_calling/downsample/", i, sep = "")) %>%
      dplyr::filter(enrichment > 1.3) %>%
      arrange(desc(enrichment))
    for (j in seq(0.1, 1, 0.1)) {
      cat(j, " ")
      ol <- findOverlaps(allEnGR,
                         head(tmp, n = round(j * nrow(tmp), digits = 0)) %>%
                           makeGRangesFromDataFrame())
      df <- rbind(df, 
                  c(i, j, length(unique(ol@to)), round(j * nrow(tmp), digits = 0)))
    }
  }
  
  df <- as.data.frame(df)
  colnames(df) <- c("id", "TopEnRate", "NumHit", "NumTop")
  df <- df %>%
    dplyr::mutate(Sample = str_replace_all(id, "Enhancers_(.+)_subRate(.+)_rep(.+).tsv.gz", "\\1"), 
                  Rate = str_replace_all(id, "Enhancers_(.+)_subRate(.+)_rep(.+).tsv.gz", "\\2"),
                  Rep = str_replace_all(id, "Enhancers_(.+)_subRate(.+)_rep(.+).tsv.gz", "\\3")) %>%
    dplyr::select(-id)
  # fwrite(df , file = "./R2.enhancer_calling/downsample/Number_of_top_enhancer_repeatly_identified.csv")
  

```

*plot*
```{r}
  # plot
  fread("./R2.enhancer_calling/downsample/Number_of_top_enhancer_repeatly_identified.csv") %>%
    dplyr::mutate(olPercent = NumHit / NumTop, 
                  Rate = as.character(Rate), 
                  TopEnRate = as.character(TopEnRate)
                  ) %>%
    dplyr::filter(str_detect(Rate, "(0.2|0.4|0.6|0.8)")) %>%
    ggplot(aes(x = TopEnRate, y = olPercent, color = Rate)) + 
      geom_beeswarm(dodge.width = 0.2, alpha = 2/3, size = 0.3, shape = 16) +
      geom_boxplot(fill = "transparent", 
                   position = position_dodge(width = 0.2), 
                   outlier.alpha = 0,
                   width = 0.8, 
                   size = 0.2) +
      # geom_violin(fill = "transparent") +
      scale_color_npg() + 
      toolkit_tyj$theme_my + 
      theme(legend.position = c(0.12, 0.25), 
            legend.title = element_blank(), 
            legend.key.size = unit(0.3, units = "cm")) + 
      xlab("Enhancers identified using sampled down data") + 
      ylab("Percentage of shared enhancers") +
      scale_y_continuous(labels = scales::percent) + 
      scale_x_discrete(labels = paste(1:10, "0%", sep = ""))
    
  toolkit_tyj$SavePlot(filename_prefix = "./R2.enhancer_calling/downsample/Rediscovery_rate_with_varied_downsample_ratio",
                       width = 8, height = 6)

```




# 10 Chip seq

three datasets: SRX1044778: SRR2046531: H4K16ac SRX361945: SRR1005422: H3K9me1 DRX066756: DRR072818: H3K4me2

```{bash}
  # chip-seq
    prefetch SRR2046531 SRR1005422
    fasterq-dump SRR2046531 SRR1005422
    wget -c ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fastq/DRA005/DRA005154/DRX066756/DRR072818.fastq.bz2
    
```

```{bash}
#!/bin/bash
#BSUB -J call                  ### set the job Name
#BSUB -q short               ### specify queue, medium/short/debug/ser
#BSUB -n 40                  ### ask for number of cores (default: 1)
#BSUB -W 360:00             ### set walltime limit: hh:mm
#BSUB -e %J.err              ### -o and -e mean append, -oo and -eo mean overwrite
#BSUB -o %J.out              ### Specify the output and error file. %J is the job-id
#BSUB -R "span[ptile=40]"    ### ask for 40 cores per node

cd /scratch/2021-04-10/bioTanyj/At_enhancer/refgenome/ChIPseq

NumThreads=25

# index reference genome sequence
# /work/bio-tanyj/soft/bowtie2-2.4.1-linux-x86_64/bowtie2-build --threads 40 ../TAIR10_no_repeat.fa ../TAIR10_no_repeat.fa


# ## ----------------------QC -----------------------------------------------
# /work/bio-tanyj/soft/fastqc_v0.11.9/FastQC/fastqc -t 3 \*.fastq -o ./
# /work/bio-tanyj/miniconda3/bin/multiqc ./

## map and call peaks

# # trim
# for i in `ls | grep -E "\*.fastq"`
# do
#     /work/bio-tanyj/soft/TrimGalore-0.6.5/trim_galore --clip_R1 5 ${i} -j 15 --length 36 -o ./trimed/
# done

# # quanlity value of SRR2046531.fastq was abnormal.
# sed 's/\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*/FKKFKKKFKKKKKKKKKAKK7FKFAFFKKKKKKKKKKKKKKFA7KFKKFK/g' SRR2046531.fastq > SRR2046531_trimmed.fq

# ## QC of trimed files
# cd ./trimed
# /work/bio-tanyj/soft/fastqc_v0.11.9/FastQC/fastqc -t 3 \*trimmed.fq -o ./
# /work/bio-tanyj/miniconda3/bin/multiqc ./

# change name to histone modification


# ------------------------mapping & calling ---------------------------------
# for i in `ls | grep -E "*.fq"`
# do
#     {
#     sample_name=${i%.fq}
#     mkdir ${sample_name}_macs_res
#     echo "=============================bowtie2 result======================================="
#     # single-end
#     # /work/bio-tanyj/soft/bowtie2-2.4.1-linux-x86_64/bowtie2 -p ${NumThreads} \
#     #     -x ../TAIR10_no_repeat.fa -U ${i} | \
#     #     samtools view -bh -@ ${NumThreads} | \
#     #     samtools sort -@ ${NumThreads} \
#     #     -o ${sample_name}_raw.bam 
#     # # paired-end
#     # bowtie2 -p 8 -x /home/wanjing/data3/ref_genomes/TAIR10/TAIR10_bowtie2/chr_num_modified/TAIR10 -1 \*_1.fastq.gz -2 \*_2.fastq.gz | samtools view -bh -@ 8 | samtools sort -@ 8 -o "$sample_name"_raw.bam 

#     echo "=============================picard rm duplicate======================================="
#     java -jar /work/bio-tanyj/soft/picard.jar MarkDuplicates I="$sample_name"_raw.bam O="$sample_name".bam M="$sample_name"_dup.txt REMOVE_DUPLICATES=true
#     /work/bio-tanyj/miniconda3/bin/samtools index -@ ${NumThreads} "$sample_name".bam

#     echo "=============================bamCoverage======================================="
#     /work/bio-tanyj/miniconda3/bin/bamCoverage --bam "$sample_name".bam --binSize 10 --normalizeUsing RPGC --effectiveGenomeSize 135000000 -o "$sample_name".bw

#     echo "=============================callpeak narrowpeak======================================="
#     /work/bio-tanyj/miniconda3/envs/macs2/bin/macs2 callpeak -t "$sample_name".bam -n "$sample_name" -g 1.35e8 -q 0.01 --outdir ./${sample_name}_macs_res

#     echo "=============================callpeak broadpeak======================================="
#     /work/bio-tanyj/miniconda3/envs/macs2/bin/macs2 callpeak -t "$sample_name".bam -n "$sample_name" --broad --broad-cutoff 0.1 -g 1.35e8 --outdir ./${sample_name}_macs_res 
#     }&
# done
# wait


## H3K16ac need call with extra parameters
for i in `ls | grep -E "H4K16ac.fq"`
do
    {
    sample_name=${i%.fq}
    mkdir ${sample_name}_macs_res
    echo "=============================bowtie2 result======================================="
    # single-end
    # /work/bio-tanyj/soft/bowtie2-2.4.1-linux-x86_64/bowtie2 -p ${NumThreads} \
    #     -x ../TAIR10_no_repeat.fa -U ${i} | \
    #     samtools view -bh -@ ${NumThreads} | \
    #     samtools sort -@ ${NumThreads} \
    #     -o ${sample_name}_raw.bam 
    # # paired-end
    # bowtie2 -p 8 -x /home/wanjing/data3/ref_genomes/TAIR10/TAIR10_bowtie2/chr_num_modified/TAIR10 -1 \*_1.fastq.gz -2 \*_2.fastq.gz | samtools view -bh -@ 8 | samtools sort -@ 8 -o "$sample_name"_raw.bam 

    # echo "=============================picard rm duplicate======================================="
    # java -jar /work/bio-tanyj/soft/picard.jar MarkDuplicates I="$sample_name"_raw.bam O="$sample_name".bam M="$sample_name"_dup.txt REMOVE_DUPLICATES=true
    # /work/bio-tanyj/miniconda3/bin/samtools index -@ ${NumThreads} "$sample_name".bam

    # echo "=============================bamCoverage======================================="
    # /work/bio-tanyj/miniconda3/bin/bamCoverage --bam "$sample_name".bam --binSize 10 --normalizeUsing RPGC --effectiveGenomeSize 135000000 -o "$sample_name".bw

    echo "=============================callpeak narrowpeak======================================="
    /work/bio-tanyj/miniconda3/envs/macs2/bin/macs2 callpeak -t "$sample_name".bam -n "$sample_name" -g 1.35e8 -q 0.01 --nomodel --extsize 147 --outdir ./${sample_name}_macs_res

    echo "=============================callpeak broadpeak======================================="
    /work/bio-tanyj/miniconda3/envs/macs2/bin/macs2 callpeak -t "$sample_name".bam -n "$sample_name" --broad --broad-cutoff 0.1 -g 1.35e8 --nomodel --extsize 147 --outdir ./${sample_name}_macs_res 
    }&
done
wait
```


# 11 GWAS
Significant associated loci and all genotype can be download from https://aragwas.1001genomes.org/.
The study was published: https://academic.oup.com/nar/article/46/D1/D1150/4559687.

Did SALs located in enhancers higher than expectation? when comparing to CDS and promoters.
obtain SNP infor (chr, position) from download vcf file of 1135 accessions.

## 11.1 load SAL and SNP.

```{r}
## load SALs
  source("D:/SUST/code/TanYongjun_code.R")
  araGWASsal <- fread("./refgenome/araGWAS_catalog/aragwas_associations_significant_permutation.csv")
  araGWASsal <- araGWASsal %>%
    dplyr::select(snp.chr, snp.position) %>%
    unique %>%
    na.omit() %>%
    as.data.frame() %>%
    dplyr::rename(seqnames = snp.chr, end = snp.position) %>%
    dplyr::mutate(start = end - 1, seqnames = str_replace_all(seqnames, "chr", ""))
  salGR <- makeGRangesFromDataFrame(araGWASsal)

# ### all SNP in GWAS from HDF5 file (NOT USED!)
  # library(rhdf5)
  # h5f <- H5Fopen("./refgenome/araGWAS_catalog/regmap_SNP_MATRIX/all_chromosomes_binary.hdf5")
  # h5f$snps %>% head()
  # 
  # h5f <- H5Fopen("./refgenome/araGWAS_catalog/1001_SNP_MATRIX_GZIP/imputed_snps_binary_gzip.hdf5")
  # h5f
  # h5f$kinship %>% as.data.frame() %>% head() %>% View()
  # 
  # h5ls("./refgenome/araGWAS_catalog/full_imputed_SNP_MATRIX/all_chromosomes_binary.hdf5")
  # 
  # h5dump("./refgenome/araGWAS_catalog/full_imputed_SNP_MATRIX/all_chromosomes_binary.hdf5", load = F)

# all SNP and small indels (get from 1001genomes, not genotype provide by aragwas).
  genoAll <- fread("./R11.GWAS/1001genomes_snp-short-indel_only_ACGTN_loci_position.tsv.gz")
  genoAllGR <- genoAll %>%
    dplyr::mutate(start = V2 - 1) %>%
    dplyr::rename(seqnames = V1, end = V2) %>%
    makeGRangesFromDataFrame()
  snp_all_GR <- list(SAL = salGR, 
                     allSNP = genoAllGR)
  
# Did all SALs in genotype?
  ol <- findOverlaps(salGR, genoAllGR)
  cat(str_c("Did all SALs contained in the genotype file? ", length(salGR) == length(unique(ol@from))))

```


## 11.2 find overlap, and fisher's exact test
*Each kind of CREs as a whole*

```{r}
## load CREs
  cre_all <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
    dplyr::filter(!(method == "cSTARR-seq" & enrichment < 1.3))
  table(cre_all$method)
  IntronicEnhancer <- fread("./refgenome/Intronic_DHS_enhancer_seedling_20210604.csv.gz") %>%
    dplyr::select(seqnames:end) %>%
    dplyr::mutate(method = "Enhancer_Meng2021", 
                  seqnames = as.character(seqnames))
  cre_new <- fread("./ReMapping20210713/At_enhancer_1.3_20210713.bed") %>%
    dplyr::select(V1:V3, V7)
  colnames(cre_new) <- colnames(IntronicEnhancer)
  cre_all <- full_join(cre_all, IntronicEnhancer) %>%
    full_join(., cre_new) %>%
    dplyr::mutate(method = str_replace_all(method, c("cSTARR-seq" = "Enhancer_STARRseq"))) %>%
    dplyr::filter(str_detect(method, "(iSTARR|PAS|Random|Promoter)", negate = T))
  
  
## each kind of CRE as a whole.
  cre_all_GR <- cre_all %>%
    split(., .$method) %>%
    map(makeGRangesFromDataFrame)
  
# calculate overlap
  df <- NULL
  for (i in names(cre_all_GR)) {
    for (j in names(snp_all_GR)) {
      cat(">>> ", i, " vs ", j, "\n")
      ol <- findOverlaps(cre_all_GR[[i]], snp_all_GR[[j]])
      df <- rbind(df, 
                  c(i, j, length(cre_all_GR[[i]]), length(snp_all_GR[[j]]), length(unique(ol@to))))
    }
  }
  df <- as.data.frame(df)
  colnames(df) <- c("CRE", "SNP", "NumCRE", "NumLoci", "NumLociInCRE")
  for (i in 3:5) {
    df[,i] <- as.numeric(df[,i])
  }
  
# fisher exact test
  tmp <- unique(df$CRE)
  names(tmp) <- tmp
  FisherP <- map_df(tmp, function(x){
      z <- df[df$CRE == x, 4:5] %>%
        as.matrix() %>%
        fisher.test()
      return(z$p.value)
      }) %>%
    t() %>%
    as.data.frame() %>%
    mutate(CRE = row.names(.), 
           sigLable = toolkit_tyj$returnAsterisk(V1),
           V1 = scientific(V1, digits = 3), 
           start  = "allSNP", 
           end = "SAL", 
           y = 0.015, 
           SNP = NA,
           CRE = factor(CRE, levels = c("Enhancer_STARRseq", 
                                        "Enhancer_Zhu2015", 
                                        "Enhancer_Wang2019", 
                                        "Enhancer_Meng2021", 
                                        "new"))) %>%
    dplyr::rename(pvalue = V1)
  rm(tmp)

# p value
  df %>%
    dplyr::mutate(Percent = as.numeric(NumLociInCRE) / as.numeric(NumLoci), 
                  CRE = factor(CRE, levels = c("Enhancer_STARRseq", 
                                               "Enhancer_Zhu2015", 
                                               "Enhancer_Wang2019", 
                                               "Enhancer_Meng2021",
                                               "new"))) %>%
    ggplot(aes(x = SNP, y = Percent, fill = SNP)) + 
      geom_col() + 
      facet_wrap(facets = "CRE") + 
      toolkit_tyj$theme_my + 
      scale_color_npg() + 
      theme(legend.position = "none") + 
      geom_text(aes(y = Percent / 2, label = percent(Percent, accuracy = 0.01)), 
                size = 2) + 
      geom_signif(data = FisherP,
                  aes(xmin =start, xmax = end, annotations = sigLable, y_position = 0.038),
                  manual = T, textsize = 2, color = "grey25", size = 0.3) +
      ggtitle("Percent of SNP or SALs located in CREs") + 
      scale_y_continuous(expand = expansion(mult = c(0, 0), add = c(0, 0.007)))
  
  # toolkit_tyj$SavePlot("./R11.GWAS/Percent_of_SNP_SALs_located_in_CREs_Five_kind_enhancers_20210716",
  #                      width = 10, height = 8, device = "png")

```

*Each enhaner Kmeans cluster*

```{r}
## load CREs
  cre_all <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq" & enrichment > 1.3, seqnames != "Mt")
  cre_all$method <- str_c("Kmeans Cluster", cre_all$clusterK4)

  
  
## each kind of CRE as a whole.
  cre_all_GR <- cre_all %>%
    split(., .$method) %>%
    map(makeGRangesFromDataFrame)
  
# calculate overlap
  df <- NULL
  for (i in names(cre_all_GR)) {
    for (j in names(snp_all_GR)) {
      cat(">>> ", i, " vs ", j, "\n")
      ol <- findOverlaps(cre_all_GR[[i]], snp_all_GR[[j]])
      df <- rbind(df, 
                  c(i, j, length(cre_all_GR[[i]]), length(snp_all_GR[[j]]), length(unique(ol@to))))
    }
  }
  df <- as.data.frame(df)
  colnames(df) <- c("CRE", "SNP", "NumCRE", "NumLoci", "NumLociInCRE")
  for (i in 3:5) {
    df[,i] <- as.numeric(df[,i])
  }
  
# fisher exact test
  tmp <- unique(df$CRE)
  names(tmp) <- tmp
  FisherP <- map_df(tmp, function(x){
      z <- df[df$CRE == x, 4:5] %>%
        as.matrix() %>%
        fisher.test()
      return(z$p.value)
      }) %>%
    t() %>%
    as.data.frame() %>%
    mutate(CRE = row.names(.), 
           sigLable = toolkit_tyj$returnAsterisk(V1),
           V1 = scientific(V1, digits = 3), 
           start  = "allSNP", 
           end = "SAL", 
           y = 0.015, 
           SNP = NA) %>%
    dplyr::rename(pvalue = V1)
  rm(tmp)

# p value
  df %>%
    dplyr::mutate(Percent = as.numeric(NumLociInCRE) / as.numeric(NumLoci)) %>%
    ggplot(aes(x = SNP, y = Percent, fill = SNP)) + 
      geom_col() + 
      facet_wrap(facets = "CRE") + 
      toolkit_tyj$theme_my + 
      scale_color_npg() + 
      theme(legend.position = "none") + 
      geom_text(aes(y = Percent / 2, label = percent(Percent, accuracy = 0.01)), 
                size = 2) + 
      geom_signif(data = FisherP,
                  aes(xmin =start, xmax = end, annotations = sigLable, y_position = 0.012),
                  manual = T, textsize = 2, color = "grey25", size = 0.3) +
      ggtitle("Percent of SNP or SALs located in CREs") + 
      scale_y_continuous(expand = expansion(mult = c(0, 0), add = c(0, 0.002)))
  
  # toolkit_tyj$SavePlot("./R11.GWAS/Percent_of_SNP_SALs_located_in_CRE_enhancer_KmeansCluster_20210608",
  #                      width = 8, height = 8, device = "png")

```


*Each enhaner activity cluster*

```{r}
## load CREs
  cre_all <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_20210510.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq" & enrichment > 1.3, seqnames != "Mt")
  cre_all$method <- str_replace_all(cre_all$ClusterActivity, "Part", "Activity Cluster ")

  
  
## each kind of CRE as a whole.
  cre_all_GR <- cre_all %>%
    split(., .$method) %>%
    map(makeGRangesFromDataFrame)
  
# calculate overlap
  df <- NULL
  for (i in names(cre_all_GR)) {
    for (j in names(snp_all_GR)) {
      cat(">>> ", i, " vs ", j, "\n")
      ol <- findOverlaps(cre_all_GR[[i]], snp_all_GR[[j]])
      df <- rbind(df, 
                  c(i, j, length(cre_all_GR[[i]]), length(snp_all_GR[[j]]), length(unique(ol@to))))
    }
  }
  df <- as.data.frame(df)
  colnames(df) <- c("CRE", "SNP", "NumCRE", "NumLoci", "NumLociInCRE")
  for (i in 3:5) {
    df[,i] <- as.numeric(df[,i])
  }
  
# fisher exact test
  tmp <- unique(df$CRE)
  names(tmp) <- tmp
  FisherP <- map_df(tmp, function(x){
      z <- df[df$CRE == x, 4:5] %>%
        as.matrix() %>%
        fisher.test()
      return(z$p.value)
      }) %>%
    t() %>%
    as.data.frame() %>%
    mutate(CRE = row.names(.), 
           sigLable = toolkit_tyj$returnAsterisk(V1),
           V1 = scientific(V1, digits = 3), 
           start  = "allSNP", 
           end = "SAL", 
           y = 0.015, 
           SNP = NA) %>%
    dplyr::rename(pvalue = V1)
  rm(tmp)

# p value
  df %>%
    dplyr::mutate(Percent = as.numeric(NumLociInCRE) / as.numeric(NumLoci)) %>%
    ggplot(aes(x = SNP, y = Percent, fill = SNP)) + 
      geom_col() + 
      facet_wrap(facets = "CRE") + 
      toolkit_tyj$theme_my + 
      scale_color_npg() + 
      theme(legend.position = "none") + 
      geom_text(aes(y = Percent / 2, label = percent(Percent, accuracy = 0.01)), 
                size = 2) + 
      geom_signif(data = FisherP,
                  aes(xmin =start, xmax = end, annotations = sigLable, y_position = 0.012),
                  manual = T, textsize = 2, color = "grey25", size = 0.3) +
      ggtitle("Percent of SNP or SALs located in CREs") + 
      scale_y_continuous(expand = expansion(mult = c(0, 0), add = c(0, 0.002)))

  # toolkit_tyj$SavePlot("./R11.GWAS/Percent_of_SNP_SALs_located_in_CRE_enhancer_ActivityCluster_20210608",
  #                      width = 8, height = 8, device = "png")

```


## test
```{r}

df <- fread("Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz")
table(df$ClusterType)
df %>%
  dplyr::filter(sampleCov < 500) %>%
  group_by(ClusterType) %>%
  count()
```


## ol between enhancer/silencer
```{r}
source("D:/SUST/code/TanYongjun_code.R")
en_AT <- fread("Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz")
si_AT1 <- fread("./others/AtSI14C56_0.8fold.bed", header = F)
si_AT2 <- fread("./others/AtSI28C1213_0.8fold.bed", header = F)
enGR <- makeGRangesFromDataFrame(en_AT)
si1GR <- makeGRangesFromDataFrame(si_AT1, seqnames.field = "V1", start.field = "V2", end.field = "V3")
si2GR <- makeGRangesFromDataFrame(si_AT2, seqnames.field = "V1", start.field = "V2", end.field = "V3")

ol1 <- findOverlaps(enGR, si1GR)
ol2 <- findOverlaps(enGR, si2GR)
ol3 <- findOverlaps(si1GR, si2GR)


```

## ol enhancers identified in 202103/202107
```{r}
source("D:/SUST/code/TanYongjun_code.R")
enhancer1 <- fread("./Enhancers_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_TE_SuperEn_20210513.csv.gz") %>%
  dplyr::mutate(padj = p.adjust(pVal, method = "BH")) %>%
  dplyr::mutate(ID = "Old")
enhancer2 <- fread("./ReMapping20210713/At_enhancer_raw_20210713.tsv") %>%
  dplyr::mutate(padj = p.adjust(pVal, method = "BH"), seqnames = str_replace_all(seqnames, "chr", "")) %>%
  dplyr::filter(padj < 0.001, enrichment > 1.3) %>%
  dplyr::mutate(ID = "new")

full_join(enhancer1, enhancer2) %>%
  ggplot(aes(x = enrichment, color = ID)) + 
    geom_density()

# enhancer2 %>%
#   fwrite("./ReMapping20210713/At_enhancer_1.3_20210713.bed", row.names = F, col.names = F, quote = F, sep = "\t")

## ol between enhancers identified in differen datasets.
enhancer1GR <- makeGRangesFromDataFrame(enhancer1)
enhancer2GR <- makeGRangesFromDataFrame(enhancer2)
ol <- findOverlaps(enhancer1GR, enhancer2GR, minoverlap = 1)
ol

enhancer1$HitByNew <- "N"
enhancer1$HitByNew[ol@from] <- "Y"
table(enhancer1$ClusterType, enhancer1$HitByNew)
table(enhancer1$ClusterActivity, enhancer1$HitByNew)

```


```{r}
  source("D:/SUST/code/TanYongjun_code.R")
  
# load
  cre_all <- fread("./Enhancers_DHS_STARRseq1.3_H3K4_subsetChrState_PCA_score_Kmeans4-7_RPKM_ActCluster_IntronicEn_TE_20220424.csv.gz") %>%
    dplyr::filter(method == "cSTARR-seq", enrichment > 1.3)
  df <- cre_all %>%
    dplyr::select(seqnames:end, sampleCov, controlCov, pVal, enrichment, 
                  clusterK4, annotation, geneId, distanceToTSS, ClusterActivity,
                  TE, TE_Family, TE_SuperFamily) %>%
    dplyr::rename(Activity_Group = ClusterActivity, TE_Name = TE, 
                  Nearest_geneID = geneId, 
                  Distance_to_Neareast_Gene_TSS = distanceToTSS,
                  Kmeans_cluster = clusterK4) %>%
    dplyr::mutate(Kmeans_cluster = str_c("C", Kmeans_cluster),
                  Kmeans_cluster = ifelse(Kmeans_cluster == "C0", NA, Kmeans_cluster ),
                  Activity_Group = str_replace_all(Activity_Group, "Part", "Act"))
  df %>%
    count(Kmeans_cluster, Activity_Group)
  df %>%
    fwrite("./Enhancer_STARRseq_Kmeans_Activity_Features_TE_20220427.csv")
  NA
```


# Cd --------

```{r}
  # 
  source("D:/SUST/code/TanYongjun_code.R")
  df <- fread("D:/R/Cd_province/all_data.txt")
  
  # fwrite(df, file = "D:/R/Cd_province/all_data.csv.gz")
  
  tmp <- df %>% 
    dplyr::select(matches("农（镉）")) %>%
    dplyr::mutate(Sample = str_c("S", row.names(.))) %>%
    pivot_longer(cols = matches("农（镉）")) %>%
    dplyr::mutate(name = str_replace_all(name, "2013农（镉）", "2013农T2（镉）"),
                  Season = str_extract(name, "T\\d"), 
                  Year = str_extract(name, "\\d\\d\\d\\d")) %>%
    na.omit() %>%
    group_by(name, Season, Year) %>%
    dplyr::mutate(Num = n())
```

# plot
```{r}
  # histgram
  tmp %>%
    dplyr::mutate(Facet_x = str_c(Season, "\n(N=", Num, ")")) %>%
    ggplot(aes(x = value)) + 
      geom_histogram(binwidth = 0.1, color = "white") + 
      facet_grid(Year ~ Season, scales = "free") + 
      toolkit_tyj$theme_my +
      scale_x_continuous(breaks = seq(0, 4, 0.5), minor_breaks = waiver(), limits = c(0, 2)) +
      xlab("Cd(mg/kg)")
  toolkit_tyj$SavePlot(filename_prefix = "D:/R/Cd_province/Hist_all_Cd_20220429", width = 10, height = 8)
  
  
# barplot
  
  df_tmp <- tmp %>%
    dplyr::mutate(Bins = cut(ifelse(value == 0, 0.000001, value), breaks = c(seq(0, 1.5, 0.1), Inf))) %>%
    group_by(name, Bins) %>%
    dplyr::mutate(Bins_N = n()) %>%
    ungroup() %>%
    dplyr::select(-c(Sample, value)) %>%
    distinct() %>%
    dplyr::mutate(Rate = Bins_N / Num,
                  label = str_c(percent(Rate, accuracy = 0.01), "\nN=", Bins_N),
                  label = percent(Rate, accuracy = 0.01)) %>%
    group_by(name) %>%
    arrange(name, Bins) %>%
    dplyr::mutate(CumRate = cumsum(Rate),
                  labelCum = percent(CumRate, accuracy = 0.01))
  
  df_tmp %>%
    ggplot(aes(x = Bins, y = Rate)) + 
      geom_col(fill = "grey45") +
      facet_grid(Year ~ Season) + 
      toolkit_tyj$theme_my +
      xlab("Cd (mg/kg)") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_y_continuous(labels = scales::percent) + 
      geom_text(aes(y = Rate, label = label), size = 4/.pt, nudge_y = 0.17, color = "grey26", angle = 90) +
      ylab("Percentage")
  
  toolkit_tyj$SavePlot(filename_prefix = "D:/R/Cd_province/Hist_all_Cd_Percentage_20220429", width = 10, height = 8)
  
  df_tmp %>%
    ggplot(aes(x = Bins, y = CumRate, fill = Bins)) + 
      geom_col(fill = "grey55") +
      facet_grid(Year ~ Season) + 
      toolkit_tyj$theme_my +
      xlab("Cd (mg/kg)") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_y_continuous(labels = scales::percent, breaks = seq(0, 1, 0.2), limits = c(0, 1.3)) + 
      geom_text(aes(y = ifelse(CumRate > 0.6, CumRate/2, CumRate + 0.25), label = labelCum), 
                size = 4/.pt, nudge_y = 0, color = "grey26", angle = 90) +
      ylab("Percentage")
  
  toolkit_tyj$SavePlot(filename_prefix = "D:/R/Cd_province/Hist_all_Cd_Cumulative_Percentage_20220429", width = 10, height = 8)
  
  
# barplot
  
  df_tmp <- tmp %>%
    # dplyr::filter(value > 0.2) %>%
    dplyr::mutate(Bins = cut(ifelse(value == 0, 0.000001, value), breaks = c(seq(0, 1.5, 0.1), Inf))) %>%
    group_by(Bins) %>%
    dplyr::mutate(Bins_N = n()) %>%
    ungroup() %>%
    dplyr::select(Bins, Bins_N) %>%
    distinct() %>%
    dplyr::mutate(Rate = Bins_N / sum(Bins_N),
                  label = str_c(percent(Rate, accuracy = 0.01), "\nN=", Bins_N),
                  label = percent(Rate, accuracy = 0.01)) %>%
    # group_by(name) %>%
    arrange(Bins) %>%
    dplyr::mutate(CumRate = cumsum(Rate),
                  labelCum = percent(CumRate, accuracy = 0.01))
  
  df_tmp %>%
    ungroup() %>%
    ggplot(aes(x = Bins, y = Rate)) + 
      geom_line(aes(x = Bins, y = CumRate), color = "steelblue", group = "line") +
      geom_col(aes(fill = Bins)) +
      toolkit_tyj$theme_my +
      xlab("Cd (mg/kg)") +
      geom_point(aes(y = CumRate), size = 0.5) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_y_continuous(labels = scales::percent, breaks = seq(0, 1, 0.2), limits = c(0, 1)) + 
      geom_text(aes(y = CumRate, label = labelCum), 
                size = 5/.pt, nudge_y = -0.05, nudge_x = 0.7, color = "grey6") +
      ylab("Percentage") +
      scale_x_discrete(labels = c(seq(0.1, 1.5, 0.1), ">1.5")) + 
      scale_fill_manual(values = c("#1a9850", "27e478",
                                   colorRampPalette(c("#fdae61", "#fd382d"))(length(unique(df_tmp$Bins))-2))) + 
      theme(legend.position = "none")
  
  toolkit_tyj$SavePlot(filename_prefix = "D:/R/Cd_province/Hist_all_Cd_Cumulative_Percentage_merge_20220429", width = 8, height = 6)
  
  NA
```

# plot
```{r}
  # expectation of putative low-Cd variates.
  tmp_exp <- tmp %>%
    ungroup() %>%
    dplyr::mutate(LowVar = "-0%")
  for (i in seq(0.05, 0.95, 0.05)) {
    tmp_exp <- tmp %>%
      ungroup() %>%
      dplyr::mutate(value = value * (1 - i),
                    LowVar = str_c("-", percent(i, accuracy = 1))) %>%
      rbind(., tmp_exp)
  }

  #
  df_tmp <- tmp_exp %>%
    dplyr::mutate(Bins = cut(ifelse(value == 0, 0.000001, value), breaks = c(seq(0, 1.5, 0.1), Inf))) %>%
    group_by(LowVar, Bins) %>%
    dplyr::mutate(Bins_N = n()) %>%
    dplyr::select(LowVar, Bins, Bins_N) %>%
    distinct() %>%
    group_by(LowVar) %>%
    dplyr::mutate(Rate = Bins_N / sum(Bins_N),
                  label = str_c(percent(Rate, accuracy = 0.01), "\nN=", Bins_N),
                  label = percent(Rate, accuracy = 0.01)) %>%
    # group_by(name) %>%
    arrange(LowVar, Bins) %>%
    dplyr::mutate(CumRate = cumsum(Rate),
                  labelCum = percent(CumRate, accuracy = 0.01))
  
  df_tmp %>%
    ungroup() %>%
    ggplot(aes(x = Bins, y = Rate)) + 
      geom_line(aes(x = Bins, y = CumRate, color = LowVar, group = LowVar)) +
      # geom_col(fill = "grey55") +
      toolkit_tyj$theme_my +
      xlab("Cd (mg/kg)") +
      geom_point(aes(y = CumRate), size = 0.5) + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) + 
      scale_y_continuous(labels = scales::percent, breaks = seq(0, 1, 0.2), limits = c(0, 1)) + 
      geom_text(aes(y = CumRate, label = labelCum), data = dplyr::filter(df_tmp, Bins == "(0.1,0.2]"),
                size = 5/.pt, nudge_y = -0.05, nudge_x = 0.7, color = "grey6") +
      ylab("Percentage") +
      # scale_x_discrete(labels = c(str_c("< ", seq(0.1, 1.5, 0.1)), ">1.5")) + 
      scale_color_brewer(palette = "YlGn")
  
  toolkit_tyj$SavePlot(filename_prefix = "D:/R/Cd_province/Hist_all_Cd_Cumulative_Percentage_Expectation_merge_20220429", width = 8, height = 6)
  
  df_tmp1 <- df_tmp %>%
     dplyr::filter(Bins == "(0.1,0.2]") %>%
    dplyr::mutate(Exceed = 1-CumRate) %>%
    ungroup() %>%
    pivot_longer(cols = c(CumRate, Exceed)) %>%
    dplyr::mutate(name = str_replace_all(name, "CumRate", "Safe (<0.2)"),
                  name = factor(name, levels = c("Exceed", "Safe (<0.2)")),
                  LowVar = factor(LowVar, levels = c("-0%", 
                                                     str_c("-", percent(seq(0.05, 0.95, 0.05), accuracy = 1)))
                                  ))
  df_tmp1 %>%
    ggplot(aes(x = LowVar, y = value, fill = name)) + 
      geom_col() +
      toolkit_tyj$theme_my +
      geom_text(aes(x = LowVar, y = value/2, label = labelCum), 
                data = dplyr::filter(df_tmp1, name == "Safe (<0.2)"),
                size = 5/.pt, angle = 90, color = "grey15") + 
      scale_fill_npg("nrc", alpha = 4/5) + 
      scale_y_continuous(labels = scales::percent) + 
      xlab("Variety") + 
      ylab("Percentage") + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            legend.key.size = unit(0.3, "cm"),
            legend.title = element_blank(),
            legend.position = "bottom",
            legend.margin = margin(-10, 0,0,0))
  
  toolkit_tyj$SavePlot(filename_prefix = "D:/R/Cd_province/Barplot_Percentage_safe_Expectation_merge_20220429", width = 8, height = 6)
  
```



